{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d4364c55",
   "metadata": {},
   "outputs": [],
   "source": [
    "from hypergraphs import *\n",
    "from kegg import *\n",
    "import pandas as pd\n",
    "import os\n",
    "import logging\n",
    "\n",
    "from Bio.KEGG.KGML.KGML_parser import read\n",
    "import Bio.KEGG.KGML.KGML_pathway\n",
    "from ast import literal_eval\n",
    "from urllib.request import urlopen\n",
    "from urllib.parse import quote_plus\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "from Levenshtein import ratio\n",
    "import scipy.stats\n",
    "import pathlib\n",
    "from qf.cc import cardon_crochemore_colored, cardon_crochemore\n",
    "import random\n",
    "\n",
    "from reportlab.lib.units import inch\n",
    "import PIL.Image\n",
    "from io import BytesIO\n",
    "import reportlab.pdfgen.canvas\n",
    "import reportlab.lib.colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6b13e6bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "organismDict = {\"hsa\": \"Homo sapiens (human)\"}\n",
    "outputDirectory = \"/Users/boldi/Desktop/pw/\"\n",
    "inputDirectory = \"../../LaTeX/Data/KEGG-Pathways/Mar2024/\"\n",
    "dataDirectory = \"../../LaTeX/Data/KEGG-Pathways\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "87cb1b1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathwaylistFilename = \"KEGG-pathways.txt\"  # In inputDirectory, csv: ID, type, name\n",
    "superpathway = \"All\"                  # Name of the superpathway\n",
    "organism = \"hsa\"\n",
    "confidence = 0.05\n",
    "\n",
    "produceGraphsWithCCcoloring = False\n",
    "produceGraphsWithAnomalies = True\n",
    "useCCcoloringForAnomalies = False\n",
    "produceAnomaliesForNonsignificantPairs = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cb8e2d96",
   "metadata": {},
   "outputs": [],
   "source": [
    "safeSuperpathway = makesafe(superpathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e6caae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pws = pd.read_csv(os.path.join(inputDirectory, \"KEGG-pathways.txt\"), \n",
    "            names=[\"Pathway ID\", \"Pathway type\", \"Pathway name\"], dtype=\"string\", keep_default_na=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "199b2487",
   "metadata": {},
   "outputs": [],
   "source": [
    "superpathwayDict = {\n",
    "    superpathway: [(x,y) for x,y in zip(pws[\"Pathway ID\"], pws[\"Pathway name\"])]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "47a38c39",
   "metadata": {},
   "outputs": [],
   "source": [
    "def words2set(words):\n",
    "    \"\"\"\n",
    "        Given a space-separated list of words, returns it as a set of words\n",
    "    \"\"\"\n",
    "    return set(words.split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e0a74d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import NamedTuple\n",
    "\n",
    "class Relation(NamedTuple):\n",
    "    \"\"\"\n",
    "        Represents a relation. It is characterized by:\n",
    "        - a type (e.g. \"PPRel\") \n",
    "        - a list of subtypes (e.g. [('activation', '-->'), ('indirect effect', '..>')])\n",
    "        - a set of components (source)\n",
    "        - a list of sets of components (target)\n",
    "    \"\"\"\n",
    "    relType: str\n",
    "    relSubtypes: list\n",
    "    source: set\n",
    "    target: list\n",
    "    sourceg: Bio.KEGG.KGML.KGML_pathway.Graphics\n",
    "    targetg: Bio.KEGG.KGML.KGML_pathway.Graphics\n",
    "        \n",
    "    @classmethod\n",
    "    def fromKEGG(cls, relation, pathway):\n",
    "        \"\"\"\n",
    "            Constructor: builds a Relation from a KEGG relation\n",
    "        \"\"\"\n",
    "        if relation.entry2.name == \"undefined\":\n",
    "            s = []\n",
    "            for component in relation.entry2.components:\n",
    "                s += [words2set(pathway.entries[component.id].name)]\n",
    "            targetg = pathway.entries[component.id].graphics[0]\n",
    "        else:\n",
    "            s = [words2set(relation.entry2.name)]\n",
    "            targetg = relation.entry2.graphics[0]\n",
    "        return cls(relation.type, relation.subtypes, words2set(relation.entry1.name), s, targetg, relation.entry1.graphics[0])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ba5d2e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsFromKEGGpathway(pathway):\n",
    "    \"\"\"\n",
    "        Given a KEGG pathway, returns the list of relations it contains (represented as Relation).\n",
    "    \"\"\"\n",
    "    return [Relation.fromKEGG(r, pathway) for r in pathway.relations]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "751bc16c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pathwayFromFile(dataDirectory, organism, pathwayID):\n",
    "    \"\"\"\n",
    "        Load and return a KEGG pathway, given the data directory, the organism specifying the subdirectory,\n",
    "        and the pathwayID (the filename should be organism+pathwayID+\"xml\".\n",
    "    \"\"\"\n",
    "    return KGML_parser.read(open(os.path.join(dataDirectory, organism, organism+pathwayID+\".xml\"), \"r\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bd27fdc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsFromFile(dataDirectory, organism, pathwayID):\n",
    "    \"\"\"\n",
    "        Given a pathway file (specified by a root dataDirectory, the name of the organism subdirectory, and the name\n",
    "        of the file [organism+pathwayID.xml]), reads it and returns the list of its relations.\n",
    "    \"\"\"\n",
    "    pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "    return relationsFromKEGGpathway(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eaa01766",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsForSuperpathway(dataDirectory, organism, superpathwayDict, superpathway):\n",
    "    \"\"\"\n",
    "        Given a superpathDict (whose keys are superpathway names and whose values are list of\n",
    "        pairs (pathwayID, pathwayName)), reads all the pathway file for a specific superpathway name\n",
    "        and returns a dictionary whose keys are the pathwayID's and whose values are the list of relations.\n",
    "    \"\"\"\n",
    "    rel = {}\n",
    "    for k,v in superpathwayDict[superpathway]:\n",
    "        #logging.info(\"Reading\", k, v)\n",
    "        rel[k] = relationsFromFile(dataDirectory, organism, k)\n",
    "    return rel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "15479a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def subtypes(rel):\n",
    "    \"\"\"\n",
    "        Given a map whose values are relations, accumulate all subtypes appearing and assign them a number.\n",
    "        The result is a map from subtype string to number.\n",
    "    \"\"\"\n",
    "    all_subtypes = set([])\n",
    "    for pathwayid, relations in rel.items():\n",
    "        for relation in relations:\n",
    "            all_subtypes |= set([str(relation.relSubtypes)])\n",
    "        \n",
    "    s = sorted(list(all_subtypes))\n",
    "    subtype2color = {v:k for k,v in enumerate(s)}\n",
    "    color2subtype = {k:v for k,v in enumerate(s)}\n",
    "\n",
    "    return subtype2color, color2subtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1e229eaf-d667-466c-8238-8205257b0ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set2can(s):\n",
    "    \"\"\"\n",
    "        Convert a set to a string in a canonical way.\n",
    "    \"\"\"\n",
    "    return str(sorted(list(s)))\n",
    "\n",
    "def can2set(c):\n",
    "    \"\"\"\n",
    "        Does the converse of set2can.\n",
    "    \"\"\"\n",
    "    return set(literal_eval(c))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75b272b9-b220-492b-adf6-509f9952842f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indices(rel):\n",
    "    \"\"\"\n",
    "        Given a map whose values are lists of relations, it considers all the relations one by one, and attributes a unique id \n",
    "        to each element (i.e., set of components appearing as source or in the target of some relation) and block (the set of\n",
    "        element id appearing as target of some relation).\n",
    "        This function returns the dictionaries to move from/to an element or block to the corresponding id.\n",
    "        Elements are string representations of sorted lists of strings.\n",
    "        Blocks are string representations of sorted lists of ints.\n",
    "    \"\"\"\n",
    "    element2id = {}\n",
    "    id2element = {}\n",
    "    block2id = {}\n",
    "    id2block = {}\n",
    "\n",
    "    for relations in rel.values():\n",
    "        for relation in relations:\n",
    "            s = set2can(relation.source)\n",
    "            if s not in element2id.keys():\n",
    "                element2id[s] = len(element2id)\n",
    "            targetset = set([])\n",
    "            for targ in relation.target:\n",
    "                t = set2can(targ)\n",
    "                if t not in element2id.keys():\n",
    "                    element2id[t] = len(element2id)\n",
    "                targetset |= set([element2id[t]])\n",
    "            ts = set2can(targetset)\n",
    "            if ts not in block2id.keys():\n",
    "                block2id[ts] = len(block2id)\n",
    "\n",
    "    for k,v in element2id.items():\n",
    "        id2element[v] = k\n",
    "    for k,v in block2id.items():\n",
    "        id2block[v] = k\n",
    "    return element2id, id2element, block2id, id2block\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f9c3f128-cf97-4bb3-8ce5-65c49f865d98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_gene_KEGG(geneID):\n",
    "    \"\"\"\n",
    "        Search for gene on KEGG. \n",
    "        \n",
    "        Returns list of names.\n",
    "    \"\"\"\n",
    "    url1 = \"https://www.kegg.jp/entry/\"+geneID\n",
    "    response1 = requests.get(url1, allow_redirects=False)    \n",
    "    soup = BeautifulSoup(response1.text, \"html.parser\")\n",
    "    tds = [tds for tds in soup.find_all(\"td\", {\"class\": \"td11 defd\"})]\n",
    "    names = tds[0].getText().strip().split(\", \")\n",
    "    return names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1631e1d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_temp_imagefilename(url):\n",
    "    \"\"\"Return filename of temporary file containing downloaded image.\n",
    "\n",
    "    Create a new temporary file to hold the image file at the passed URL\n",
    "    and return the filename.\n",
    "    \"\"\"\n",
    "    img = urlopen(url).read()\n",
    "    im = PIL.Image.open(BytesIO(img))\n",
    "    # im.transpose(Image.FLIP_TOP_BOTTOM)\n",
    "    f = tempfile.NamedTemporaryFile(delete=False, suffix=\".png\")\n",
    "    fname = f.name\n",
    "    f.close()\n",
    "    im.save(fname, \"PNG\")\n",
    "    return fname\n",
    "\n",
    "\n",
    "#def enhance_method(klass, method_name, replacement):\n",
    "#    'replace a method with an enhanced version'\n",
    "#    method = getattr(klass, method_name)\n",
    "#    def enhanced(*args, **kwds): return replacement(*args, **kwds)\n",
    "#    setattr(klass, method_name, enhanced)\n",
    "\n",
    "def new_draw(self, filename, relations, elaborateArc = None):\n",
    "        \"\"\"Add the map elements to the drawing.\"\"\"\n",
    "        # Instantiate the drawing, first\n",
    "        # size x_max, y_max for now - we can add margins, later\n",
    "        if self.import_imagemap:\n",
    "            # We're drawing directly on the image, so we set the canvas to the\n",
    "            # same size as the image\n",
    "            if os.path.isfile(self.pathway.image):\n",
    "                imfilename = self.pathway.image\n",
    "            else:\n",
    "                imfilename = get_temp_imagefilename(self.pathway.image)\n",
    "            im = PIL.Image.open(imfilename)\n",
    "            cwidth, cheight = im.size\n",
    "        else:\n",
    "            # No image, so we set the canvas size to accommodate visible\n",
    "            # elements\n",
    "            cwidth, cheight = (self.pathway.bounds[1][0], self.pathway.bounds[1][1])\n",
    "        # Instantiate canvas\n",
    "        self.drawing = reportlab.pdfgen.canvas.Canvas(\n",
    "            filename,\n",
    "            bottomup=0,\n",
    "            pagesize=(\n",
    "                cwidth * (1 + 2 * self.margins[0]),\n",
    "                cheight * (1 + 2 * self.margins[1]),\n",
    "            ),\n",
    "        )\n",
    "        self.drawing.setFont(self.fontname, self.fontsize)\n",
    "        # Transform the canvas to add the margins\n",
    "        self.drawing.translate(\n",
    "            self.margins[0] * self.pathway.bounds[1][0],\n",
    "            self.margins[1] * self.pathway.bounds[1][1],\n",
    "        )\n",
    "        # Add the map image, if required\n",
    "        if self.import_imagemap:\n",
    "            self.drawing.saveState()\n",
    "            self.drawing.scale(1, -1)\n",
    "            self.drawing.translate(0, -cheight)\n",
    "            self.drawing.drawImage(imfilename, 0, 0)\n",
    "            self.drawing.restoreState()\n",
    "        # Add the reactions, compounds and maps\n",
    "        # Maps go on first, to be overlaid by more information.\n",
    "        # By default, they're slightly transparent.\n",
    "        if self.show_maps:\n",
    "            self.__add_maps()\n",
    "        if self.show_reaction_entries:\n",
    "            KGMLCanvas._KGMLCanvas__add_reaction_entries(self)\n",
    "        if self.show_orthologs:\n",
    "            KGMLCanvas._KGMLCanvas__add_orthologs(self)\n",
    "        if self.show_compounds:\n",
    "            KGMLCanvas._KGMLCanvas__add_compounds(self)\n",
    "        if self.show_genes:\n",
    "            KGMLCanvas._KGMLCanvas__add_genes(self)\n",
    "        # TODO: complete draw_relations code\n",
    "        # if self.draw_relations:\n",
    "        #    self.__add_relations()\n",
    "        # Write the pathway map to PDF\n",
    "        if elaborateArc is not None:\n",
    "            for relation in relations:\n",
    "                elaborateArc(relation, self.drawing)\n",
    "        self.drawing.save()\n",
    "        \n",
    "#enhance_method(KGMLCanvas, 'draw',  new_draw)\n",
    "setattr(KGMLCanvas, 'new_draw', new_draw)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0a11bbae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build all data needed for the relations of the superpathway\n",
    "rel = relationsForSuperpathway(dataDirectory, organism, superpathwayDict, superpathway)\n",
    "element2id, id2element, block2id, id2block = indices(rel)\n",
    "subtype2color, color2subtype = subtypes(rel)\n",
    "elements = list(set.union(*[can2set(k) for k in element2id.keys()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a1be8c2d-c5da-4e04-bf67-7e43b1a71b7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading gene names\n"
     ]
    }
   ],
   "source": [
    "# name2gene is a function mapping names to gene IDs (hsa:xxxx)\n",
    "# e.g. KCNE3, BRGDA6 etc are all mapped to hsa:10008 (see https://www.kegg.jp/entry/hsa:10008)\n",
    "# This map was produced by scraping kegg, but this has been done only once.\n",
    "# After that, we just read a pkl file\n",
    "\n",
    "name2keggFilename = os.path.join(inputDirectory, \"name2kegg.pkl\")\n",
    "\n",
    "if os.path.exists(name2keggFilename):\n",
    "    print(\"Reading gene names\")\n",
    "    with open(name2keggFilename, \"rb\") as handle:\n",
    "        name2gene = pickle.load(handle)\n",
    "else:\n",
    "    print(\"Producing gene names\")\n",
    "    d = {}\n",
    "    nelements = len(elements)\n",
    "    for count,element in enumerate(elements):\n",
    "        if count % 10 == 0:\n",
    "            print(f\"{count}/{nelements}\")\n",
    "        if element.startswith(\"hsa\"):\n",
    "            names = search_gene_KEGG(element)\n",
    "            for name in names:\n",
    "                d[name] = element\n",
    "    with open(name2keggFilename, \"wb\") as handle:\n",
    "        pickle.dump(d, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "    name2gene = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "74d19326",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading patient data\n"
     ]
    }
   ],
   "source": [
    "# Read patient data\n",
    "print(\"Reading patient data\")\n",
    "genes = pd.read_csv(os.path.join(inputDirectory, \"genes.csv\"), dtype={\"name\": str})\n",
    "patients = genes[\"name\"].values.tolist()\n",
    "classification = pd.read_csv(os.path.join(inputDirectory, \"classification.csv\"), dtype={\"name\": str})\n",
    "\n",
    "patient2class = {patient: classification.loc[classification[\"name\"]==patient][\"classification\"].values.flatten().tolist()[0] for patient in patients}\n",
    "numHealthy = sum([x == 'healthy' for x in patient2class.values()])\n",
    "numDiseased = sum([x == 'diseased' for x in patient2class.values()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "34acce33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dividing genes into two frames (one for healthy individuals and one for diseased individuals)\n",
    "genesHealthy = genes.loc[genes[\"name\"].apply(lambda x: patient2class[x])==\"healthy\"]\n",
    "genesDiseased = genes.loc[genes[\"name\"].apply(lambda x: patient2class[x])==\"diseased\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1c90f901-1413-4356-bfa8-faf25fc494df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building the relation graph\n",
      "- Edge (9,26) of type [('compound', 90)] for ['00010'] appears as [('compound', 91)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 91)]\n",
      "- Edge (26,4) of type [('compound', 93)] for ['00010', '00010'] appears as [('compound', 94)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 94)]\n",
      "- Edge (4,1) of type [('compound', 94)] for ['00010'] appears as [('compound', 93)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 93)]\n",
      "- Edge (4,6) of type [('compound', 94)] for ['00010', '00010'] appears as [('compound', 93)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 93)]\n",
      "- Edge (4,1) of type [('compound', 93)] for ['00010', '00010'] appears as [('compound', 94)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 94)]\n",
      "- Edge (33,39) of type [('compound', 330)] for ['00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (39,46) of type [('compound', 331)] for ['00230'] appears as [('compound', 326)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 326)]\n",
      "- Edge (39,48) of type [('compound', 331)] for ['00230'] appears as [('compound', 326)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 326)]\n",
      "- Edge (46,48) of type [('compound', 331)] for ['00230'] appears as [('compound', 326)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 326)]\n",
      "- Edge (33,37) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (33,39) of type [('compound', 325)] for ['00230', '00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (37,39) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (39,48) of type [('compound', 326)] for ['00230', '00230'] appears as [('compound', 366)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 366)]\n",
      "- Edge (48,37) of type [('compound', 332)] for ['00230'] appears as [('compound', 355)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 355)]\n",
      "- Edge (40,39) of type [('compound', 365)] for ['00230'] appears as [('compound', 357)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 357)]\n",
      "- Edge (33,58) of type [('compound', 324)] for ['00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (33,33) of type [('compound', 324)] for ['00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (33,35) of type [('compound', 324)] for ['00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (40,35) of type [('compound', 360)] for ['00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (58,35) of type [('compound', 360)] for ['00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (33,60) of type [('compound', 380)] for ['00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (53,78) of type [('compound', 320)] for ['00230'] appears as [('compound', 356)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 356)]\n",
      "- Edge (33,60) of type [('compound', 325)] for ['00230', '00230'] appears as [('compound', 390)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 390)]\n",
      "- Edge (35,60) of type [('compound', 380)] for ['00230'] appears as [('compound', 390)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 390)]\n",
      "- Edge (68,60) of type [('compound', 380)] for ['00230'] appears as [('compound', 390)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 390)]\n",
      "- Edge (58,35) of type [('compound', 362)] for ['00230', '00230'] appears as [('compound', 324)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 324)]\n",
      "- Edge (58,33) of type [('compound', 362)] for ['00230'] appears as [('compound', 324)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 324)]\n",
      "- Edge (58,35) of type [('compound', 324)] for ['00230', '00230', '00230'] appears as [('compound', 322)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 322)]\n",
      "- Edge (47,39) of type [('compound', 325)] for ['00230'] appears as [('compound', 321)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 321)]\n",
      "- Edge (35,33) of type [('compound', 324)] for ['00230'] appears as [('compound', 369)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 369)]\n",
      "- Edge (33,33) of type [('compound', 362)] for ['00230', '00230'] appears as [('compound', 369)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 369)]\n",
      "- Edge (52,33) of type [('compound', 325)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (52,39) of type [('compound', 325)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (53,37) of type [('compound', 332)] for ['00230'] appears as [('compound', 355)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 355)]\n",
      "- Edge (56,33) of type [('compound', 368)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (44,49) of type [('compound', 377)] for ['00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (63,37) of type [('compound', 365)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (63,39) of type [('compound', 365)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (56,34) of type [('compound', 330)] for ['00230'] appears as [('compound', 392)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 392)]\n",
      "- Edge (56,37) of type [('compound', 330)] for ['00230'] appears as [('compound', 392)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 392)]\n",
      "- Edge (34,37) of type [('compound', 330)] for ['00230'] appears as [('compound', 392)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 392)]\n",
      "- Edge (53,39) of type [('compound', 320)] for ['00230'] appears as [('compound', 367)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 367)]\n",
      "- Edge (53,37) of type [('compound', 355)] for ['00230', '00230'] appears as [('compound', 343)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 343)]\n",
      "- Edge (53,50) of type [('compound', 332)] for ['00230', '00230'] appears as [('compound', 343)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 343)]\n",
      "- Edge (53,64) of type [('compound', 355)] for ['00230'] appears as [('compound', 343)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 343)]\n",
      "- Edge (76,37) of type [('compound', 392)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (77,33) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (77,63) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (77,37) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (77,39) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (33,39) of type [('compound', 365)] for ['00230', '00230', '00230'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (35,33) of type [('compound', 369)] for ['00230', '00230'] appears as [('compound', 194)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 194)]\n",
      "- Edge (92,78) of type [('compound', 221)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (33,33) of type [('compound', 369)] for ['00230', '00230', '00230'] appears as [('compound', 194)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 194)]\n",
      "- Edge (33,93) of type [('compound', 213)] for ['00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (33,88) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (88,95) of type [('compound', 212)] for ['00240'] appears as [('compound', 188)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 188)]\n",
      "- Edge (99,78) of type [('compound', 218)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (90,88) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (39,95) of type [('compound', 212)] for ['00240'] appears as [('compound', 188)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 188)]\n",
      "- Edge (91,78) of type [('compound', 218)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (33,93) of type [('compound', 185)] for ['00240', '00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (39,60) of type [('compound', 365)] for ['00230'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (33,39) of type [('compound', 213)] for ['00230', '00230', '00230', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (90,93) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (33,39) of type [('compound', 195)] for ['00230', '00230', '00230', '00240', '00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (97,78) of type [('compound', 219)] for ['00240'] appears as [('compound', 196)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 196)]\n",
      "- Edge (88,93) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (63,39) of type [('compound', 330)] for ['00230', '00230'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (60,33) of type [('compound', 365)] for ['00230'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (90,39) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (33,60) of type [('compound', 390)] for ['00230', '00230', '00230'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (35,60) of type [('compound', 390)] for ['00230', '00230'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (60,39) of type [('compound', 325)] for ['00230'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (101,103) of type [('compound', 210)] for ['00240'] appears as [('compound', 202)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 202)]\n",
      "- Edge (60,33) of type [('compound', 191)] for ['00230', '00240'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (90,39) of type [('compound', 187)] for ['00240', '00240'] appears as [('compound', 221)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 221)]\n",
      "- Edge (103,104) of type [('compound', 209)] for ['00240'] appears as [('compound', 203)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 203)]\n",
      "- Edge (33,63) of type [('compound', 369)] for ['00230'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (35,90) of type [('compound', 186)] for ['00240'] appears as [('compound', 222)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 222)]\n",
      "- Edge (35,63) of type [('compound', 369)] for ['00230'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (33,33) of type [('compound', 194)] for ['00230', '00230', '00230', '00240'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (88,60) of type [('compound', 187)] for ['00240'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (35,33) of type [('compound', 194)] for ['00230', '00230', '00240'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (35,61) of type [('compound', 380)] for ['00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (33,61) of type [('compound', 380)] for ['00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (33,60) of type [('compound', 187)] for ['00230', '00230', '00230', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (60,61) of type [('compound', 380)] for ['00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (60,39) of type [('compound', 213)] for ['00230', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (39,95) of type [('compound', 188)] for ['00240', '00240'] appears as [('compound', 220)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 220)]\n",
      "- Edge (35,60) of type [('compound', 191)] for ['00230', '00230', '00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (100,97) of type [('compound', 196)] for ['00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (35,60) of type [('compound', 185)] for ['00230', '00230', '00240', '00240'] appears as [('compound', 193)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 193)]\n",
      "- Edge (33,60) of type [('compound', 195)] for ['00230', '00230', '00230', '00240', '00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (35,93) of type [('compound', 185)] for ['00240'] appears as [('compound', 193)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 193)]\n",
      "- Edge (33,60) of type [('compound', 185)] for ['00230', '00230', '00230', '00240', '00240', '00240'] appears as [('compound', 193)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 193)]\n",
      "- Edge (90,105) of type [('compound', 221)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (33,93) of type [('compound', 187)] for ['00240', '00240', '00240'] appears as [('compound', 193)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 193)]\n",
      "- Edge (98,33) of type [('compound', 185)] for ['00240'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (99,39) of type [('compound', 195)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (91,39) of type [('compound', 195)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (39,93) of type [('compound', 187)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (33,58) of type [('compound', 362)] for ['00230', '00230'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (33,33) of type [('compound', 186)] for ['00230', '00230', '00230', '00240', '00240'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (91,100) of type [('compound', 195)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (39,92) of type [('compound', 221)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (39,95) of type [('compound', 220)] for ['00240', '00240', '00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (35,58) of type [('compound', 222)] for ['00240'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (39,96) of type [('compound', 188)] for ['00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (33,93) of type [('compound', 193)] for ['00240', '00240', '00240', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (106,106) of type [('compound', 181)] for ['00240'] appears as [('compound', 180)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 180)]\n",
      "- Edge (35,33) of type [('compound', 186)] for ['00230', '00230', '00240', '00240'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (39,97) of type [('compound', 196)] for ['00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (98,93) of type [('compound', 185)] for ['00240'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (35,90) of type [('compound', 222)] for ['00240', '00240'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (95,78) of type [('compound', 219)] for ['00240'] appears as [('compound', 220)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 220)]\n",
      "- Edge (33,93) of type [('compound', 195)] for ['00240', '00240', '00240', '00240', '00240'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (35,91) of type [('compound', 194)] for ['00240'] appears as [('compound', 217)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 217)]\n",
      "- Edge (35,93) of type [('compound', 193)] for ['00240', '00240'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (35,58) of type [('compound', 214)] for ['00240', '00240'] appears as [('compound', 217)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 217)]\n",
      "- Edge (39,93) of type [('compound', 195)] for ['00240', '00240'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (159,165) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04062. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (483,485) of type [('activation', '-->')] for ['04062'] appears as [('binding/association', '---')] for 04062. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (483,131) of type [('activation', '-->')] for ['04062'] appears as [('binding/association', '---')] for 04062. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (165,187) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->')] for 04066. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,144) of type [('compound', 151), ('activation', '-->')] for ['04062'] appears as [('activation', '-->')] for 04066. Overwriting with type [('compound', 151), ('activation', '-->')]\n",
      "- Edge (568,569) of type [('compound', 49)] for ['04070'] appears as [('compound', 48)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 48)]\n",
      "- Edge (573,571) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (569,571) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (568,571) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (584,585) of type [('compound', 52)] for ['04070'] appears as [('compound', 235)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 235)]\n",
      "- Edge (585,584) of type [('compound', 234)] for ['04070'] appears as [('compound', 68)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 68)]\n",
      "- Edge (579,578) of type [('compound', 61)] for ['04070'] appears as [('compound', 56)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 56)]\n",
      "- Edge (588,587) of type [('compound', 75)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (588,575) of type [('compound', 75)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (492,573) of type [('compound', 73)] for ['04070'] appears as [('compound', 66)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 66)]\n",
      "- Edge (578,578) of type [('compound', 60)] for ['04070', '04070'] appears as [('compound', 231)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 231)]\n",
      "- Edge (590,586) of type [('compound', 66)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (586,573) of type [('compound', 73)] for ['04070'] appears as [('compound', 66)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 66)]\n",
      "- Edge (568,569) of type [('compound', 48)] for ['04070', '04070'] appears as [('compound', 72)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 72)]\n",
      "- Edge (574,587) of type [('compound', 74)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (573,568) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (568,569) of type [('compound', 72)] for ['04070', '04070', '04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (577,579) of type [('compound', 61)] for ['04070'] appears as [('compound', 56)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 56)]\n",
      "- Edge (581,577) of type [('compound', 62)] for ['04070'] appears as [('compound', 55)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 55)]\n",
      "- Edge (573,587) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (572,581) of type [('compound', 55)] for ['04070'] appears as [('compound', 63)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 63)]\n",
      "- Edge (587,569) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (570,594) of type [('compound', 66)] for ['04070'] appears as [('compound', 49)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 49)]\n",
      "- Edge (129,613) of type [('expression', '-->')] for ['04115'] appears as [('activation', '-->')] for 04115. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (489,494) of type [] for ['04062'] appears as [('activation', '-->')] for 04151. Overwriting with type [('activation', '-->')]\n",
      "- Edge (144,140) of type [('phosphorylation', '+p'), ('indirect effect', '..>')] for ['04062'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (204,157) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04151. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (157,500) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062'] appears as [('activation', '-->')] for 04151. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (528,165) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (533,553) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (553,554) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (120,121) of type [] for ['04010'] appears as [('activation', '-->')] for 04210. Overwriting with type [('activation', '-->')]\n",
      "- Edge (136,140) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04210. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (533,544) of type [('inhibition', '--|')] for ['04066'] appears as [('inhibition', '--|'), ('phosphorylation', '+p')] for 04218. Overwriting with type [('inhibition', '--|'), ('phosphorylation', '+p')]\n",
      "- Edge (573,494) of type [('inhibition', '--|'), ('dephosphorylation', '-p')] for ['04151'] appears as [('inhibition', '--|')] for 04218. Overwriting with type [('inhibition', '--|'), ('dephosphorylation', '-p')]\n",
      "- Edge (739,524) of type [('activation', '-->')] for ['04213'] appears as [('inhibition', '--|')] for 04218. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (129,224) of type [('expression', '-->')] for ['04115', '04115'] appears as [('activation', '-->')] for 04218. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (161,129) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04218. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (202,626) of type [('expression', '-->'), ('indirect effect', '..>')] for ['04218'] appears as [('expression', '-->')] for 04218. Overwriting with type [('expression', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (792,780) of type [('activation', '-->')] for ['04310'] appears as [('inhibition', '--|')] for 04310. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (129,623) of type [('expression', '-->')] for ['04115', '04115'] appears as [('indirect effect', '..>')] for 04310. \n",
      "****Conflict! Rewriting as [('indirect effect', '..>')]\n",
      "- Edge (141,758) of type [('activation', '-->'), ('dephosphorylation', '-p')] for ['04218'] appears as [('dephosphorylation', '-p')] for 04310. Overwriting with type [('activation', '-->'), ('dephosphorylation', '-p')]\n",
      "- Edge (499,114) of type [('activation', '-->')] for ['04310'] appears as [('phosphorylation', '+p'), ('indirect effect', '..>')] for 04510. \n",
      "****Conflict! Rewriting as [('phosphorylation', '+p'), ('indirect effect', '..>')]\n",
      "- Edge (492,144) of type [('compound', 151), ('activation', '-->')] for ['04062', '04066'] appears as [('compound', 24)] for 04510. \n",
      "****Conflict! Rewriting as [('compound', 24)]\n",
      "- Edge (507,204) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04510. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (663,144) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04210'] appears as [('phosphorylation', '+p')] for 04510. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (484,486) of type [('binding/association', '---')] for ['04062'] appears as [('phosphorylation', '+p')] for 04510. \n",
      "****Conflict! Rewriting as [('phosphorylation', '+p')]\n",
      "- Edge (114,706) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04210'] appears as [('phosphorylation', '+p')] for 04510. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (159,165) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04010', '04062'] appears as [('phosphorylation', '+p')] for 04510. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (165,563) of type [('expression', '-->'), ('indirect effect', '..>')] for ['04210'] appears as [('phosphorylation', '+p')] for 04510. \n",
      "****Conflict! Rewriting as [('phosphorylation', '+p')]\n",
      "- Edge (486,131) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04510. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (852,714) of type [('binding/association', '---')] for ['04510'] appears as [('activation', '-->')] for 04611. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,144) of type [('compound', 24)] for ['04062', '04066', '04510'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 04611. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (165,184) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 04611. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (716,482) of type [('dissociation', '-+-')] for ['04210'] appears as [('inhibition', '--|')] for 04620. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (142,117) of type [('activation', '-->')] for ['04010'] appears as [('binding/association', '---')] for 04620. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (528,165) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04151', '04210', '04218'] appears as [('phosphorylation', '+p')] for 04620. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (770,161) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04218'] appears as [('phosphorylation', '+p')] for 04620. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (144,533) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (553,554) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04151'] appears as [('activation', '-->')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (492,663) of type [('compound', 24)] for ['04510'] appears as [('compound', 71), ('activation', '-->')] for 04910. \n",
      "****Conflict! Rewriting as [('compound', 71), ('activation', '-->')]\n",
      "- Edge (157,500) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062', '04151'] appears as [('activation', '-->')] for 04910. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (165,187) of type [('activation', '-->')] for ['04010', '04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (165,124) of type [('phosphorylation', '+p')] for ['04010', '04510'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (187,545) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (736,742) of type [('activation', '-->')] for ['04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (736,649) of type [('activation', '-->')] for ['04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (649,492) of type [('activation', '-->')] for ['04151', '04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (961,716) of type [('phosphorylation', '+p')] for ['04620'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (716,482) of type [('inhibition', '--|')] for ['04210', '04620'] appears as [('dissociation', '-+-')] for 04931. \n",
      "****Conflict! Rewriting as [('dissociation', '-+-')]\n",
      "- Edge (649,492) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04213', '04931'] appears as [('activation', '-->')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (649,492) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04213', '04931', '04931'] appears as [('activation', '-->')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (492,503) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062'] appears as [('activation', '-->')] for 04931. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (1012,649) of type [('inhibition', '--|')] for ['04931'] appears as [('inhibition', '--|'), ('indirect effect', '..>')] for 04931. Overwriting with type [('inhibition', '--|'), ('indirect effect', '..>')]\n",
      "- Edge (547,548) of type [('activation', '-->')] for ['04611', '04931'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 04931. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (800,148) of type [('compound', 12)] for ['04310'] appears as [('compound', 277), ('activation', '-->')] for 05146. \n",
      "****Conflict! Rewriting as [('compound', 277), ('activation', '-->')]\n",
      "- Edge (513,859) of type [('inhibition', '--|'), ('phosphorylation', '+p')] for ['04510'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (492,144) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062', '04066', '04510', '04611', '04620'] appears as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')]\n",
      "- Edge (893,897) of type [('compound', 131)] for ['04512'] appears as [('binding/association', '---')] for 05205. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (639,728) of type [('activation', '-->')] for ['04213'] appears as [('binding/association', '---')] for 05205. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (165,124) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04010', '04510', '04910'] appears as [('phosphorylation', '+p')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (728,492) of type [('activation', '-->')] for ['04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (144,533) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04910'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (553,554) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04151', '04910'] appears as [('activation', '-->')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (553,670) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151'] appears as [('activation', '-->')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (492,663) of type [('compound', 71), ('activation', '-->')] for ['04510', '04910'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (663,553) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04910'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (512,513) of type [('activation', '-->')] for ['04062', '04510', '04611'] appears as [('binding/association', '---')] for 05205. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (1129,165) of type [('activation', '-->'), ('indirect effect', '..>')] for ['05205'] appears as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')]\n",
      "- Edge (561,1123) of type [('inhibition', '--|')] for ['05205'] appears as [('activation', '-->')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (539,482) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04066'] appears as [('activation', '-->')] for 05321. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (649,492) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04213', '04931', '04931', '04931'] appears as [('missing interaction', '-/-')] for 05415. \n",
      "****Conflict! Rewriting as [('missing interaction', '-/-')]\n",
      "- Edge (144,659) of type [('inhibition', '--|'), ('phosphorylation', '+p')] for ['04151', '04510', '04910', '04931'] appears as [('missing interaction', '-/-')] for 05415. \n",
      "****Conflict! Rewriting as [('missing interaction', '-/-')]\n",
      "- Edge (547,548) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04611', '04931', '04931'] appears as [('activation', '-->')] for 05415. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (492,494) of type [('activation', '-->')] for ['04151', '04210', '04213', '04213', '04218', '04931'] appears as [('missing interaction', '-/-')] for 05415. \n",
      "****Conflict! Rewriting as [('missing interaction', '-/-')]\n",
      "- Edge (145,142) of type [('binding/association', '---')] for ['04010'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (144,659) of type [('missing interaction', '-/-')] for ['04151', '04510', '04910', '04931', '05415'] appears as [('inhibition', '--|'), ('phosphorylation', '+p')] for 05417. \n",
      "****Conflict! Rewriting as [('inhibition', '--|'), ('phosphorylation', '+p')]\n",
      "- Edge (509,764) of type [('activation', '-->')] for ['04218', '04970'] appears as [('binding/association', '---')] for 05417. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (482,547) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04931'] appears as [('repression', '--|')] for 05417. \n",
      "****Conflict! Rewriting as [('repression', '--|')]\n",
      "- Edge (482,402) of type [('expression', '-->')] for ['04620'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (699,700) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04210'] appears as [('activation', '-->')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (482,1209) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (482,1210) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (482,1212) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (482,390) of type [('expression', '-->')] for ['04620', '05417', '05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (482,1213) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (482,300) of type [('expression', '-->')] for ['04620', '04931', '04931', '05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (1136,512) of type [('binding/association', '---')] for ['05205'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (482,271) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (482,183) of type [('expression', '-->')] for ['04620', '04931', '04931', '05321', '05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (482,402) of type [('activation', '-->')] for ['04620', '05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (482,1215) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (492,494) of type [('missing interaction', '-/-')] for ['04151', '04210', '04213', '04213', '04218', '04931', '05415'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (482,191) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (482,390) of type [('activation', '-->')] for ['04620', '05417', '05417', '05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (482,300) of type [('activation', '-->')] for ['04620', '04931', '04931', '05417', '05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (663,144) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04210', '04510', '04910', '04931'] appears as [('activation', '-->')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (482,183) of type [('activation', '-->')] for ['04620', '04931', '04931', '05321', '05417', '05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (114,115) of type [('phosphorylation', '+p')] for ['04010', '04620'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (468,595) of type [('indirect effect', '..>')] for ['04115'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05417. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (963,114) of type [('phosphorylation', '+p')] for ['04620'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (117,140) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "Wrap: more than 148 colors needed\n",
      "Wrap: more than 148 colors needed\n",
      "Wrap: more than 148 colors needed\n",
      "Wrap: more than 148 colors needed\n"
     ]
    }
   ],
   "source": [
    "# Building the graph of all relations contained in rel values\n",
    "# Nodes are IDs, and node x correspond to the set of genes (and/or compound) id2element[x]\n",
    "print(\"Building the relation graph\")\n",
    "G = nx.DiGraph() \n",
    "for pathwayID, relations in rel.items():\n",
    "    for relation in relations:\n",
    "        source = element2id[set2can(relation.source)]\n",
    "        for targ in relation.target:\n",
    "            target = element2id[set2can(targ)]\n",
    "            newType = str(relation.relSubtypes)\n",
    "            color = subtype2color[newType]\n",
    "            pl = []\n",
    "            if G.has_edge(source, target):\n",
    "                d = G.get_edge_data(source, target)\n",
    "                if d[\"color\"] != color:\n",
    "                    oldType = color2subtype[d[\"color\"]]\n",
    "                    print(\"- Edge ({},{}) of type {} for {} appears as {} for {}. \".format(\n",
    "                        source, target, color2subtype[d[\"color\"]], d[\"pathwayIDs\"], \n",
    "                        str(relation.relSubtypes), pathwayID\n",
    "                    ), end=\"\")\n",
    "                    if oldType[1:-1] in newType[1:-1]:\n",
    "                        print(\"Overwriting with type {}\".format(newType))\n",
    "                    elif newType[1:-1] in oldType[1:-1]:\n",
    "                        print(\"Overwriting with type {}\".format(oldType))\n",
    "                        color = subtype2color[oldType]\n",
    "                    else:\n",
    "                        print(\"\\n****Conflict! Rewriting as {}\".format(newType))\n",
    "                pl = d[\"pathwayIDs\"]\n",
    "            G.add_edge(source, target, color=color, pathwayIDs=pl + [pathwayID])\n",
    "            \n",
    "\n",
    "# Computing the minimum fibres\n",
    "cc=cardon_crochemore_colored(G)\n",
    "\n",
    "colors = list(matplotlib.colors.CSS4_COLORS.values())\n",
    "usedColors = 0\n",
    "\n",
    "cc2fibresize = {}\n",
    "cc2fibre = {}\n",
    "ccnontrivial = {}\n",
    "nontrivialclass2col = {}\n",
    "\n",
    "fibres = []\n",
    "node2fibre = {}\n",
    "for v in set(cc.values()):\n",
    "    fibre = [x for x in G.nodes() if cc[x] == v]\n",
    "    fibres += [fibre]\n",
    "    for x in fibre:\n",
    "        node2fibre[x] = fibre\n",
    "    cc2fibre[v] = fibre\n",
    "    cc2fibresize[v] = len(fibre)\n",
    "    if len(fibre) > 0:\n",
    "        ccnontrivial[v] = cc[v]\n",
    "        nontrivialclass2col[v] = colors[usedColors]\n",
    "        usedColors = (usedColors + 1 ) % len(colors)\n",
    "        if usedColors == 0:\n",
    "            print(\"Wrap: more than\", len(colors), \"colors needed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d51dbd7f-35b9-444a-9238-3af727c06411",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping genes to columns\n"
     ]
    }
   ],
   "source": [
    "# Build node2columns, mapping each node of G to the list of genes corresponding to it\n",
    "# for which we know a value (in the genes dataframe), sorted\n",
    "#\n",
    "\n",
    "print(\"Mapping genes to columns\")\n",
    "namedGenes = set(name2gene.values())     \n",
    "columns = set(genes.columns.values.tolist()[2:])\n",
    "node2columns = {}\n",
    "for node in G.nodes():\n",
    "    nodeGenes = can2set(id2element[node])\n",
    "    s = []\n",
    "    for gene in nodeGenes & namedGenes:\n",
    "        for k,v in name2gene.items():\n",
    "            if v == gene:\n",
    "                if k in columns:\n",
    "                    s += [k]\n",
    "    node2columns[node]=sorted(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "05b96661",
   "metadata": {},
   "outputs": [],
   "source": [
    "if produceGraphsWithCCcoloring:\n",
    "    s = \"\"\n",
    "    for pair in superpathwayDict[superpathway]:\n",
    "        pathwayID = pair[0]\n",
    "        pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "        canvas = KGMLCanvas(pathway)\n",
    "        for k in pathway.entries:\n",
    "            name = pathway.entries[k].name\n",
    "            canonicalName = set2can(words2set(name))\n",
    "            if canonicalName in element2id.keys():\n",
    "                node = element2id[canonicalName]\n",
    "                klass = cc[node]\n",
    "                if klass in nontrivialclass2col:\n",
    "                    pathway.entries[k].graphics[0].bgcolor = nontrivialclass2col[klass]\n",
    "                    pathway.entries[k].graphics[0].fgcolor = nontrivialclass2col[klass]\n",
    "                else:\n",
    "                    pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "                    pathway.enries[k].graphics[0].fgcolor = \"#FFFFFF\"\n",
    "                s += pathwayID + \"\\t\" + str(node) + \"\\t\" + str(klass) +\"\\t\"+pathway.entries[k].graphics[0].name+\"\\t\"+canonicalName + \"\\n\"\n",
    "        canvas.import_imagemap = True\n",
    "        pdfName = organism + pathwayID + \".pdf\"\n",
    "        pathlib.Path(os.path.join(outputDirectory, organism, safeSuperpathway)).mkdir(parents=True, exist_ok=True)\n",
    "        canvas.draw(os.path.join(outputDirectory, organism, safeSuperpathway, pdfName))\n",
    "    with open(os.path.join(outputDirectory, organism, safeSuperpathway, \"all.tsv\"), \"wt\") as file:\n",
    "        file.write(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "73ce2fd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def columns2nodes(columns, node2columns, ignoreEmpty = True):\n",
    "    \"\"\"\n",
    "        Returns a map from columns to the nodes where the column appears.\n",
    "    \"\"\"\n",
    "    res = {} \n",
    "    for c in columns:\n",
    "        s = set([])\n",
    "        for no, co in node2columns.items():\n",
    "            if c in co:\n",
    "                s |= set([no])\n",
    "        if len(s) > 0 or not ignoreEmpty:\n",
    "            res[c] = s\n",
    "    return res\n",
    "\n",
    "c2n = columns2nodes(columns, node2columns)\n",
    "usableColumns = list(c2n.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "00742eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def corrForPair(genes, set1, set2, correlationFunction = lambda x,y: scipy.stats.kendalltau(x,y).statistic):\n",
    "    \"\"\"\n",
    "        Given a dataframe genes (i.e., a set of patients), and two sets of column names (set1, set2), \n",
    "        compute the correlation function\n",
    "        for every pair of columns in set1 and set2 and return the average statistic and its std.\n",
    "    \"\"\"\n",
    "    taus = []\n",
    "    for gene1 in set1:\n",
    "        for gene2 in set2:\n",
    "            expr1 = genes.loc[:,gene1].values.tolist()\n",
    "            expr2 = genes.loc[:,gene2].values.tolist()\n",
    "            if len(set(expr1)) == 1 or len(set(expr2)) == 1:\n",
    "                continue\n",
    "            taus += [correlationFunction(expr1, expr2).statistic]\n",
    "    if len(taus) < 2:\n",
    "        return np.nan, np.nan\n",
    "    return pd.DataFrame(taus).mean().values[0], pd.DataFrame(taus).std().values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b3232726",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sampleDeltaDist(genes, set1, set2, numberOfRows1, numberOfSamples = 500, correlationFunction = lambda x,y: scipy.stats.kendalltau(x,y).statistic):\n",
    "    \"\"\"\n",
    "        Repeatedly (for numberOfSamples times) do the following:\n",
    "        - select numberOfRows1 rows of genes, and then compute tauForPair on those rows for set1 and set2\n",
    "        - do the same for remaining rows\n",
    "        - take the difference between the two resulting set of taus.\n",
    "        Return the mean and std of differences.\n",
    "    \"\"\"\n",
    "    diffs = []\n",
    "    \n",
    "    for i in range(numberOfSamples):\n",
    "        genes1 = random.sample(range(len(genes)), numberOfRows1)\n",
    "        genes2 = list(set(range(len(genes)))-set(genes1))\n",
    "        mt1, mstd1 = corrForPair(genes.iloc[genes1], set1, set2, correlationFunction)\n",
    "        if np.isnan(mp1):\n",
    "            continue\n",
    "        mt2, mstd2 = corrForPair(genes.iloc[genes2], set1, set2, correlationFunction)\n",
    "        if np.isnan(mp2):\n",
    "            continue\n",
    "        diffs += [mt1-mt2]\n",
    "    return diffs\n",
    "\n",
    "\n",
    "def deltaForPair(genes1, genes2, set1, set2, correlationFunction = lambda x,y: scipy.stats.kendalltau(x,y).statistic):\n",
    "    \"\"\"\n",
    "        Compute corrForPair for set1 and set2 over the set of rows genes1 and genes2, and return\n",
    "        the difference of taus.\n",
    "    \"\"\"\n",
    "    mt1, mstd1 = corrForPair(genes1, set1, set2, correlationFunction)\n",
    "    mt2, mstd2 = corrForPair(genes2, set1, set2, correlationFunction)\n",
    "    return mt1-mt2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2d3473f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Producing anomalies\n",
      "0.00% done\n",
      "0.39% done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "correlationFunction = lambda x,y:  scipy.stats.spearmanr(x,y).statistic\n",
    "correlationFunctionName = \"spearmanrho\"\n",
    "relAnomalies = os.path.join(inputDirectory, \"anomalies-\" + correlationFunctionName + \".csv\")\n",
    "\n",
    "if os.path.exists(relAnomalies):\n",
    "    print(\"Reading anomalies\")\n",
    "    relcorrel = pd.read_csv(relAnomalies)\n",
    "else:\n",
    "    print(\"Producing anomalies\")\n",
    "    relcorrel = pd.DataFrame(columns=[\n",
    "        \"Source gene(s)\",\n",
    "        \"Target gene(s)\",\n",
    "        \"Pathway ID\",\n",
    "        \"Relation type\",\n",
    "        \"Correlation for healthy\",\n",
    "        \"Correlation for diseased\",\n",
    "        \"Correlation difference (H-D)\",\n",
    "        \"Expected correlation difference\",\n",
    "        \"P-value\",\n",
    "        \"Comment\"\n",
    "        ])\n",
    "    numberOfEdges = G.number_of_edges()\n",
    "    doneEdges = 0\n",
    "    for s,t,d in G.edges(data=True):\n",
    "        if doneEdges % 10 == 0:\n",
    "            print(\"{:.2f}% done\".format(100 * doneEdges / numberOfEdges))\n",
    "        doneEdges += 1\n",
    "        subtypeName = [k for k,v in subtype2color.items() if d[\"color\"]==v][0]\n",
    "        sg = node2columns[s]\n",
    "        tg = node2columns[t]\n",
    "        if len(sg) == 0 or len(tg) == 0:\n",
    "            continue\n",
    "        avgh, stdh = corrForPair(genesHealthy, sg, tg, correlationFunction=correlationFunction)\n",
    "        avgd, stdd = corrForPair(genesDiseased, sg, tg, correlationFunction=correlationFunction)\n",
    "        taudiff = avgh - avgd\n",
    "        ttn = sampleDeltaDist(genes, sg, tg, numHealthy, correlationFunction=correlationFunction)\n",
    "        tt = [x for x in ttn if ~np.isnan(x)]\n",
    "        if len(tt) == 0:\n",
    "            continue\n",
    "        ttmean = pd.DataFrame(tt).mean().values[0]\n",
    "        ecdft = scipy.stats.ecdf(tt)  # Empirical CDF\n",
    "        if taudiff < 0:\n",
    "            pvalue = ecdft.cdf.evaluate(taudiff)\n",
    "            comment = \"More correlated than for healthy individuals\"\n",
    "        else:\n",
    "            pvalue = 1 - ecdft.cdf.evaluate(taudiff)    \n",
    "            comment = \"Less correlated than for healthy individuals\"\n",
    "        if produceAnomaliesForNonsignificantPairs or produpvalue <= confidence:\n",
    "            relcorrel.loc[len(relcorrel)] = [\n",
    "                str(sg),\n",
    "                str(tg),\n",
    "                d[\"pathwayIDs\"],\n",
    "                subtypeName,\n",
    "                avgh, \n",
    "                avgd,\n",
    "                taudiff,\n",
    "                ttmean,\n",
    "                pvalue,\n",
    "                comment\n",
    "            ]\n",
    "    relcorrel = relcorrel.sort_values(by=[\"P-value\"], ascending=True)\n",
    "    relcorrel.to_csv(relAnomalies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e517681e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Anomalies by type\")\n",
    "statisticallySignificant = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence)   # Statistically significant\n",
    "]\n",
    "\n",
    "failingActivation = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"activation\")) & # Activation  \n",
    "    (relcorrel[\"Correlation difference (H-D)\"]>0) # Less than healthy\n",
    "]\n",
    "\n",
    "failingInhibition = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"inhibition\")) & # Inhibition\n",
    "    (relcorrel[\"Correlation difference (H-D)\"]<0) # More than healthy\n",
    "]\n",
    "\n",
    "excessiveActivation = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"activation\")) & # Activation  \n",
    "    (relcorrel[\"Correlation difference (H-D)\"]<0) # More than healthy\n",
    "]\n",
    "\n",
    "excessiveInhibition = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"inhibition\")) & # Inhibition\n",
    "    (relcorrel[\"Correlation difference (H-D)\"]>0) # Less than healthy\n",
    "]\n",
    "\n",
    "otherTypesOfRelations = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    ~(relcorrel[\"Relation type\"].str.contains(\"activation\")) & # No activation  \n",
    "    ~(relcorrel[\"Relation type\"].str.contains(\"inhibition\"))   # No inhibition \n",
    "]\n",
    "\n",
    "print(\"P-value < {:.2f}: \".format(100 * confidence / 100), len(statisticallySignificant))\n",
    "print(\"Failing activations (green):\", len(failingActivation))\n",
    "print(\"Failing inhibitions (blue):\", len(failingInhibition))\n",
    "print(\"Excessive activations (red):\", len(excessiveActivation))\n",
    "print(\"Excessive inhibitions (orange):\", len(excessiveInhibition))\n",
    "print(\"Other types of relations (yellow):\", len(otherTypesOfRelations))\n",
    "print(\"Total:\", len(failingActivation)+len(failingInhibition)+len(excessiveActivation)+len(excessiveInhibition)+len(otherTypesOfRelations))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2eec52e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[-0.1605292450249678,\n",
       " 0.14975082186964828,\n",
       " 0.09727994462633172,\n",
       " 0.3240312421286076,\n",
       " -0.23044038108235876,\n",
       " 0.03338489304447943,\n",
       " -0.39402541652159706,\n",
       " -0.3803968261327082,\n",
       " 0.17535109153211897,\n",
       " -0.21385777820094493,\n",
       " 0.17246375480913334,\n",
       " -0.20765279980064555,\n",
       " 0.1728306382170895,\n",
       " -0.39349967894042276,\n",
       " -0.06217319945594946,\n",
       " 0.17466758767501955,\n",
       " -0.0417330913850789,\n",
       " 0.15049138327449013,\n",
       " 0.1665258681779001,\n",
       " 0.11953482074568816,\n",
       " -0.0903156133346941,\n",
       " 0.25039908477145645,\n",
       " 0.3089560699986381,\n",
       " -0.2502289246118846,\n",
       " -0.3588843569873766,\n",
       " -0.13212019277210216,\n",
       " -0.17154204335931805,\n",
       " -0.3222983194616919,\n",
       " -0.10614653891251623,\n",
       " 0.11945020713924845,\n",
       " -0.18653689268923299,\n",
       " 0.2893566110885536,\n",
       " 0.18039519493853723,\n",
       " 0.01111111111111111,\n",
       " 0.31908652243634705,\n",
       " -0.14289936635875478,\n",
       " 0.12779734154389577,\n",
       " 0.14772715214805793,\n",
       " 0.13946702840433883,\n",
       " 0.18952688001782703,\n",
       " -0.05152156891982226,\n",
       " -0.1536060418848146,\n",
       " -0.24642796953555815,\n",
       " -0.22422398967859974,\n",
       " 0.15708070452290357,\n",
       " -0.08862278417760289,\n",
       " 0.18117912537947822,\n",
       " -0.2268270770328142,\n",
       " 0.2565460277670126,\n",
       " 0.29338935694376816,\n",
       " -0.24427831117452245,\n",
       " 0.10760524116260115,\n",
       " -0.2005568002856269,\n",
       " -0.4052860220243814,\n",
       " -0.4556296442026382,\n",
       " 0.24573433973519387,\n",
       " 0.35385238230474336,\n",
       " -0.23996235882194317,\n",
       " 0.2277018975335675,\n",
       " 0.011128297681493145,\n",
       " -0.19887093109355397,\n",
       " -0.11670949281793547,\n",
       " -0.06207098488582596,\n",
       " -0.04987053502147783,\n",
       " 0.01111111111111111,\n",
       " -0.06207629545421801,\n",
       " -0.2960603195248383,\n",
       " 0.4313527749076412,\n",
       " -0.2801314246856127,\n",
       " 0.19891334715401401,\n",
       " 0.1917072541334155,\n",
       " -0.1719041093409137,\n",
       " 0.22485356952189262,\n",
       " 0.2253928214978938,\n",
       " -0.061999727395305786,\n",
       " -0.4221362215519936,\n",
       " 0.26445779716787066,\n",
       " 0.18883849663574992,\n",
       " -0.3803968261327082,\n",
       " -0.2177063502345598,\n",
       " -0.2540373632560117,\n",
       " 0.058222529225006515,\n",
       " 0.2060459619187188,\n",
       " -0.2682414815376495,\n",
       " 0.3344389034961457,\n",
       " 0.1394034965511986,\n",
       " 0.22833895134619747,\n",
       " 0.20630578139817762,\n",
       " -0.011180339887498949,\n",
       " -0.3326147295458609,\n",
       " 0.31418567403528397,\n",
       " 0.39204766186200074,\n",
       " 0.2564354551590313,\n",
       " 0.08055169670971271,\n",
       " -0.062101579945570945,\n",
       " 0.392321817742906,\n",
       " -0.19447372533059998,\n",
       " 0.2760795549587042,\n",
       " -0.1430110300828384,\n",
       " 0.0500434982893525,\n",
       " 0.17880572488733176,\n",
       " 0.16669439453224766,\n",
       " -0.021614606808287386,\n",
       " -0.23374385771204942,\n",
       " -0.38840999983420615,\n",
       " -0.17410085645431805,\n",
       " -0.3642011599819127,\n",
       " 0.23563672242991174,\n",
       " -0.39491600154800244,\n",
       " -0.12985689604837894,\n",
       " -0.11665003647306169,\n",
       " 0.09746028852245811,\n",
       " -0.17849914872339298,\n",
       " 0.34420513866107944,\n",
       " 0.09734429939923661,\n",
       " 0.1896601985237749,\n",
       " -0.23414168830968995,\n",
       " -0.14289936635875478,\n",
       " -0.19854773575141438,\n",
       " 0.2256498068033591,\n",
       " -0.021611113534963713,\n",
       " -0.38487373508208506,\n",
       " -0.00546449957962231,\n",
       " -0.17233226052711442,\n",
       " -0.28058107064673427,\n",
       " 0.2771424317518624,\n",
       " 0.20460034006103084,\n",
       " -0.18999842457073682,\n",
       " 0.26776836204031934,\n",
       " -0.00939675292092211,\n",
       " 0.4042276513038637,\n",
       " -0.1975935269441103,\n",
       " 0.404736391326171,\n",
       " -0.26986589812137385,\n",
       " -0.3218769728059042,\n",
       " 0.0028147124805533823,\n",
       " 0.011145564251507057,\n",
       " -0.18799134564310838,\n",
       " -0.13095036610225785,\n",
       " 0.29853597637451923,\n",
       " -0.12632221925315154,\n",
       " 0.2452029888164201,\n",
       " 0.25665327510703595,\n",
       " -0.011145564251507057,\n",
       " 0.43032569496257833,\n",
       " -0.28879725425059266,\n",
       " 0.3913273801782829,\n",
       " -0.2406404671582223,\n",
       " -0.4524445306996524,\n",
       " 0.3087167911677978,\n",
       " 0.0028216989227445467,\n",
       " 0.3244748485791777,\n",
       " 0.28396169396820703,\n",
       " -0.25040749171337184,\n",
       " 0.01111111111111111,\n",
       " 0.2399629095970629,\n",
       " 0.02774914702901172,\n",
       " -0.15911500499414344,\n",
       " -0.10261239595707045,\n",
       " -0.2239000446668623,\n",
       " 0.46950948740621345,\n",
       " -0.17306280732697246,\n",
       " -0.2339423111327797,\n",
       " 0.24025974025974026,\n",
       " 0.09712348156025435,\n",
       " -0.10618404343457463,\n",
       " 0.33738583641333936,\n",
       " 0.10749147109876975,\n",
       " 0.0500434982893525,\n",
       " 0.29338935694376816,\n",
       " -0.3484029000826902,\n",
       " -0.35480292347259335,\n",
       " 0.1293259935504979,\n",
       " 0.13956625511470397,\n",
       " -0.19439671683981175,\n",
       " 0.0029640828250181023,\n",
       " 0.2053264689598151,\n",
       " 0.41724375660966817,\n",
       " 0.22023677813839265,\n",
       " -0.4029694084100962,\n",
       " -0.26004689175377793,\n",
       " -0.20032973371014473,\n",
       " -0.38079395724302817,\n",
       " -0.322819055743902,\n",
       " 0.32459463956484824,\n",
       " 0.13389625972088873,\n",
       " 0.15908589581388266,\n",
       " -0.1937408151738464,\n",
       " -0.26998327179779497,\n",
       " 0.0723346267923666,\n",
       " 0.14151701383707832,\n",
       " 0.2238631519570014,\n",
       " -0.3382663370011983,\n",
       " -0.18792765260974312,\n",
       " 0.05002949001563663,\n",
       " -0.2408547815129669,\n",
       " -0.4362706802682732,\n",
       " -0.3320677124741052,\n",
       " 0.055185875536909464,\n",
       " -0.4024341478841671,\n",
       " 0.2017580130037816,\n",
       " -0.08847967493479109,\n",
       " 0.15843389763547314,\n",
       " -0.27113098182960405,\n",
       " -0.17049869702740011,\n",
       " -0.35846504526361267,\n",
       " -0.05044974734318092,\n",
       " 0.3642021528901843,\n",
       " 0.29882945526820726,\n",
       " -0.13797755217164226,\n",
       " -0.1786943003832723,\n",
       " -0.24398169767007183,\n",
       " 0.0803592155141961,\n",
       " 0.1579698546924509,\n",
       " 0.07492065545758919,\n",
       " -0.1422461272309552,\n",
       " 0.3246390258128079,\n",
       " -0.13147171750745304,\n",
       " 0.11925977093056642,\n",
       " 0.2229997621322205,\n",
       " -0.29617301026218035,\n",
       " 0.28981835542676226,\n",
       " -0.3122560529957745,\n",
       " -0.23025237628669845,\n",
       " 0.08862766837598918,\n",
       " -0.3775028738628892,\n",
       " -0.4281003561011584,\n",
       " -0.021614606808287386,\n",
       " -0.2466164517036805,\n",
       " 0.21156004996584687,\n",
       " 0.1886172069548474,\n",
       " 0.14926670285032886,\n",
       " 0.0873570938727074,\n",
       " 0.4828803187095696,\n",
       " 0.2920948092277542,\n",
       " 0.2831305388354483,\n",
       " 0.35044059302286323,\n",
       " -0.15966591122918308,\n",
       " 0.1719041093409137,\n",
       " -0.2556522145790442,\n",
       " -0.4108465259457845,\n",
       " 0.10554627495657275,\n",
       " -0.027675997844038205,\n",
       " 0.058222529225006515,\n",
       " 0.224750891534175,\n",
       " -0.13215215625051993,\n",
       " 0.25373072883168984,\n",
       " -0.13559074456241796,\n",
       " 0.1804860649103287,\n",
       " 0.34012031569988455,\n",
       " 0.1588312969209859,\n",
       " -0.36310632624000905,\n",
       " -0.2146328333027206,\n",
       " -0.3224496348305391,\n",
       " -0.3121416956601221,\n",
       " -0.07527276571396392,\n",
       " 0.08664651580950446,\n",
       " 0.08055169670971271,\n",
       " 0.0500434982893525,\n",
       " -0.35576041977255635,\n",
       " -0.26004689175377793,\n",
       " 0.10292915806279637,\n",
       " -0.254040970526607,\n",
       " 0.04999169857931076,\n",
       " 0.11945020713924845,\n",
       " -0.04542036258154457,\n",
       " -0.2272435710604329,\n",
       " -0.08608199595189851,\n",
       " -0.08945676550323428,\n",
       " 0.18158430982523543,\n",
       " 0.04992613059444879,\n",
       " 0.23608254127076303,\n",
       " 0.03333333333333333,\n",
       " 0.019407509741668837,\n",
       " 0.3152359316077862,\n",
       " -0.3481406334596244,\n",
       " -0.14242535963690067,\n",
       " -0.251062751333177,\n",
       " 0.01111111111111111,\n",
       " 0.058222529225006515,\n",
       " -0.3066063650326517,\n",
       " 0.33402950937439146,\n",
       " 0.019441962882064782,\n",
       " -0.14202566736460015,\n",
       " 0.25665327510703595,\n",
       " 0.1421019325843148,\n",
       " -0.021614606808287386,\n",
       " 0.03211499021563735,\n",
       " -0.15913282265699066,\n",
       " 0.41627261488646794,\n",
       " 0.15160635049303095,\n",
       " 0.20434429376682306,\n",
       " -0.2507665178820482,\n",
       " -0.08940553638175322,\n",
       " 0.3779063551847508,\n",
       " 0.12959078526388795,\n",
       " -0.2963083863007511,\n",
       " 0.22761908869295616,\n",
       " -0.3945920136355232,\n",
       " -0.3233585528589189,\n",
       " 0.3152359316077862,\n",
       " -0.2963286029761791,\n",
       " -0.2701953132753041,\n",
       " 0.36605658775753047,\n",
       " -0.2703017381024182,\n",
       " 0.12783936636504337,\n",
       " -0.2963286029761791,\n",
       " 0.2538229595957179,\n",
       " 0.011162911443718952,\n",
       " 0.43001100628561545,\n",
       " 0.03604857645259864,\n",
       " -0.13797755217164226,\n",
       " 0.302564524265467,\n",
       " 0.40565572244494935,\n",
       " -0.28960537393097613,\n",
       " -0.20749123908977618,\n",
       " -0.03483171220790085,\n",
       " 0.14456760224535797,\n",
       " 0.15049138327449013,\n",
       " -0.45272636018627366,\n",
       " 0.3379860351770726,\n",
       " 0.23709844893573795,\n",
       " -0.41850794515868084,\n",
       " 0.19869434933632701,\n",
       " 0.3247147349749698,\n",
       " -0.2596014027826284,\n",
       " -0.28001706734996035,\n",
       " 0.25623491554508515,\n",
       " 0.08068021453208879,\n",
       " -0.2675457666329173,\n",
       " 0.25585728404337393,\n",
       " -0.08602133259875835,\n",
       " 0.11101947647281493,\n",
       " 0.16155497752231723,\n",
       " 0.1788312381558294,\n",
       " 0.3771551370631937,\n",
       " 0.14739057429274816,\n",
       " -0.20054459224128035,\n",
       " -0.39402541652159706,\n",
       " 0.10266697366945096,\n",
       " -0.021624805161535712,\n",
       " 0.22049790817744963,\n",
       " 0.2722390183180996,\n",
       " 0.15903623309479065,\n",
       " -0.10610317248240983,\n",
       " -0.42051490265699326,\n",
       " 0.041785132961714586,\n",
       " -0.06568217028598644,\n",
       " 0.14225733171444002,\n",
       " -0.01943170044890881,\n",
       " -0.4216236109038688,\n",
       " 0.07204884399925604,\n",
       " -0.17396210154561947,\n",
       " -0.10547196938851068,\n",
       " -0.28611598548158623,\n",
       " -0.3586044622370645,\n",
       " 0.20361705831429727,\n",
       " 0.30883627901296756,\n",
       " 0.26677532048516106,\n",
       " -0.17205230110559142,\n",
       " -0.28600956065447225,\n",
       " -0.16953693111920445,\n",
       " 0.20885332341775623,\n",
       " 0.17154204335931805,\n",
       " 0.11123256083021205,\n",
       " -0.08891341577976619,\n",
       " -0.19962698307484433,\n",
       " 0.11932990114657421,\n",
       " 0.3349789291555689,\n",
       " -0.1702846534237119,\n",
       " 0.019452323632085516,\n",
       " 0.08871320430902181,\n",
       " -0.13233903046141188,\n",
       " 0.08876916662521822,\n",
       " 0.10947767153697097,\n",
       " -0.06207098488582596,\n",
       " 0.08040129364380078,\n",
       " 0.20630578139817762,\n",
       " 0.08079089532916793,\n",
       " -0.1024223836749254,\n",
       " -0.08931052639439299,\n",
       " -0.15389356844685415,\n",
       " -0.1716624058008212,\n",
       " 0.11099138884744043,\n",
       " -0.2697498243041546,\n",
       " -0.1703558207909037,\n",
       " -0.2800228120473487,\n",
       " 0.19707940204588278,\n",
       " 0.45650739297192844,\n",
       " -0.4043043317691966,\n",
       " -0.26965439278019177,\n",
       " 0.12716437589306553,\n",
       " -0.2669734429692655,\n",
       " 0.04992613059444879,\n",
       " -0.07534420383221213,\n",
       " 0.48248442005145503,\n",
       " 0.20637672813973768,\n",
       " -0.43039156723616095,\n",
       " 0.11932990114657421,\n",
       " -0.4103168928927632,\n",
       " 0.16679347565363778,\n",
       " -0.00930269549124705,\n",
       " 0.35136559896228897,\n",
       " 0.01943872235525243,\n",
       " -0.22692244852967808,\n",
       " -0.06304974880384685,\n",
       " 0.0028286677446475623,\n",
       " 0.298201423675574,\n",
       " 0.3919678501405024,\n",
       " -0.2130983191250612,\n",
       " 0.44499369421314816,\n",
       " 0.2452029888164201,\n",
       " 0.2203007924728525,\n",
       " -0.22830511991482544,\n",
       " 0.2749774376060589,\n",
       " 0.10755516413213503,\n",
       " 0.11096337230000869,\n",
       " -0.14230572022148047,\n",
       " 0.18420458516531654,\n",
       " -0.42655951750321275,\n",
       " -0.11763015745769988,\n",
       " -0.46131396717082096,\n",
       " 0.04989822006626043,\n",
       " -0.3696995507106915,\n",
       " -0.005461241283324271,\n",
       " 0.08046576385780255,\n",
       " 0.41694800436546664,\n",
       " -0.15812799492176818,\n",
       " -0.2437899926848096]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4917654",
   "metadata": {},
   "outputs": [],
   "source": [
    "def elaborateArc(relation, sdrawing):\n",
    "    source= str(node2columns[element2id[set2can(relation.source)]])\n",
    "    for tg in relation.target:\n",
    "        target = str(node2columns[element2id[set2can(tg)]])\n",
    "        r = relcorrel.loc[(relcorrel[\"Source gene(s)\"] == source) & (relcorrel[\"Target gene(s)\"] == target)]\n",
    "        if len(r) == 0:\n",
    "            continue\n",
    "        if len(r) > 1:\n",
    "            pass\n",
    "            #print(f\"({source}, {target}) appearing more than once in relcorrel\")\n",
    "            #print(list(r.to_dict()[\"Correlation difference (H-D)\"].values()))\n",
    "        rd = r.to_dict()\n",
    "        pvalue = list(rd[\"P-value\"].values())[0]\n",
    "        corrdiff = list(rd[\"Correlation difference (H-D)\"].values())[0]\n",
    "        relationType = list(rd[\"Relation type\"].values())[0]\n",
    "        if pvalue < confidence:        \n",
    "            if \"activation\" in relationType:\n",
    "                if corrdiff > 0:\n",
    "                    col = reportlab.lib.colors.green\n",
    "                else:\n",
    "                    col = reportlab.lib.colors.red\n",
    "            elif \"inhibition\" in relationType:\n",
    "                if corrdiff > 0:\n",
    "                    col = reportlab.lib.colors.orange\n",
    "                else:\n",
    "                    col = reportlab.lib.colors.blue\n",
    "            else:\n",
    "                col = reportlab.lib.colors.yellow\n",
    "            if np.isnan(corrdif):\n",
    "                print(f\"Correlation for {source} vs. {target} is {corrdiff}\")\n",
    "            else:\n",
    "                sdrawing.setLineWidth(3*math.exp(abs(corrdiff)*3))\n",
    "                sdrawing.setStrokeColor(col)\n",
    "                sdrawing.line(relation.sourceg.x, relation.sourceg.y, relation.targetg.x, relation.targetg.y)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32b16bfb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if produceGraphsWithAnomalies:\n",
    "    print(\"Producing anomaly PDFs\")\n",
    "    s = \"\"\n",
    "    for pair in superpathwayDict[superpathway]:\n",
    "        pathwayID = pair[0]\n",
    "        pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "\n",
    "        relations = []\n",
    "        for relation in rel[pathwayID]:\n",
    "            relations += [relation]\n",
    "\n",
    "        canvas = KGMLCanvas(pathway)\n",
    "\n",
    "        if useCCcoloringForAnomalies:\n",
    "            for k in pathway.entries:\n",
    "                name = pathway.entries[k].name\n",
    "                canonicalName = set2can(words2set(name))\n",
    "                if canonicalName in element2id.keys():\n",
    "                    node = element2id[canonicalName]\n",
    "                    klass = cc[node]\n",
    "                    if klass in nontrivialclass2col:\n",
    "                        pathway.entries[k].graphics[0].bgcolor = nontrivialclass2col[klass]\n",
    "                        pathway.entries[k].graphics[0].fgcolor = nontrivialclass2col[klass]\n",
    "                    else:\n",
    "                        pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "                        pathway.entries[k].graphics[0].fgcolor = \"#FFFFFF\"\n",
    "                    s += pathwayID + \"\\t\" + str(node) + \"\\t\" + str(klass) +\"\\t\"+pathway.entries[k].graphics[0].name+\"\\t\"+canonicalName + \"\\n\"\n",
    "\n",
    "        canvas.import_imagemap = True\n",
    "        pdfName = organism + pathwayID + \".pdf\"\n",
    "        pathlib.Path(os.path.join(outputDirectory, organism, safeSuperpathway + \"-anomalies\")).mkdir(parents=True, exist_ok=True)\n",
    "        canvas.new_draw(os.path.join(outputDirectory, organism, safeSuperpathway+ \"-anomalies\", pdfName), relations, elaborateArc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e63a371",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Computing anomalies by pathway\")\n",
    "pathwayAnomalies = pd.DataFrame(columns=[\n",
    "        \"Pathway ID\",\n",
    "        \"Pathway\",\n",
    "        \"Average square difference\",\n",
    "        \"Count\",\n",
    "        \"Relative count\"\n",
    "        ])\n",
    "for pair in superpathwayDict[superpathway]:\n",
    "    pathwayID = pair[0]\n",
    "    pathwayName= pair[1]\n",
    "    ssqd = 0\n",
    "    count = 0\n",
    "    countSig = 0\n",
    "    for relation in rel[pathwayID]:\n",
    "        source = str(node2columns[element2id[set2can(relation.source)]])\n",
    "        for targ in relation.target:\n",
    "            target = str(node2columns[element2id[set2can(targ)]])\n",
    "            tt = relcorrel.loc[(relcorrel[\"Source gene(s)\"] == source) & (relcorrel[\"Target gene(s)\"] == target)]\n",
    "            if len(tt) == 1 and tt[\"P-value\"].to_list()[0] <= confidence:\n",
    "                cd = tt[\"Correlation difference (H-D)\"].to_list()[0]\n",
    "                ssqd += cd * cd\n",
    "                countSig += 1\n",
    "            count += 1\n",
    "    if count > 0:\n",
    "        pathwayAnomalies.loc[len(pathwayAnomalies)]=[pathwayID, pathwayName, ssqd, countSig, countSig / count]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff411669-e133-48ac-86b9-dcc350d4d957",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Sorting pathways by number of anomalies and writing it out as a csv\")\n",
    "relPathwayAnomalies = os.path.join(inputDirectory, \"pathway-anomalies-\" + correlationFunctionName + \".csv\")\n",
    "pathwayAnomalies = pathwayAnomalies.sort_values(by=[\"Count\"], ascending=False)\n",
    "pathwayAnomalies.to_csv(relPathwayAnomalies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cae2695",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Extracting genes involved in top-k anomalous pathways and appearing as columns\")\n",
    "nTopAnomPathway = 7\n",
    "topAnomPathway = pathwayAnomalies.head(nTopAnomPathway)[\"Pathway ID\"].to_list()\n",
    "anomPathwayGenes = set([])\n",
    "for pathwayId in topAnomPathway:\n",
    "    for relation in rel[pathwayId]:\n",
    "        anomPathwayGenes |= relation.source \n",
    "        for t in relation.target:\n",
    "            anomPathwayGenes |= t\n",
    "anomPathwayGeneNames = [name for name,gene in name2gene.items() if gene in anomPathwayGenes and name in columns]\n",
    "print(len(anomPathwayGeneNames), \"genes involved\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1e14ecf",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Computing correlations for anomalous-pathway genes in the two populations\")\n",
    "healthyMatrix = genesHealthy[anomPathwayGeneNames].to_numpy() #Rows are healthy patients, columns are genes appearing in anomalous pathways\n",
    "diseasedMatrix = genesDiseased[anomPathwayGeneNames].to_numpy() #Rows are diseased patients, columns are genes appearing in anomalous pathways\n",
    "\n",
    "healthyCorr = np.corrcoef(healthyMatrix, rowvar=False) #Correlation between anomalous genes in healthy individuals\n",
    "diseasedCorr = np.corrcoef(diseasedMatrix, rowvar=False) #Correlation between anomalous genes in diseased individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "f3b26aa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sklearn.cluster\n",
    "bh = sklearn.cluster.k_means(np.square(healthyCorr), n_clusters=9)[1]\n",
    "bd = sklearn.cluster.k_means(np.square(diseasedCorr), n_clusters=9)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "e28a768a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_values([28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 59, 59, 59, 60, 60, 60, 60, 60, 60, 60, 60, 60, 61, 61, 61, 61, 61, 61, 61, 61, 61, 62, 62, 62, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 65, 65, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 70, 70, 70, 70, 70, 70, 71, 71, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 73, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 74, 78, 78, 78, 78, 78, 78, 78, 78, 79, 79, 79, 80, 80, 80, 80, 82, 82, 82, 82, 82, 90, 90, 90, 93, 94, 97, 99, 100, 100, 100, 101, 101, 101, 101, 101, 101, 101, 101, 101, 102, 102, 102, 103, 108, 108, 108, 108, 108, 108, 108, 108, 108, 108, 109, 109, 109, 109, 109, 109, 109, 109, 109, 109, 109, 109, 112, 112, 112, 112, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 114, 117, 117, 117, 117, 117, 117, 117, 117, 117, 117, 117, 117, 117, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 118, 121, 121, 122, 122, 122, 122, 122, 122, 122, 122, 122, 122, 122, 125, 125, 126, 126, 126, 126, 126, 126, 126, 126, 126, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 130, 130, 130, 130, 130, 130, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 138, 138, 138, 138, 138, 138, 138, 138, 138, 138, 138, 142, 142, 142, 142, 142, 142, 142, 145, 151, 151, 151, 151, 154, 154, 154, 154, 154, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 156, 157, 157, 161, 161, 161, 161, 161, 177, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 251, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 301, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 318, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 331, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 334, 335, 335, 336, 336, 336, 336, 336, 336, 336, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 344, 353, 353, 353, 353, 353, 353, 365, 365, 365, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 371, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 375, 384, 384, 384, 384, 384, 384, 384])\n"
     ]
    }
   ],
   "source": [
    "import sklearn.cluster\n",
    "import operator\n",
    "\n",
    "gene2fr = defaultdict(lambda: 0)\n",
    "\n",
    "for k in range(10, 11):\n",
    "    bh = sklearn.cluster.k_means(np.square(healthyCorr), n_clusters=k)[1]\n",
    "    bd = sklearn.cluster.k_means(np.square(diseasedCorr), n_clusters=k)[1]\n",
    "\n",
    "    t = 0\n",
    "    tc = 0\n",
    "    for i1,g1 in enumerate(anomPathwayGeneNames):\n",
    "        for i2,g2 in enumerate(anomPathwayGeneNames):\n",
    "            if i1<i2:\n",
    "                if (bh[i1]==bh[i2]) != (bd[i1]==bd[i2]):\n",
    "                    t += 1\n",
    "                    #print(t, g1, g2)\n",
    "                    gene2fr[g1] += 1\n",
    "                    gene2fr[g2] += 1\n",
    "                else:\n",
    "                    tc += 1\n",
    "sorted_dict = dict(sorted(gene2fr.items(), key=operator.itemgetter(1)))\n",
    "print(sorted_dict.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93581742",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "686f5ac6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbhcs, dbhlabels = dbscan(np.square(healthyCorr), metric=\"precomputed\", eps=0.8)\n",
    "len(set(dbhlabels))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "84e6a2f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[6.067391 , 8.8757504, 5.6641085, ..., 6.707134 , 5.450995 ,\n",
       "        5.850196 ],\n",
       "       [5.795969 , 9.0911432, 5.3878315, ..., 7.09882  , 5.382501 ,\n",
       "        5.92895  ],\n",
       "       [5.91032  , 8.6722996, 5.9592795, ..., 6.446724 , 5.665595 ,\n",
       "        6.42348  ],\n",
       "       ...,\n",
       "       [5.95303  , 8.1330288, 5.5837515, ..., 6.263437 , 5.730636 ,\n",
       "        6.453225 ],\n",
       "       [5.912833 , 8.9221486, 5.6921765, ..., 6.278591 , 5.583134 ,\n",
       "        6.19882  ],\n",
       "       [6.087491 , 8.7737798, 5.7816135, ..., 5.899442 , 5.98729  ,\n",
       "        6.231638 ]])"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genesHealthy[names].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "7a2c013a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MAP3K19',\n",
       " 'CCDC115',\n",
       " '649030',\n",
       " 'POP7',\n",
       " 'POLD3',\n",
       " 'SNTN',\n",
       " 'MICAL1',\n",
       " 'MATN3',\n",
       " '646804',\n",
       " 'PIGL',\n",
       " '729046',\n",
       " '400752',\n",
       " 'KCNG1',\n",
       " 'KLC2',\n",
       " 'NUDT16L1',\n",
       " 'METTL21C',\n",
       " '100128260',\n",
       " '440928',\n",
       " '730060',\n",
       " '653242',\n",
       " 'PRKAR2A',\n",
       " 'TACC3',\n",
       " 'ZNF98',\n",
       " '389033',\n",
       " 'ZNF699',\n",
       " 'ABHD3',\n",
       " 'TMEM231',\n",
       " '646769',\n",
       " '389286',\n",
       " '100128098',\n",
       " 'RDH11',\n",
       " 'RAB3B',\n",
       " 'PIEZO1',\n",
       " 'SCX',\n",
       " 'MYSM1',\n",
       " 'PARP12',\n",
       " 'ZUFSP',\n",
       " 'NAALADL1',\n",
       " '653111',\n",
       " 'SIRT1',\n",
       " 'HACD1',\n",
       " 'CSMD1',\n",
       " 'TANC2',\n",
       " 'PCM1',\n",
       " 'TARDBP',\n",
       " 'GEMIN8',\n",
       " 'SELE',\n",
       " 'MYF6',\n",
       " 'DCN',\n",
       " 'NOS1AP',\n",
       " '644105',\n",
       " '652483',\n",
       " 'MRE11A',\n",
       " '730358',\n",
       " 'DYDC2',\n",
       " 'TMEM161B',\n",
       " 'PEG10',\n",
       " 'SP140',\n",
       " 'ACTN1',\n",
       " 'AGAP3',\n",
       " 'VCX2',\n",
       " 'POLH',\n",
       " 'SIN3B',\n",
       " '284185',\n",
       " '645635',\n",
       " '100131726',\n",
       " 'EPAS1',\n",
       " 'SYT9',\n",
       " '653080',\n",
       " 'PADI1',\n",
       " 'AMBN',\n",
       " '727759',\n",
       " 'POPDC2',\n",
       " '647009',\n",
       " '650114',\n",
       " 'PARP14',\n",
       " 'SPX',\n",
       " '647389',\n",
       " 'ENPEP',\n",
       " 'TMEM190',\n",
       " 'STEAP1',\n",
       " 'NPR1',\n",
       " 'PPP2R3A',\n",
       " '594838',\n",
       " 'MED27',\n",
       " 'PPIE',\n",
       " '100129396',\n",
       " 'ACTR2',\n",
       " '8123',\n",
       " '645968',\n",
       " 'ADH1A',\n",
       " 'WIZ',\n",
       " '653401',\n",
       " 'RFXAP',\n",
       " '728476',\n",
       " 'ATF5',\n",
       " 'KIAA0895',\n",
       " 'ZNF79',\n",
       " 'FAM83F',\n",
       " 'CAV2',\n",
       " '100129063',\n",
       " 'CCT8',\n",
       " 'CLN3',\n",
       " 'COX11',\n",
       " 'ZFP64',\n",
       " '11213',\n",
       " 'TMEM55A',\n",
       " 'TFAP2E',\n",
       " 'HMGCLL1',\n",
       " 'C1orf106',\n",
       " 'POLR2J2',\n",
       " '100131960',\n",
       " 'AADACL4',\n",
       " 'BTC',\n",
       " 'NR1D1',\n",
       " '652966',\n",
       " 'CENPBD1',\n",
       " 'CDCA4',\n",
       " 'TMEM140',\n",
       " '441737',\n",
       " 'PDHB',\n",
       " 'SLC25A48',\n",
       " 'CNTNAP1',\n",
       " 'ZP3',\n",
       " 'MORF4L2',\n",
       " 'MIER1',\n",
       " '100129667',\n",
       " 'CTSK',\n",
       " 'TMEM74',\n",
       " 'UMOD',\n",
       " 'FASTKD3',\n",
       " '440313',\n",
       " 'MMGT1',\n",
       " 'ZNF250',\n",
       " 'RPL28',\n",
       " '1268',\n",
       " 'PKHD1L1',\n",
       " '651361',\n",
       " 'HLA-DRB1',\n",
       " 'PIK3C3',\n",
       " 'TMOD1',\n",
       " 'ZNF333',\n",
       " 'FOXG1',\n",
       " 'KIRREL',\n",
       " '80307',\n",
       " 'RCAN2',\n",
       " 'CXorf57',\n",
       " '6085',\n",
       " 'SOX15',\n",
       " 'CEP83',\n",
       " '727935',\n",
       " 'CLTA',\n",
       " 'COMMD7',\n",
       " '441120',\n",
       " 'OCM2',\n",
       " '645722',\n",
       " '646942',\n",
       " 'PRKRA',\n",
       " 'VHL',\n",
       " '442366',\n",
       " 'NOD2',\n",
       " '56664',\n",
       " '729433',\n",
       " '647995',\n",
       " 'NKX2-8',\n",
       " '441763',\n",
       " 'KLKB1',\n",
       " 'RNPEPL1',\n",
       " 'C6orf136',\n",
       " 'DCAF13',\n",
       " 'METTL1',\n",
       " 'KIFAP3',\n",
       " 'WIPF2',\n",
       " '652456',\n",
       " '654056',\n",
       " 'CDK8',\n",
       " 'FAM209A',\n",
       " '728790',\n",
       " '100130079',\n",
       " '440704',\n",
       " 'PPM1D',\n",
       " 'PLSCR3',\n",
       " 'TTR',\n",
       " 'TFIP11',\n",
       " 'LSR',\n",
       " '648659',\n",
       " '100129002',\n",
       " '643152',\n",
       " '644000',\n",
       " 'CA2',\n",
       " 'SPP1',\n",
       " 'HP',\n",
       " '728014',\n",
       " 'PEX11B',\n",
       " '114796',\n",
       " 'GTF3C5',\n",
       " 'GLB1L',\n",
       " 'PPFIBP1',\n",
       " 'ZWILCH',\n",
       " '400389',\n",
       " '388153',\n",
       " '653113',\n",
       " 'ST8SIA3',\n",
       " 'GNA14',\n",
       " 'PANX2',\n",
       " 'PTPN1',\n",
       " 'OR1J2',\n",
       " '644640',\n",
       " '100127922',\n",
       " 'HLA-DRA',\n",
       " 'DYNC2LI1',\n",
       " 'TEX2',\n",
       " 'CCDC58',\n",
       " 'MDP1',\n",
       " 'CCHCR1',\n",
       " '644280',\n",
       " 'STRN',\n",
       " 'RIC3',\n",
       " 'IL12RB2',\n",
       " '442442',\n",
       " '646057',\n",
       " '643797',\n",
       " 'SLC9A3R1',\n",
       " 'NT5M',\n",
       " '100134868',\n",
       " 'MTRF1',\n",
       " 'PITX2',\n",
       " 'SPACA9',\n",
       " '285444',\n",
       " 'SMARCA1',\n",
       " '642443',\n",
       " 'IFNG',\n",
       " 'AOX1',\n",
       " '574480',\n",
       " 'SPOPL',\n",
       " 'EDEM1',\n",
       " '54577',\n",
       " 'A4GNT',\n",
       " '100132763',\n",
       " '650339',\n",
       " 'GRXCR2',\n",
       " 'OAZ2',\n",
       " '692057',\n",
       " 'ZNF124',\n",
       " '728693',\n",
       " 'SIGLEC14',\n",
       " 'PDE4B',\n",
       " '641959',\n",
       " 'FAM102A',\n",
       " '642655',\n",
       " 'F10',\n",
       " 'POLR3H',\n",
       " 'HECTD3',\n",
       " 'PHLDA1',\n",
       " '388759',\n",
       " 'ABCC8',\n",
       " '100128304',\n",
       " 'GALNT15',\n",
       " 'ABCF1',\n",
       " 'COG3',\n",
       " 'GPHB5',\n",
       " 'MRPL11',\n",
       " 'KLHL34',\n",
       " 'PCDH1',\n",
       " '10778',\n",
       " '728368',\n",
       " 'ZNF160',\n",
       " 'OR51T1',\n",
       " 'THY1',\n",
       " 'JUNB',\n",
       " '653994',\n",
       " 'MAN2C1',\n",
       " 'IFI44L',\n",
       " 'CTAGE5',\n",
       " 'RAB11FIP4',\n",
       " 'RAD17',\n",
       " 'NAT10',\n",
       " 'PRNP',\n",
       " 'AGPAT1',\n",
       " 'TMEM131',\n",
       " 'FTH1',\n",
       " 'RTN4RL2',\n",
       " 'TMPRSS13',\n",
       " 'KCTD6',\n",
       " 'UTP14C',\n",
       " '389599',\n",
       " 'ZNF385D',\n",
       " 'SOCS2',\n",
       " '2681',\n",
       " 'KIAA0930',\n",
       " 'PLCH1',\n",
       " 'TLCD2',\n",
       " 'ATAD2',\n",
       " '26783',\n",
       " 'TMEM19',\n",
       " 'BCAS2',\n",
       " 'MIF4GD',\n",
       " '646515',\n",
       " 'FBXO32',\n",
       " 'TCF15',\n",
       " 'OPRD1',\n",
       " 'CHST12',\n",
       " 'SEC23A',\n",
       " 'INO80B',\n",
       " '646774',\n",
       " 'TMCO3',\n",
       " 'FBN2',\n",
       " 'SDC1',\n",
       " 'ATRNL1',\n",
       " '647210',\n",
       " 'KCNH4',\n",
       " '643313',\n",
       " 'TUBA8',\n",
       " 'RPSA',\n",
       " 'VAPA',\n",
       " 'F5',\n",
       " 'JUN',\n",
       " 'UBE2E2',\n",
       " 'ZNF207',\n",
       " 'NAPRT',\n",
       " 'TP53INP2',\n",
       " '641844',\n",
       " 'PLXNB1',\n",
       " '100129878',\n",
       " 'DOCK8',\n",
       " 'SPTLC3',\n",
       " '400145',\n",
       " '100192420',\n",
       " 'CD70',\n",
       " 'RGL4',\n",
       " '649752',\n",
       " 'CHORDC1',\n",
       " '100129566',\n",
       " 'WNT10B',\n",
       " '677807',\n",
       " 'FAM192A',\n",
       " '283970',\n",
       " 'NRIP1',\n",
       " '728831',\n",
       " 'FBP2',\n",
       " 'RSRP1',\n",
       " 'MSS51',\n",
       " 'CPXCR1',\n",
       " '649702',\n",
       " 'B2M',\n",
       " 'ZNF317',\n",
       " '648342',\n",
       " 'EGFR',\n",
       " 'PDE9A',\n",
       " '729269',\n",
       " 'GPR42',\n",
       " 'TMEM233',\n",
       " '648130',\n",
       " 'KCTD7',\n",
       " '399959',\n",
       " 'HOXA13',\n",
       " 'AMMECR1L',\n",
       " '645460',\n",
       " '644224',\n",
       " '440482',\n",
       " 'NLE1',\n",
       " 'TBC1D22A',\n",
       " 'RPS18',\n",
       " '652743',\n",
       " '652226',\n",
       " 'C10orf11',\n",
       " 'C1orf50',\n",
       " 'KEAP1',\n",
       " '647797',\n",
       " 'ABCD1',\n",
       " 'TMEM67',\n",
       " 'C9orf66',\n",
       " 'IGFN1',\n",
       " 'PRRG1',\n",
       " 'ACYP1',\n",
       " 'CHMP4A',\n",
       " 'TYRP1',\n",
       " 'USP1',\n",
       " 'RPL32',\n",
       " 'CCL28',\n",
       " 'CTSV',\n",
       " 'TSSC4',\n",
       " '100132119',\n",
       " 'CNGA3',\n",
       " 'PZP',\n",
       " 'MAP2K5',\n",
       " 'PDE3B',\n",
       " 'KCNK18',\n",
       " 'CTNNB1',\n",
       " 'SOD1',\n",
       " 'CADPS',\n",
       " 'MYOF',\n",
       " 'NGLY1',\n",
       " '651524',\n",
       " 'IMPACT',\n",
       " 'TMEM45A',\n",
       " '652479',\n",
       " 'RALGPS1',\n",
       " 'HCN3',\n",
       " '727773',\n",
       " 'MRPL16',\n",
       " 'KYAT1',\n",
       " 'ITGA7',\n",
       " 'HNMT',\n",
       " 'POLRMT',\n",
       " '339760',\n",
       " '79804',\n",
       " 'MZT2A',\n",
       " 'INO80E',\n",
       " '728673',\n",
       " '730077',\n",
       " 'NFRKB',\n",
       " 'HERC4',\n",
       " 'SAMD9',\n",
       " 'ZNF442',\n",
       " 'UBR5',\n",
       " '652624',\n",
       " 'GDF15',\n",
       " 'PALMD',\n",
       " '79970',\n",
       " '730357',\n",
       " '100128016',\n",
       " '100131138',\n",
       " 'USH2A',\n",
       " 'EMC9',\n",
       " 'ANKRD27',\n",
       " 'PRICKLE2',\n",
       " 'AZIN2',\n",
       " 'SHISA3',\n",
       " 'PTPN11',\n",
       " 'FAM217B',\n",
       " 'TTL',\n",
       " 'GMCL1',\n",
       " 'CIAO1',\n",
       " 'NKAP',\n",
       " 'TRAPPC13',\n",
       " 'SCN4A',\n",
       " '100129977',\n",
       " '100132973',\n",
       " 'RPS20',\n",
       " 'GGA3',\n",
       " 'MED17',\n",
       " '100132485',\n",
       " '646372',\n",
       " 'MAEA',\n",
       " 'MRPL12',\n",
       " 'ZDHHC11',\n",
       " 'SLC38A6',\n",
       " 'MED4',\n",
       " 'CDC26',\n",
       " 'GALT',\n",
       " 'SSX3',\n",
       " 'DNAJB1',\n",
       " 'ANKRD13D',\n",
       " 'HNRNPU',\n",
       " 'HIC1',\n",
       " 'RBL2',\n",
       " '388654',\n",
       " '2576',\n",
       " 'RHOU',\n",
       " '401375',\n",
       " 'CEBPA',\n",
       " 'AKAP9',\n",
       " 'ADNP2',\n",
       " 'ICOS',\n",
       " 'RHOT1',\n",
       " '285941',\n",
       " '152667',\n",
       " '650517',\n",
       " 'FNBP4',\n",
       " 'POR',\n",
       " 'TMEM138',\n",
       " '729954',\n",
       " 'PIP4K2C',\n",
       " 'HSFX2',\n",
       " 'AKR1B1',\n",
       " 'LYL1',\n",
       " '647585',\n",
       " '100133854',\n",
       " 'LPAR3',\n",
       " 'MYC',\n",
       " '645146',\n",
       " '85009',\n",
       " 'BMS1',\n",
       " '388182',\n",
       " 'DENND1A',\n",
       " '649813',\n",
       " 'TMEM39B',\n",
       " 'CYP4F3',\n",
       " 'ASB15',\n",
       " 'PLEKHG3',\n",
       " '729960',\n",
       " '642456',\n",
       " 'MNDA',\n",
       " 'C1orf158',\n",
       " 'TAGLN',\n",
       " '729208',\n",
       " '100129484',\n",
       " 'PRELID3A',\n",
       " '100132319',\n",
       " 'FLOT1',\n",
       " 'OR5F1',\n",
       " '400214',\n",
       " '400955',\n",
       " '574442',\n",
       " 'DPAGT1',\n",
       " 'PID1',\n",
       " 'VPS53',\n",
       " 'KDM4B',\n",
       " 'DNAJC4',\n",
       " 'FGFR1OP',\n",
       " 'KIF26B',\n",
       " '441520',\n",
       " '1292',\n",
       " 'F13A1',\n",
       " 'DPP7',\n",
       " '80154',\n",
       " 'VTA1',\n",
       " 'NAV2',\n",
       " '441007',\n",
       " '645730',\n",
       " '642097',\n",
       " 'FAM47C',\n",
       " '645662',\n",
       " 'ST6GALNAC3',\n",
       " 'ENKUR',\n",
       " 'SP9',\n",
       " '730183',\n",
       " 'GRPEL1',\n",
       " 'CPA5',\n",
       " 'SMAD6',\n",
       " 'CGRRF1',\n",
       " 'JAKMIP1',\n",
       " 'CD3D',\n",
       " 'PRKCD',\n",
       " 'MPZL1',\n",
       " '100132499',\n",
       " '401011',\n",
       " 'ARRDC5',\n",
       " '100131335',\n",
       " 'YIPF3',\n",
       " 'MAPK8IP3',\n",
       " '649422',\n",
       " '401002',\n",
       " 'MBLAC2',\n",
       " 'YEATS4',\n",
       " 'MFSD9',\n",
       " '664618',\n",
       " 'PRICKLE4',\n",
       " '319139',\n",
       " '644590',\n",
       " 'HRC',\n",
       " '648638',\n",
       " 'KLHL41',\n",
       " 'SLC35A3',\n",
       " '650098',\n",
       " 'RAB3A',\n",
       " 'RASSF6',\n",
       " 'FBXW5',\n",
       " 'SGPL1',\n",
       " 'SLC10A3',\n",
       " 'MFSD1',\n",
       " 'PSMA2',\n",
       " 'DEFB104B',\n",
       " '652534',\n",
       " 'SLC33A1',\n",
       " 'GUCA1B',\n",
       " 'KNTC1',\n",
       " '51275',\n",
       " 'C2orf42',\n",
       " 'ZMYM4',\n",
       " '286094',\n",
       " '100131372',\n",
       " '5786',\n",
       " '728779',\n",
       " '647987',\n",
       " 'IL10RB',\n",
       " 'ASB10',\n",
       " 'ATP5SL',\n",
       " 'GPR137B',\n",
       " 'PLEKHG5',\n",
       " '441212',\n",
       " '220115',\n",
       " '51306',\n",
       " 'TUBB',\n",
       " 'SMIM4',\n",
       " 'TTLL7',\n",
       " 'NEK6',\n",
       " 'SMC4',\n",
       " 'POLR3E',\n",
       " '653147',\n",
       " '2282',\n",
       " 'WNT6',\n",
       " 'KREMEN2',\n",
       " 'TJP3',\n",
       " 'AKR1C3',\n",
       " '388438',\n",
       " 'MRPL23',\n",
       " 'TRMT61B',\n",
       " 'SERPINB6',\n",
       " 'UTP4',\n",
       " 'GGCT',\n",
       " 'KIT',\n",
       " '100134098',\n",
       " 'CD8A',\n",
       " 'ELP2',\n",
       " 'RAD51B',\n",
       " 'KIAA0586',\n",
       " 'LUZP2',\n",
       " '100128562',\n",
       " 'LARS',\n",
       " 'EXOC8',\n",
       " '100133585',\n",
       " 'KCNH8',\n",
       " '100134584',\n",
       " '100129637',\n",
       " 'TRMT1',\n",
       " 'DPY19L1',\n",
       " '653450',\n",
       " 'RNF146',\n",
       " 'JOSD1',\n",
       " '84809',\n",
       " '100130812',\n",
       " 'SUZ12',\n",
       " '152302',\n",
       " 'AGR3',\n",
       " '729668',\n",
       " 'PSMF1',\n",
       " 'SLC35B2',\n",
       " 'HAMP',\n",
       " 'TCAF1',\n",
       " 'RPS15',\n",
       " 'ZNF511',\n",
       " 'NXN',\n",
       " 'ARL17A',\n",
       " 'NLRX1',\n",
       " 'NUTM2F',\n",
       " 'TRAF6',\n",
       " '642381',\n",
       " 'RPL7',\n",
       " 'DMRTC1B',\n",
       " 'DENND6A',\n",
       " 'RHCE',\n",
       " 'FAM98B',\n",
       " '400446',\n",
       " 'ERP44',\n",
       " '729143',\n",
       " 'PBX4',\n",
       " 'SCNM1',\n",
       " '730284',\n",
       " 'C18orf21',\n",
       " 'CD93',\n",
       " 'GPAM',\n",
       " 'CD101',\n",
       " 'P2RX3',\n",
       " 'CPPED1',\n",
       " '100132205',\n",
       " '653506',\n",
       " 'BCL2',\n",
       " 'CLIC1',\n",
       " '388289',\n",
       " '26172',\n",
       " '648622',\n",
       " 'ZNF77',\n",
       " 'ZNF718',\n",
       " '652408',\n",
       " '401224',\n",
       " 'PYGL',\n",
       " '728263',\n",
       " 'FGD2',\n",
       " 'SGSM2',\n",
       " 'PDGFD',\n",
       " '121906',\n",
       " 'SNRPN',\n",
       " 'DOHH',\n",
       " '89837',\n",
       " 'MGAT2',\n",
       " '151457',\n",
       " 'KLRC3',\n",
       " 'ORC1',\n",
       " '650921',\n",
       " 'FAM134A',\n",
       " 'BCAP31',\n",
       " '401242',\n",
       " '440013',\n",
       " 'CLPS',\n",
       " 'TSKU',\n",
       " 'PDSS2',\n",
       " '2497',\n",
       " '652606',\n",
       " 'TIMM8A',\n",
       " '645636',\n",
       " '731186',\n",
       " 'HIC2',\n",
       " 'NKX1-1',\n",
       " '647086',\n",
       " 'HMGN5',\n",
       " 'RERE',\n",
       " 'KDM2B',\n",
       " 'CD5',\n",
       " 'SPON1',\n",
       " 'KDM1A',\n",
       " 'CCL23',\n",
       " '100132717',\n",
       " 'ORC4',\n",
       " 'MAGEA9B',\n",
       " '441124',\n",
       " 'ACSF3',\n",
       " '390226',\n",
       " 'NDUFA6',\n",
       " 'PRR15L',\n",
       " '441013',\n",
       " 'STPG2',\n",
       " 'SLX4',\n",
       " '100190939',\n",
       " 'NUDT21',\n",
       " 'POLR2D',\n",
       " 'CEACAM3',\n",
       " 'C1GALT1C1',\n",
       " 'RBSN',\n",
       " '139046',\n",
       " '644171',\n",
       " 'DUSP14',\n",
       " 'ECHDC3',\n",
       " 'THADA',\n",
       " 'METTL21B',\n",
       " 'ATP23',\n",
       " 'CENPT',\n",
       " 'DIABLO',\n",
       " '401288',\n",
       " '440366',\n",
       " '100132354',\n",
       " 'F8A3',\n",
       " 'FAF1',\n",
       " 'WAC',\n",
       " '651450',\n",
       " 'GLI1',\n",
       " 'CMSS1',\n",
       " '646509',\n",
       " 'OR52E6',\n",
       " '649618',\n",
       " 'IQCK',\n",
       " 'NCKIPSD',\n",
       " '644929',\n",
       " 'CRTAP',\n",
       " 'MFSD7',\n",
       " '651952',\n",
       " 'RTCA',\n",
       " 'BLCAP',\n",
       " 'LRMP',\n",
       " 'PREPL',\n",
       " 'TESPA1',\n",
       " 'MTIF3',\n",
       " 'FZD10',\n",
       " '100129424',\n",
       " '441843',\n",
       " 'TOLLIP',\n",
       " 'TFPI2',\n",
       " 'MFAP3L',\n",
       " 'USH1G',\n",
       " 'BNIP1',\n",
       " 'CLEC4G',\n",
       " 'BAX',\n",
       " 'CXCR5',\n",
       " 'ZNF860',\n",
       " 'TXNRD3NB',\n",
       " 'FIS1',\n",
       " 'GNG8',\n",
       " 'EVI2A',\n",
       " '651398',\n",
       " 'ZKSCAN3',\n",
       " '729751',\n",
       " 'GSG1',\n",
       " 'SAMD4A',\n",
       " 'NEK4',\n",
       " 'MSI1',\n",
       " '644044',\n",
       " 'C8orf86',\n",
       " '84054',\n",
       " '642398',\n",
       " 'CAPRIN2',\n",
       " '260341',\n",
       " 'C5orf51',\n",
       " '100124537',\n",
       " 'SMUG1',\n",
       " 'MFSD11',\n",
       " '55147',\n",
       " 'ZDHHC8',\n",
       " 'ZNF556',\n",
       " 'MRPS18B',\n",
       " 'SRCAP',\n",
       " 'GPR3',\n",
       " 'IFI30',\n",
       " 'USP35',\n",
       " 'FHL2',\n",
       " 'SIRPD',\n",
       " 'ANKRD9',\n",
       " 'RRM1',\n",
       " 'ZNF781',\n",
       " 'TTC17',\n",
       " 'TYK2',\n",
       " '283521',\n",
       " '388339',\n",
       " '644076',\n",
       " '389105',\n",
       " 'FXR2',\n",
       " 'PRB1',\n",
       " 'RALY',\n",
       " 'CNIH1',\n",
       " 'CHMP7',\n",
       " 'C8orf58',\n",
       " 'SLC39A9',\n",
       " '100132091',\n",
       " 'LUC7L2',\n",
       " '8193',\n",
       " 'RPL10L',\n",
       " '650463',\n",
       " 'RGMA',\n",
       " 'SIRT3',\n",
       " '649248',\n",
       " '441958',\n",
       " 'ELF3',\n",
       " 'C2CD4C',\n",
       " '727987',\n",
       " 'HEXA',\n",
       " '55370',\n",
       " 'TRUB1',\n",
       " 'EMX2',\n",
       " 'RPL35',\n",
       " 'CLDND2',\n",
       " 'REEP4',\n",
       " 'PPP1R14C',\n",
       " 'PEX2',\n",
       " 'TM2D3',\n",
       " 'WDR91',\n",
       " 'ALCAM',\n",
       " 'HLTF',\n",
       " 'PBLD',\n",
       " '650930',\n",
       " 'CBX7',\n",
       " 'FUNDC2',\n",
       " 'ASH2L',\n",
       " 'FAM13C',\n",
       " 'PCDH18',\n",
       " 'SCRN1',\n",
       " 'NCOA1',\n",
       " '645904',\n",
       " 'ANXA11',\n",
       " '645848',\n",
       " 'MAP1LC3C',\n",
       " 'NIT1',\n",
       " '341965',\n",
       " '100131094',\n",
       " 'TSPAN13',\n",
       " '100132632',\n",
       " 'TRIM21',\n",
       " 'NCOR2',\n",
       " 'PTGFRN',\n",
       " 'C16orf92',\n",
       " 'CASZ1',\n",
       " 'RGS6',\n",
       " 'RNF145',\n",
       " 'ZC3HAV1',\n",
       " 'POLR3D',\n",
       " 'TNNI1',\n",
       " 'PLPP2',\n",
       " 'RPS29',\n",
       " '645058',\n",
       " 'C1S',\n",
       " '400946',\n",
       " 'MYOC',\n",
       " 'CABP2',\n",
       " 'SPATA3',\n",
       " '55025',\n",
       " '442572',\n",
       " 'DNAJC5',\n",
       " '645153',\n",
       " 'BYSL',\n",
       " '648600',\n",
       " 'ENY2',\n",
       " 'METTL6',\n",
       " 'NIPAL3',\n",
       " 'ZNF408',\n",
       " '729950',\n",
       " 'NCCRP1',\n",
       " 'BCL9',\n",
       " 'SKIV2L',\n",
       " 'TPMT',\n",
       " 'OTUB1',\n",
       " 'RGS18',\n",
       " 'FTSJ1',\n",
       " '100129191',\n",
       " 'RPLP0',\n",
       " 'GREM2',\n",
       " 'PDS5A',\n",
       " '646626',\n",
       " '100127899',\n",
       " '90925',\n",
       " 'GLO1',\n",
       " 'STK11',\n",
       " 'ZNF296',\n",
       " 'MNAT1',\n",
       " '388007',\n",
       " 'YAE1D1',\n",
       " 'PDLIM3',\n",
       " 'GCLM',\n",
       " 'C19orf52',\n",
       " '163720',\n",
       " 'EDEM2',\n",
       " '100190986',\n",
       " '401537',\n",
       " 'PRSS57',\n",
       " 'TRIB2',\n",
       " 'RPS27A',\n",
       " 'GSR',\n",
       " 'HMGA2',\n",
       " 'FGFBP2',\n",
       " 'CCDC69',\n",
       " 'CCDC93',\n",
       " '730415',\n",
       " 'MBD3',\n",
       " 'BPNT1',\n",
       " '646836',\n",
       " '730092',\n",
       " 'MED11',\n",
       " 'RPS3A',\n",
       " 'SLC16A7',\n",
       " 'TRMT61A',\n",
       " '440030',\n",
       " 'PRR14L',\n",
       " 'MAP3K7CL',\n",
       " 'RGS2',\n",
       " 'CDK1',\n",
       " 'ANK2',\n",
       " 'CSRNP3',\n",
       " '728701',\n",
       " 'HSD17B14',\n",
       " 'TNIP1',\n",
       " '649076',\n",
       " 'ZNF775',\n",
       " '121952',\n",
       " 'MATK',\n",
       " 'LMO7',\n",
       " '340206',\n",
       " 'PLCL1',\n",
       " 'SDCCAG3',\n",
       " 'PPT2',\n",
       " '85287',\n",
       " 'ADGRE1',\n",
       " 'PLCB3',\n",
       " 'ZNF22',\n",
       " 'NSFL1C',\n",
       " 'CENPH',\n",
       " 'RNF214',\n",
       " 'IL18R1',\n",
       " '283547',\n",
       " '554203',\n",
       " 'OGG1',\n",
       " '652427',\n",
       " 'HIBCH',\n",
       " 'PHKA1',\n",
       " 'MTFMT',\n",
       " '652900',\n",
       " 'ARG2',\n",
       " '642678',\n",
       " 'LMNB1',\n",
       " '51214',\n",
       " 'ESAM',\n",
       " 'OR10AD1',\n",
       " '646103',\n",
       " '147172',\n",
       " 'C1orf116',\n",
       " 'CEP44',\n",
       " '390735',\n",
       " '729677',\n",
       " '727735',\n",
       " 'PIGV',\n",
       " 'GTPBP8',\n",
       " '100129203',\n",
       " 'OR1N1',\n",
       " 'UNC13A',\n",
       " 'LSM2',\n",
       " '402066',\n",
       " 'BRIP1',\n",
       " '90024',\n",
       " '692099',\n",
       " '100302280',\n",
       " 'SH2D1A',\n",
       " 'COLGALT2',\n",
       " 'SLC25A44',\n",
       " 'CIB2',\n",
       " 'NT5C',\n",
       " 'TRAPPC9',\n",
       " 'METTL8',\n",
       " '402251',\n",
       " '644345',\n",
       " '728127',\n",
       " 'AGAP6',\n",
       " 'RRP15',\n",
       " '645648',\n",
       " ...}"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36fbda30",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qf",
   "language": "python",
   "name": "qf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
