{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d4364c55",
   "metadata": {},
   "outputs": [],
   "source": [
    "from hypergraphs import *\n",
    "from kegg import *\n",
    "import pandas as pd\n",
    "import os\n",
    "import logging\n",
    "\n",
    "from Bio.KEGG.KGML.KGML_parser import read\n",
    "import Bio.KEGG.KGML.KGML_pathway\n",
    "from ast import literal_eval\n",
    "from urllib.request import urlopen\n",
    "from urllib.parse import quote_plus\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "from Levenshtein import ratio\n",
    "import scipy.stats\n",
    "import pathlib\n",
    "from qf.cc import cardon_crochemore_colored, cardon_crochemore\n",
    "import random\n",
    "\n",
    "from reportlab.lib.units import inch\n",
    "import PIL.Image\n",
    "from io import BytesIO\n",
    "import reportlab.pdfgen.canvas\n",
    "import reportlab.lib.colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6b13e6bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "organismDict = {\"hsa\": \"Homo sapiens (human)\"}\n",
    "outputDirectory = \"/Users/boldi/Desktop/pw/\"\n",
    "inputDirectory = \"../../LaTeX/Data/KEGG-Pathways/Mar2024/\"\n",
    "dataDirectory = \"../../LaTeX/Data/KEGG-Pathways\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "87cb1b1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathwaylistFilename = \"KEGG-pathways.txt\"  # In inputDirectory, csv: ID, type, name\n",
    "superpathway = \"All\"                  # Name of the superpathway\n",
    "organism = \"hsa\"\n",
    "confidence = 0.05\n",
    "\n",
    "produceGraphsWithCCcoloring = False\n",
    "produceGraphsWithAnomalies = True\n",
    "useCCcoloringForAnomalies = False\n",
    "produceAnomaliesForNonsignificantPairs = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cb8e2d96",
   "metadata": {},
   "outputs": [],
   "source": [
    "safeSuperpathway = makesafe(superpathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e6caae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pws = pd.read_csv(os.path.join(inputDirectory, \"KEGG-pathways.txt\"), \n",
    "            names=[\"Pathway ID\", \"Pathway type\", \"Pathway name\"], dtype=\"string\", keep_default_na=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "199b2487",
   "metadata": {},
   "outputs": [],
   "source": [
    "superpathwayDict = {\n",
    "    superpathway: [(x,y) for x,y in zip(pws[\"Pathway ID\"], pws[\"Pathway name\"])]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "47a38c39",
   "metadata": {},
   "outputs": [],
   "source": [
    "def words2set(words):\n",
    "    \"\"\"\n",
    "        Given a space-separated list of words, returns it as a set of words\n",
    "    \"\"\"\n",
    "    return set(words.split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e0a74d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import NamedTuple\n",
    "\n",
    "class Relation(NamedTuple):\n",
    "    \"\"\"\n",
    "        Represents a relation. It is characterized by:\n",
    "        - a type (e.g. \"PPRel\") \n",
    "        - a list of subtypes (e.g. [('activation', '-->'), ('indirect effect', '..>')])\n",
    "        - a set of components (source)\n",
    "        - a list of sets of components (target)\n",
    "    \"\"\"\n",
    "    relType: str\n",
    "    relSubtypes: list\n",
    "    source: set\n",
    "    target: list\n",
    "    sourceg: Bio.KEGG.KGML.KGML_pathway.Graphics\n",
    "    targetg: Bio.KEGG.KGML.KGML_pathway.Graphics\n",
    "        \n",
    "    @classmethod\n",
    "    def fromKEGG(cls, relation, pathway):\n",
    "        \"\"\"\n",
    "            Constructor: builds a Relation from a KEGG relation\n",
    "        \"\"\"\n",
    "        if relation.entry2.name == \"undefined\":\n",
    "            s = []\n",
    "            for component in relation.entry2.components:\n",
    "                s += [words2set(pathway.entries[component.id].name)]\n",
    "            targetg = pathway.entries[component.id].graphics[0]\n",
    "        else:\n",
    "            s = [words2set(relation.entry2.name)]\n",
    "            targetg = relation.entry2.graphics[0]\n",
    "        return cls(relation.type, relation.subtypes, words2set(relation.entry1.name), s, targetg, relation.entry1.graphics[0])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ba5d2e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsFromKEGGpathway(pathway):\n",
    "    \"\"\"\n",
    "        Given a KEGG pathway, returns the list of relations it contains (represented as Relation).\n",
    "    \"\"\"\n",
    "    return [Relation.fromKEGG(r, pathway) for r in pathway.relations]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "751bc16c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pathwayFromFile(dataDirectory, organism, pathwayID):\n",
    "    \"\"\"\n",
    "        Load and return a KEGG pathway, given the data directory, the organism specifying the subdirectory,\n",
    "        and the pathwayID (the filename should be organism+pathwayID+\"xml\".\n",
    "    \"\"\"\n",
    "    return KGML_parser.read(open(os.path.join(dataDirectory, organism, organism+pathwayID+\".xml\"), \"r\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bd27fdc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsFromFile(dataDirectory, organism, pathwayID):\n",
    "    \"\"\"\n",
    "        Given a pathway file (specified by a root dataDirectory, the name of the organism subdirectory, and the name\n",
    "        of the file [organism+pathwayID.xml]), reads it and returns the list of its relations.\n",
    "    \"\"\"\n",
    "    pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "    return relationsFromKEGGpathway(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eaa01766",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsForSuperpathway(dataDirectory, organism, superpathwayDict, superpathway):\n",
    "    \"\"\"\n",
    "        Given a superpathDict (whose keys are superpathway names and whose values are list of\n",
    "        pairs (pathwayID, pathwayName)), reads all the pathway file for a specific superpathway name\n",
    "        and returns a dictionary whose keys are the pathwayID's and whose values are the list of relations.\n",
    "    \"\"\"\n",
    "    rel = {}\n",
    "    for k,v in superpathwayDict[superpathway]:\n",
    "        #logging.info(\"Reading\", k, v)\n",
    "        rel[k] = relationsFromFile(dataDirectory, organism, k)\n",
    "    return rel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "15479a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def subtypes(rel):\n",
    "    \"\"\"\n",
    "        Given a map whose values are relations, accumulate all subtypes appearing and assign them a number.\n",
    "        The result is a map from subtype string to number.\n",
    "    \"\"\"\n",
    "    all_subtypes = set([])\n",
    "    for pathwayid, relations in rel.items():\n",
    "        for relation in relations:\n",
    "            all_subtypes |= set([str(relation.relSubtypes)])\n",
    "        \n",
    "    s = sorted(list(all_subtypes))\n",
    "    subtype2color = {v:k for k,v in enumerate(s)}\n",
    "    color2subtype = {k:v for k,v in enumerate(s)}\n",
    "\n",
    "    return subtype2color, color2subtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1e229eaf-d667-466c-8238-8205257b0ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set2can(s):\n",
    "    \"\"\"\n",
    "        Convert a set to a string in a canonical way.\n",
    "    \"\"\"\n",
    "    return str(sorted(list(s)))\n",
    "\n",
    "def can2set(c):\n",
    "    \"\"\"\n",
    "        Does the converse of set2can.\n",
    "    \"\"\"\n",
    "    return set(literal_eval(c))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75b272b9-b220-492b-adf6-509f9952842f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indices(rel):\n",
    "    \"\"\"\n",
    "        Given a map whose values are lists of relations, it considers all the relations one by one, and attributes a unique id \n",
    "        to each element (i.e., set of components appearing as source or in the target of some relation) and block (the set of\n",
    "        element id appearing as target of some relation).\n",
    "        This function returns the dictionaries to move from/to an element or block to the corresponding id.\n",
    "        Elements are string representations of sorted lists of strings.\n",
    "        Blocks are string representations of sorted lists of ints.\n",
    "    \"\"\"\n",
    "    element2id = {}\n",
    "    id2element = {}\n",
    "    block2id = {}\n",
    "    id2block = {}\n",
    "\n",
    "    for relations in rel.values():\n",
    "        for relation in relations:\n",
    "            s = set2can(relation.source)\n",
    "            if s not in element2id.keys():\n",
    "                element2id[s] = len(element2id)\n",
    "            targetset = set([])\n",
    "            for targ in relation.target:\n",
    "                t = set2can(targ)\n",
    "                if t not in element2id.keys():\n",
    "                    element2id[t] = len(element2id)\n",
    "                targetset |= set([element2id[t]])\n",
    "            ts = set2can(targetset)\n",
    "            if ts not in block2id.keys():\n",
    "                block2id[ts] = len(block2id)\n",
    "\n",
    "    for k,v in element2id.items():\n",
    "        id2element[v] = k\n",
    "    for k,v in block2id.items():\n",
    "        id2block[v] = k\n",
    "    return element2id, id2element, block2id, id2block\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f9c3f128-cf97-4bb3-8ce5-65c49f865d98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_gene_KEGG(geneID):\n",
    "    \"\"\"\n",
    "        Search for gene on KEGG. \n",
    "        \n",
    "        Returns list of names.\n",
    "    \"\"\"\n",
    "    url1 = \"https://www.kegg.jp/entry/\"+geneID\n",
    "    response1 = requests.get(url1, allow_redirects=False)    \n",
    "    soup = BeautifulSoup(response1.text, \"html.parser\")\n",
    "    tds = [tds for tds in soup.find_all(\"td\", {\"class\": \"td11 defd\"})]\n",
    "    names = tds[0].getText().strip().split(\", \")\n",
    "    return names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1631e1d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_temp_imagefilename(url):\n",
    "    \"\"\"Return filename of temporary file containing downloaded image.\n",
    "\n",
    "    Create a new temporary file to hold the image file at the passed URL\n",
    "    and return the filename.\n",
    "    \"\"\"\n",
    "    img = urlopen(url).read()\n",
    "    im = PIL.Image.open(BytesIO(img))\n",
    "    # im.transpose(Image.FLIP_TOP_BOTTOM)\n",
    "    f = tempfile.NamedTemporaryFile(delete=False, suffix=\".png\")\n",
    "    fname = f.name\n",
    "    f.close()\n",
    "    im.save(fname, \"PNG\")\n",
    "    return fname\n",
    "\n",
    "\n",
    "#def enhance_method(klass, method_name, replacement):\n",
    "#    'replace a method with an enhanced version'\n",
    "#    method = getattr(klass, method_name)\n",
    "#    def enhanced(*args, **kwds): return replacement(*args, **kwds)\n",
    "#    setattr(klass, method_name, enhanced)\n",
    "\n",
    "def new_draw(self, filename, relations, elaborateArc = None):\n",
    "        \"\"\"Add the map elements to the drawing.\"\"\"\n",
    "        # Instantiate the drawing, first\n",
    "        # size x_max, y_max for now - we can add margins, later\n",
    "        if self.import_imagemap:\n",
    "            # We're drawing directly on the image, so we set the canvas to the\n",
    "            # same size as the image\n",
    "            if os.path.isfile(self.pathway.image):\n",
    "                imfilename = self.pathway.image\n",
    "            else:\n",
    "                imfilename = get_temp_imagefilename(self.pathway.image)\n",
    "            im = PIL.Image.open(imfilename)\n",
    "            cwidth, cheight = im.size\n",
    "        else:\n",
    "            # No image, so we set the canvas size to accommodate visible\n",
    "            # elements\n",
    "            cwidth, cheight = (self.pathway.bounds[1][0], self.pathway.bounds[1][1])\n",
    "        # Instantiate canvas\n",
    "        self.drawing = reportlab.pdfgen.canvas.Canvas(\n",
    "            filename,\n",
    "            bottomup=0,\n",
    "            pagesize=(\n",
    "                cwidth * (1 + 2 * self.margins[0]),\n",
    "                cheight * (1 + 2 * self.margins[1]),\n",
    "            ),\n",
    "        )\n",
    "        self.drawing.setFont(self.fontname, self.fontsize)\n",
    "        # Transform the canvas to add the margins\n",
    "        self.drawing.translate(\n",
    "            self.margins[0] * self.pathway.bounds[1][0],\n",
    "            self.margins[1] * self.pathway.bounds[1][1],\n",
    "        )\n",
    "        # Add the map image, if required\n",
    "        if self.import_imagemap:\n",
    "            self.drawing.saveState()\n",
    "            self.drawing.scale(1, -1)\n",
    "            self.drawing.translate(0, -cheight)\n",
    "            self.drawing.drawImage(imfilename, 0, 0)\n",
    "            self.drawing.restoreState()\n",
    "        # Add the reactions, compounds and maps\n",
    "        # Maps go on first, to be overlaid by more information.\n",
    "        # By default, they're slightly transparent.\n",
    "        if self.show_maps:\n",
    "            self.__add_maps()\n",
    "        if self.show_reaction_entries:\n",
    "            KGMLCanvas._KGMLCanvas__add_reaction_entries(self)\n",
    "        if self.show_orthologs:\n",
    "            KGMLCanvas._KGMLCanvas__add_orthologs(self)\n",
    "        if self.show_compounds:\n",
    "            KGMLCanvas._KGMLCanvas__add_compounds(self)\n",
    "        if self.show_genes:\n",
    "            KGMLCanvas._KGMLCanvas__add_genes(self)\n",
    "        # TODO: complete draw_relations code\n",
    "        # if self.draw_relations:\n",
    "        #    self.__add_relations()\n",
    "        # Write the pathway map to PDF\n",
    "        if elaborateArc is not None:\n",
    "            for relation in relations:\n",
    "                elaborateArc(relation, self.drawing)\n",
    "        self.drawing.save()\n",
    "        \n",
    "#enhance_method(KGMLCanvas, 'draw',  new_draw)\n",
    "setattr(KGMLCanvas, 'new_draw', new_draw)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0a11bbae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build all data needed for the relations of the superpathway\n",
    "rel = relationsForSuperpathway(dataDirectory, organism, superpathwayDict, superpathway)\n",
    "element2id, id2element, block2id, id2block = indices(rel)\n",
    "subtype2color, color2subtype = subtypes(rel)\n",
    "elements = list(set.union(*[can2set(k) for k in element2id.keys()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a1be8c2d-c5da-4e04-bf67-7e43b1a71b7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading gene names\n"
     ]
    }
   ],
   "source": [
    "# name2gene is a function mapping names to gene IDs (hsa:xxxx)\n",
    "# e.g. KCNE3, BRGDA6 etc are all mapped to hsa:10008 (see https://www.kegg.jp/entry/hsa:10008)\n",
    "# This map was produced by scraping kegg, but this has been done only once.\n",
    "# After that, we just read a pkl file\n",
    "\n",
    "name2keggFilename = os.path.join(inputDirectory, \"name2kegg.pkl\")\n",
    "\n",
    "if os.path.exists(name2keggFilename):\n",
    "    print(\"Reading gene names\")\n",
    "    with open(name2keggFilename, \"rb\") as handle:\n",
    "        name2gene = pickle.load(handle)\n",
    "else:\n",
    "    print(\"Producing gene names\")\n",
    "    d = {}\n",
    "    nelements = len(elements)\n",
    "    for count,element in enumerate(elements):\n",
    "        if count % 10 == 0:\n",
    "            print(f\"{count}/{nelements}\")\n",
    "        if element.startswith(\"hsa\"):\n",
    "            names = search_gene_KEGG(element)\n",
    "            for name in names:\n",
    "                d[name] = element\n",
    "    with open(name2keggFilename, \"wb\") as handle:\n",
    "        pickle.dump(d, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "    name2gene = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "74d19326",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading patient data\n"
     ]
    }
   ],
   "source": [
    "# Read patient data\n",
    "print(\"Reading patient data\")\n",
    "genes = pd.read_csv(os.path.join(inputDirectory, \"genes.csv\"), dtype={\"name\": str})\n",
    "patients = genes[\"name\"].values.tolist()\n",
    "classification = pd.read_csv(os.path.join(inputDirectory, \"classification.csv\"), dtype={\"name\": str})\n",
    "\n",
    "patient2class = {patient: classification.loc[classification[\"name\"]==patient][\"classification\"].values.flatten().tolist()[0] for patient in patients}\n",
    "numHealthy = sum([x == 'healthy' for x in patient2class.values()])\n",
    "numDiseased = sum([x == 'diseased' for x in patient2class.values()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "34acce33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dividing genes into two frames (one for healthy individuals and one for diseased individuals)\n",
    "genesHealthy = genes.loc[genes[\"name\"].apply(lambda x: patient2class[x])==\"healthy\"]\n",
    "genesDiseased = genes.loc[genes[\"name\"].apply(lambda x: patient2class[x])==\"diseased\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1c90f901-1413-4356-bfa8-faf25fc494df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building the relation graph\n",
      "- Edge (4,1) of type [('compound', 93)] for ['00010'] appears as [('compound', 94)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 94)]\n",
      "- Edge (4,2) of type [('compound', 93)] for ['00010'] appears as [('compound', 94)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 94)]\n",
      "- Edge (0,21) of type [('compound', 90)] for ['00010'] appears as [('compound', 91)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 91)]\n",
      "- Edge (21,4) of type [('compound', 93)] for ['00010'] appears as [('compound', 94)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 94)]\n",
      "- Edge (21,4) of type [('compound', 94)] for ['00010', '00010', '00010'] appears as [('compound', 93)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 93)]\n",
      "- Edge (40,38) of type [('compound', 320)] for ['00230'] appears as [('compound', 356)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 356)]\n",
      "- Edge (40,56) of type [('compound', 343)] for ['00230'] appears as [('compound', 355)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 355)]\n",
      "- Edge (54,46) of type [('compound', 368)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (54,55) of type [('compound', 392)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (54,56) of type [('compound', 392)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (55,56) of type [('compound', 392)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (40,51) of type [('compound', 343)] for ['00230', '00230', '00230', '00230'] appears as [('compound', 332)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 332)]\n",
      "- Edge (40,56) of type [('compound', 355)] for ['00230', '00230'] appears as [('compound', 332)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 332)]\n",
      "- Edge (61,56) of type [('compound', 392)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (34,46) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (34,50) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (34,56) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (34,57) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (46,56) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (46,57) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (50,56) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (50,57) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (56,57) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (57,52) of type [('compound', 331)] for ['00230'] appears as [('compound', 366)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 366)]\n",
      "- Edge (52,56) of type [('compound', 332)] for ['00230'] appears as [('compound', 355)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 355)]\n",
      "- Edge (40,62) of type [('compound', 343)] for ['00230'] appears as [('compound', 355)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 355)]\n",
      "- Edge (41,57) of type [('compound', 365)] for ['00230'] appears as [('compound', 357)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 357)]\n",
      "- Edge (58,57) of type [('compound', 330)] for ['00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (46,57) of type [('compound', 365)] for ['00230', '00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (48,46) of type [('compound', 369)] for ['00230'] appears as [('compound', 324)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 324)]\n",
      "- Edge (57,69) of type [('compound', 331)] for ['00230'] appears as [('compound', 326)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 326)]\n",
      "- Edge (57,52) of type [('compound', 366)] for ['00230', '00230'] appears as [('compound', 326)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 326)]\n",
      "- Edge (69,52) of type [('compound', 331)] for ['00230'] appears as [('compound', 326)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 326)]\n",
      "- Edge (46,45) of type [('compound', 390)] for ['00230'] appears as [('compound', 380)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 380)]\n",
      "- Edge (46,46) of type [('compound', 369)] for ['00230'] appears as [('compound', 324)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 324)]\n",
      "- Edge (48,45) of type [('compound', 390)] for ['00230'] appears as [('compound', 380)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 380)]\n",
      "- Edge (49,45) of type [('compound', 390)] for ['00230'] appears as [('compound', 380)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 380)]\n",
      "- Edge (42,53) of type [('compound', 325)] for ['00230'] appears as [('compound', 377)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 377)]\n",
      "- Edge (73,48) of type [('compound', 360)] for ['00230'] appears as [('compound', 324)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 324)]\n",
      "- Edge (73,48) of type [('compound', 324)] for ['00230', '00230'] appears as [('compound', 322)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 322)]\n",
      "- Edge (37,57) of type [('compound', 325)] for ['00230'] appears as [('compound', 321)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 321)]\n",
      "- Edge (58,46) of type [('compound', 325)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (46,45) of type [('compound', 380)] for ['00230', '00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (40,57) of type [('compound', 367)] for ['00230'] appears as [('compound', 320)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 320)]\n",
      "- Edge (46,73) of type [('compound', 324)] for ['00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (46,46) of type [('compound', 324)] for ['00230', '00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (46,48) of type [('compound', 324)] for ['00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (41,48) of type [('compound', 360)] for ['00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (73,46) of type [('compound', 324)] for ['00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (73,48) of type [('compound', 322)] for ['00230', '00230', '00230'] appears as [('compound', 362)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 362)]\n",
      "- Edge (48,50) of type [('compound', 369)] for ['00230'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (46,46) of type [('compound', 362)] for ['00230', '00230', '00230'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (48,46) of type [('compound', 324)] for ['00230', '00230'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (48,70) of type [('compound', 380)] for ['00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (46,70) of type [('compound', 380)] for ['00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (45,70) of type [('compound', 380)] for ['00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (48,45) of type [('compound', 380)] for ['00230', '00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (46,45) of type [('compound', 325)] for ['00230', '00230', '00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (96,46) of type [('compound', 185)] for ['00240'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (46,73) of type [('compound', 362)] for ['00230', '00230'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (46,46) of type [('compound', 186)] for ['00230', '00230', '00230', '00240'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (57,89) of type [('compound', 221)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (57,94) of type [('compound', 220)] for ['00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (91,91) of type [('compound', 181)] for ['00240'] appears as [('compound', 180)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 180)]\n",
      "- Edge (48,46) of type [('compound', 186)] for ['00230', '00230', '00240'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (48,73) of type [('compound', 214)] for ['00240'] appears as [('compound', 217)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 217)]\n",
      "- Edge (94,38) of type [('compound', 220)] for ['00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (46,57) of type [('compound', 325)] for ['00230', '00230', '00230'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (48,103) of type [('compound', 217)] for ['00240'] appears as [('compound', 194)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 194)]\n",
      "- Edge (48,46) of type [('compound', 214)] for ['00230', '00230', '00240', '00240'] appears as [('compound', 194)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 194)]\n",
      "- Edge (89,38) of type [('compound', 221)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (57,94) of type [('compound', 219)] for ['00240', '00240'] appears as [('compound', 212)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 212)]\n",
      "- Edge (46,46) of type [('compound', 214)] for ['00230', '00230', '00230', '00240', '00240'] appears as [('compound', 194)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 194)]\n",
      "- Edge (46,106) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (106,94) of type [('compound', 212)] for ['00240'] appears as [('compound', 188)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 188)]\n",
      "- Edge (103,57) of type [('compound', 218)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (57,97) of type [('compound', 219)] for ['00240'] appears as [('compound', 188)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 188)]\n",
      "- Edge (103,98) of type [('compound', 218)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (102,38) of type [('compound', 218)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (101,106) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (57,94) of type [('compound', 212)] for ['00240', '00240', '00240'] appears as [('compound', 188)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 188)]\n",
      "- Edge (103,38) of type [('compound', 218)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (57,45) of type [('compound', 365)] for ['00230'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (46,57) of type [('compound', 213)] for ['00230', '00230', '00230', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (46,57) of type [('compound', 195)] for ['00230', '00230', '00230', '00240', '00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (99,38) of type [('compound', 219)] for ['00240'] appears as [('compound', 196)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 196)]\n",
      "- Edge (50,57) of type [('compound', 365)] for ['00230', '00230'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (102,57) of type [('compound', 218)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (45,46) of type [('compound', 365)] for ['00230'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (101,57) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (46,45) of type [('compound', 185)] for ['00230', '00230', '00230', '00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (57,99) of type [('compound', 219)] for ['00240'] appears as [('compound', 196)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 196)]\n",
      "- Edge (48,45) of type [('compound', 185)] for ['00230', '00230', '00240'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (98,99) of type [('compound', 219)] for ['00240'] appears as [('compound', 196)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 196)]\n",
      "- Edge (101,88) of type [('compound', 218)] for ['00240'] appears as [('compound', 221)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 221)]\n",
      "- Edge (45,57) of type [('compound', 325)] for ['00230'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (100,107) of type [('compound', 210)] for ['00240'] appears as [('compound', 202)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 202)]\n",
      "- Edge (45,46) of type [('compound', 191)] for ['00230', '00240'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (48,101) of type [('compound', 214)] for ['00240'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (101,57) of type [('compound', 187)] for ['00240', '00240'] appears as [('compound', 221)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 221)]\n",
      "- Edge (107,92) of type [('compound', 209)] for ['00240'] appears as [('compound', 203)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 203)]\n",
      "- Edge (48,73) of type [('compound', 217)] for ['00240', '00240'] appears as [('compound', 222)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 222)]\n",
      "- Edge (46,50) of type [('compound', 369)] for ['00230'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (48,101) of type [('compound', 186)] for ['00240', '00240'] appears as [('compound', 222)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 222)]\n",
      "- Edge (106,45) of type [('compound', 187)] for ['00240'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (46,45) of type [('compound', 187)] for ['00230', '00230', '00230', '00240', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (45,57) of type [('compound', 213)] for ['00230', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (48,45) of type [('compound', 191)] for ['00230', '00230', '00240', '00240'] appears as [('compound', 193)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 193)]\n",
      "- Edge (46,45) of type [('compound', 195)] for ['00230', '00230', '00230', '00240', '00240', '00240'] appears as [('compound', 193)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 193)]\n",
      "- Edge (46,108) of type [('compound', 193)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (46,108) of type [('compound', 195)] for ['00240', '00240'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (48,108) of type [('compound', 193)] for ['00240'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (57,108) of type [('compound', 195)] for ['00240'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (46,108) of type [('compound', 191)] for ['00240', '00240', '00240'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (96,108) of type [('compound', 191)] for ['00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (48,108) of type [('compound', 191)] for ['00240', '00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (46,108) of type [('compound', 213)] for ['00240', '00240', '00240', '00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (46,108) of type [('compound', 185)] for ['00240', '00240', '00240', '00240', '00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (101,108) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (106,108) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (57,108) of type [('compound', 213)] for ['00240', '00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (138,176) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04062. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (518,504) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04062. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (518,216) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04062. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (176,189) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->')] for 04066. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (487,166) of type [('compound', 151), ('activation', '-->')] for ['04062'] appears as [('activation', '-->')] for 04066. Overwriting with type [('compound', 151), ('activation', '-->')]\n",
      "- Edge (569,572) of type [('compound', 52)] for ['04070'] appears as [('compound', 235)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 235)]\n",
      "- Edge (572,569) of type [('compound', 234)] for ['04070'] appears as [('compound', 68)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 68)]\n",
      "- Edge (576,575) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (575,577) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (582,583) of type [('compound', 66)] for ['04070'] appears as [('compound', 49)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 49)]\n",
      "- Edge (574,577) of type [('compound', 75)] for ['04070'] appears as [('compound', 48)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 48)]\n",
      "- Edge (574,577) of type [('compound', 48)] for ['04070', '04070'] appears as [('compound', 49)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 49)]\n",
      "- Edge (588,575) of type [('compound', 75)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (588,571) of type [('compound', 75)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (487,576) of type [('compound', 73)] for ['04070'] appears as [('compound', 66)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 66)]\n",
      "- Edge (576,589) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (585,585) of type [('compound', 60)] for ['04070'] appears as [('compound', 231)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 231)]\n",
      "- Edge (577,589) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (574,589) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (573,575) of type [('compound', 75)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (587,568) of type [('compound', 66)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (584,585) of type [('compound', 56)] for ['04070'] appears as [('compound', 61)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 61)]\n",
      "- Edge (585,585) of type [('compound', 231)] for ['04070', '04070'] appears as [('compound', 60)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 60)]\n",
      "- Edge (576,574) of type [('compound', 75)] for ['04070'] appears as [('compound', 72)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 72)]\n",
      "- Edge (574,577) of type [('compound', 49)] for ['04070', '04070', '04070'] appears as [('compound', 72)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 72)]\n",
      "- Edge (580,591) of type [('compound', 55)] for ['04070'] appears as [('compound', 63)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 63)]\n",
      "- Edge (586,584) of type [('compound', 56)] for ['04070'] appears as [('compound', 61)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 61)]\n",
      "- Edge (591,586) of type [('compound', 55)] for ['04070'] appears as [('compound', 62)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 62)]\n",
      "- Edge (568,576) of type [('compound', 73)] for ['04070'] appears as [('compound', 66)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 66)]\n",
      "- Edge (159,608) of type [('expression', '-->')] for ['04115'] appears as [('activation', '-->')] for 04115. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (482,501) of type [] for ['04062'] appears as [('activation', '-->')] for 04151. Overwriting with type [('activation', '-->')]\n",
      "- Edge (166,133) of type [('phosphorylation', '+p'), ('indirect effect', '..>')] for ['04062'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (197,198) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04151. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (198,493) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062'] appears as [('activation', '-->')] for 04151. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (540,176) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (544,546) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (546,548) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (155,156) of type [] for ['04010'] appears as [('activation', '-->')] for 04210. Overwriting with type [('activation', '-->')]\n",
      "- Edge (130,133) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04210. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (110,159) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04218. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (544,538) of type [('inhibition', '--|')] for ['04066'] appears as [('inhibition', '--|'), ('phosphorylation', '+p')] for 04218. Overwriting with type [('inhibition', '--|'), ('phosphorylation', '+p')]\n",
      "- Edge (576,501) of type [('inhibition', '--|'), ('dephosphorylation', '-p')] for ['04151'] appears as [('inhibition', '--|')] for 04218. Overwriting with type [('inhibition', '--|'), ('dephosphorylation', '-p')]\n",
      "- Edge (737,519) of type [('activation', '-->')] for ['04213'] appears as [('inhibition', '--|')] for 04218. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (159,207) of type [('expression', '-->')] for ['04115', '04115'] appears as [('activation', '-->')] for 04218. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (124,623) of type [('expression', '-->'), ('indirect effect', '..>')] for ['04218'] appears as [('expression', '-->')] for 04218. Overwriting with type [('expression', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (792,785) of type [('activation', '-->')] for ['04310'] appears as [('inhibition', '--|')] for 04310. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (165,775) of type [('activation', '-->'), ('dephosphorylation', '-p')] for ['04218'] appears as [('dephosphorylation', '-p')] for 04310. Overwriting with type [('activation', '-->'), ('dephosphorylation', '-p')]\n",
      "- Edge (159,619) of type [('expression', '-->')] for ['04115', '04115'] appears as [('indirect effect', '..>')] for 04310. \n",
      "****Conflict! Rewriting as [('indirect effect', '..>')]\n",
      "- Edge (487,166) of type [('compound', 151), ('activation', '-->')] for ['04062', '04066'] appears as [('compound', 24)] for 04510. \n",
      "****Conflict! Rewriting as [('compound', 24)]\n",
      "- Edge (485,486) of type [('binding/association', '---')] for ['04062'] appears as [('phosphorylation', '+p')] for 04510. \n",
      "****Conflict! Rewriting as [('phosphorylation', '+p')]\n",
      "- Edge (120,718) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04210'] appears as [('phosphorylation', '+p')] for 04510. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (138,176) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04010', '04062'] appears as [('phosphorylation', '+p')] for 04510. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (176,557) of type [('expression', '-->'), ('indirect effect', '..>')] for ['04210'] appears as [('phosphorylation', '+p')] for 04510. \n",
      "****Conflict! Rewriting as [('phosphorylation', '+p')]\n",
      "- Edge (486,216) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04510. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,120) of type [('activation', '-->')] for ['04310'] appears as [('phosphorylation', '+p'), ('indirect effect', '..>')] for 04510. \n",
      "****Conflict! Rewriting as [('phosphorylation', '+p'), ('indirect effect', '..>')]\n",
      "- Edge (497,197) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04510. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (669,166) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04210'] appears as [('phosphorylation', '+p')] for 04510. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (176,228) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 04611. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (841,704) of type [('binding/association', '---')] for ['04510'] appears as [('activation', '-->')] for 04611. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (487,166) of type [('compound', 24)] for ['04062', '04066', '04510'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 04611. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (722,521) of type [('dissociation', '-+-')] for ['04210'] appears as [('inhibition', '--|')] for 04620. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (127,116) of type [('activation', '-->')] for ['04010'] appears as [('binding/association', '---')] for 04620. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (540,176) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04151', '04210', '04218'] appears as [('phosphorylation', '+p')] for 04620. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (757,110) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04218'] appears as [('phosphorylation', '+p')] for 04620. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (176,189) of type [('activation', '-->')] for ['04010', '04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (176,157) of type [('phosphorylation', '+p')] for ['04010', '04510'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (189,539) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (734,741) of type [('activation', '-->')] for ['04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (166,544) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (546,548) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04151'] appears as [('activation', '-->')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (487,669) of type [('compound', 24)] for ['04510'] appears as [('compound', 71), ('activation', '-->')] for 04910. \n",
      "****Conflict! Rewriting as [('compound', 71), ('activation', '-->')]\n",
      "- Edge (198,493) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062', '04151'] appears as [('activation', '-->')] for 04910. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (954,722) of type [('phosphorylation', '+p')] for ['04620'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (722,521) of type [('inhibition', '--|')] for ['04210', '04620'] appears as [('dissociation', '-+-')] for 04931. \n",
      "****Conflict! Rewriting as [('dissociation', '-+-')]\n",
      "- Edge (487,496) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062'] appears as [('activation', '-->')] for 04931. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (734,652) of type [('activation', '-->')] for ['04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (536,542) of type [('activation', '-->')] for ['04611', '04931'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 04931. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (1018,652) of type [('inhibition', '--|'), ('indirect effect', '..>')] for ['04931'] appears as [('inhibition', '--|')] for 04931. Overwriting with type [('inhibition', '--|'), ('indirect effect', '..>')]\n",
      "- Edge (652,487) of type [('activation', '-->')] for ['04151', '04213', '04931', '04931'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (796,145) of type [('compound', 12)] for ['04310'] appears as [('compound', 277), ('activation', '-->')] for 05146. \n",
      "****Conflict! Rewriting as [('compound', 277), ('activation', '-->')]\n",
      "- Edge (874,875) of type [('compound', 131)] for ['04512'] appears as [('binding/association', '---')] for 05205. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (176,157) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04010', '04510', '04910'] appears as [('phosphorylation', '+p')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (506,510) of type [('activation', '-->')] for ['04062', '04510', '04611'] appears as [('binding/association', '---')] for 05205. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (487,166) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062', '04066', '04510', '04611', '04620'] appears as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')]\n",
      "- Edge (510,850) of type [('inhibition', '--|'), ('phosphorylation', '+p')] for ['04510'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (635,727) of type [('activation', '-->')] for ['04213'] appears as [('binding/association', '---')] for 05205. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (558,1089) of type [('activation', '-->')] for ['05205'] appears as [('inhibition', '--|')] for 05205. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (727,487) of type [('activation', '-->')] for ['04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (166,544) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04910'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (546,548) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04151', '04910'] appears as [('activation', '-->')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (546,691) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151'] appears as [('activation', '-->')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (487,669) of type [('compound', 71), ('activation', '-->')] for ['04510', '04910'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (669,546) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04910'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (1122,176) of type [('activation', '-->'), ('indirect effect', '..>')] for ['05205'] appears as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')]\n",
      "- Edge (529,521) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04066'] appears as [('activation', '-->')] for 05321. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (487,501) of type [('activation', '-->')] for ['04151', '04210', '04213', '04213', '04218', '04931'] appears as [('missing interaction', '-/-')] for 05415. \n",
      "****Conflict! Rewriting as [('missing interaction', '-/-')]\n",
      "- Edge (652,487) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04213', '04931', '04931', '04931'] appears as [('missing interaction', '-/-')] for 05415. \n",
      "****Conflict! Rewriting as [('missing interaction', '-/-')]\n",
      "- Edge (166,682) of type [('inhibition', '--|'), ('phosphorylation', '+p')] for ['04151', '04510', '04910', '04931'] appears as [('missing interaction', '-/-')] for 05415. \n",
      "****Conflict! Rewriting as [('missing interaction', '-/-')]\n",
      "- Edge (536,542) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04611', '04931', '04931'] appears as [('activation', '-->')] for 05415. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (418,642) of type [('indirect effect', '..>')] for ['04115'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05417. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (955,120) of type [('phosphorylation', '+p')] for ['04620'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (120,152) of type [('phosphorylation', '+p')] for ['04010', '04620'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (116,133) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (167,127) of type [('binding/association', '---')] for ['04010'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (166,682) of type [('missing interaction', '-/-')] for ['04151', '04510', '04910', '04931', '05415'] appears as [('inhibition', '--|'), ('phosphorylation', '+p')] for 05417. \n",
      "****Conflict! Rewriting as [('inhibition', '--|'), ('phosphorylation', '+p')]\n",
      "- Edge (502,748) of type [('activation', '-->')] for ['04218', '04970'] appears as [('binding/association', '---')] for 05417. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (521,536) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04931'] appears as [('repression', '--|')] for 05417. \n",
      "****Conflict! Rewriting as [('repression', '--|')]\n",
      "- Edge (521,450) of type [('expression', '-->')] for ['04620'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (715,716) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04210'] appears as [('activation', '-->')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (521,1208) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (521,1209) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (521,1210) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (521,321) of type [('expression', '-->')] for ['04620'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (521,1211) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (521,353) of type [('expression', '-->')] for ['04620', '04931', '04931'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (1095,506) of type [('binding/association', '---')] for ['05205'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (521,447) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (521,187) of type [('expression', '-->')] for ['04620', '04931', '04931', '05321'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (521,450) of type [('activation', '-->')] for ['04620', '05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (521,1213) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (487,501) of type [('missing interaction', '-/-')] for ['04151', '04210', '04213', '04213', '04218', '04931', '05415'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (521,191) of type [('activation', '-->')] for ['05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (521,321) of type [('activation', '-->')] for ['04620', '05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (521,353) of type [('activation', '-->')] for ['04620', '04931', '04931', '05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "- Edge (669,166) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04210', '04510', '04910', '04931'] appears as [('activation', '-->')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (521,187) of type [('activation', '-->')] for ['04620', '04931', '04931', '05321', '05417'] appears as [('expression', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('expression', '-->')]\n",
      "Wrap: more than 148 colors needed\n",
      "Wrap: more than 148 colors needed\n",
      "Wrap: more than 148 colors needed\n",
      "Wrap: more than 148 colors needed\n"
     ]
    }
   ],
   "source": [
    "# Building the graph of all relations contained in rel values\n",
    "# Nodes are IDs, and node x correspond to the set of genes (and/or compound) id2element[x]\n",
    "print(\"Building the relation graph\")\n",
    "G = nx.DiGraph() \n",
    "for pathwayID, relations in rel.items():\n",
    "    for relation in relations:\n",
    "        source = element2id[set2can(relation.source)]\n",
    "        for targ in relation.target:\n",
    "            target = element2id[set2can(targ)]\n",
    "            newType = str(relation.relSubtypes)\n",
    "            color = subtype2color[newType]\n",
    "            pl = []\n",
    "            if G.has_edge(source, target):\n",
    "                d = G.get_edge_data(source, target)\n",
    "                if d[\"color\"] != color:\n",
    "                    oldType = color2subtype[d[\"color\"]]\n",
    "                    print(\"- Edge ({},{}) of type {} for {} appears as {} for {}. \".format(\n",
    "                        source, target, color2subtype[d[\"color\"]], d[\"pathwayIDs\"], \n",
    "                        str(relation.relSubtypes), pathwayID\n",
    "                    ), end=\"\")\n",
    "                    if oldType[1:-1] in newType[1:-1]:\n",
    "                        print(\"Overwriting with type {}\".format(newType))\n",
    "                    elif newType[1:-1] in oldType[1:-1]:\n",
    "                        print(\"Overwriting with type {}\".format(oldType))\n",
    "                        color = subtype2color[oldType]\n",
    "                    else:\n",
    "                        print(\"\\n****Conflict! Rewriting as {}\".format(newType))\n",
    "                pl = d[\"pathwayIDs\"]\n",
    "            G.add_edge(source, target, color=color, pathwayIDs=pl + [pathwayID])\n",
    "            \n",
    "\n",
    "# Computing the minimum fibres\n",
    "cc=cardon_crochemore_colored(G)\n",
    "\n",
    "colors = list(matplotlib.colors.CSS4_COLORS.values())\n",
    "usedColors = 0\n",
    "\n",
    "cc2fibresize = {}\n",
    "cc2fibre = {}\n",
    "ccnontrivial = {}\n",
    "nontrivialclass2col = {}\n",
    "\n",
    "fibres = []\n",
    "node2fibre = {}\n",
    "for v in set(cc.values()):\n",
    "    fibre = [x for x in G.nodes() if cc[x] == v]\n",
    "    fibres += [fibre]\n",
    "    for x in fibre:\n",
    "        node2fibre[x] = fibre\n",
    "    cc2fibre[v] = fibre\n",
    "    cc2fibresize[v] = len(fibre)\n",
    "    if len(fibre) > 0:\n",
    "        ccnontrivial[v] = cc[v]\n",
    "        nontrivialclass2col[v] = colors[usedColors]\n",
    "        usedColors = (usedColors + 1 ) % len(colors)\n",
    "        if usedColors == 0:\n",
    "            print(\"Wrap: more than\", len(colors), \"colors needed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d51dbd7f-35b9-444a-9238-3af727c06411",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping genes to columns\n"
     ]
    }
   ],
   "source": [
    "# Build node2columns, mapping each node of G to the list of genes corresponding to it\n",
    "# for which we know a value (in the genes dataframe), sorted\n",
    "#\n",
    "\n",
    "print(\"Mapping genes to columns\")\n",
    "namedGenes = set(name2gene.values())     \n",
    "columns = set(genes.columns.values.tolist()[2:])\n",
    "node2columns = {}\n",
    "for node in G.nodes():\n",
    "    nodeGenes = can2set(id2element[node])\n",
    "    s = []\n",
    "    for gene in nodeGenes & namedGenes:\n",
    "        for k,v in name2gene.items():\n",
    "            if v == gene:\n",
    "                if k in columns:\n",
    "                    s += [k]\n",
    "    node2columns[node]=sorted(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "05b96661",
   "metadata": {},
   "outputs": [],
   "source": [
    "if produceGraphsWithCCcoloring:\n",
    "    s = \"\"\n",
    "    for pair in superpathwayDict[superpathway]:\n",
    "        pathwayID = pair[0]\n",
    "        pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "        canvas = KGMLCanvas(pathway)\n",
    "        for k in pathway.entries:\n",
    "            name = pathway.entries[k].name\n",
    "            canonicalName = set2can(words2set(name))\n",
    "            if canonicalName in element2id.keys():\n",
    "                node = element2id[canonicalName]\n",
    "                klass = cc[node]\n",
    "                if klass in nontrivialclass2col:\n",
    "                    pathway.entries[k].graphics[0].bgcolor = nontrivialclass2col[klass]\n",
    "                    pathway.entries[k].graphics[0].fgcolor = nontrivialclass2col[klass]\n",
    "                else:\n",
    "                    pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "                    pathway.enries[k].graphics[0].fgcolor = \"#FFFFFF\"\n",
    "                s += pathwayID + \"\\t\" + str(node) + \"\\t\" + str(klass) +\"\\t\"+pathway.entries[k].graphics[0].name+\"\\t\"+canonicalName + \"\\n\"\n",
    "        canvas.import_imagemap = True\n",
    "        pdfName = organism + pathwayID + \".pdf\"\n",
    "        pathlib.Path(os.path.join(outputDirectory, organism, safeSuperpathway)).mkdir(parents=True, exist_ok=True)\n",
    "        canvas.draw(os.path.join(outputDirectory, organism, safeSuperpathway, pdfName))\n",
    "    with open(os.path.join(outputDirectory, organism, safeSuperpathway, \"all.tsv\"), \"wt\") as file:\n",
    "        file.write(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "73ce2fd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def columns2nodes(columns, node2columns, ignoreEmpty = True):\n",
    "    \"\"\"\n",
    "        Returns a map from columns to the nodes where the column appears.\n",
    "    \"\"\"\n",
    "    res = {} \n",
    "    for c in columns:\n",
    "        s = set([])\n",
    "        for no, co in node2columns.items():\n",
    "            if c in co:\n",
    "                s |= set([no])\n",
    "        if len(s) > 0 or not ignoreEmpty:\n",
    "            res[c] = s\n",
    "    return res\n",
    "\n",
    "c2n = columns2nodes(columns, node2columns)\n",
    "usableColumns = list(c2n.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "00742eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def corrForPair(genes, set1, set2, correlationFunction = lambda x,y: scipy.stats.kendalltau(x,y).statistic):\n",
    "    \"\"\"\n",
    "        Given a dataframe genes (i.e., a set of patients), and two sets of column names (set1, set2), \n",
    "        compute the correlation function\n",
    "        for every pair of columns in set1 and set2 and return the average statistic and its std.\n",
    "    \"\"\"\n",
    "    taus = []\n",
    "    for gene1 in set1:\n",
    "        for gene2 in set2:\n",
    "            expr1 = genes.loc[:,gene1].values.tolist()\n",
    "            expr2 = genes.loc[:,gene2].values.tolist()\n",
    "            if len(set(expr1)) == 1 or len(set(expr2)) == 1:\n",
    "                continue\n",
    "            taus += [correlationFunction(expr1, expr2).statistic]\n",
    "    if len(taus) < 2:\n",
    "        return np.nan, np.nan\n",
    "    return pd.DataFrame(taus).mean().values[0], pd.DataFrame(taus).std().values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b3232726",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sampleDeltaDist(genes, set1, set2, numberOfRows1, numberOfSamples = 500, correlationFunction = lambda x,y: scipy.stats.kendalltau(x,y).statistic):\n",
    "    \"\"\"\n",
    "        Repeatedly (for numberOfSamples times) do the following:\n",
    "        - select numberOfRows1 rows of genes, and then compute tauForPair on those rows for set1 and set2\n",
    "        - do the same for remaining rows\n",
    "        - take the difference between the two resulting set of taus.\n",
    "        Return the mean and std of differences.\n",
    "    \"\"\"\n",
    "    diffs = []\n",
    "    \n",
    "    for i in range(numberOfSamples):\n",
    "        genes1 = random.sample(range(len(genes)), numberOfRows1)\n",
    "        genes2 = list(set(range(len(genes)))-set(genes1))\n",
    "        mt1, mstd1 = corrForPair(genes.iloc[genes1], set1, set2, correlationFunction)\n",
    "        if np.isnan(mp1):\n",
    "            continue\n",
    "        mt2, mstd2 = corrForPair(genes.iloc[genes2], set1, set2, correlationFunction)\n",
    "        if np.isnan(mp2):\n",
    "            continue\n",
    "        diffs += [mt1-mt2]\n",
    "    return diffs\n",
    "\n",
    "\n",
    "def deltaForPair(genes1, genes2, set1, set2, correlationFunction = lambda x,y: scipy.stats.kendalltau(x,y).statistic):\n",
    "    \"\"\"\n",
    "        Compute corrForPair for set1 and set2 over the set of rows genes1 and genes2, and return\n",
    "        the difference of taus.\n",
    "    \"\"\"\n",
    "    mt1, mstd1 = corrForPair(genes1, set1, set2, correlationFunction)\n",
    "    mt2, mstd2 = corrForPair(genes2, set1, set2, correlationFunction)\n",
    "    return mt1-mt2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2d3473f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading anomalies\n"
     ]
    }
   ],
   "source": [
    "correlationFunction = lambda x,y:  scipy.stats.spearmanr(x,y).statistic\n",
    "correlationFunctionName = \"spearmanrho\"\n",
    "relAnomalies = os.path.join(inputDirectory, \"anomalies-\" + correlationFunctionName + \".csv\")\n",
    "\n",
    "if os.path.exists(relAnomalies):\n",
    "    print(\"Reading anomalies\")\n",
    "    relcorrel = pd.read_csv(relAnomalies)\n",
    "else:\n",
    "    print(\"Producing anomalies\")\n",
    "    relcorrel = pd.DataFrame(columns=[\n",
    "        \"Source gene(s)\",\n",
    "        \"Target gene(s)\",\n",
    "        \"Pathway ID\",\n",
    "        \"Relation type\",\n",
    "        \"Correlation for healthy\",\n",
    "        \"Correlation for diseased\",\n",
    "        \"Correlation difference (H-D)\",\n",
    "        \"Expected correlation difference\",\n",
    "        \"P-value\",\n",
    "        \"Comment\"\n",
    "        ])\n",
    "    numberOfEdges = G.number_of_edges()\n",
    "    doneEdges = 0\n",
    "    for s,t,d in G.edges(data=True):\n",
    "        if doneEdges % 10 == 0:\n",
    "            print(\"{:.2f}% done\".format(100 * doneEdges / numberOfEdges))\n",
    "        doneEdges += 1\n",
    "        subtypeName = [k for k,v in subtype2color.items() if d[\"color\"]==v][0]\n",
    "        sg = node2columns[s]\n",
    "        tg = node2columns[t]\n",
    "        if len(sg) == 0 or len(tg) == 0:\n",
    "            continue\n",
    "        avgh, stdh = corrForPair(genesHealthy, sg, tg, correlationFunction=correlationFunction)\n",
    "        avgd, stdd = corrForPair(genesDiseased, sg, tg, correlationFunction=correlationFunction)\n",
    "        taudiff = avgh - avgd\n",
    "        ttn = sampleDeltaDist(genes, sg, tg, numHealthy, correlationFunction=correlationFunction)\n",
    "        tt = [x for x in ttn if ~np.isnan(x)]\n",
    "        if len(tt) == 0:\n",
    "            continue\n",
    "        ttmean = pd.DataFrame(tt).mean().values[0]\n",
    "        ecdft = scipy.stats.ecdf(tt)  # Empirical CDF\n",
    "        if taudiff < 0:\n",
    "            pvalue = ecdft.cdf.evaluate(taudiff)\n",
    "            comment = \"More correlated than for healthy individuals\"\n",
    "        else:\n",
    "            pvalue = 1 - ecdft.cdf.evaluate(taudiff)    \n",
    "            comment = \"Less correlated than for healthy individuals\"\n",
    "        if produceAnomaliesForNonsignificantPairs or produpvalue <= confidence:\n",
    "            relcorrel.loc[len(relcorrel)] = [\n",
    "                str(sg),\n",
    "                str(tg),\n",
    "                d[\"pathwayIDs\"],\n",
    "                subtypeName,\n",
    "                avgh, \n",
    "                avgd,\n",
    "                taudiff,\n",
    "                ttmean,\n",
    "                pvalue,\n",
    "                comment\n",
    "            ]\n",
    "    relcorrel = relcorrel.sort_values(by=[\"P-value\"], ascending=True)\n",
    "    relcorrel.to_csv(relAnomalies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e517681e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anomalies by type\n",
      "P-value < 0.05:  93\n",
      "Failing activations (green): 21\n",
      "Failing inhibitions (blue): 5\n",
      "Excessive activations (red): 13\n",
      "Excessive inhibitions (orange): 9\n",
      "Other types of relations (yellow): 43\n",
      "Total: 91\n"
     ]
    }
   ],
   "source": [
    "print(\"Anomalies by type\")\n",
    "statisticallySignificant = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence)   # Statistically significant\n",
    "]\n",
    "\n",
    "failingActivation = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"activation\")) & # Activation  \n",
    "    (relcorrel[\"Correlation difference (H-D)\"]>0) # Less than healthy\n",
    "]\n",
    "\n",
    "failingInhibition = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"inhibition\")) & # Inhibition\n",
    "    (relcorrel[\"Correlation difference (H-D)\"]<0) # More than healthy\n",
    "]\n",
    "\n",
    "excessiveActivation = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"activation\")) & # Activation  \n",
    "    (relcorrel[\"Correlation difference (H-D)\"]<0) # More than healthy\n",
    "]\n",
    "\n",
    "excessiveInhibition = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"inhibition\")) & # Inhibition\n",
    "    (relcorrel[\"Correlation difference (H-D)\"]>0) # Less than healthy\n",
    "]\n",
    "\n",
    "otherTypesOfRelations = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    ~(relcorrel[\"Relation type\"].str.contains(\"activation\")) & # No activation  \n",
    "    ~(relcorrel[\"Relation type\"].str.contains(\"inhibition\"))   # No inhibition \n",
    "]\n",
    "\n",
    "print(\"P-value < {:.2f}: \".format(100 * confidence / 100), len(statisticallySignificant))\n",
    "print(\"Failing activations (green):\", len(failingActivation))\n",
    "print(\"Failing inhibitions (blue):\", len(failingInhibition))\n",
    "print(\"Excessive activations (red):\", len(excessiveActivation))\n",
    "print(\"Excessive inhibitions (orange):\", len(excessiveInhibition))\n",
    "print(\"Other types of relations (yellow):\", len(otherTypesOfRelations))\n",
    "print(\"Total:\", len(failingActivation)+len(failingInhibition)+len(excessiveActivation)+len(excessiveInhibition)+len(otherTypesOfRelations))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eec52e3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d4917654",
   "metadata": {},
   "outputs": [],
   "source": [
    "def elaborateArc(relation, sdrawing):\n",
    "    source= str(node2columns[element2id[set2can(relation.source)]])\n",
    "    for tg in relation.target:\n",
    "        target = str(node2columns[element2id[set2can(tg)]])\n",
    "        r = relcorrel.loc[(relcorrel[\"Source gene(s)\"] == source) & (relcorrel[\"Target gene(s)\"] == target)]\n",
    "        if len(r) == 0:\n",
    "            continue\n",
    "        if len(r) > 1:\n",
    "            pass\n",
    "            #print(f\"({source}, {target}) appearing more than once in relcorrel\")\n",
    "            #print(list(r.to_dict()[\"Correlation difference (H-D)\"].values()))\n",
    "        rd = r.to_dict()\n",
    "        pvalue = list(rd[\"P-value\"].values())[0]\n",
    "        corrdiff = list(rd[\"Correlation difference (H-D)\"].values())[0]\n",
    "        relationType = list(rd[\"Relation type\"].values())[0]\n",
    "        if pvalue < confidence:        \n",
    "            if \"activation\" in relationType:\n",
    "                if corrdiff > 0:\n",
    "                    col = reportlab.lib.colors.green\n",
    "                else:\n",
    "                    col = reportlab.lib.colors.red\n",
    "            elif \"inhibition\" in relationType:\n",
    "                if corrdiff > 0:\n",
    "                    col = reportlab.lib.colors.orange\n",
    "                else:\n",
    "                    col = reportlab.lib.colors.blue\n",
    "            else:\n",
    "                col = reportlab.lib.colors.yellow\n",
    "            if np.isnan(corrdiff):\n",
    "                print(f\"Correlation for {source} vs. {target} is {corrdiff}\")\n",
    "            else:\n",
    "                sdrawing.setLineWidth(3*math.exp(abs(corrdiff)*3))\n",
    "                sdrawing.setStrokeColor(col)\n",
    "                sdrawing.line(relation.sourceg.x, relation.sourceg.y, relation.targetg.x, relation.targetg.y)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "32b16bfb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Producing anomaly PDFs\n",
      "Correlation for ['TP53'] vs. ['AIFM2', 'PRG3'] is nan\n",
      "Correlation for ['SOST'] vs. ['HBM', 'LRP5', 'LRP6'] is nan\n",
      "Correlation for ['IFNB1'] vs. ['IFNAR1', 'IFNAR2'] is nan\n",
      "Correlation for ['IRF7'] vs. ['IFNA5', 'IFNB1'] is nan\n",
      "Correlation for ['IRF3'] vs. ['IFNA5', 'IFNB1'] is nan\n"
     ]
    }
   ],
   "source": [
    "if produceGraphsWithAnomalies:\n",
    "    print(\"Producing anomaly PDFs\")\n",
    "    s = \"\"\n",
    "    for pair in superpathwayDict[superpathway]:\n",
    "        pathwayID = pair[0]\n",
    "        pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "\n",
    "        relations = []\n",
    "        for relation in rel[pathwayID]:\n",
    "            relations += [relation]\n",
    "\n",
    "        canvas = KGMLCanvas(pathway)\n",
    "\n",
    "        if useCCcoloringForAnomalies:\n",
    "            for k in pathway.entries:\n",
    "                name = pathway.entries[k].name\n",
    "                canonicalName = set2can(words2set(name))\n",
    "                if canonicalName in element2id.keys():\n",
    "                    node = element2id[canonicalName]\n",
    "                    klass = cc[node]\n",
    "                    if klass in nontrivialclass2col:\n",
    "                        pathway.entries[k].graphics[0].bgcolor = nontrivialclass2col[klass]\n",
    "                        pathway.entries[k].graphics[0].fgcolor = nontrivialclass2col[klass]\n",
    "                    else:\n",
    "                        pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "                        pathway.entries[k].graphics[0].fgcolor = \"#FFFFFF\"\n",
    "                    s += pathwayID + \"\\t\" + str(node) + \"\\t\" + str(klass) +\"\\t\"+pathway.entries[k].graphics[0].name+\"\\t\"+canonicalName + \"\\n\"\n",
    "\n",
    "        canvas.import_imagemap = True\n",
    "        pdfName = organism + pathwayID + \".pdf\"\n",
    "        pathlib.Path(os.path.join(outputDirectory, organism, safeSuperpathway + \"-anomalies\")).mkdir(parents=True, exist_ok=True)\n",
    "        canvas.new_draw(os.path.join(outputDirectory, organism, safeSuperpathway+ \"-anomalies\", pdfName), relations, elaborateArc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8e63a371",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing anomalies by pathway\n"
     ]
    }
   ],
   "source": [
    "print(\"Computing anomalies by pathway\")\n",
    "pathwayAnomalies = pd.DataFrame(columns=[\n",
    "        \"Pathway ID\",\n",
    "        \"Pathway\",\n",
    "        \"Average square difference\",\n",
    "        \"Count\",\n",
    "        \"Relative count\"\n",
    "        ])\n",
    "for pair in superpathwayDict[superpathway]:\n",
    "    pathwayID = pair[0]\n",
    "    pathwayName= pair[1]\n",
    "    ssqd = 0\n",
    "    count = 0\n",
    "    countSig = 0\n",
    "    for relation in rel[pathwayID]:\n",
    "        source = str(node2columns[element2id[set2can(relation.source)]])\n",
    "        for targ in relation.target:\n",
    "            target = str(node2columns[element2id[set2can(targ)]])\n",
    "            tt = relcorrel.loc[(relcorrel[\"Source gene(s)\"] == source) & (relcorrel[\"Target gene(s)\"] == target)]\n",
    "            if len(tt) == 1 and tt[\"P-value\"].to_list()[0] <= confidence:\n",
    "                cd = tt[\"Correlation difference (H-D)\"].to_list()[0]\n",
    "                if not np.isnan(cd):\n",
    "                    ssqd += cd * cd\n",
    "                    countSig += 1\n",
    "            count += 1\n",
    "    if count > 0:\n",
    "        pathwayAnomalies.loc[len(pathwayAnomalies)]=[pathwayID, pathwayName, ssqd, countSig, countSig / count]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ff411669-e133-48ac-86b9-dcc350d4d957",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sorting pathways by number of anomalies and writing it out as a csv\n"
     ]
    }
   ],
   "source": [
    "print(\"Sorting pathways by number of anomalies and writing it out as a csv\")\n",
    "relPathwayAnomalies = os.path.join(inputDirectory, \"pathway-anomalies-\" + correlationFunctionName + \".csv\")\n",
    "pathwayAnomalies = pathwayAnomalies.sort_values(by=[\"Count\"], ascending=False)\n",
    "pathwayAnomalies.to_csv(relPathwayAnomalies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3cae2695",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting genes involved in top-k anomalous pathways and appearing as columns\n",
      "1770 genes involved\n",
      "Genes involved by pathway\n",
      "\t05417: 195\n",
      "\t00010: 65\n",
      "\t04010: 290\n",
      "\t00240: 57\n",
      "\t04218: 142\n",
      "\t04620: 87\n",
      "\t04512: 84\n",
      "\t04060: 258\n",
      "\t05205: 197\n",
      "\t04151: 304\n",
      "\t04062: 188\n",
      "\t04931: 92\n",
      "\t04910: 131\n",
      "\t04310: 164\n",
      "\t04210: 126\n",
      "\t04115: 73\n",
      "\t04611: 122\n",
      "\t04510: 206\n",
      "\t00230: 126\n",
      "\t04066: 100\n",
      "\t04213: 56\n",
      "\t04975: 11\n",
      "\t05146: 36\n",
      "\t05321: 43\n",
      "\t05415: 101\n",
      "\t04970: 52\n",
      "\t04974: 3\n",
      "\t04979: 35\n",
      "\t04145: 16\n",
      "\t04070: 86\n"
     ]
    }
   ],
   "source": [
    "print(\"Extracting genes involved in top-k anomalous pathways and appearing as columns\")\n",
    "nTopAnomPathway = len(pathwayAnomalies)\n",
    "topAnomPathway = pathwayAnomalies.head(nTopAnomPathway)[\"Pathway ID\"].to_list()\n",
    "anomPathwayGenes = set([])\n",
    "anomPathwayGenesByPathway = defaultdict(lambda: set([]))\n",
    "for pathwayId in topAnomPathway:\n",
    "    for relation in rel[pathwayId]:\n",
    "        anomPathwayGenes |= relation.source \n",
    "        anomPathwayGenesByPathway[pathwayId] |= relation.source \n",
    "        for t in relation.target:\n",
    "            anomPathwayGenes |= t\n",
    "            anomPathwayGenesByPathway[pathwayId] |= t\n",
    "anomPathwayGeneNames = [name for name,gene in name2gene.items() if gene in anomPathwayGenes and name in columns]\n",
    "print(len(anomPathwayGeneNames), \"genes involved\")\n",
    "anomPathwayGenesByPathwayNames = {}\n",
    "print(\"Genes involved by pathway\")\n",
    "for pathwayId in topAnomPathway:\n",
    "    anomPathwayGenesByPathwayNames[pathwayId] = [name for name,gene in name2gene.items() if gene in anomPathwayGenesByPathway[pathwayId] and name in columns]\n",
    "    t = len(anomPathwayGenesByPathwayNames[pathwayId])\n",
    "    print(f\"\\t{pathwayId}: {t}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c1e14ecf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing correlations for anomalous-pathway genes in the two populations\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/boldi/.local/share/virtualenvs/HyperFibrations-l1h_g1LE/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n"
     ]
    }
   ],
   "source": [
    "print(\"Computing correlations for anomalous-pathway genes in the two populations\")\n",
    "healthyMatrix = genesHealthy[anomPathwayGeneNames].to_numpy() #Rows are healthy patients, columns are genes appearing in anomalous pathways\n",
    "diseasedMatrix = genesDiseased[anomPathwayGeneNames].to_numpy() #Rows are diseased patients, columns are genes appearing in anomalous pathways\n",
    "\n",
    "healthyCorr = np.corrcoef(healthyMatrix, rowvar=False) #Correlation between anomalous genes in healthy individuals\n",
    "diseasedCorr = np.corrcoef(diseasedMatrix, rowvar=False) #Correlation between anomalous genes in diseased individuals\n",
    "healthyCorrByPathway = {}\n",
    "diseasedCorrByPathway = {}\n",
    "for pathwayId in topAnomPathway:\n",
    "    healthyMatrix = genesHealthy[anomPathwayGenesByPathwayNames[pathwayId]].to_numpy() #Rows are healthy patients, columns are genes appearing in anomalous pathways\n",
    "    diseasedMatrix = genesDiseased[anomPathwayGenesByPathwayNames[pathwayId]].to_numpy() #Rows are healthy patients, columns are genes appearing in anomalous pathways\n",
    "    healthyCorrByPathway[pathwayId] = np.corrcoef(healthyMatrix, rowvar=False)\n",
    "    diseasedCorrByPathway[pathwayId] = np.corrcoef(diseasedMatrix, rowvar=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "36fbda30",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(data=healthyCorr, columns=anomPathwayGeneNames).to_csv(\"/tmp/matrices/healthy-all.csv\")\n",
    "pd.DataFrame(data=diseasedCorr, columns=anomPathwayGeneNames).to_csv(\"/tmp/matrices/diseased-all.csv\")\n",
    "for pathwayId in topAnomPathway:\n",
    "    pd.DataFrame(data=healthyCorrByPathway[pathwayId], columns=anomPathwayGenesByPathwayNames[pathwayId]).to_csv(\"/tmp/matrices/healthy-\" + pathwayId + \".csv\")\n",
    "    pd.DataFrame(data=diseasedCorrByPathway[pathwayId], columns=anomPathwayGenesByPathwayNames[pathwayId]).to_csv(\"/tmp/matrices/diseased-\" + pathwayId + \".csv\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3edb119d",
   "metadata": {},
   "outputs": [],
   "source": [
    "feb = pd.read_csv(\"../../LaTeX/Data/KEGG-Pathways/Feb2024/pathway-anomalies-spearmanrho.csv\").drop([\"Count\", \"Relative count\", \"Unnamed: 0\"], axis=1).set_index(\"Pathway ID\")\n",
    "mar = pd.read_csv(\"../../LaTeX/Data/KEGG-Pathways/Mar2024/pathway-anomalies-spearmanrho.csv\").drop([\"Count\", \"Relative count\", \"Unnamed: 0\"], axis=1).set_index(\"Pathway ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "0b6ca0f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "both = feb.join(mar, on=\"Pathway ID\", lsuffix=\"_feb\", rsuffix=\"_mar\")\n",
    "both[\"SS\"] = both[\"Average square difference_feb\"] * both[\"Average square difference_mar\"]\n",
    "both = both.sort_values(by=[\"SS\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "dc74aec4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway_feb</th>\n",
       "      <th>Average square difference_feb</th>\n",
       "      <th>Pathway_mar</th>\n",
       "      <th>Average square difference_mar</th>\n",
       "      <th>SS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathway ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5417</th>\n",
       "      <td>Lipid and atherosclerosis</td>\n",
       "      <td>0.625288</td>\n",
       "      <td>Lipid and atherosclerosis</td>\n",
       "      <td>1.962452</td>\n",
       "      <td>1.227097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4010</th>\n",
       "      <td>MAPK signaling pathway</td>\n",
       "      <td>0.828898</td>\n",
       "      <td>MAPK signaling pathway</td>\n",
       "      <td>0.869938</td>\n",
       "      <td>0.721090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4218</th>\n",
       "      <td>Cellular senescence</td>\n",
       "      <td>0.609202</td>\n",
       "      <td>Cellular senescence</td>\n",
       "      <td>0.690115</td>\n",
       "      <td>0.420420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4931</th>\n",
       "      <td>Insulin resistance</td>\n",
       "      <td>0.470507</td>\n",
       "      <td>Insulin resistance</td>\n",
       "      <td>0.746365</td>\n",
       "      <td>0.351169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5205</th>\n",
       "      <td>Proteoglycans in cancer</td>\n",
       "      <td>0.610077</td>\n",
       "      <td>Proteoglycans in cancer</td>\n",
       "      <td>0.428428</td>\n",
       "      <td>0.261374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4060</th>\n",
       "      <td>Cytokine-cytokine receptor interaction</td>\n",
       "      <td>0.253874</td>\n",
       "      <td>Cytokine-cytokine receptor interaction</td>\n",
       "      <td>0.875745</td>\n",
       "      <td>0.222329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4115</th>\n",
       "      <td>p53 signaling pathway</td>\n",
       "      <td>0.331166</td>\n",
       "      <td>p53 signaling pathway</td>\n",
       "      <td>0.622686</td>\n",
       "      <td>0.206212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4310</th>\n",
       "      <td>Wnt signaling pathway</td>\n",
       "      <td>0.413964</td>\n",
       "      <td>Wnt signaling pathway</td>\n",
       "      <td>0.429984</td>\n",
       "      <td>0.177998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>0.409587</td>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>0.424209</td>\n",
       "      <td>0.173751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4620</th>\n",
       "      <td>Toll-like receptor signaling pathway</td>\n",
       "      <td>0.180650</td>\n",
       "      <td>Toll-like receptor signaling pathway</td>\n",
       "      <td>0.935194</td>\n",
       "      <td>0.168943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>0.725467</td>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>0.226474</td>\n",
       "      <td>0.164300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4062</th>\n",
       "      <td>Chemokine signaling pathway</td>\n",
       "      <td>0.393130</td>\n",
       "      <td>Chemokine signaling pathway</td>\n",
       "      <td>0.390351</td>\n",
       "      <td>0.153458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4512</th>\n",
       "      <td>ECM-receptor interaction</td>\n",
       "      <td>0.193051</td>\n",
       "      <td>ECM-receptor interaction</td>\n",
       "      <td>0.773692</td>\n",
       "      <td>0.149362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4151</th>\n",
       "      <td>PI3K-Akt signaling pathway</td>\n",
       "      <td>0.349093</td>\n",
       "      <td>PI3K-Akt signaling pathway</td>\n",
       "      <td>0.387528</td>\n",
       "      <td>0.135283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4210</th>\n",
       "      <td>Apoptosis</td>\n",
       "      <td>0.239570</td>\n",
       "      <td>Apoptosis</td>\n",
       "      <td>0.546943</td>\n",
       "      <td>0.131031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5321</th>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>0.541599</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>0.210383</td>\n",
       "      <td>0.113943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4910</th>\n",
       "      <td>Insulin signaling pathway</td>\n",
       "      <td>0.269982</td>\n",
       "      <td>Insulin signaling pathway</td>\n",
       "      <td>0.358141</td>\n",
       "      <td>0.096691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>0.105384</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>0.698681</td>\n",
       "      <td>0.073630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4611</th>\n",
       "      <td>Platelet activation</td>\n",
       "      <td>0.232881</td>\n",
       "      <td>Platelet activation</td>\n",
       "      <td>0.280477</td>\n",
       "      <td>0.065318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4510</th>\n",
       "      <td>Focal adhesion</td>\n",
       "      <td>0.258699</td>\n",
       "      <td>Focal adhesion</td>\n",
       "      <td>0.248795</td>\n",
       "      <td>0.064363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4213</th>\n",
       "      <td>Longevity regulating pathway - multiple species</td>\n",
       "      <td>0.201630</td>\n",
       "      <td>Longevity regulating pathway - multiple species</td>\n",
       "      <td>0.314748</td>\n",
       "      <td>0.063463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5415</th>\n",
       "      <td>Diabetic cardiomyopathy</td>\n",
       "      <td>0.156905</td>\n",
       "      <td>Diabetic cardiomyopathy</td>\n",
       "      <td>0.148089</td>\n",
       "      <td>0.023236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4975</th>\n",
       "      <td>Fat digestion and absorption</td>\n",
       "      <td>0.087746</td>\n",
       "      <td>Fat digestion and absorption</td>\n",
       "      <td>0.056692</td>\n",
       "      <td>0.004975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4066</th>\n",
       "      <td>HIF-1 signaling pathway</td>\n",
       "      <td>0.021271</td>\n",
       "      <td>HIF-1 signaling pathway</td>\n",
       "      <td>0.163584</td>\n",
       "      <td>0.003480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4070</th>\n",
       "      <td>Phosphatidylinositol signaling system</td>\n",
       "      <td>0.554156</td>\n",
       "      <td>Phosphatidylinositol signaling system</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4970</th>\n",
       "      <td>Salivary secretion</td>\n",
       "      <td>0.187899</td>\n",
       "      <td>Salivary secretion</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4145</th>\n",
       "      <td>Phagosome</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Phagosome</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4974</th>\n",
       "      <td>Protein digestion and absorption</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Protein digestion and absorption</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4979</th>\n",
       "      <td>Cholesterol metabolism</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Cholesterol metabolism</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5146</th>\n",
       "      <td>Amoebiasis</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Amoebiasis</td>\n",
       "      <td>0.226558</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Pathway_feb   \n",
       "Pathway ID                                                    \n",
       "5417                              Lipid and atherosclerosis  \\\n",
       "4010                                 MAPK signaling pathway   \n",
       "4218                                    Cellular senescence   \n",
       "4931                                     Insulin resistance   \n",
       "5205                                Proteoglycans in cancer   \n",
       "4060                 Cytokine-cytokine receptor interaction   \n",
       "4115                                  p53 signaling pathway   \n",
       "4310                                  Wnt signaling pathway   \n",
       "240                                   Pyrimidine metabolism   \n",
       "4620                   Toll-like receptor signaling pathway   \n",
       "230                                       Purine metabolism   \n",
       "4062                            Chemokine signaling pathway   \n",
       "4512                               ECM-receptor interaction   \n",
       "4151                             PI3K-Akt signaling pathway   \n",
       "4210                                              Apoptosis   \n",
       "5321                             Inflammatory bowel disease   \n",
       "4910                              Insulin signaling pathway   \n",
       "10                             Glycolysis / Gluconeogenesis   \n",
       "4611                                    Platelet activation   \n",
       "4510                                         Focal adhesion   \n",
       "4213        Longevity regulating pathway - multiple species   \n",
       "5415                                Diabetic cardiomyopathy   \n",
       "4975                           Fat digestion and absorption   \n",
       "4066                                HIF-1 signaling pathway   \n",
       "4070                  Phosphatidylinositol signaling system   \n",
       "4970                                     Salivary secretion   \n",
       "4145                                              Phagosome   \n",
       "4974                       Protein digestion and absorption   \n",
       "4979                                 Cholesterol metabolism   \n",
       "5146                                             Amoebiasis   \n",
       "\n",
       "            Average square difference_feb   \n",
       "Pathway ID                                  \n",
       "5417                             0.625288  \\\n",
       "4010                             0.828898   \n",
       "4218                             0.609202   \n",
       "4931                             0.470507   \n",
       "5205                             0.610077   \n",
       "4060                             0.253874   \n",
       "4115                             0.331166   \n",
       "4310                             0.413964   \n",
       "240                              0.409587   \n",
       "4620                             0.180650   \n",
       "230                              0.725467   \n",
       "4062                             0.393130   \n",
       "4512                             0.193051   \n",
       "4151                             0.349093   \n",
       "4210                             0.239570   \n",
       "5321                             0.541599   \n",
       "4910                             0.269982   \n",
       "10                               0.105384   \n",
       "4611                             0.232881   \n",
       "4510                             0.258699   \n",
       "4213                             0.201630   \n",
       "5415                             0.156905   \n",
       "4975                             0.087746   \n",
       "4066                             0.021271   \n",
       "4070                             0.554156   \n",
       "4970                             0.187899   \n",
       "4145                             0.000000   \n",
       "4974                             0.000000   \n",
       "4979                             0.000000   \n",
       "5146                             0.000000   \n",
       "\n",
       "                                                Pathway_mar   \n",
       "Pathway ID                                                    \n",
       "5417                              Lipid and atherosclerosis  \\\n",
       "4010                                 MAPK signaling pathway   \n",
       "4218                                    Cellular senescence   \n",
       "4931                                     Insulin resistance   \n",
       "5205                                Proteoglycans in cancer   \n",
       "4060                 Cytokine-cytokine receptor interaction   \n",
       "4115                                  p53 signaling pathway   \n",
       "4310                                  Wnt signaling pathway   \n",
       "240                                   Pyrimidine metabolism   \n",
       "4620                   Toll-like receptor signaling pathway   \n",
       "230                                       Purine metabolism   \n",
       "4062                            Chemokine signaling pathway   \n",
       "4512                               ECM-receptor interaction   \n",
       "4151                             PI3K-Akt signaling pathway   \n",
       "4210                                              Apoptosis   \n",
       "5321                             Inflammatory bowel disease   \n",
       "4910                              Insulin signaling pathway   \n",
       "10                             Glycolysis / Gluconeogenesis   \n",
       "4611                                    Platelet activation   \n",
       "4510                                         Focal adhesion   \n",
       "4213        Longevity regulating pathway - multiple species   \n",
       "5415                                Diabetic cardiomyopathy   \n",
       "4975                           Fat digestion and absorption   \n",
       "4066                                HIF-1 signaling pathway   \n",
       "4070                  Phosphatidylinositol signaling system   \n",
       "4970                                     Salivary secretion   \n",
       "4145                                              Phagosome   \n",
       "4974                       Protein digestion and absorption   \n",
       "4979                                 Cholesterol metabolism   \n",
       "5146                                             Amoebiasis   \n",
       "\n",
       "            Average square difference_mar        SS  \n",
       "Pathway ID                                           \n",
       "5417                             1.962452  1.227097  \n",
       "4010                             0.869938  0.721090  \n",
       "4218                             0.690115  0.420420  \n",
       "4931                             0.746365  0.351169  \n",
       "5205                             0.428428  0.261374  \n",
       "4060                             0.875745  0.222329  \n",
       "4115                             0.622686  0.206212  \n",
       "4310                             0.429984  0.177998  \n",
       "240                              0.424209  0.173751  \n",
       "4620                             0.935194  0.168943  \n",
       "230                              0.226474  0.164300  \n",
       "4062                             0.390351  0.153458  \n",
       "4512                             0.773692  0.149362  \n",
       "4151                             0.387528  0.135283  \n",
       "4210                             0.546943  0.131031  \n",
       "5321                             0.210383  0.113943  \n",
       "4910                             0.358141  0.096691  \n",
       "10                               0.698681  0.073630  \n",
       "4611                             0.280477  0.065318  \n",
       "4510                             0.248795  0.064363  \n",
       "4213                             0.314748  0.063463  \n",
       "5415                             0.148089  0.023236  \n",
       "4975                             0.056692  0.004975  \n",
       "4066                             0.163584  0.003480  \n",
       "4070                             0.000000  0.000000  \n",
       "4970                             0.000000  0.000000  \n",
       "4145                             0.000000  0.000000  \n",
       "4974                             0.000000  0.000000  \n",
       "4979                             0.000000  0.000000  \n",
       "5146                             0.226558  0.000000  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "both"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7496cda1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qf",
   "language": "python",
   "name": "qf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
