{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d4364c55",
   "metadata": {},
   "outputs": [],
   "source": [
    "from hypergraphs import *\n",
    "from kegg import *\n",
    "import pandas as pd\n",
    "import os\n",
    "import logging\n",
    "\n",
    "from Bio.KEGG.KGML.KGML_parser import read\n",
    "import Bio.KEGG.KGML.KGML_pathway\n",
    "from ast import literal_eval\n",
    "from urllib.request import urlopen\n",
    "from urllib.parse import quote_plus\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "from Levenshtein import ratio\n",
    "import scipy.stats\n",
    "import pathlib\n",
    "from qf.cc import cardon_crochemore_colored, cardon_crochemore\n",
    "import random\n",
    "\n",
    "from reportlab.lib.units import inch\n",
    "import PIL.Image\n",
    "from io import BytesIO\n",
    "import reportlab.pdfgen.canvas\n",
    "import reportlab.lib.colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6b13e6bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "organismDict = {\"hsa\": \"Homo sapiens (human)\"}\n",
    "outputDirectory = \"/Users/boldi/Desktop/pw/\"\n",
    "inputDirectory = \"../../LaTeX/Data/KEGG-Pathways/Feb2024/\"\n",
    "dataDirectory = \"../../LaTeX/Data/KEGG-Pathways\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "87cb1b1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathwaylistFilename = \"KEGG-pathways.txt\"  # In inputDirectory, csv: ID, type, name\n",
    "superpathway = \"All\"                  # Name of the superpathway\n",
    "organism = \"hsa\"\n",
    "confidence = 0.05\n",
    "\n",
    "produceGraphsWithCCcoloring = False\n",
    "produceGraphsWithAnomalies = True\n",
    "useCCcoloringForAnomalies = False\n",
    "produceAnomaliesForNonsignificantPairs = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cb8e2d96",
   "metadata": {},
   "outputs": [],
   "source": [
    "safeSuperpathway = makesafe(superpathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e6caae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pws = pd.read_csv(os.path.join(inputDirectory, \"KEGG-pathways.txt\"), \n",
    "            names=[\"Pathway ID\", \"Pathway type\", \"Pathway name\"], dtype=\"string\", keep_default_na=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "199b2487",
   "metadata": {},
   "outputs": [],
   "source": [
    "superpathwayDict = {\n",
    "    superpathway: [(x,y) for x,y in zip(pws[\"Pathway ID\"], pws[\"Pathway name\"])]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "47a38c39",
   "metadata": {},
   "outputs": [],
   "source": [
    "def words2set(words):\n",
    "    \"\"\"\n",
    "        Given a space-separated list of words, returns it as a set of words\n",
    "    \"\"\"\n",
    "    return set(words.split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e0a74d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import NamedTuple\n",
    "\n",
    "class Relation(NamedTuple):\n",
    "    \"\"\"\n",
    "        Represents a relation. It is characterized by:\n",
    "        - a type (e.g. \"PPRel\") \n",
    "        - a list of subtypes (e.g. [('activation', '-->'), ('indirect effect', '..>')])\n",
    "        - a set of components (source)\n",
    "        - a list of sets of components (target)\n",
    "    \"\"\"\n",
    "    relType: str\n",
    "    relSubtypes: list\n",
    "    source: set\n",
    "    target: list\n",
    "    sourceg: Bio.KEGG.KGML.KGML_pathway.Graphics\n",
    "    targetg: Bio.KEGG.KGML.KGML_pathway.Graphics\n",
    "        \n",
    "    @classmethod\n",
    "    def fromKEGG(cls, relation, pathway):\n",
    "        \"\"\"\n",
    "            Constructor: builds a Relation from a KEGG relation\n",
    "        \"\"\"\n",
    "        if relation.entry2.name == \"undefined\":\n",
    "            s = []\n",
    "            for component in relation.entry2.components:\n",
    "                s += [words2set(pathway.entries[component.id].name)]\n",
    "            targetg = pathway.entries[component.id].graphics[0]\n",
    "        else:\n",
    "            s = [words2set(relation.entry2.name)]\n",
    "            targetg = relation.entry2.graphics[0]\n",
    "        return cls(relation.type, relation.subtypes, words2set(relation.entry1.name), s, targetg, relation.entry1.graphics[0])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ba5d2e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsFromKEGGpathway(pathway):\n",
    "    \"\"\"\n",
    "        Given a KEGG pathway, returns the list of relations it contains (represented as Relation).\n",
    "    \"\"\"\n",
    "    return [Relation.fromKEGG(r, pathway) for r in pathway.relations]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "751bc16c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pathwayFromFile(dataDirectory, organism, pathwayID):\n",
    "    \"\"\"\n",
    "        Load and return a KEGG pathway, given the data directory, the organism specifying the subdirectory,\n",
    "        and the pathwayID (the filename should be organism+pathwayID+\"xml\".\n",
    "    \"\"\"\n",
    "    return KGML_parser.read(open(os.path.join(dataDirectory, organism, organism+pathwayID+\".xml\"), \"r\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bd27fdc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsFromFile(dataDirectory, organism, pathwayID):\n",
    "    \"\"\"\n",
    "        Given a pathway file (specified by a root dataDirectory, the name of the organism subdirectory, and the name\n",
    "        of the file [organism+pathwayID.xml]), reads it and returns the list of its relations.\n",
    "    \"\"\"\n",
    "    pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "    return relationsFromKEGGpathway(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eaa01766",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsForSuperpathway(dataDirectory, organism, superpathwayDict, superpathway):\n",
    "    \"\"\"\n",
    "        Given a superpathDict (whose keys are superpathway names and whose values are list of\n",
    "        pairs (pathwayID, pathwayName)), reads all the pathway file for a specific superpathway name\n",
    "        and returns a dictionary whose keys are the pathwayID's and whose values are the list of relations.\n",
    "    \"\"\"\n",
    "    rel = {}\n",
    "    for k,v in superpathwayDict[superpathway]:\n",
    "        #logging.info(\"Reading\", k, v)\n",
    "        rel[k] = relationsFromFile(dataDirectory, organism, k)\n",
    "    return rel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "15479a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def subtypes(rel):\n",
    "    \"\"\"\n",
    "        Given a map whose values are relations, accumulate all subtypes appearing and assign them a number.\n",
    "        The result is a map from subtype string to number.\n",
    "    \"\"\"\n",
    "    all_subtypes = set([])\n",
    "    for pathwayid, relations in rel.items():\n",
    "        for relation in relations:\n",
    "            all_subtypes |= set([str(relation.relSubtypes)])\n",
    "        \n",
    "    s = sorted(list(all_subtypes))\n",
    "    subtype2color = {v:k for k,v in enumerate(s)}\n",
    "    color2subtype = {k:v for k,v in enumerate(s)}\n",
    "\n",
    "    return subtype2color, color2subtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1e229eaf-d667-466c-8238-8205257b0ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set2can(s):\n",
    "    \"\"\"\n",
    "        Convert a set to a string in a canonical way.\n",
    "    \"\"\"\n",
    "    return str(sorted(list(s)))\n",
    "\n",
    "def can2set(c):\n",
    "    \"\"\"\n",
    "        Does the converse of set2can.\n",
    "    \"\"\"\n",
    "    return set(literal_eval(c))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75b272b9-b220-492b-adf6-509f9952842f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indices(rel):\n",
    "    \"\"\"\n",
    "        Given a map whose values are lists of relations, it considers all the relations one by one, and attributes a unique id \n",
    "        to each element (i.e., set of components appearing as source or in the target of some relation) and block (the set of\n",
    "        element id appearing as target of some relation).\n",
    "        This function returns the dictionaries to move from/to an element or block to the corresponding id.\n",
    "        Elements are string representations of sorted lists of strings.\n",
    "        Blocks are string representations of sorted lists of ints.\n",
    "    \"\"\"\n",
    "    element2id = {}\n",
    "    id2element = {}\n",
    "    block2id = {}\n",
    "    id2block = {}\n",
    "\n",
    "    for relations in rel.values():\n",
    "        for relation in relations:\n",
    "            s = set2can(relation.source)\n",
    "            if s not in element2id.keys():\n",
    "                element2id[s] = len(element2id)\n",
    "            targetset = set([])\n",
    "            for targ in relation.target:\n",
    "                t = set2can(targ)\n",
    "                if t not in element2id.keys():\n",
    "                    element2id[t] = len(element2id)\n",
    "                targetset |= set([element2id[t]])\n",
    "            ts = set2can(targetset)\n",
    "            if ts not in block2id.keys():\n",
    "                block2id[ts] = len(block2id)\n",
    "\n",
    "    for k,v in element2id.items():\n",
    "        id2element[v] = k\n",
    "    for k,v in block2id.items():\n",
    "        id2block[v] = k\n",
    "    return element2id, id2element, block2id, id2block\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f9c3f128-cf97-4bb3-8ce5-65c49f865d98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_gene_KEGG(geneID):\n",
    "    \"\"\"\n",
    "        Search for gene on KEGG. \n",
    "        \n",
    "        Returns list of names.\n",
    "    \"\"\"\n",
    "    url1 = \"https://www.kegg.jp/entry/\"+geneID\n",
    "    response1 = requests.get(url1, allow_redirects=False)    \n",
    "    soup = BeautifulSoup(response1.text, \"html.parser\")\n",
    "    tds = [tds for tds in soup.find_all(\"td\", {\"class\": \"td11 defd\"})]\n",
    "    names = tds[0].getText().strip().split(\", \")\n",
    "    return names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1631e1d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_temp_imagefilename(url):\n",
    "    \"\"\"Return filename of temporary file containing downloaded image.\n",
    "\n",
    "    Create a new temporary file to hold the image file at the passed URL\n",
    "    and return the filename.\n",
    "    \"\"\"\n",
    "    img = urlopen(url).read()\n",
    "    im = PIL.Image.open(BytesIO(img))\n",
    "    # im.transpose(Image.FLIP_TOP_BOTTOM)\n",
    "    f = tempfile.NamedTemporaryFile(delete=False, suffix=\".png\")\n",
    "    fname = f.name\n",
    "    f.close()\n",
    "    im.save(fname, \"PNG\")\n",
    "    return fname\n",
    "\n",
    "\n",
    "#def enhance_method(klass, method_name, replacement):\n",
    "#    'replace a method with an enhanced version'\n",
    "#    method = getattr(klass, method_name)\n",
    "#    def enhanced(*args, **kwds): return replacement(*args, **kwds)\n",
    "#    setattr(klass, method_name, enhanced)\n",
    "\n",
    "def new_draw(self, filename, relations, elaborateArc = None):\n",
    "        \"\"\"Add the map elements to the drawing.\"\"\"\n",
    "        # Instantiate the drawing, first\n",
    "        # size x_max, y_max for now - we can add margins, later\n",
    "        if self.import_imagemap:\n",
    "            # We're drawing directly on the image, so we set the canvas to the\n",
    "            # same size as the image\n",
    "            if os.path.isfile(self.pathway.image):\n",
    "                imfilename = self.pathway.image\n",
    "            else:\n",
    "                imfilename = get_temp_imagefilename(self.pathway.image)\n",
    "            im = PIL.Image.open(imfilename)\n",
    "            cwidth, cheight = im.size\n",
    "        else:\n",
    "            # No image, so we set the canvas size to accommodate visible\n",
    "            # elements\n",
    "            cwidth, cheight = (self.pathway.bounds[1][0], self.pathway.bounds[1][1])\n",
    "        # Instantiate canvas\n",
    "        self.drawing = reportlab.pdfgen.canvas.Canvas(\n",
    "            filename,\n",
    "            bottomup=0,\n",
    "            pagesize=(\n",
    "                cwidth * (1 + 2 * self.margins[0]),\n",
    "                cheight * (1 + 2 * self.margins[1]),\n",
    "            ),\n",
    "        )\n",
    "        self.drawing.setFont(self.fontname, self.fontsize)\n",
    "        # Transform the canvas to add the margins\n",
    "        self.drawing.translate(\n",
    "            self.margins[0] * self.pathway.bounds[1][0],\n",
    "            self.margins[1] * self.pathway.bounds[1][1],\n",
    "        )\n",
    "        # Add the map image, if required\n",
    "        if self.import_imagemap:\n",
    "            self.drawing.saveState()\n",
    "            self.drawing.scale(1, -1)\n",
    "            self.drawing.translate(0, -cheight)\n",
    "            self.drawing.drawImage(imfilename, 0, 0)\n",
    "            self.drawing.restoreState()\n",
    "        # Add the reactions, compounds and maps\n",
    "        # Maps go on first, to be overlaid by more information.\n",
    "        # By default, they're slightly transparent.\n",
    "        if self.show_maps:\n",
    "            self.__add_maps()\n",
    "        if self.show_reaction_entries:\n",
    "            KGMLCanvas._KGMLCanvas__add_reaction_entries(self)\n",
    "        if self.show_orthologs:\n",
    "            KGMLCanvas._KGMLCanvas__add_orthologs(self)\n",
    "        if self.show_compounds:\n",
    "            KGMLCanvas._KGMLCanvas__add_compounds(self)\n",
    "        if self.show_genes:\n",
    "            KGMLCanvas._KGMLCanvas__add_genes(self)\n",
    "        # TODO: complete draw_relations code\n",
    "        # if self.draw_relations:\n",
    "        #    self.__add_relations()\n",
    "        # Write the pathway map to PDF\n",
    "        if elaborateArc is not None:\n",
    "            for relation in relations:\n",
    "                elaborateArc(relation, self.drawing)\n",
    "        self.drawing.save()\n",
    "        \n",
    "#enhance_method(KGMLCanvas, 'draw',  new_draw)\n",
    "setattr(KGMLCanvas, 'new_draw', new_draw)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0a11bbae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build all data needed for the relations of the superpathway\n",
    "rel = relationsForSuperpathway(dataDirectory, organism, superpathwayDict, superpathway)\n",
    "element2id, id2element, block2id, id2block = indices(rel)\n",
    "subtype2color, color2subtype = subtypes(rel)\n",
    "elements = list(set.union(*[can2set(k) for k in element2id.keys()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a1be8c2d-c5da-4e04-bf67-7e43b1a71b7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading gene names\n"
     ]
    }
   ],
   "source": [
    "# name2gene is a function mapping names to gene IDs (hsa:xxxx)\n",
    "# e.g. KCNE3, BRGDA6 etc are all mapped to hsa:10008 (see https://www.kegg.jp/entry/hsa:10008)\n",
    "# This map was produced by scraping kegg, but this has been done only once.\n",
    "# After that, we just read a pkl file\n",
    "\n",
    "name2keggFilename = os.path.join(inputDirectory, \"name2kegg.pkl\")\n",
    "\n",
    "if os.path.exists(name2keggFilename):\n",
    "    print(\"Reading gene names\")\n",
    "    with open(name2keggFilename, \"rb\") as handle:\n",
    "        name2gene = pickle.load(handle)\n",
    "else:\n",
    "    print(\"Producing gene names\")\n",
    "    d = {}\n",
    "    nelements = len(elements)\n",
    "    for count,element in enumerate(elements):\n",
    "        if count % 10 == 0:\n",
    "            print(f\"{count}/{nelements}\")\n",
    "        if element.startswith(\"hsa\"):\n",
    "            names = search_gene_KEGG(element)\n",
    "            for name in names:\n",
    "                d[name] = element\n",
    "    with open(name2keggFilename, \"wb\") as handle:\n",
    "        pickle.dump(d, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "    name2gene = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "74d19326",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading patient data\n"
     ]
    }
   ],
   "source": [
    "# Read patient data\n",
    "print(\"Reading patient data\")\n",
    "genes = pd.read_csv(os.path.join(inputDirectory, \"genes.csv\"), dtype={\"name\": str})\n",
    "patients = genes[\"name\"].values.tolist()\n",
    "classification = pd.read_csv(os.path.join(inputDirectory, \"classification.csv\"), dtype={\"name\": str})\n",
    "\n",
    "patient2class = {patient: classification.loc[classification[\"name\"]==patient][\"classification\"].values.flatten().tolist()[0] for patient in patients}\n",
    "numHealthy = sum([x == 'healthy' for x in patient2class.values()])\n",
    "numDiseased = sum([x == 'diseased' for x in patient2class.values()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "34acce33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dividing genes into two frames (one for healthy individuals and one for diseased individuals)\n",
    "genesHealthy = genes.loc[genes[\"name\"].apply(lambda x: patient2class[x])==\"healthy\"]\n",
    "genesDiseased = genes.loc[genes[\"name\"].apply(lambda x: patient2class[x])==\"diseased\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1c90f901-1413-4356-bfa8-faf25fc494df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building the relation graph\n",
      "- Edge (0,1) of type [('compound', 94)] for ['00010', '00010'] appears as [('compound', 93)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 93)]\n",
      "- Edge (0,3) of type [('compound', 94)] for ['00010'] appears as [('compound', 93)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 93)]\n",
      "- Edge (0,3) of type [('compound', 93)] for ['00010', '00010'] appears as [('compound', 94)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 94)]\n",
      "- Edge (22,30) of type [('compound', 90)] for ['00010'] appears as [('compound', 91)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 91)]\n",
      "- Edge (30,0) of type [('compound', 93)] for ['00010', '00010'] appears as [('compound', 94)] for 00010. \n",
      "****Conflict! Rewriting as [('compound', 94)]\n",
      "- Edge (59,62) of type [('compound', 330)] for ['00230'] appears as [('compound', 392)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 392)]\n",
      "- Edge (67,63) of type [('compound', 355)] for ['00230'] appears as [('compound', 343)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 343)]\n",
      "- Edge (73,63) of type [('compound', 392)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (71,36) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (71,60) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (48,49) of type [('compound', 377)] for ['00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (59,63) of type [('compound', 392)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (62,63) of type [('compound', 392)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (67,72) of type [('compound', 343)] for ['00230', '00230', '00230', '00230'] appears as [('compound', 332)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 332)]\n",
      "- Edge (67,63) of type [('compound', 343)] for ['00230', '00230'] appears as [('compound', 332)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 332)]\n",
      "- Edge (42,36) of type [('compound', 330)] for ['00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (42,64) of type [('compound', 330)] for ['00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (36,64) of type [('compound', 330)] for ['00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (64,80) of type [('compound', 331)] for ['00230'] appears as [('compound', 326)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 326)]\n",
      "- Edge (64,66) of type [('compound', 331)] for ['00230'] appears as [('compound', 326)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 326)]\n",
      "- Edge (80,66) of type [('compound', 331)] for ['00230'] appears as [('compound', 326)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 326)]\n",
      "- Edge (36,63) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (36,64) of type [('compound', 325)] for ['00230', '00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (60,63) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (60,64) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (63,64) of type [('compound', 330)] for ['00230'] appears as [('compound', 365)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 365)]\n",
      "- Edge (64,66) of type [('compound', 326)] for ['00230', '00230'] appears as [('compound', 366)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 366)]\n",
      "- Edge (66,63) of type [('compound', 332)] for ['00230'] appears as [('compound', 355)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 355)]\n",
      "- Edge (67,65) of type [('compound', 343)] for ['00230'] appears as [('compound', 355)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 355)]\n",
      "- Edge (35,64) of type [('compound', 365)] for ['00230'] appears as [('compound', 357)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 357)]\n",
      "- Edge (35,38) of type [('compound', 362)] for ['00230'] appears as [('compound', 360)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 360)]\n",
      "- Edge (40,38) of type [('compound', 362)] for ['00230'] appears as [('compound', 360)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 360)]\n",
      "- Edge (36,83) of type [('compound', 380)] for ['00230'] appears as [('compound', 325)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 325)]\n",
      "- Edge (67,84) of type [('compound', 320)] for ['00230'] appears as [('compound', 356)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 356)]\n",
      "- Edge (36,83) of type [('compound', 325)] for ['00230', '00230'] appears as [('compound', 390)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 390)]\n",
      "- Edge (38,83) of type [('compound', 380)] for ['00230'] appears as [('compound', 390)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 390)]\n",
      "- Edge (55,83) of type [('compound', 380)] for ['00230'] appears as [('compound', 390)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 390)]\n",
      "- Edge (36,40) of type [('compound', 362)] for ['00230'] appears as [('compound', 324)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 324)]\n",
      "- Edge (36,38) of type [('compound', 362)] for ['00230'] appears as [('compound', 324)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 324)]\n",
      "- Edge (36,36) of type [('compound', 362)] for ['00230'] appears as [('compound', 324)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 324)]\n",
      "- Edge (40,38) of type [('compound', 360)] for ['00230', '00230'] appears as [('compound', 324)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 324)]\n",
      "- Edge (40,36) of type [('compound', 362)] for ['00230'] appears as [('compound', 324)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 324)]\n",
      "- Edge (40,38) of type [('compound', 324)] for ['00230', '00230', '00230'] appears as [('compound', 322)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 322)]\n",
      "- Edge (68,64) of type [('compound', 325)] for ['00230'] appears as [('compound', 321)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 321)]\n",
      "- Edge (67,64) of type [('compound', 367)] for ['00230'] appears as [('compound', 320)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 320)]\n",
      "- Edge (59,36) of type [('compound', 330)] for ['00230'] appears as [('compound', 368)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 368)]\n",
      "- Edge (38,36) of type [('compound', 324)] for ['00230'] appears as [('compound', 369)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 369)]\n",
      "- Edge (36,36) of type [('compound', 324)] for ['00230', '00230'] appears as [('compound', 369)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 369)]\n",
      "- Edge (71,63) of type [('compound', 365)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (71,64) of type [('compound', 365)] for ['00230'] appears as [('compound', 330)] for 00230. \n",
      "****Conflict! Rewriting as [('compound', 330)]\n",
      "- Edge (38,36) of type [('compound', 369)] for ['00230', '00230'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (36,40) of type [('compound', 324)] for ['00230', '00230'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (36,64) of type [('compound', 365)] for ['00230', '00230', '00230'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (38,93) of type [('compound', 217)] for ['00240'] appears as [('compound', 194)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 194)]\n",
      "- Edge (36,91) of type [('compound', 191)] for ['00240'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (38,36) of type [('compound', 214)] for ['00230', '00230', '00240'] appears as [('compound', 194)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 194)]\n",
      "- Edge (90,91) of type [('compound', 191)] for ['00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (36,36) of type [('compound', 369)] for ['00230', '00230', '00230'] appears as [('compound', 194)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 194)]\n",
      "- Edge (38,91) of type [('compound', 191)] for ['00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (36,91) of type [('compound', 213)] for ['00240', '00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (94,95) of type [('compound', 212)] for ['00240'] appears as [('compound', 188)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 188)]\n",
      "- Edge (64,95) of type [('compound', 212)] for ['00240'] appears as [('compound', 188)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 188)]\n",
      "- Edge (36,91) of type [('compound', 185)] for ['00240', '00240', '00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (36,64) of type [('compound', 213)] for ['00230', '00230', '00230', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (92,91) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (94,91) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (64,91) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (100,102) of type [('compound', 210)] for ['00240'] appears as [('compound', 202)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 202)]\n",
      "- Edge (92,64) of type [('compound', 213)] for ['00240'] appears as [('compound', 221)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 221)]\n",
      "- Edge (102,103) of type [('compound', 209)] for ['00240'] appears as [('compound', 203)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 203)]\n",
      "- Edge (36,91) of type [('compound', 187)] for ['00240', '00240', '00240', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (64,95) of type [('compound', 188)] for ['00240', '00240'] appears as [('compound', 220)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 220)]\n",
      "- Edge (98,99) of type [('compound', 196)] for ['00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (92,104) of type [('compound', 221)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (89,64) of type [('compound', 195)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (93,64) of type [('compound', 195)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (93,98) of type [('compound', 195)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (64,105) of type [('compound', 221)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (64,95) of type [('compound', 220)] for ['00240', '00240', '00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (64,96) of type [('compound', 188)] for ['00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (106,106) of type [('compound', 181)] for ['00240'] appears as [('compound', 180)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 180)]\n",
      "- Edge (64,99) of type [('compound', 196)] for ['00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (95,84) of type [('compound', 220)] for ['00240'] appears as [('compound', 219)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 219)]\n",
      "- Edge (105,84) of type [('compound', 221)] for ['00240'] appears as [('compound', 218)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 218)]\n",
      "- Edge (36,94) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (89,84) of type [('compound', 218)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (92,94) of type [('compound', 213)] for ['00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (93,84) of type [('compound', 218)] for ['00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (64,83) of type [('compound', 365)] for ['00230'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (36,64) of type [('compound', 195)] for ['00230', '00230', '00230', '00240', '00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (99,84) of type [('compound', 219)] for ['00240'] appears as [('compound', 196)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 196)]\n",
      "- Edge (60,64) of type [('compound', 365)] for ['00230', '00230'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (83,36) of type [('compound', 365)] for ['00230'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (92,64) of type [('compound', 221)] for ['00240', '00240'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (36,83) of type [('compound', 390)] for ['00230', '00230', '00230'] appears as [('compound', 187)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 187)]\n",
      "- Edge (38,83) of type [('compound', 390)] for ['00230', '00230'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (83,64) of type [('compound', 325)] for ['00230'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (83,36) of type [('compound', 191)] for ['00230', '00240'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (38,92) of type [('compound', 214)] for ['00240'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (38,40) of type [('compound', 217)] for ['00240'] appears as [('compound', 222)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 222)]\n",
      "- Edge (36,60) of type [('compound', 369)] for ['00230'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (38,92) of type [('compound', 186)] for ['00240', '00240'] appears as [('compound', 222)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 222)]\n",
      "- Edge (38,60) of type [('compound', 369)] for ['00230'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (36,36) of type [('compound', 194)] for ['00230', '00230', '00230', '00240'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (94,83) of type [('compound', 187)] for ['00240'] appears as [('compound', 213)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 213)]\n",
      "- Edge (38,36) of type [('compound', 194)] for ['00230', '00230', '00240', '00240'] appears as [('compound', 186)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 186)]\n",
      "- Edge (38,52) of type [('compound', 380)] for ['00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (36,52) of type [('compound', 380)] for ['00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (36,83) of type [('compound', 187)] for ['00230', '00230', '00230', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (83,52) of type [('compound', 380)] for ['00230'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (83,64) of type [('compound', 213)] for ['00230', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (38,83) of type [('compound', 191)] for ['00230', '00230', '00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (38,83) of type [('compound', 185)] for ['00230', '00230', '00240', '00240'] appears as [('compound', 193)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 193)]\n",
      "- Edge (36,83) of type [('compound', 195)] for ['00230', '00230', '00230', '00240', '00240'] appears as [('compound', 185)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 185)]\n",
      "- Edge (38,91) of type [('compound', 185)] for ['00240', '00240'] appears as [('compound', 193)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 193)]\n",
      "- Edge (36,83) of type [('compound', 185)] for ['00230', '00230', '00230', '00240', '00240', '00240'] appears as [('compound', 193)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 193)]\n",
      "- Edge (36,91) of type [('compound', 195)] for ['00240', '00240', '00240', '00240', '00240'] appears as [('compound', 193)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 193)]\n",
      "- Edge (90,36) of type [('compound', 185)] for ['00240'] appears as [('compound', 191)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 191)]\n",
      "- Edge (64,91) of type [('compound', 187)] for ['00240', '00240'] appears as [('compound', 195)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 195)]\n",
      "- Edge (36,36) of type [('compound', 186)] for ['00230', '00230', '00230', '00240', '00240'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (38,40) of type [('compound', 222)] for ['00240', '00240'] appears as [('compound', 214)] for 00240. \n",
      "****Conflict! Rewriting as [('compound', 214)]\n",
      "- Edge (487,489) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04062. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (487,177) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04062. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (129,161) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04062. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (161,204) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->')] for 04066. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (484,183) of type [('compound', 151), ('activation', '-->')] for ['04062'] appears as [('activation', '-->')] for 04066. Overwriting with type [('compound', 151), ('activation', '-->')]\n",
      "- Edge (569,573) of type [('compound', 48)] for ['04070'] appears as [('compound', 49)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 49)]\n",
      "- Edge (579,576) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (573,576) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (569,576) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (590,591) of type [('compound', 52)] for ['04070'] appears as [('compound', 235)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 235)]\n",
      "- Edge (591,590) of type [('compound', 234)] for ['04070'] appears as [('compound', 68)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 68)]\n",
      "- Edge (569,573) of type [('compound', 49)] for ['04070', '04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (587,579) of type [('compound', 66)] for ['04070'] appears as [('compound', 73)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 73)]\n",
      "- Edge (571,570) of type [('compound', 49)] for ['04070'] appears as [('compound', 66)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 66)]\n",
      "- Edge (575,583) of type [('compound', 61)] for ['04070'] appears as [('compound', 56)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 56)]\n",
      "- Edge (579,592) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (592,573) of type [('compound', 72)] for ['04070'] appears as [('compound', 75)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 75)]\n",
      "- Edge (593,592) of type [('compound', 75)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (593,574) of type [('compound', 75)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (484,579) of type [('compound', 73)] for ['04070'] appears as [('compound', 66)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 66)]\n",
      "- Edge (583,583) of type [('compound', 60)] for ['04070', '04070'] appears as [('compound', 231)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 231)]\n",
      "- Edge (581,592) of type [('compound', 75)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (586,587) of type [('compound', 66)] for ['04070'] appears as [('compound', 74)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 74)]\n",
      "- Edge (579,569) of type [('compound', 75)] for ['04070'] appears as [('compound', 72)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 72)]\n",
      "- Edge (569,573) of type [('compound', 75)] for ['04070', '04070', '04070'] appears as [('compound', 72)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 72)]\n",
      "- Edge (577,575) of type [('compound', 61)] for ['04070'] appears as [('compound', 56)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 56)]\n",
      "- Edge (580,577) of type [('compound', 62)] for ['04070'] appears as [('compound', 55)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 55)]\n",
      "- Edge (578,580) of type [('compound', 55)] for ['04070'] appears as [('compound', 63)] for 04070. \n",
      "****Conflict! Rewriting as [('compound', 63)]\n",
      "- Edge (176,625) of type [('expression', '-->')] for ['04115'] appears as [('activation', '-->')] for 04115. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (496,499) of type [] for ['04062'] appears as [('activation', '-->')] for 04151. Overwriting with type [('activation', '-->')]\n",
      "- Edge (183,114) of type [('phosphorylation', '+p'), ('indirect effect', '..>')] for ['04062'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (149,150) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04151. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (150,505) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062'] appears as [('activation', '-->')] for 04151. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (536,161) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (540,561) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (561,562) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04151. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (113,114) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04210. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (118,119) of type [] for ['04010'] appears as [('activation', '-->')] for 04210. Overwriting with type [('activation', '-->')]\n",
      "- Edge (540,552) of type [('inhibition', '--|')] for ['04066'] appears as [('inhibition', '--|'), ('phosphorylation', '+p')] for 04218. Overwriting with type [('inhibition', '--|'), ('phosphorylation', '+p')]\n",
      "- Edge (579,499) of type [('inhibition', '--|'), ('dephosphorylation', '-p')] for ['04151'] appears as [('inhibition', '--|')] for 04218. Overwriting with type [('inhibition', '--|'), ('dephosphorylation', '-p')]\n",
      "- Edge (728,488) of type [('activation', '-->')] for ['04213'] appears as [('inhibition', '--|')] for 04218. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (176,228) of type [('expression', '-->')] for ['04115', '04115'] appears as [('activation', '-->')] for 04218. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (214,600) of type [('expression', '-->'), ('indirect effect', '..>')] for ['04218'] appears as [('expression', '-->')] for 04218. Overwriting with type [('expression', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (189,176) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04218. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (796,780) of type [('activation', '-->')] for ['04310'] appears as [('inhibition', '--|')] for 04310. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (176,634) of type [('expression', '-->')] for ['04115', '04115'] appears as [('indirect effect', '..>')] for 04310. \n",
      "****Conflict! Rewriting as [('indirect effect', '..>')]\n",
      "- Edge (181,769) of type [('activation', '-->'), ('dephosphorylation', '-p')] for ['04218'] appears as [('dephosphorylation', '-p')] for 04310. Overwriting with type [('activation', '-->'), ('dephosphorylation', '-p')]\n",
      "- Edge (493,177) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04510. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (483,163) of type [('activation', '-->')] for ['04310'] appears as [('phosphorylation', '+p'), ('indirect effect', '..>')] for 04510. \n",
      "****Conflict! Rewriting as [('phosphorylation', '+p'), ('indirect effect', '..>')]\n",
      "- Edge (484,183) of type [('compound', 151), ('activation', '-->')] for ['04062', '04066'] appears as [('compound', 24)] for 04510. \n",
      "****Conflict! Rewriting as [('compound', 24)]\n",
      "- Edge (511,149) of type [('binding/association', '---')] for ['04062'] appears as [('activation', '-->')] for 04510. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (649,183) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04210'] appears as [('phosphorylation', '+p')] for 04510. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (494,493) of type [('binding/association', '---')] for ['04062'] appears as [('phosphorylation', '+p')] for 04510. \n",
      "****Conflict! Rewriting as [('phosphorylation', '+p')]\n",
      "- Edge (163,697) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04210'] appears as [('phosphorylation', '+p')] for 04510. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (129,161) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04010', '04062'] appears as [('phosphorylation', '+p')] for 04510. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (161,527) of type [('expression', '-->'), ('indirect effect', '..>')] for ['04210'] appears as [('phosphorylation', '+p')] for 04510. \n",
      "****Conflict! Rewriting as [('phosphorylation', '+p')]\n",
      "- Edge (856,709) of type [('binding/association', '---')] for ['04510'] appears as [('activation', '-->')] for 04611. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (484,183) of type [('compound', 24)] for ['04062', '04066', '04510'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 04611. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (161,202) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 04611. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (711,492) of type [('dissociation', '-+-')] for ['04210'] appears as [('inhibition', '--|')] for 04620. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (115,116) of type [('activation', '-->')] for ['04010'] appears as [('binding/association', '---')] for 04620. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (536,161) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04151', '04210', '04218'] appears as [('phosphorylation', '+p')] for 04620. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (772,189) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04218'] appears as [('phosphorylation', '+p')] for 04620. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (183,540) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (561,562) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04151'] appears as [('activation', '-->')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (484,649) of type [('compound', 24)] for ['04510'] appears as [('compound', 71), ('activation', '-->')] for 04910. \n",
      "****Conflict! Rewriting as [('compound', 71), ('activation', '-->')]\n",
      "- Edge (150,505) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062', '04151'] appears as [('activation', '-->')] for 04910. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (161,204) of type [('activation', '-->')] for ['04010', '04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (161,154) of type [('phosphorylation', '+p')] for ['04010', '04510'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (204,553) of type [('activation', '-->')] for ['04066'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (741,726) of type [('activation', '-->')] for ['04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04910. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (711,492) of type [('inhibition', '--|')] for ['04210', '04620'] appears as [('dissociation', '-+-')] for 04931. \n",
      "****Conflict! Rewriting as [('dissociation', '-+-')]\n",
      "- Edge (555,556) of type [('activation', '-->')] for ['04611'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 04931. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (555,556) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04611', '04931'] appears as [('activation', '-->')] for 04931. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (741,685) of type [('activation', '-->')] for ['04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (685,484) of type [('activation', '-->')] for ['04151', '04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (685,484) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04213', '04931'] appears as [('activation', '-->')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (954,711) of type [('phosphorylation', '+p')] for ['04620'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (484,481) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062'] appears as [('activation', '-->')] for 04931. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (685,484) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04213', '04931', '04931'] appears as [('activation', '-->')] for 04931. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (1024,685) of type [('inhibition', '--|')] for ['04931'] appears as [('inhibition', '--|'), ('indirect effect', '..>')] for 04931. Overwriting with type [('inhibition', '--|'), ('indirect effect', '..>')]\n",
      "- Edge (804,184) of type [('compound', 12)] for ['04310'] appears as [('compound', 277), ('activation', '-->')] for 05146. \n",
      "****Conflict! Rewriting as [('compound', 277), ('activation', '-->')]\n",
      "- Edge (1089,161) of type [('activation', '-->'), ('indirect effect', '..>')] for ['05205'] appears as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')]\n",
      "- Edge (528,1116) of type [('activation', '-->')] for ['05205'] appears as [('inhibition', '--|')] for 05205. \n",
      "****Conflict! Rewriting as [('inhibition', '--|')]\n",
      "- Edge (891,892) of type [('compound', 131)] for ['04512'] appears as [('binding/association', '---')] for 05205. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (516,517) of type [('activation', '-->')] for ['04062', '04510', '04611'] appears as [('binding/association', '---')] for 05205. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (484,183) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04062', '04066', '04510', '04611', '04620'] appears as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p'), ('indirect effect', '..>')]\n",
      "- Edge (517,836) of type [('inhibition', '--|'), ('phosphorylation', '+p')] for ['04510'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (596,734) of type [('activation', '-->')] for ['04213'] appears as [('binding/association', '---')] for 05205. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (161,154) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04010', '04510', '04910'] appears as [('phosphorylation', '+p')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (734,484) of type [('activation', '-->')] for ['04213'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (183,540) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04910'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (561,562) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04066', '04151', '04910'] appears as [('activation', '-->')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (561,679) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151'] appears as [('activation', '-->')] for 05205. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (484,649) of type [('compound', 71), ('activation', '-->')] for ['04510', '04910'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (649,561) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04910'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05205. \n",
      "****Conflict! Rewriting as [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (546,492) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04066'] appears as [('activation', '-->')] for 05321. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (555,556) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04611', '04931', '04931'] appears as [('activation', '-->')] for 05415. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "- Edge (484,499) of type [('activation', '-->')] for ['04151', '04210', '04213', '04213', '04218', '04931'] appears as [('missing interaction', '-/-')] for 05415. \n",
      "****Conflict! Rewriting as [('missing interaction', '-/-')]\n",
      "- Edge (685,484) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04213', '04931', '04931', '04931'] appears as [('missing interaction', '-/-')] for 05415. \n",
      "****Conflict! Rewriting as [('missing interaction', '-/-')]\n",
      "- Edge (183,665) of type [('inhibition', '--|'), ('phosphorylation', '+p')] for ['04151', '04510', '04910', '04931'] appears as [('missing interaction', '-/-')] for 05415. \n",
      "****Conflict! Rewriting as [('missing interaction', '-/-')]\n",
      "- Edge (116,114) of type [('phosphorylation', '+p')] for ['04010'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (1126,516) of type [('binding/association', '---')] for ['05205'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (956,163) of type [('phosphorylation', '+p')] for ['04620'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (484,499) of type [('missing interaction', '-/-')] for ['04151', '04210', '04213', '04213', '04218', '04931', '05415'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (124,115) of type [('binding/association', '---')] for ['04010'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (183,665) of type [('missing interaction', '-/-')] for ['04151', '04510', '04910', '04931', '05415'] appears as [('inhibition', '--|'), ('phosphorylation', '+p')] for 05417. \n",
      "****Conflict! Rewriting as [('inhibition', '--|'), ('phosphorylation', '+p')]\n",
      "- Edge (513,773) of type [('activation', '-->')] for ['04218', '04970'] appears as [('binding/association', '---')] for 05417. \n",
      "****Conflict! Rewriting as [('binding/association', '---')]\n",
      "- Edge (492,555) of type [('activation', '-->'), ('indirect effect', '..>')] for ['04931'] appears as [('repression', '--|')] for 05417. \n",
      "****Conflict! Rewriting as [('repression', '--|')]\n",
      "- Edge (492,1194) of type [('expression', '-->')] for ['05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,1196) of type [('expression', '-->')] for ['05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,1197) of type [('expression', '-->')] for ['05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,1199) of type [('expression', '-->')] for ['05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,254) of type [('expression', '-->')] for ['05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,256) of type [('expression', '-->')] for ['04620', '05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (724,725) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04210'] appears as [('activation', '-->')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (492,1200) of type [('expression', '-->')] for ['05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,145) of type [('expression', '-->')] for ['05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,353) of type [('expression', '-->')] for ['04620', '05417', '05417', '05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,402) of type [('expression', '-->')] for ['04620', '04931', '04931', '05417', '05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (492,142) of type [('expression', '-->')] for ['04620', '04931', '04931', '05321', '05417', '05417'] appears as [('activation', '-->')] for 05417. \n",
      "****Conflict! Rewriting as [('activation', '-->')]\n",
      "- Edge (649,183) of type [('activation', '-->'), ('phosphorylation', '+p')] for ['04151', '04210', '04510', '04910', '04931'] appears as [('activation', '-->')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (163,171) of type [('phosphorylation', '+p')] for ['04010', '04620'] appears as [('activation', '-->'), ('phosphorylation', '+p')] for 05417. Overwriting with type [('activation', '-->'), ('phosphorylation', '+p')]\n",
      "- Edge (252,606) of type [('indirect effect', '..>')] for ['04115'] appears as [('activation', '-->'), ('indirect effect', '..>')] for 05417. Overwriting with type [('activation', '-->'), ('indirect effect', '..>')]\n",
      "Wrap: more than 148 colors needed\n",
      "Wrap: more than 148 colors needed\n",
      "Wrap: more than 148 colors needed\n",
      "Wrap: more than 148 colors needed\n"
     ]
    }
   ],
   "source": [
    "# Building the graph of all relations contained in rel values\n",
    "# Nodes are IDs, and node x correspond to the set of genes (and/or compound) id2element[x]\n",
    "print(\"Building the relation graph\")\n",
    "G = nx.DiGraph() \n",
    "for pathwayID, relations in rel.items():\n",
    "    for relation in relations:\n",
    "        source = element2id[set2can(relation.source)]\n",
    "        for targ in relation.target:\n",
    "            target = element2id[set2can(targ)]\n",
    "            newType = str(relation.relSubtypes)\n",
    "            color = subtype2color[newType]\n",
    "            pl = []\n",
    "            if G.has_edge(source, target):\n",
    "                d = G.get_edge_data(source, target)\n",
    "                if d[\"color\"] != color:\n",
    "                    oldType = color2subtype[d[\"color\"]]\n",
    "                    print(\"- Edge ({},{}) of type {} for {} appears as {} for {}. \".format(\n",
    "                        source, target, color2subtype[d[\"color\"]], d[\"pathwayIDs\"], \n",
    "                        str(relation.relSubtypes), pathwayID\n",
    "                    ), end=\"\")\n",
    "                    if oldType[1:-1] in newType[1:-1]:\n",
    "                        print(\"Overwriting with type {}\".format(newType))\n",
    "                    elif newType[1:-1] in oldType[1:-1]:\n",
    "                        print(\"Overwriting with type {}\".format(oldType))\n",
    "                        color = subtype2color[oldType]\n",
    "                    else:\n",
    "                        print(\"\\n****Conflict! Rewriting as {}\".format(newType))\n",
    "                pl = d[\"pathwayIDs\"]\n",
    "            G.add_edge(source, target, color=color, pathwayIDs=pl + [pathwayID])\n",
    "            \n",
    "\n",
    "# Computing the minimum fibres\n",
    "cc=cardon_crochemore_colored(G)\n",
    "\n",
    "colors = list(matplotlib.colors.CSS4_COLORS.values())\n",
    "usedColors = 0\n",
    "\n",
    "cc2fibresize = {}\n",
    "cc2fibre = {}\n",
    "ccnontrivial = {}\n",
    "nontrivialclass2col = {}\n",
    "\n",
    "fibres = []\n",
    "node2fibre = {}\n",
    "for v in set(cc.values()):\n",
    "    fibre = [x for x in G.nodes() if cc[x] == v]\n",
    "    fibres += [fibre]\n",
    "    for x in fibre:\n",
    "        node2fibre[x] = fibre\n",
    "    cc2fibre[v] = fibre\n",
    "    cc2fibresize[v] = len(fibre)\n",
    "    if len(fibre) > 0:\n",
    "        ccnontrivial[v] = cc[v]\n",
    "        nontrivialclass2col[v] = colors[usedColors]\n",
    "        usedColors = (usedColors + 1 ) % len(colors)\n",
    "        if usedColors == 0:\n",
    "            print(\"Wrap: more than\", len(colors), \"colors needed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d51dbd7f-35b9-444a-9238-3af727c06411",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping genes to columns\n"
     ]
    }
   ],
   "source": [
    "# Build node2columns, mapping each node of G to the list of genes corresponding to it\n",
    "# for which we know a value (in the genes dataframe), sorted\n",
    "#\n",
    "\n",
    "print(\"Mapping genes to columns\")\n",
    "namedGenes = set(name2gene.values())     \n",
    "columns = set(genes.columns.values.tolist()[2:])\n",
    "node2columns = {}\n",
    "for node in G.nodes():\n",
    "    nodeGenes = can2set(id2element[node])\n",
    "    s = []\n",
    "    for gene in nodeGenes & namedGenes:\n",
    "        for k,v in name2gene.items():\n",
    "            if v == gene:\n",
    "                if k in columns:\n",
    "                    s += [k]\n",
    "    node2columns[node]=sorted(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "05b96661",
   "metadata": {},
   "outputs": [],
   "source": [
    "if produceGraphsWithCCcoloring:\n",
    "    s = \"\"\n",
    "    for pair in superpathwayDict[superpathway]:\n",
    "        pathwayID = pair[0]\n",
    "        pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "        canvas = KGMLCanvas(pathway)\n",
    "        for k in pathway.entries:\n",
    "            name = pathway.entries[k].name\n",
    "            canonicalName = set2can(words2set(name))\n",
    "            if canonicalName in element2id.keys():\n",
    "                node = element2id[canonicalName]\n",
    "                klass = cc[node]\n",
    "                if klass in nontrivialclass2col:\n",
    "                    pathway.entries[k].graphics[0].bgcolor = nontrivialclass2col[klass]\n",
    "                    pathway.entries[k].graphics[0].fgcolor = nontrivialclass2col[klass]\n",
    "                else:\n",
    "                    pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "                    pathway.enries[k].graphics[0].fgcolor = \"#FFFFFF\"\n",
    "                s += pathwayID + \"\\t\" + str(node) + \"\\t\" + str(klass) +\"\\t\"+pathway.entries[k].graphics[0].name+\"\\t\"+canonicalName + \"\\n\"\n",
    "        canvas.import_imagemap = True\n",
    "        pdfName = organism + pathwayID + \".pdf\"\n",
    "        pathlib.Path(os.path.join(outputDirectory, organism, safeSuperpathway)).mkdir(parents=True, exist_ok=True)\n",
    "        canvas.draw(os.path.join(outputDirectory, organism, safeSuperpathway, pdfName))\n",
    "    with open(os.path.join(outputDirectory, organism, safeSuperpathway, \"all.tsv\"), \"wt\") as file:\n",
    "        file.write(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "73ce2fd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def columns2nodes(columns, node2columns, ignoreEmpty = True):\n",
    "    \"\"\"\n",
    "        Returns a map from columns to the nodes where the column appears.\n",
    "    \"\"\"\n",
    "    res = {} \n",
    "    for c in columns:\n",
    "        s = set([])\n",
    "        for no, co in node2columns.items():\n",
    "            if c in co:\n",
    "                s |= set([no])\n",
    "        if len(s) > 0 or not ignoreEmpty:\n",
    "            res[c] = s\n",
    "    return res\n",
    "\n",
    "c2n = columns2nodes(columns, node2columns)\n",
    "usableColumns = list(c2n.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "00742eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def corrForPair(genes, set1, set2, correlationFunction = lambda x,y: scipy.stats.kendalltau(x,y).statistic):\n",
    "    \"\"\"\n",
    "        Given a dataframe genes (i.e., a set of patients), and two sets of column names (set1, set2), \n",
    "        compute the correlation function\n",
    "        for every pair of columns in set1 and set2 and return the average statistic and its std.\n",
    "    \"\"\"\n",
    "    taus = []\n",
    "    for gene1 in set1:\n",
    "        for gene2 in set2:\n",
    "            expr1 = genes.loc[:,gene1].values.tolist()\n",
    "            expr2 = genes.loc[:,gene2].values.tolist()\n",
    "            if len(set(expr1)) == 1 or len(set(expr2)) == 1:\n",
    "                continue\n",
    "            taus += [correlationFunction(expr1, expr2).statistic]\n",
    "    if len(taus) < 2:\n",
    "        return np.nan, np.nan\n",
    "    return pd.DataFrame(taus).mean().values[0], pd.DataFrame(taus).std().values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b3232726",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sampleDeltaDist(genes, set1, set2, numberOfRows1, numberOfSamples = 500, correlationFunction = lambda x,y: scipy.stats.kendalltau(x,y).statistic):\n",
    "    \"\"\"\n",
    "        Repeatedly (for numberOfSamples times) do the following:\n",
    "        - select numberOfRows1 rows of genes, and then compute tauForPair on those rows for set1 and set2\n",
    "        - do the same for remaining rows\n",
    "        - take the difference between the two resulting set of taus.\n",
    "        Return the mean and std of differences.\n",
    "    \"\"\"\n",
    "    diffs = []\n",
    "    \n",
    "    for i in range(numberOfSamples):\n",
    "        genes1 = random.sample(range(len(genes)), numberOfRows1)\n",
    "        genes2 = list(set(range(len(genes)))-set(genes1))\n",
    "        mt1, mstd1 = corrForPair(genes.iloc[genes1], set1, set2, correlationFunction)\n",
    "        if np.isnan(mp1):\n",
    "            continue\n",
    "        mt2, mstd2 = corrForPair(genes.iloc[genes2], set1, set2, correlationFunction)\n",
    "        if np.isnan(mp2):\n",
    "            continue\n",
    "        diffs += [mt1-mt2]\n",
    "    return diffs\n",
    "\n",
    "\n",
    "def deltaForPair(genes1, genes2, set1, set2, correlationFunction = lambda x,y: scipy.stats.kendalltau(x,y).statistic):\n",
    "    \"\"\"\n",
    "        Compute corrForPair for set1 and set2 over the set of rows genes1 and genes2, and return\n",
    "        the difference of taus.\n",
    "    \"\"\"\n",
    "    mt1, mstd1 = corrForPair(genes1, set1, set2, correlationFunction)\n",
    "    mt2, mstd2 = corrForPair(genes2, set1, set2, correlationFunction)\n",
    "    return mt1-mt2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2d3473f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading anomalies\n"
     ]
    }
   ],
   "source": [
    "correlationFunction = lambda x,y:  scipy.stats.spearmanr(x,y).statistic\n",
    "correlationFunctionName = \"spearmanrho\"\n",
    "relAnomalies = os.path.join(inputDirectory, \"anomalies-\" + correlationFunctionName + \".csv\")\n",
    "\n",
    "if os.path.exists(relAnomalies):\n",
    "    print(\"Reading anomalies\")\n",
    "    relcorrel = pd.read_csv(relAnomalies)\n",
    "else:\n",
    "    print(\"Producing anomalies\")\n",
    "    relcorrel = pd.DataFrame(columns=[\n",
    "        \"Source gene(s)\",\n",
    "        \"Target gene(s)\",\n",
    "        \"Pathway ID\",\n",
    "        \"Relation type\",\n",
    "        \"Correlation for healthy\",\n",
    "        \"Correlation for diseased\",\n",
    "        \"Correlation difference (H-D)\",\n",
    "        \"Expected correlation difference\",\n",
    "        \"P-value\",\n",
    "        \"Comment\"\n",
    "        ])\n",
    "    numberOfEdges = G.number_of_edges()\n",
    "    doneEdges = 0\n",
    "    for s,t,d in G.edges(data=True):\n",
    "        if doneEdges % 10 == 0:\n",
    "            print(\"{:.2f}% done\".format(100 * doneEdges / numberOfEdges))\n",
    "        doneEdges += 1\n",
    "        subtypeName = [k for k,v in subtype2color.items() if d[\"color\"]==v][0]\n",
    "        sg = node2columns[s]\n",
    "        tg = node2columns[t]\n",
    "        if len(sg) == 0 or len(tg) == 0:\n",
    "            continue\n",
    "        avgh, stdh = corrForPair(genesHealthy, sg, tg, correlationFunction=correlationFunction)\n",
    "        avgd, stdd = corrForPair(genesDiseased, sg, tg, correlationFunction=correlationFunction)\n",
    "        taudiff = avgh - avgd\n",
    "        ttn = sampleDeltaDist(genes, sg, tg, numHealthy, correlationFunction=correlationFunction)\n",
    "        tt = [x for x in ttn if ~np.isnan(x)]\n",
    "        if len(tt) == 0:\n",
    "            continue\n",
    "        ttmean = pd.DataFrame(tt).mean().values[0]\n",
    "        ecdft = scipy.stats.ecdf(tt)  # Empirical CDF\n",
    "        if taudiff < 0:\n",
    "            pvalue = ecdft.cdf.evaluate(taudiff)\n",
    "            comment = \"More correlated than for healthy individuals\"\n",
    "        else:\n",
    "            pvalue = 1 - ecdft.cdf.evaluate(taudiff)    \n",
    "            comment = \"Less correlated than for healthy individuals\"\n",
    "        if produceAnomaliesForNonsignificantPairs or produpvalue <= confidence:\n",
    "            relcorrel.loc[len(relcorrel)] = [\n",
    "                str(sg),\n",
    "                str(tg),\n",
    "                d[\"pathwayIDs\"],\n",
    "                subtypeName,\n",
    "                avgh, \n",
    "                avgd,\n",
    "                taudiff,\n",
    "                ttmean,\n",
    "                pvalue,\n",
    "                comment\n",
    "            ]\n",
    "    relcorrel = relcorrel.sort_values(by=[\"P-value\"], ascending=True)\n",
    "    relcorrel.to_csv(relAnomalies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e517681e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anomalies by type\n",
      "P-value < 0.05:  101\n",
      "Failing activations (green): 15\n",
      "Failing inhibitions (blue): 10\n",
      "Excessive activations (red): 17\n",
      "Excessive inhibitions (orange): 7\n",
      "Other types of relations (yellow): 52\n",
      "Total: 101\n"
     ]
    }
   ],
   "source": [
    "print(\"Anomalies by type\")\n",
    "statisticallySignificant = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence)   # Statistically significant\n",
    "]\n",
    "\n",
    "failingActivation = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"activation\")) & # Activation  \n",
    "    (relcorrel[\"Correlation difference (H-D)\"]>0) # Less than healthy\n",
    "]\n",
    "\n",
    "failingInhibition = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"inhibition\")) & # Inhibition\n",
    "    (relcorrel[\"Correlation difference (H-D)\"]<0) # More than healthy\n",
    "]\n",
    "\n",
    "excessiveActivation = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"activation\")) & # Activation  \n",
    "    (relcorrel[\"Correlation difference (H-D)\"]<0) # More than healthy\n",
    "]\n",
    "\n",
    "excessiveInhibition = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"inhibition\")) & # Inhibition\n",
    "    (relcorrel[\"Correlation difference (H-D)\"]>0) # Less than healthy\n",
    "]\n",
    "\n",
    "otherTypesOfRelations = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    ~(relcorrel[\"Relation type\"].str.contains(\"activation\")) & # No activation  \n",
    "    ~(relcorrel[\"Relation type\"].str.contains(\"inhibition\"))   # No inhibition \n",
    "]\n",
    "\n",
    "print(\"P-value < {:.2f}: \".format(100 * confidence / 100), len(statisticallySignificant))\n",
    "print(\"Failing activations (green):\", len(failingActivation))\n",
    "print(\"Failing inhibitions (blue):\", len(failingInhibition))\n",
    "print(\"Excessive activations (red):\", len(excessiveActivation))\n",
    "print(\"Excessive inhibitions (orange):\", len(excessiveInhibition))\n",
    "print(\"Other types of relations (yellow):\", len(otherTypesOfRelations))\n",
    "print(\"Total:\", len(failingActivation)+len(failingInhibition)+len(excessiveActivation)+len(excessiveInhibition)+len(otherTypesOfRelations))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eec52e3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d4917654",
   "metadata": {},
   "outputs": [],
   "source": [
    "def elaborateArc(relation, sdrawing):\n",
    "    source= str(node2columns[element2id[set2can(relation.source)]])\n",
    "    for tg in relation.target:\n",
    "        target = str(node2columns[element2id[set2can(tg)]])\n",
    "        r = relcorrel.loc[(relcorrel[\"Source gene(s)\"] == source) & (relcorrel[\"Target gene(s)\"] == target)]\n",
    "        if len(r) == 0:\n",
    "            continue\n",
    "        if len(r) > 1:\n",
    "            pass\n",
    "            #print(f\"({source}, {target}) appearing more than once in relcorrel\")\n",
    "            #print(list(r.to_dict()[\"Correlation difference (H-D)\"].values()))\n",
    "        rd = r.to_dict()\n",
    "        pvalue = list(rd[\"P-value\"].values())[0]\n",
    "        corrdiff = list(rd[\"Correlation difference (H-D)\"].values())[0]\n",
    "        relationType = list(rd[\"Relation type\"].values())[0]\n",
    "        if pvalue < confidence:        \n",
    "            if \"activation\" in relationType:\n",
    "                if corrdiff > 0:\n",
    "                    col = reportlab.lib.colors.green\n",
    "                else:\n",
    "                    col = reportlab.lib.colors.red\n",
    "            elif \"inhibition\" in relationType:\n",
    "                if corrdiff > 0:\n",
    "                    col = reportlab.lib.colors.orange\n",
    "                else:\n",
    "                    col = reportlab.lib.colors.blue\n",
    "            else:\n",
    "                col = reportlab.lib.colors.yellow\n",
    "            if np.isnan(corrdiff):\n",
    "                print(f\"Correlation for {source} vs. {target} is {corrdiff}\")\n",
    "            else:\n",
    "                sdrawing.setLineWidth(3*math.exp(abs(corrdiff)*3))\n",
    "                sdrawing.setStrokeColor(col)\n",
    "                sdrawing.line(relation.sourceg.x, relation.sourceg.y, relation.targetg.x, relation.targetg.y)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "32b16bfb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Producing anomaly PDFs\n"
     ]
    }
   ],
   "source": [
    "if produceGraphsWithAnomalies:\n",
    "    print(\"Producing anomaly PDFs\")\n",
    "    s = \"\"\n",
    "    for pair in superpathwayDict[superpathway]:\n",
    "        pathwayID = pair[0]\n",
    "        pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "\n",
    "        relations = []\n",
    "        for relation in rel[pathwayID]:\n",
    "            relations += [relation]\n",
    "\n",
    "        canvas = KGMLCanvas(pathway)\n",
    "\n",
    "        if useCCcoloringForAnomalies:\n",
    "            for k in pathway.entries:\n",
    "                name = pathway.entries[k].name\n",
    "                canonicalName = set2can(words2set(name))\n",
    "                if canonicalName in element2id.keys():\n",
    "                    node = element2id[canonicalName]\n",
    "                    klass = cc[node]\n",
    "                    if klass in nontrivialclass2col:\n",
    "                        pathway.entries[k].graphics[0].bgcolor = nontrivialclass2col[klass]\n",
    "                        pathway.entries[k].graphics[0].fgcolor = nontrivialclass2col[klass]\n",
    "                    else:\n",
    "                        pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "                        pathway.entries[k].graphics[0].fgcolor = \"#FFFFFF\"\n",
    "                    s += pathwayID + \"\\t\" + str(node) + \"\\t\" + str(klass) +\"\\t\"+pathway.entries[k].graphics[0].name+\"\\t\"+canonicalName + \"\\n\"\n",
    "\n",
    "        canvas.import_imagemap = True\n",
    "        pdfName = organism + pathwayID + \".pdf\"\n",
    "        pathlib.Path(os.path.join(outputDirectory, organism, safeSuperpathway + \"-anomalies\")).mkdir(parents=True, exist_ok=True)\n",
    "        canvas.new_draw(os.path.join(outputDirectory, organism, safeSuperpathway+ \"-anomalies\", pdfName), relations, elaborateArc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8e63a371",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing anomalies by pathway\n"
     ]
    }
   ],
   "source": [
    "print(\"Computing anomalies by pathway\")\n",
    "pathwayAnomalies = pd.DataFrame(columns=[\n",
    "        \"Pathway ID\",\n",
    "        \"Pathway\",\n",
    "        \"Average square difference\",\n",
    "        \"Count\",\n",
    "        \"Relative count\"\n",
    "        ])\n",
    "for pair in superpathwayDict[superpathway]:\n",
    "    pathwayID = pair[0]\n",
    "    pathwayName= pair[1]\n",
    "    ssqd = 0\n",
    "    count = 0\n",
    "    countSig = 0\n",
    "    for relation in rel[pathwayID]:\n",
    "        source = str(node2columns[element2id[set2can(relation.source)]])\n",
    "        for targ in relation.target:\n",
    "            target = str(node2columns[element2id[set2can(targ)]])\n",
    "            tt = relcorrel.loc[(relcorrel[\"Source gene(s)\"] == source) & (relcorrel[\"Target gene(s)\"] == target)]\n",
    "            if len(tt) == 1 and tt[\"P-value\"].to_list()[0] <= confidence:\n",
    "                cd = tt[\"Correlation difference (H-D)\"].to_list()[0]\n",
    "                ssqd += cd * cd\n",
    "                countSig += 1\n",
    "            count += 1\n",
    "    if count > 0:\n",
    "        pathwayAnomalies.loc[len(pathwayAnomalies)]=[pathwayID, pathwayName, ssqd, countSig, countSig / count]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ff411669-e133-48ac-86b9-dcc350d4d957",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sorting pathways by number of anomalies and writing it out as a csv\n"
     ]
    }
   ],
   "source": [
    "print(\"Sorting pathways by number of anomalies and writing it out as a csv\")\n",
    "relPathwayAnomalies = os.path.join(inputDirectory, \"pathway-anomalies-\" + correlationFunctionName + \".csv\")\n",
    "pathwayAnomalies = pathwayAnomalies.sort_values(by=[\"Count\"], ascending=False)\n",
    "pathwayAnomalies.to_csv(relPathwayAnomalies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3cae2695",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting genes involved in top-k anomalous pathways and appearing as columns\n",
      "1537 genes involved\n",
      "Genes involved by pathway\n",
      "\t00230: 102\n",
      "\t04070: 78\n",
      "\t00240: 49\n",
      "\t05205: 178\n",
      "\t04010: 252\n",
      "\t04062: 166\n",
      "\t04310: 144\n",
      "\t04910: 128\n",
      "\t04510: 185\n",
      "\t04218: 134\n",
      "\t04931: 88\n",
      "\t04611: 115\n",
      "\t04512: 70\n",
      "\t05417: 182\n",
      "\t04213: 59\n",
      "\t04210: 120\n",
      "\t05321: 37\n",
      "\t04151: 267\n",
      "\t04620: 84\n",
      "\t04115: 68\n",
      "\t04970: 48\n",
      "\t04060: 201\n",
      "\t00010: 62\n",
      "\t04066: 98\n",
      "\t04975: 10\n",
      "\t05415: 99\n",
      "\t04145: 15\n",
      "\t04974: 2\n",
      "\t04979: 31\n",
      "\t05146: 35\n"
     ]
    }
   ],
   "source": [
    "print(\"Extracting genes involved in top-k anomalous pathways and appearing as columns\")\n",
    "nTopAnomPathway = len(pathwayAnomalies)\n",
    "topAnomPathway = pathwayAnomalies.head(nTopAnomPathway)[\"Pathway ID\"].to_list()\n",
    "anomPathwayGenes = set([])\n",
    "anomPathwayGenesByPathway = defaultdict(lambda: set([]))\n",
    "for pathwayId in topAnomPathway:\n",
    "    for relation in rel[pathwayId]:\n",
    "        anomPathwayGenes |= relation.source \n",
    "        anomPathwayGenesByPathway[pathwayId] |= relation.source \n",
    "        for t in relation.target:\n",
    "            anomPathwayGenes |= t\n",
    "            anomPathwayGenesByPathway[pathwayId] |= t\n",
    "anomPathwayGeneNames = [name for name,gene in name2gene.items() if gene in anomPathwayGenes and name in columns]\n",
    "print(len(anomPathwayGeneNames), \"genes involved\")\n",
    "anomPathwayGenesByPathwayNames = {}\n",
    "print(\"Genes involved by pathway\")\n",
    "for pathwayId in topAnomPathway:\n",
    "    anomPathwayGenesByPathwayNames[pathwayId] = [name for name,gene in name2gene.items() if gene in anomPathwayGenesByPathway[pathwayId] and name in columns]\n",
    "    t = len(anomPathwayGenesByPathwayNames[pathwayId])\n",
    "    print(f\"\\t{pathwayId}: {t}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c1e14ecf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing correlations for anomalous-pathway genes in the two populations\n"
     ]
    }
   ],
   "source": [
    "print(\"Computing correlations for anomalous-pathway genes in the two populations\")\n",
    "healthyMatrix = genesHealthy[anomPathwayGeneNames].to_numpy() #Rows are healthy patients, columns are genes appearing in anomalous pathways\n",
    "diseasedMatrix = genesDiseased[anomPathwayGeneNames].to_numpy() #Rows are diseased patients, columns are genes appearing in anomalous pathways\n",
    "\n",
    "healthyCorr = np.corrcoef(healthyMatrix, rowvar=False) #Correlation between anomalous genes in healthy individuals\n",
    "diseasedCorr = np.corrcoef(diseasedMatrix, rowvar=False) #Correlation between anomalous genes in diseased individuals\n",
    "healthyCorrByPathway = {}\n",
    "diseasedCorrByPathway = {}\n",
    "for pathwayId in topAnomPathway:\n",
    "    healthyMatrix = genesHealthy[anomPathwayGenesByPathwayNames[pathwayId]].to_numpy() #Rows are healthy patients, columns are genes appearing in anomalous pathways\n",
    "    diseasedMatrix = genesDiseased[anomPathwayGenesByPathwayNames[pathwayId]].to_numpy() #Rows are healthy patients, columns are genes appearing in anomalous pathways\n",
    "    healthyCorrByPathway[pathwayId] = np.corrcoef(healthyMatrix, rowvar=False)\n",
    "    diseasedCorrByPathway[pathwayId] = np.corrcoef(diseasedMatrix, rowvar=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "36fbda30",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(data=healthyCorr, columns=anomPathwayGeneNames).to_csv(\"/tmp/matrices/healthy-all.csv\")\n",
    "pd.DataFrame(data=diseasedCorr, columns=anomPathwayGeneNames).to_csv(\"/tmp/matrices/diseased-all.csv\")\n",
    "for pathwayId in topAnomPathway:\n",
    "    pd.DataFrame(data=healthyCorrByPathway[pathwayId], columns=anomPathwayGenesByPathwayNames[pathwayId]).to_csv(\"/tmp/matrices/healthy-\" + pathwayId + \".csv\")\n",
    "    pd.DataFrame(data=diseasedCorrByPathway[pathwayId], columns=anomPathwayGenesByPathwayNames[pathwayId]).to_csv(\"/tmp/matrices/diseased-\" + pathwayId + \".csv\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3edb119d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qf",
   "language": "python",
   "name": "qf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
