{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d4364c55",
   "metadata": {},
   "outputs": [],
   "source": [
    "from hypergraphs import *\n",
    "from kegg import *\n",
    "import pandas as pd\n",
    "import os\n",
    "import logging\n",
    "\n",
    "from Bio.KEGG.KGML.KGML_parser import read\n",
    "import Bio.KEGG.KGML.KGML_pathway\n",
    "from ast import literal_eval\n",
    "from urllib.request import urlopen\n",
    "from urllib.parse import quote_plus\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "from Levenshtein import ratio\n",
    "import scipy.stats\n",
    "import pathlib\n",
    "from qf.cc import cardon_crochemore_colored, cardon_crochemore\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6b13e6bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "organismDict = {\"hsa\": \"Homo sapiens (human)\"}\n",
    "outputDirectory = \"/Users/boldi/Desktop/pw/\"\n",
    "inputDirectory = \"../../LaTeX/Data/KEGG-Pathways/Nov2023/\"\n",
    "dataDirectory = \"../../LaTeX/Data/KEGG-Pathways\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e6caae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pws = pd.read_csv(os.path.join(inputDirectory, \"KEGG-pathways.txt\"), header=0, \n",
    "            names=[\"Pathway ID\", \"Pathway type\", \"Pathway name\"], dtype=\"string\", keep_default_na=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "199b2487",
   "metadata": {},
   "outputs": [],
   "source": [
    "superpathwayDict = {\n",
    "    \"Involved\": [(x,y) for x,y in zip(pws[\"Pathway ID\"], pws[\"Pathway name\"])]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bce1e1d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "superpathway = \"Involved\"\n",
    "organism = \"hsa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cb8e2d96",
   "metadata": {},
   "outputs": [],
   "source": [
    "safeSuperpathway = makesafe(superpathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "47a38c39",
   "metadata": {},
   "outputs": [],
   "source": [
    "def words2set(words):\n",
    "    \"\"\"\n",
    "        Given a space-separated list of words, returns it as a set of words\n",
    "    \"\"\"\n",
    "    return set(words.split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8e0a74d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import NamedTuple\n",
    "\n",
    "class Relation(NamedTuple):\n",
    "    \"\"\"\n",
    "        Represents a relation. It is characterized by:\n",
    "        - a type (e.g. \"PPRel\") \n",
    "        - a list of subtypes (e.g. [('activation', '-->'), ('indirect effect', '..>')])\n",
    "        - a set of components (source)\n",
    "        - a list of sets of components (target)\n",
    "    \"\"\"\n",
    "    relType: str\n",
    "    relSubtypes: list\n",
    "    source: set\n",
    "    target: list\n",
    "    sourceg: Bio.KEGG.KGML.KGML_pathway.Graphics\n",
    "    targetg: Bio.KEGG.KGML.KGML_pathway.Graphics\n",
    "        \n",
    "    @classmethod\n",
    "    def fromKEGG(cls, relation, pathway):\n",
    "        \"\"\"\n",
    "            Constructor: builds a Relation from a KEGG relation\n",
    "        \"\"\"\n",
    "        if relation.entry2.name == \"undefined\":\n",
    "            s = []\n",
    "            for component in relation.entry2.components:\n",
    "                s += [words2set(pathway.entries[component.id].name)]\n",
    "            targetg = pathway.entries[component.id].graphics[0]\n",
    "        else:\n",
    "            s = [words2set(relation.entry2.name)]\n",
    "            targetg = relation.entry2.graphics[0]\n",
    "        return cls(relation.type, relation.subtypes, words2set(relation.entry1.name), s, targetg, relation.entry1.graphics[0])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5ba5d2e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsFromKEGGpathway(pathway):\n",
    "    \"\"\"\n",
    "        Given a KEGG pathway, returns the list of relations it contains (represented as Relation).\n",
    "    \"\"\"\n",
    "    return [Relation.fromKEGG(r, pathway) for r in pathway.relations]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "751bc16c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pathwayFromFile(dataDirectory, organism, pathwayID):\n",
    "    \"\"\"\n",
    "        Load and return a KEGG pathway, given the data directory, the organism specifying the subdirectory,\n",
    "        and the pathwayID (the filename should be organism+pathwayID+\"xml\".\n",
    "    \"\"\"\n",
    "    return KGML_parser.read(open(os.path.join(dataDirectory, organism, organism+pathwayID+\".xml\"), \"r\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bd27fdc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsFromFile(dataDirectory, organism, pathwayID):\n",
    "    \"\"\"\n",
    "        Given a pathway file (specified by a root dataDirectory, the name of the organism subdirectory, and the name\n",
    "        of the file [organism+pathwayID.xml]), reads it and returns the list of its relations.\n",
    "    \"\"\"\n",
    "    pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "    return relationsFromKEGGpathway(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "eaa01766",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsForSuperpathway(dataDirectory, organism, superpathwayDict, superpathway):\n",
    "    \"\"\"\n",
    "        Given a superpathDict (whose keys are superpathway names and whose values are list of\n",
    "        pairs (pathwayID, pathwayName)), reads all the pathway file for a specific superpathway name\n",
    "        and returns a dictionary whose keys are the pathwayID's and whose values are the list of relations.\n",
    "    \"\"\"\n",
    "    rel = {}\n",
    "    for k,v in superpathwayDict[superpathway]:\n",
    "        #logging.info(\"Reading\", k, v)\n",
    "        rel[k] = relationsFromFile(dataDirectory, organism, k)\n",
    "    return rel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "15479a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def subtypes(rel):\n",
    "    \"\"\"\n",
    "        Given a map whose values are relations, accumulate all subtypes appearing and assign them a number.\n",
    "        The result is a map from subtype string to number.\n",
    "    \"\"\"\n",
    "    all_subtypes = set([])\n",
    "    for pathwayid, relations in rel.items():\n",
    "        for relation in relations:\n",
    "            all_subtypes |= set([str(relation.relSubtypes)])\n",
    "        \n",
    "    s = sorted(list(all_subtypes))\n",
    "    subtype2color = {v:k for k,v in enumerate(s)}\n",
    "\n",
    "    return subtype2color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1e229eaf-d667-466c-8238-8205257b0ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set2can(s):\n",
    "    \"\"\"\n",
    "        Convert a set to a string in a canonical way.\n",
    "    \"\"\"\n",
    "    return str(sorted(list(s)))\n",
    "\n",
    "def can2set(c):\n",
    "    \"\"\"\n",
    "        Does the converse of set2can.\n",
    "    \"\"\"\n",
    "    return set(literal_eval(c))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "75b272b9-b220-492b-adf6-509f9952842f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indices(rel):\n",
    "    \"\"\"\n",
    "        Given a map whose values are lists of relations, it considers all the relations one by one, and attributes a unique id \n",
    "        to each element (i.e., set of components appearing as source or in the target of some relation) and block (the set of\n",
    "        element id appearing as target of some relation).\n",
    "        This function returns the dictionaries to move from/to an element or block to the corresponding id.\n",
    "        Elements are string representations of sorted lists of strings.\n",
    "        Blocks are string representations of sorted lists of ints.\n",
    "    \"\"\"\n",
    "    element2id = {}\n",
    "    id2element = {}\n",
    "    block2id = {}\n",
    "    id2block = {}\n",
    "\n",
    "    for relations in rel.values():\n",
    "        for relation in relations:\n",
    "            s = set2can(relation.source)\n",
    "            if s not in element2id.keys():\n",
    "                element2id[s] = len(element2id)\n",
    "            targetset = set([])\n",
    "            for targ in relation.target:\n",
    "                t = set2can(targ)\n",
    "                if t not in element2id.keys():\n",
    "                    element2id[t] = len(element2id)\n",
    "                targetset |= set([element2id[t]])\n",
    "            ts = set2can(targetset)\n",
    "            if ts not in block2id.keys():\n",
    "                block2id[ts] = len(block2id)\n",
    "\n",
    "    for k,v in element2id.items():\n",
    "        id2element[v] = k\n",
    "    for k,v in block2id.items():\n",
    "        id2block[v] = k\n",
    "    return element2id, id2element, block2id, id2block\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f9c3f128-cf97-4bb3-8ce5-65c49f865d98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_gene_KEGG(geneID):\n",
    "    \"\"\"\n",
    "        Search for gene on KEGG. \n",
    "        \n",
    "        Returns list of names.\n",
    "    \"\"\"\n",
    "    url1 = \"https://www.kegg.jp/entry/\"+geneID\n",
    "    response1 = requests.get(url1, allow_redirects=False)    \n",
    "    soup = BeautifulSoup(response1.text, \"html.parser\")\n",
    "    tds = [tds for tds in soup.find_all(\"td\", {\"class\": \"td11 defd\"})]\n",
    "    names = tds[0].getText().strip().split(\", \")\n",
    "    return names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0a11bbae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build all data needed for the relations of the superpathway\n",
    "rel = relationsForSuperpathway(dataDirectory, organism, superpathwayDict, superpathway)\n",
    "element2id, id2element, block2id, id2block = indices(rel)\n",
    "subtype2color = subtypes(rel)\n",
    "elements = list(set.union(*[can2set(k) for k in element2id.keys()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a1be8c2d-c5da-4e04-bf67-7e43b1a71b7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading gene names\n"
     ]
    }
   ],
   "source": [
    "# name2gene is a function mapping names to gene IDs (hsa:xxxx)\n",
    "# e.g. KCNE3, BRGDA6 etc are all mapped to hsa:10008 (see https://www.kegg.jp/entry/hsa:10008)\n",
    "# This map was produced by scraping kegg, but this has been done only once.\n",
    "# After that, we just read a pkl file\n",
    "\n",
    "name2keggFilename = os.path.join(inputDirectory, \"name2kegg.pkl\")\n",
    "\n",
    "if os.path.exists(name2keggFilename):\n",
    "    print(\"Reading gene names\")\n",
    "    with open(name2keggFilename, \"rb\") as handle:\n",
    "        name2gene = pickle.load(handle)\n",
    "else:\n",
    "    print(\"Producing gene names\")\n",
    "    d = {}\n",
    "    nelements = len(elements)\n",
    "    for count,element in enumerate(elements):\n",
    "        if count % 10 == 0:\n",
    "            print(f\"{count}/{nelements}\")\n",
    "        if element.startswith(\"hsa\"):\n",
    "            names = search_gene_KEGG(element)\n",
    "            for name in names:\n",
    "                d[name] = element\n",
    "    with open(name2keggFilename, \"wb\") as handle:\n",
    "        pickle.dump(d, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "    name2gene = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "74d19326",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading patient data\n"
     ]
    }
   ],
   "source": [
    "# Read patient data\n",
    "print(\"Reading patient data\")\n",
    "genes = pd.read_csv(os.path.join(inputDirectory, \"genes.csv\"))\n",
    "patients = genes[\"name\"].values.tolist()\n",
    "classification = pd.read_csv(os.path.join(inputDirectory, \"classification.csv\"))\n",
    "\n",
    "patient2class = {patient: classification.loc[classification[\"name\"]==patient][\"classification\"].values.flatten().tolist()[0] for patient in patients}\n",
    "numHealthy = sum([x == 'healthy' for x in patient2class.values()])\n",
    "numDiseased = sum([x == 'diseased' for x in patient2class.values()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "34acce33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dividing genes into two frames (one for healthy individuals and one for diseased individuals)\n",
    "genesHealthy = genes.loc[genes[\"name\"].apply(lambda x: patient2class[x])==\"healthy\"]\n",
    "genesDiseased = genes.loc[genes[\"name\"].apply(lambda x: patient2class[x])==\"diseased\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1c90f901-1413-4356-bfa8-faf25fc494df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building the relation graph\n"
     ]
    }
   ],
   "source": [
    "# Building the graph of all relations contained in rel values\n",
    "# Nodes are IDs, and node x correspond to the set of genes (and/or compound) id2element[x]\n",
    "print(\"Building the relation graph\")\n",
    "G = nx.DiGraph() \n",
    "for relations in rel.values():\n",
    "    for relation in relations:\n",
    "        source = element2id[set2can(relation.source)]\n",
    "        for targ in relation.target:\n",
    "            target = element2id[set2can(targ)]\n",
    "            G.add_edge(source, target, color=subtype2color[str(relation.relSubtypes)])\n",
    "            \n",
    "\n",
    "# Computing the minimum fibres\n",
    "cc=cardon_crochemore_colored(G)\n",
    "\n",
    "colors = list(matplotlib.colors.CSS4_COLORS.values())\n",
    "usedColors = 0\n",
    "\n",
    "cc2fibresize = {}\n",
    "cc2fibre = {}\n",
    "ccnontrivial = {}\n",
    "nontrivialclass2col = {}\n",
    "\n",
    "fibres = []\n",
    "node2fibre = {}\n",
    "for v in set(cc.values()):\n",
    "    fibre = [x for x in G.nodes() if cc[x] == v]\n",
    "    fibres += [fibre]\n",
    "    for x in fibre:\n",
    "        node2fibre[x] = fibre\n",
    "    cc2fibre[v] = fibre\n",
    "    cc2fibresize[v] = len(fibre)\n",
    "    if len(fibre) > 0:\n",
    "        ccnontrivial[v] = cc[v]\n",
    "        nontrivialclass2col[v] = colors[usedColors]\n",
    "        usedColors = (usedColors + 1 ) % len(colors)\n",
    "        if usedColors == 0:\n",
    "            print(\"Wrap: more than\", len(colors), \"colors needed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d51dbd7f-35b9-444a-9238-3af727c06411",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping genes to columns\n"
     ]
    }
   ],
   "source": [
    "# Build node2columns, mapping each node of G to the list of genes corresponding to it\n",
    "# for which we know a value (in the genes dataframe), sorted\n",
    "#\n",
    "\n",
    "print(\"Mapping genes to columns\")\n",
    "namedGenes = set(name2gene.values())     \n",
    "columns = set(genes.columns.values.tolist()[2:])\n",
    "node2columns = {}\n",
    "for node in G.nodes():\n",
    "    nodeGenes = can2set(id2element[node])\n",
    "    s = []\n",
    "    for gene in nodeGenes & namedGenes:\n",
    "        for k,v in name2gene.items():\n",
    "            if v == gene:\n",
    "                if k in columns:\n",
    "                    s += [k]\n",
    "    node2columns[node]=sorted(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "05b96661",
   "metadata": {},
   "outputs": [],
   "source": [
    "s = \"\"\n",
    "for pair in superpathwayDict[superpathway]:\n",
    "    pathwayID = pair[0]\n",
    "    pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "    canvas = KGMLCanvas(pathway)\n",
    "    for k in pathway.entries:\n",
    "        name = pathway.entries[k].name\n",
    "        canonicalName = set2can(words2set(name))\n",
    "        if canonicalName in element2id.keys():\n",
    "            node = element2id[canonicalName]\n",
    "            klass = cc[node]\n",
    "            if klass in nontrivialclass2col:\n",
    "                pathway.entries[k].graphics[0].bgcolor = nontrivialclass2col[klass]\n",
    "                pathway.entries[k].graphics[0].fgcolor = nontrivialclass2col[klass]\n",
    "            else:\n",
    "                pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "                pathway.enries[k].graphics[0].fgcolor = \"#FFFFFF\"\n",
    "            s += pathwayID + \"\\t\" + str(node) + \"\\t\" + str(klass) +\"\\t\"+pathway.entries[k].graphics[0].name+\"\\t\"+canonicalName + \"\\n\"\n",
    "    canvas.import_imagemap = True\n",
    "    pdfName = organism + pathwayID + \".pdf\"\n",
    "    pathlib.Path(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway)).mkdir(parents=True, exist_ok=True)\n",
    "    canvas.draw(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway, pdfName))\n",
    "with open(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway, \"all.tsv\"), \"wt\") as file:\n",
    "    file.write(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "73ce2fd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def columns2nodes(columns, node2columns, ignoreEmpty = True):\n",
    "    \"\"\"\n",
    "        Returns a map from columns to the nodes where the column appears.\n",
    "    \"\"\"\n",
    "    res = {} \n",
    "    for c in columns:\n",
    "        s = set([])\n",
    "        for no, co in node2columns.items():\n",
    "            if c in co:\n",
    "                s |= set([no])\n",
    "        if len(s) > 0 or not ignoreEmpty:\n",
    "            res[c] = s\n",
    "    return res\n",
    "\n",
    "c2n = columns2nodes(columns, node2columns)\n",
    "usableColumns = list(c2n.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "00742eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def corrForPair(genes, set1, set2, correlationFunction = lambda x: scipy.stats.kendalltau(x).statistic):\n",
    "    \"\"\"\n",
    "        Given a dataframe genes (i.e., a set of patients), and two sets of column names (set1, set2), \n",
    "        compute the correlation function\n",
    "        for every pair of columns in set1 and set2 and return the average statistic and its std.\n",
    "    \"\"\"\n",
    "    taus = []\n",
    "    for gene1 in set1:\n",
    "        for gene2 in set2:\n",
    "            expr1 = genes.loc[:,gene1].values.tolist()\n",
    "            expr2 = genes.loc[:,gene2].values.tolist()\n",
    "            taus += [scipy.stats.kendalltau(expr1, expr2).statistic]\n",
    "    return pd.DataFrame(taus).mean().values[0], pd.DataFrame(taus).std().values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b3232726",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sampleDeltaDist(genes, set1, set2, numberOfRows1, numberOfSamples = 100, correlationFunction = lambda x: scipy.stats.kendalltau(x).statistic):\n",
    "    \"\"\"\n",
    "        Repeatedly (for numberOfSamples times) do the following:\n",
    "        - select numberOfRows1 rows of genes, and then compute tauForPair on those rows for set1 and set2\n",
    "        - do the same for remaining rows\n",
    "        - take the difference between the two resulting set of taus.\n",
    "        Return the mean and std of differences.\n",
    "    \"\"\"\n",
    "    diffs = []\n",
    "    \n",
    "    for i in range(numberOfSamples):\n",
    "        genes1 = random.sample(range(len(genes)), numberOfRows1)\n",
    "        genes2 = list(set(range(len(genes)))-set(genes1))\n",
    "        mt1, mstd1 = corrForPair(genes.iloc[genes1], set1, set2, correlationFunction)\n",
    "        mt2, mstd2 = corrForPair(genes.iloc[genes2], set1, set2, correlationFunction)\n",
    "        diffs += [mt1-mt2]\n",
    "    return diffs\n",
    "\n",
    "\n",
    "def deltaForPair(genes1, genes2, set1, set2, correlationFunction = lambda x: scipy.stats.kendalltau(x).statistic):\n",
    "    \"\"\"\n",
    "        Compute corrForPair for set1 and set2 over the set of rows genes1 and genes2, and return\n",
    "        the difference of taus.\n",
    "    \"\"\"\n",
    "    mt1, mstd1 = corrForPair(genes1, set1, set2, correlationFunction)\n",
    "    mt2, mstd2 = corrForPair(genes2, set1, set2, correlationFunction)\n",
    "    return mt1-mt2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2d3473f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading anomalies\n"
     ]
    }
   ],
   "source": [
    "correlationFunction = lambda x:  scipy.stats.spearmanr(x).statistic\n",
    "correlationFunctionName = \"spearmanrho\"\n",
    "relAnomalies = os.path.join(inputDirectory, \"anomalies-\" + correlationFunctionName + \".csv\")\n",
    "\n",
    "if os.path.exists(relAnomalies):\n",
    "    print(\"Reading anomalies\")\n",
    "    relcorrel = pd.read_csv(relAnomalies)\n",
    "else:\n",
    "    print(\"Producing anomalies\")\n",
    "    relcorrel = pd.DataFrame(columns=[\n",
    "        \"Souce gene(s)\",\n",
    "        \"Target gene(s)\",\n",
    "        \"Source node\",\n",
    "        \"Target node\",\n",
    "        \"Relation type\",\n",
    "        \"Correlation for healthy\",\n",
    "        \"Correlation for diseased\",\n",
    "        \"Correlation difference (H-D)\",\n",
    "        \"Expected correlation difference\",\n",
    "        \"P-value\",\n",
    "        \"Comment\"\n",
    "        ])\n",
    "    numberOfEdges = G.number_of_edges()\n",
    "    doneEdges = 0\n",
    "    for s,t,d in G.edges(data=True):\n",
    "        if doneEdges % 10 == 0:\n",
    "            print(\"{:.2f}% done\".format(100 * doneEdges / numberOfEdges))\n",
    "        doneEdges += 1\n",
    "        subtypeName = [k for k,v in subtype2color.items() if d[\"color\"]==v][0]\n",
    "        sg = node2columns[s]\n",
    "        tg = node2columns[t]\n",
    "        if len(sg) == 0 or len(tg) == 0:\n",
    "            continue\n",
    "        avgh, stdh = corrForPair(genesHealthy, sg, tg, correlationFunction=correlationFunction)\n",
    "        avgd, stdd = corrForPair(genesDiseased, sg, tg, correlationFunction=correlationFunction)\n",
    "        taudiff = avgh - avgd\n",
    "        tt = sampleDeltaDist(genes, sg, tg, numHealthy, correlationFunction=correlationFunction)\n",
    "        ttmean = pd.DataFrame(tt).mean().values[0]\n",
    "        ecdft = scipy.stats.ecdf(tt)  # Empirical CDF\n",
    "        if taudiff < 0:\n",
    "            pvalue = ecdft.cdf.evaluate(taudiff)\n",
    "            comment = \"More correlated than for healthy individuals\"\n",
    "        else:\n",
    "            pvalue = 1 - ecdft.cdf.evaluate(taudiff)    \n",
    "            comment = \"Less correlated than for healthy individuals\"\n",
    "        relcorrel.loc[len(relcorrel)] = [\n",
    "            sg,\n",
    "            tg,\n",
    "            s,\n",
    "            t,\n",
    "            subtypeName,\n",
    "            avgh, \n",
    "            avgd,\n",
    "            taudiff,\n",
    "            ttmean,\n",
    "            pvalue,\n",
    "            comment\n",
    "        ]\n",
    "    relcorrel.to_csv(relAnomalies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e517681e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P-value < 0.05:  17\n",
      "Failing activations (green): 5\n",
      "Failing inhibitions (blue): 1\n",
      "Excessive activations (red): 7\n",
      "Excessive inhibitions (orange): 0\n",
      "Other types of relations (yellow): 4\n",
      "Total: 17\n"
     ]
    }
   ],
   "source": [
    "confidence=0.05\n",
    "\n",
    "statisticallySignificant = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence)   # Statistically significant\n",
    "]\n",
    "\n",
    "failingActivation = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"activation\")) & # Activation  \n",
    "    (relcorrel[\"Correlation difference (H-D)\"]>0) # Less than healthy\n",
    "]\n",
    "\n",
    "failingInhibition = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"inhibition\")) & # Inhibition\n",
    "    (relcorrel[\"Correlation difference (H-D)\"]<0) # More than healthy\n",
    "]\n",
    "\n",
    "excessiveActivation = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"activation\")) & # Activation  \n",
    "    (relcorrel[\"Correlation difference (H-D)\"]<0) # More than healthy\n",
    "]\n",
    "\n",
    "excessiveInhibition = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    (relcorrel[\"Relation type\"].str.contains(\"inhibition\")) & # Inhibition\n",
    "    (relcorrel[\"Correlation difference (H-D)\"]>0) # Less than healthy\n",
    "]\n",
    "\n",
    "otherTypesOfRelations = relcorrel.loc[\n",
    "    (relcorrel[\"P-value\"]<confidence) &   # Statistically significant\n",
    "    ~(relcorrel[\"Relation type\"].str.contains(\"activation\")) & # No activation  \n",
    "    ~(relcorrel[\"Relation type\"].str.contains(\"inhibition\"))   # No inhibition \n",
    "]\n",
    "\n",
    "print(\"P-value < {:.2f}: \".format(100 * confidence / 100), len(statisticallySignificant))\n",
    "print(\"Failing activations (green):\", len(failingActivation))\n",
    "print(\"Failing inhibitions (blue):\", len(failingInhibition))\n",
    "print(\"Excessive activations (red):\", len(excessiveActivation))\n",
    "print(\"Excessive inhibitions (orange):\", len(excessiveInhibition))\n",
    "print(\"Other types of relations (yellow):\", len(otherTypesOfRelations))\n",
    "print(\"Total:\", len(failingActivation)+len(failingInhibition)+len(excessiveActivation)+len(excessiveInhibition)+len(otherTypesOfRelations))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "32b16bfb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from reportlab.lib.units import inch\n",
    "import PIL.Image\n",
    "from io import BytesIO\n",
    "import reportlab.pdfgen.canvas\n",
    "import reportlab.lib.colors\n",
    "\n",
    "def elaborateArc(relation, sdrawing):\n",
    "    source = element2id[set2can(relation.source)]\n",
    "    for tg in relation.target:\n",
    "        target = element2id[set2can(tg)]\n",
    "        r = relcorrel.loc[(relcorrel[\"Source node\"] == source) & (relcorrel[\"Target node\"] == target)]\n",
    "        if len(r) == 0:\n",
    "            continue\n",
    "        if len(r) > 1:\n",
    "            print(f\"({s}, {t}) appearing more than once in relcorrel\")\n",
    "            continue\n",
    "        rd = r.to_dict()\n",
    "        pvalue = list(rd[\"P-value\"].values())[0]\n",
    "        corrdiff = list(rd[\"Correlation difference (H-D)\"].values())[0]\n",
    "        relationType = list(rd[\"Relation type\"].values())[0]\n",
    "        if pvalue < confidence:        \n",
    "            if \"activation\" in relationType:\n",
    "                if corrdiff > 0:\n",
    "                    col = reportlab.lib.colors.green\n",
    "                else:\n",
    "                    col = reportlab.lib.colors.red\n",
    "            elif \"inhibition\" in relationType:\n",
    "                if corrdiff > 0:\n",
    "                    col = reportlab.lib.colors.orange\n",
    "                else:\n",
    "                    col = reportlab.lib.colors.blue\n",
    "            else:\n",
    "                col = reportlab.lib.colors.yellow\n",
    "            sdrawing.setLineWidth(3*math.exp(abs(corrdiff)*3))\n",
    "            sdrawing.setStrokeColor(col)\n",
    "            sdrawing.line(relation.sourceg.x, relation.sourceg.y, relation.targetg.x, relation.targetg.y)        \n",
    "\n",
    "#from __future__ import nested_scopes\n",
    "\n",
    "def get_temp_imagefilename(url):\n",
    "    \"\"\"Return filename of temporary file containing downloaded image.\n",
    "\n",
    "    Create a new temporary file to hold the image file at the passed URL\n",
    "    and return the filename.\n",
    "    \"\"\"\n",
    "    img = urlopen(url).read()\n",
    "    im = PIL.Image.open(BytesIO(img))\n",
    "    # im.transpose(Image.FLIP_TOP_BOTTOM)\n",
    "    f = tempfile.NamedTemporaryFile(delete=False, suffix=\".png\")\n",
    "    fname = f.name\n",
    "    f.close()\n",
    "    im.save(fname, \"PNG\")\n",
    "    return fname\n",
    "\n",
    "\n",
    "#def enhance_method(klass, method_name, replacement):\n",
    "#    'replace a method with an enhanced version'\n",
    "#    method = getattr(klass, method_name)\n",
    "#    def enhanced(*args, **kwds): return replacement(*args, **kwds)\n",
    "#    setattr(klass, method_name, enhanced)\n",
    "\n",
    "def new_draw(self, filename, relations, elaborateArc = None):\n",
    "        \"\"\"Add the map elements to the drawing.\"\"\"\n",
    "        # Instantiate the drawing, first\n",
    "        # size x_max, y_max for now - we can add margins, later\n",
    "        if self.import_imagemap:\n",
    "            # We're drawing directly on the image, so we set the canvas to the\n",
    "            # same size as the image\n",
    "            if os.path.isfile(self.pathway.image):\n",
    "                imfilename = self.pathway.image\n",
    "            else:\n",
    "                imfilename = get_temp_imagefilename(self.pathway.image)\n",
    "            im = PIL.Image.open(imfilename)\n",
    "            cwidth, cheight = im.size\n",
    "        else:\n",
    "            # No image, so we set the canvas size to accommodate visible\n",
    "            # elements\n",
    "            cwidth, cheight = (self.pathway.bounds[1][0], self.pathway.bounds[1][1])\n",
    "        # Instantiate canvas\n",
    "        self.drawing = reportlab.pdfgen.canvas.Canvas(\n",
    "            filename,\n",
    "            bottomup=0,\n",
    "            pagesize=(\n",
    "                cwidth * (1 + 2 * self.margins[0]),\n",
    "                cheight * (1 + 2 * self.margins[1]),\n",
    "            ),\n",
    "        )\n",
    "        self.drawing.setFont(self.fontname, self.fontsize)\n",
    "        # Transform the canvas to add the margins\n",
    "        self.drawing.translate(\n",
    "            self.margins[0] * self.pathway.bounds[1][0],\n",
    "            self.margins[1] * self.pathway.bounds[1][1],\n",
    "        )\n",
    "        # Add the map image, if required\n",
    "        if self.import_imagemap:\n",
    "            self.drawing.saveState()\n",
    "            self.drawing.scale(1, -1)\n",
    "            self.drawing.translate(0, -cheight)\n",
    "            self.drawing.drawImage(imfilename, 0, 0)\n",
    "            self.drawing.restoreState()\n",
    "        # Add the reactions, compounds and maps\n",
    "        # Maps go on first, to be overlaid by more information.\n",
    "        # By default, they're slightly transparent.\n",
    "        if self.show_maps:\n",
    "            self.__add_maps()\n",
    "        if self.show_reaction_entries:\n",
    "            KGMLCanvas._KGMLCanvas__add_reaction_entries(self)\n",
    "        if self.show_orthologs:\n",
    "            KGMLCanvas._KGMLCanvas__add_orthologs(self)\n",
    "        if self.show_compounds:\n",
    "            KGMLCanvas._KGMLCanvas__add_compounds(self)\n",
    "        if self.show_genes:\n",
    "            KGMLCanvas._KGMLCanvas__add_genes(self)\n",
    "        # TODO: complete draw_relations code\n",
    "        # if self.draw_relations:\n",
    "        #    self.__add_relations()\n",
    "        # Write the pathway map to PDF\n",
    "        if elaborateArc is not None:\n",
    "            for relation in relations:\n",
    "                elaborateArc(relation, self.drawing)\n",
    "        self.drawing.save()\n",
    "        \n",
    "#enhance_method(KGMLCanvas, 'draw',  new_draw)\n",
    "setattr(KGMLCanvas, 'new_draw', new_draw)\n",
    "\n",
    "s = \"\"\n",
    "for pair in superpathwayDict[superpathway]:\n",
    "    pathwayID = pair[0]\n",
    "    pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "    \n",
    "    relations = []\n",
    "    for relation in rel[pathwayID]:\n",
    "        relations += [relation]\n",
    "\n",
    "    canvas = KGMLCanvas(pathway)\n",
    "\n",
    "    for k in pathway.entries:\n",
    "        name = pathway.entries[k].name\n",
    "        canonicalName = set2can(words2set(name))\n",
    "        if canonicalName in element2id.keys():\n",
    "            node = element2id[canonicalName]\n",
    "            klass = cc[node]\n",
    "            if klass in nontrivialclass2col:\n",
    "                pathway.entries[k].graphics[0].bgcolor = nontrivialclass2col[klass]\n",
    "                pathway.entries[k].graphics[0].fgcolor = nontrivialclass2col[klass]\n",
    "            else:\n",
    "                pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "                pathway.enries[k].graphics[0].fgcolor = \"#FFFFFF\"\n",
    "            s += pathwayID + \"\\t\" + str(node) + \"\\t\" + str(klass) +\"\\t\"+pathway.entries[k].graphics[0].name+\"\\t\"+canonicalName + \"\\n\"\n",
    "    \n",
    "            \n",
    "    canvas.import_imagemap = True\n",
    "    pdfName = organism + pathwayID + \".pdf\"\n",
    "    pathlib.Path(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway + \"-col\")).mkdir(parents=True, exist_ok=True)\n",
    "    canvas.new_draw(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway+ \"-col\", pdfName), relations, elaborateArc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "73c98af8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathwayAnomalies = pd.DataFrame(columns=[\n",
    "        \"Pathway ID\",\n",
    "        \"Pathway\",\n",
    "        \"Average square difference\",\n",
    "        ])\n",
    "for pair in superpathwayDict[superpathway]:\n",
    "    pathwayID = pair[0]\n",
    "    pathwayName= pair[1]\n",
    "    ssqd = 0\n",
    "    count = 0\n",
    "    for relation in rel[pathwayID]:\n",
    "        source = element2id[set2can(relation.source)]\n",
    "        for targ in relation.target:\n",
    "            target = element2id[set2can(targ)]\n",
    "            tt = relcorrel[(relcorrel[\"Source node\"]==source) & (relcorrel[\"Target node\"]==target)]\n",
    "            if len(tt) == 1 and tt[\"P-value\"].to_list()[0] <= confidence:\n",
    "                cd = tt[\"Correlation difference (H-D)\"].to_list()[0]\n",
    "                ssqd += cd * cd\n",
    "            count += 1\n",
    "    if count > 0:\n",
    "        pathwayAnomalies.loc[len(pathwayAnomalies)]=[pathwayID, pathwayName, ssqd]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ff411669-e133-48ac-86b9-dcc350d4d957",
   "metadata": {},
   "outputs": [],
   "source": [
    "relPathwayAnomalies = os.path.join(inputDirectory, \"pathway-anomalies-\" + correlationFunctionName + \".csv\")\n",
    "pathwayAnomalies.sort_values(by=[\"Average square difference\"], ascending=False).to_csv(relPathwayAnomalies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a24f06ae-b8e0-4eb5-a9fd-3fff0b777d17",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qf",
   "language": "python",
   "name": "qf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
