{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d4364c55",
   "metadata": {},
   "outputs": [],
   "source": [
    "from hypergraphs import *\n",
    "from kegg import *\n",
    "import pandas as pd\n",
    "import os\n",
    "import logging\n",
    "\n",
    "from Bio.KEGG.KGML.KGML_parser import read\n",
    "import Bio.KEGG.KGML.KGML_pathway\n",
    "from ast import literal_eval\n",
    "from urllib.request import urlopen\n",
    "from urllib.parse import quote_plus\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "from Levenshtein import ratio\n",
    "import scipy.stats\n",
    "import pathlib\n",
    "from qf.cc import cardon_crochemore_colored, cardon_crochemore\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6b13e6bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "organismDict = {\"hsa\": \"Homo sapiens (human)\"}\n",
    "outputDirectory = \"/Users/boldi/Desktop/pw/\"\n",
    "inputDirectory = \"../../LaTeX/Data/KEGG-Pathways/Nov2023/\"\n",
    "dataDirectory = \"../../LaTeX/Data/KEGG-Pathways\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3e6caae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pws = pd.read_csv(os.path.join(inputDirectory, \"KEGG-pathways.txt\"), header=0, \n",
    "            names=[\"Pathway ID\", \"Pathway type\", \"Pathway name\"], dtype=\"string\", keep_default_na=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "199b2487",
   "metadata": {},
   "outputs": [],
   "source": [
    "superpathwayDict = {\n",
    "    \"Involved\": [(x,y) for x,y in zip(pws[\"Pathway ID\"], pws[\"Pathway name\"])]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bce1e1d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "superpathway = \"Involved\"\n",
    "organism = \"hsa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cb8e2d96",
   "metadata": {},
   "outputs": [],
   "source": [
    "safeSuperpathway = makesafe(superpathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "47a38c39",
   "metadata": {},
   "outputs": [],
   "source": [
    "def words2set(words):\n",
    "    \"\"\"\n",
    "        Given a space-separated list of words, returns it as a set of words\n",
    "    \"\"\"\n",
    "    return set(words.split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e0a74d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import NamedTuple\n",
    "\n",
    "class Relation(NamedTuple):\n",
    "    \"\"\"\n",
    "        Represents a relation. It is characterized by:\n",
    "        - a type (e.g. \"PPRel\") \n",
    "        - a list of subtypes (e.g. [('activation', '-->'), ('indirect effect', '..>')])\n",
    "        - a set of components (source)\n",
    "        - a list of sets of components (target)\n",
    "    \"\"\"\n",
    "    relType: str\n",
    "    relSubtypes: list\n",
    "    source: set\n",
    "    target: list\n",
    "    sourceg: Bio.KEGG.KGML.KGML_pathway.Graphics\n",
    "    targetg: Bio.KEGG.KGML.KGML_pathway.Graphics\n",
    "        \n",
    "    @classmethod\n",
    "    def fromKEGG(cls, relation, pathway):\n",
    "        \"\"\"\n",
    "            Constructor: builds a Relation from a KEGG relation\n",
    "        \"\"\"\n",
    "        if relation.entry2.name == \"undefined\":\n",
    "            s = []\n",
    "            for component in relation.entry2.components:\n",
    "                s += [words2set(pathway.entries[component.id].name)]\n",
    "            targetg = pathway.entries[component.id].graphics[0]\n",
    "        else:\n",
    "            s = [words2set(relation.entry2.name)]\n",
    "            targetg = relation.entry2.graphics[0]\n",
    "        return cls(relation.type, relation.subtypes, words2set(relation.entry1.name), s, targetg, relation.entry1.graphics[0])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ba5d2e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsFromKEGGpathway(pathway):\n",
    "    \"\"\"\n",
    "        Given a KEGG pathway, returns the list of relations it contains (represented as Relation).\n",
    "    \"\"\"\n",
    "    return [Relation.fromKEGG(r, pathway) for r in pathway.relations]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "751bc16c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pathwayFromFile(dataDirectory, organism, pathwayID):\n",
    "    \"\"\"\n",
    "        Load and return a KEGG pathway, given the data directory, the organism specifying the subdirectory,\n",
    "        and the pathwayID (the filename should be organism+pathwayID+\"xml\".\n",
    "    \"\"\"\n",
    "    return KGML_parser.read(open(os.path.join(dataDirectory, organism, organism+pathwayID+\".xml\"), \"r\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bd27fdc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsFromFile(dataDirectory, organism, pathwayID):\n",
    "    \"\"\"\n",
    "        Given a pathway file (specified by a root dataDirectory, the name of the organism subdirectory, and the name\n",
    "        of the file [organism+pathwayID.xml]), reads it and returns the list of its relations.\n",
    "    \"\"\"\n",
    "    pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "    return relationsFromKEGGpathway(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eaa01766",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relationsForSuperpathway(dataDirectory, organism, superpathwayDict, superpathway):\n",
    "    \"\"\"\n",
    "        Given a superpathDict (whose keys are superpathway names and whose values are list of\n",
    "        pairs (pathwayID, pathwayName)), reads all the pathway file for a specific superpathway name\n",
    "        and returns a dictionary whose keys are the pathwayID's and whose values are the list of relations.\n",
    "    \"\"\"\n",
    "    rel = {}\n",
    "    for k,v in superpathwayDict[superpathway]:\n",
    "        #logging.info(\"Reading\", k, v)\n",
    "        rel[k] = relationsFromFile(dataDirectory, organism, k)\n",
    "    return rel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "15479a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def subtypes(rel):\n",
    "    \"\"\"\n",
    "        Given a map whose values are relations, accumulate all subtypes appearing and assign them a number.\n",
    "        The result is a map from subtype string to number.\n",
    "    \"\"\"\n",
    "    all_subtypes = set([])\n",
    "    for pathwayid, relations in rel.items():\n",
    "        for relation in relations:\n",
    "            all_subtypes |= set([str(relation.relSubtypes)])\n",
    "        \n",
    "    s = sorted(list(all_subtypes))\n",
    "    subtype2color = {v:k for k,v in enumerate(s)}\n",
    "\n",
    "    return subtype2color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1e229eaf-d667-466c-8238-8205257b0ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set2can(s):\n",
    "    \"\"\"\n",
    "        Convert a set to a string in a canonical way.\n",
    "    \"\"\"\n",
    "    return str(sorted(list(s)))\n",
    "\n",
    "def can2set(c):\n",
    "    \"\"\"\n",
    "        Does the converse of set2can.\n",
    "    \"\"\"\n",
    "    return set(literal_eval(c))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75b272b9-b220-492b-adf6-509f9952842f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indices(rel):\n",
    "    \"\"\"\n",
    "        Given a map whose values are lists of relations, it considers all the relations one by one, and attributes a unique id \n",
    "        to each element (i.e., set of components appearing as source or in the target of some relation) and block (the set of\n",
    "        element id appearing as target of some relation).\n",
    "        This function returns the dictionaries to move from/to an element or block to the corresponding id.\n",
    "        Elements are string representations of sorted lists of strings.\n",
    "        Blocks are string representations of sorted lists of ints.\n",
    "    \"\"\"\n",
    "    element2id = {}\n",
    "    id2element = {}\n",
    "    block2id = {}\n",
    "    id2block = {}\n",
    "\n",
    "    for relations in rel.values():\n",
    "        for relation in relations:\n",
    "            s = set2can(relation.source)\n",
    "            if s not in element2id.keys():\n",
    "                element2id[s] = len(element2id)\n",
    "            targetset = set([])\n",
    "            for targ in relation.target:\n",
    "                t = set2can(targ)\n",
    "                if t not in element2id.keys():\n",
    "                    element2id[t] = len(element2id)\n",
    "                targetset |= set([element2id[t]])\n",
    "            ts = set2can(targetset)\n",
    "            if ts not in block2id.keys():\n",
    "                block2id[ts] = len(block2id)\n",
    "\n",
    "    for k,v in element2id.items():\n",
    "        id2element[v] = k\n",
    "    for k,v in block2id.items():\n",
    "        id2block[v] = k\n",
    "    return element2id, id2element, block2id, id2block\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f9c3f128-cf97-4bb3-8ce5-65c49f865d98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_gene_KEGG(geneID):\n",
    "    \"\"\"\n",
    "        Search for gene on KEGG. \n",
    "        \n",
    "        Returns list of names.\n",
    "    \"\"\"\n",
    "    url1 = \"https://www.kegg.jp/entry/\"+geneID\n",
    "    response1 = requests.get(url1, allow_redirects=False)    \n",
    "    soup = BeautifulSoup(response1.text, \"html.parser\")\n",
    "    tds = [tds for tds in soup.find_all(\"td\", {\"class\": \"td11 defd\"})]\n",
    "    names = tds[0].getText().strip().split(\", \")\n",
    "    return names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0a11bbae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build all data needed for the relations of the superpathway\n",
    "rel = relationsForSuperpathway(dataDirectory, organism, superpathwayDict, superpathway)\n",
    "element2id, id2element, block2id, id2block = indices(rel)\n",
    "subtype2color = subtypes(rel)\n",
    "elements = list(set.union(*[can2set(k) for k in element2id.keys()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a1be8c2d-c5da-4e04-bf67-7e43b1a71b7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading gene names\n"
     ]
    }
   ],
   "source": [
    "# name2gene is a function mapping names to gene IDs (hsa:xxxx)\n",
    "# e.g. KCNE3, BRGDA6 etc are all mapped to hsa:10008 (see https://www.kegg.jp/entry/hsa:10008)\n",
    "# This map was produced by scraping kegg, but this has been done only once.\n",
    "# After that, we just read a pkl file\n",
    "\n",
    "name2keggFilename = os.path.join(inputDirectory, \"name2kegg.pkl\")\n",
    "\n",
    "if os.path.exists(name2keggFilename):\n",
    "    print(\"Reading gene names\")\n",
    "    with open(name2keggFilename, \"rb\") as handle:\n",
    "        name2gene = pickle.load(handle)\n",
    "else:\n",
    "    print(\"Producing gene names\")\n",
    "    d = {}\n",
    "    nelements = len(elements)\n",
    "    for count,element in enumerate(elements):\n",
    "        if count % 10 == 0:\n",
    "            print(f\"{count}/{nelements}\")\n",
    "        if element.startswith(\"hsa\"):\n",
    "            names = search_gene_KEGG(element)\n",
    "            for name in names:\n",
    "                d[name] = element\n",
    "    with open(name2keggFilename, \"wb\") as handle:\n",
    "        pickle.dump(d, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "    name2gene = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "74d19326",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading patient data\n"
     ]
    }
   ],
   "source": [
    "# Read patient data\n",
    "print(\"Reading patient data\")\n",
    "genes = pd.read_csv(os.path.join(inputDirectory, \"genes.csv\"))\n",
    "patients = genes[\"name\"].values.tolist()\n",
    "classification = pd.read_csv(os.path.join(inputDirectory, \"classification.csv\"))\n",
    "\n",
    "patient2class = {patient: classification.loc[classification[\"name\"]==patient][\"classification\"].values.flatten().tolist()[0] for patient in patients}\n",
    "numHealthy = sum([x == 'healthy' for x in patient2class.values()])\n",
    "numDiseased = sum([x == 'diseased' for x in patient2class.values()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "34acce33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dividing genes into two frames (one for healthy individuals and one for diseased individuals)\n",
    "genesHealthy = genes.loc[genes[\"name\"].apply(lambda x: patient2class[x])==\"healthy\"]\n",
    "genesDiseased = genes.loc[genes[\"name\"].apply(lambda x: patient2class[x])==\"diseased\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1c90f901-1413-4356-bfa8-faf25fc494df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building the relation graph\n"
     ]
    }
   ],
   "source": [
    "# Building the graph of all relations contained in rel values\n",
    "# Nodes are IDs, and node x correspond to the set of genes (and/or compound) id2element[x]\n",
    "print(\"Building the relation graph\")\n",
    "G = nx.DiGraph() \n",
    "for relations in rel.values():\n",
    "    for relation in relations:\n",
    "        source = element2id[set2can(relation.source)]\n",
    "        for targ in relation.target:\n",
    "            target = element2id[set2can(targ)]\n",
    "            G.add_edge(source, target, color=subtype2color[str(relation.relSubtypes)])\n",
    "            \n",
    "\n",
    "# Computing the minimum fibres\n",
    "cc=cardon_crochemore_colored(G)\n",
    "\n",
    "colors = list(matplotlib.colors.CSS4_COLORS.values())\n",
    "usedColors = 0\n",
    "\n",
    "cc2fibresize = {}\n",
    "cc2fibre = {}\n",
    "ccnontrivial = {}\n",
    "nontrivialclass2col = {}\n",
    "\n",
    "fibres = []\n",
    "node2fibre = {}\n",
    "for v in set(cc.values()):\n",
    "    fibre = [x for x in G.nodes() if cc[x] == v]\n",
    "    fibres += [fibre]\n",
    "    for x in fibre:\n",
    "        node2fibre[x] = fibre\n",
    "    cc2fibre[v] = fibre\n",
    "    cc2fibresize[v] = len(fibre)\n",
    "    if len(fibre) > 0:\n",
    "        ccnontrivial[v] = cc[v]\n",
    "        nontrivialclass2col[v] = colors[usedColors]\n",
    "        usedColors += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d51dbd7f-35b9-444a-9238-3af727c06411",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping genes to columns\n"
     ]
    }
   ],
   "source": [
    "# Build node2columns, mapping each node of G to the list of genes corresponding to it\n",
    "# for which we know a value (in the genes dataframe), sorted\n",
    "#\n",
    "\n",
    "print(\"Mapping genes to columns\")\n",
    "namedGenes = set(name2gene.values())     \n",
    "columns = set(genes.columns.values.tolist()[2:])\n",
    "node2columns = {}\n",
    "for node in G.nodes():\n",
    "    nodeGenes = can2set(id2element[node])\n",
    "    s = []\n",
    "    for gene in nodeGenes & namedGenes:\n",
    "        for k,v in name2gene.items():\n",
    "            if v == gene:\n",
    "                if k in columns:\n",
    "                    s += [k]\n",
    "    node2columns[node]=sorted(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "05b96661",
   "metadata": {},
   "outputs": [],
   "source": [
    "s = \"\"\n",
    "for pair in superpathwayDict[superpathway]:\n",
    "    pathwayID = pair[0]\n",
    "    pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "    canvas = KGMLCanvas(pathway)\n",
    "    for k in pathway.entries:\n",
    "        name = pathway.entries[k].name\n",
    "        canonicalName = set2can(words2set(name))\n",
    "        if canonicalName in element2id.keys():\n",
    "            node = element2id[canonicalName]\n",
    "            klass = cc[node]\n",
    "            if klass in nontrivialclass2col:\n",
    "                pathway.entries[k].graphics[0].bgcolor = nontrivialclass2col[klass]\n",
    "                pathway.entries[k].graphics[0].fgcolor = nontrivialclass2col[klass]\n",
    "            else:\n",
    "                pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "                pathway.enries[k].graphics[0].fgcolor = \"#FFFFFF\"\n",
    "            s += pathwayID + \"\\t\" + str(node) + \"\\t\" + str(klass) +\"\\t\"+pathway.entries[k].graphics[0].name+\"\\t\"+canonicalName + \"\\n\"\n",
    "    canvas.import_imagemap = True\n",
    "    pdfName = organism + pathwayID + \".pdf\"\n",
    "    pathlib.Path(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway)).mkdir(parents=True, exist_ok=True)\n",
    "    canvas.draw(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway, pdfName))\n",
    "with open(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway, \"all.tsv\"), \"wt\") as file:\n",
    "    file.write(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "73ce2fd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def columns2nodes(columns, node2columns, ignoreEmpty = True):\n",
    "    \"\"\"\n",
    "        Returns a map from columns to the nodes where the column appears.\n",
    "    \"\"\"\n",
    "    res = {} \n",
    "    for c in columns:\n",
    "        s = set([])\n",
    "        for no, co in node2columns.items():\n",
    "            if c in co:\n",
    "                s |= set([no])\n",
    "        if len(s) > 0 or not ignoreEmpty:\n",
    "            res[c] = s\n",
    "    return res\n",
    "\n",
    "c2n = columns2nodes(columns, node2columns)\n",
    "usableColumns = list(c2n.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "17ab34ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def areRelated(column1, column2, column2nodes, node2color, inDegThreshold = 0):\n",
    "    nodes1 = column2nodes[column1]\n",
    "    nodes2 = column2nodes[column2]\n",
    "    colors1 = set([node2color[node] for node in nodes1])\n",
    "    colors2 = set([node2color[node] for node in nodes2])\n",
    "    commonColors = colors1 & colors2\n",
    "    if len(commonColors) == 0:\n",
    "        return False\n",
    "    maxInDeg = max([G.in_degree(cc2fibre[c][0]) for c in commonColors])\n",
    "    return maxInDeg >= inDegThreshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "00742eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tauForPair(genes, set1, set2):\n",
    "    \"\"\"\n",
    "        Given a dataframe genes (i.e., a set of patients), and two sets of column names (set1, set2), \n",
    "        build the Kendall's tau\n",
    "        for every pair of columns in set1 and set2 and return the average statistic and its std.\n",
    "    \"\"\"\n",
    "    taus = []\n",
    "    for gene1 in set1:\n",
    "        for gene2 in set2:\n",
    "            expr1 = genes.loc[:,gene1].values.tolist()\n",
    "            expr2 = genes.loc[:,gene2].values.tolist()\n",
    "            taus += [scipy.stats.kendalltau(expr1, expr2).statistic]\n",
    "    return pd.DataFrame(taus).mean().values[0], pd.DataFrame(taus).std().values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b3232726",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sampleTauForPairDiffsDist(genes, set1, set2, numberOfRows1, numberOfSamples = 100):\n",
    "    \"\"\"\n",
    "        Repeatedly (for numberOfSamples times) do the following:\n",
    "        - select numberOfRows1 rows of genes, and then compute tauForPair on those rows for set1 and set2\n",
    "        - do the same for remaining rows\n",
    "        - take the difference between the two resulting set of taus.\n",
    "        Return the mean and std of differences.\n",
    "    \"\"\"\n",
    "    diffs = []\n",
    "    \n",
    "    for i in range(numberOfSamples):\n",
    "        genes1 = random.sample(range(len(genes)), numberOfRows1)\n",
    "        genes2 = list(set(range(len(genes)))-set(genes1))\n",
    "        mt1, mstd1 = tauForPair(genes.iloc[genes1], set1, set2)\n",
    "        mt2, mstd2 = tauForPair(genes.iloc[genes2], set1, set2)\n",
    "        diffs += [mt1-mt2]\n",
    "    return diffs\n",
    "\n",
    "\n",
    "def tauForPairDiff(genes1, genes2, set1, set2):\n",
    "    \"\"\"\n",
    "        Compute tauForPair for set1 and set2 over the set of rows genes1 and genes2, and return\n",
    "        the difference of taus.\n",
    "    \"\"\"\n",
    "    mt1, mstd1 = tauForPair(genes1, set1, set2)\n",
    "    mt2, mstd2 = tauForPair(genes2, set1, set2)\n",
    "    return mt1-mt2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "2d3473f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[('phosphorylation', '+p')]]: (['PAK1', 'PAK4', 'PAK5', 'PAK6'], ['MAP2K1']) --- -0.1533323674297305 vs. 0.024255288322225434±0.13933715409632205->Diseased are significantly more correlated\n",
      "[[('phosphorylation', '+p')]]: (['RAF1'], ['MAP2K1']) --- -0.444566792234135 vs. -0.0007215299913068884±0.18076883712194708->Diseased are significantly more correlated\n",
      "[[('inhibition', '--|'), ('phosphorylation', '+p')]]: (['RAF1'], ['BAD']) --- -0.27238481599536357 vs. 0.007866125760649104±0.16362551421237403->Diseased are significantly more correlated\n",
      "[[('phosphorylation', '+p')]]: (['MAPK1', 'MAPK3'], ['BCL2']) --- -0.1897131266299623 vs. 0.007467400753404837±0.13190435339302747->Diseased are significantly more correlated\n",
      "[[('activation', '-->'), ('indirect effect', '..>')]]: (['MAPK1', 'MAPK3'], ['PLA2G4A', 'PLA2G4C', 'PLA2G4D', 'PLA2G4E', 'PLA2G4F']) --- 0.08977472756562416 vs. 0.0034779829158909027±0.07347794710741956->Healthy are significantly more correlated\n",
      "[[('expression', '-->')]]: (['ELK1'], ['CCND1', 'CCND2', 'CCND3']) --- -0.1337585671709467 vs. -0.004129098255935704±0.12026580130872594->Diseased are significantly more correlated\n",
      "[[('inhibition', '--|'), ('phosphorylation', '+p')]]: (['AKT1', 'AKT2', 'AKT3'], ['GSK3B']) --- -0.10588235294117647 vs. -0.004719791364821799±0.08559331300780897->Diseased are significantly more correlated\n",
      "[[('activation', '-->'), ('phosphorylation', '+p')]]: (['AKT1', 'AKT2', 'AKT3'], ['NOS3']) --- -0.14890369941079878 vs. 0.00283512025499854±0.08930246773926713->Diseased are significantly more correlated\n",
      "[[('indirect effect', '..>')]]: (['PTK2'], ['CAPN2']) --- 0.2824108953926398 vs. 0.002257316719791382±0.20497094394761997->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['PTK2'], ['GRB2']) --- -0.28861199652274705 vs. -0.003259345117357284±0.1923630532508135->Diseased are significantly more correlated\n",
      "[[('phosphorylation', '+p')]]: (['PTK2'], ['BCAR1']) --- 0.09921761808171542 vs. -0.06831527093596061±0.1673116692770267->Healthy are significantly more correlated\n",
      "[[('activation', '-->'), ('phosphorylation', '+p')]]: (['FYN', 'SLK'], ['FCER1G']) --- -0.23309185743262822 vs. -0.009312662996233002±0.13995325663661437->Diseased are significantly more correlated\n",
      "[[('indirect effect', '..>')]]: (['PXN'], ['ACTB', 'ACTG1']) --- 0.20483917705013044 vs. 0.02371022891915391±0.17007418016061698->Healthy are significantly more correlated\n",
      "[[('binding/association', '---')]]: (['PXN'], ['PARVA', 'PARVB', 'PARVG']) --- 0.221172606973824 vs. 0.005978170578576272±0.11758567379558348->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['ACTB', 'ACTG1'], ['ITGA2B', 'ITGB3']) --- -0.15859171254708782 vs. -0.0014322867303212465±0.13478903179383858->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['RAC1', 'RAC2', 'RAC3'], ['PAK1', 'PAK4', 'PAK5', 'PAK6']) --- -0.07612769245629285 vs. -0.0022640780450111225±0.07012598251352173->Diseased are significantly more correlated\n",
      "[[('phosphorylation', '+p')]]: (['MAPK10', 'MAPK8', 'MAPK9'], ['JUN']) --- 0.1631990727325413 vs. 0.02188969380855792±0.12988753473807166->Healthy are significantly more correlated\n",
      "[[('expression', '-->')]]: (['JUN'], ['CCND1', 'CCND2', 'CCND3']) --- 0.12011543414837661 vs. -0.0028407312431295052±0.1077673895847358->Healthy are significantly more correlated\n",
      "[[('compound', 24)]]: (['PIK3CA', 'PIK3CB', 'PIK3CD', 'PIK3R1', 'PIK3R2', 'PIK3R3'], ['PDPK1']) --- 0.11282719984545542 vs. 0.0001414881708257143±0.05999572111384802->Healthy are significantly more correlated\n",
      "[[('activation', '-->'), ('indirect effect', '..>')]]: (['PIK3CA', 'PIK3CB', 'PIK3CD', 'PIK3R1', 'PIK3R2', 'PIK3R3'], ['PLCG2']) --- -0.052883222254419 vs. 0.007355031967301937±0.05196768484648384->Diseased are significantly more correlated\n",
      "[[('activation', '-->'), ('indirect effect', '..>')]]: (['PIK3CA', 'PIK3CB', 'PIK3CD', 'PIK3R1', 'PIK3R2', 'PIK3R3'], ['LCP2']) --- -0.0594900028977108 vs. -0.008409743028601124±0.04635413964461296->Diseased are significantly more correlated\n",
      "[[('binding/association', '---')]]: (['ILK'], ['PXN']) --- -0.2592871631411185 vs. -0.011049550854824697±0.16520652606436678->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['RHOA'], ['ROCK1', 'ROCK2']) --- 0.14212796029516256 vs. 0.005662213352074579±0.11412210277690142->Healthy are significantly more correlated\n",
      "[[('inhibition', '--|')]]: (['PRKCA', 'PRKCB'], ['EGLN1', 'EGLN2', 'EGLN3']) --- -0.10072442770211533 vs. 0.0006791268231430488±0.07260543057450737->Diseased are significantly more correlated\n",
      "[[('binding/association', '---')]]: (['ACTN1', 'ACTN4'], ['ITGA1', 'ITGA10', 'ITGA11', 'ITGA2', 'ITGA2B', 'ITGA3', 'ITGA4', 'ITGA5', 'ITGA7', 'ITGA8', 'ITGA9', 'ITGAV']) --- -0.04122540065405135 vs. 0.00792780866037518±0.039437251581894046->Diseased are significantly more correlated\n",
      "[[('binding/association', '---')]]: (['ZYX'], ['ACTN1', 'ACTN4']) --- -0.18800347725297023 vs. -0.008770318425751851±0.10962210782176755->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['CDC42'], ['PAK1', 'PAK4', 'PAK5', 'PAK6']) --- -0.2326137351492321 vs. -0.00651666183714863±0.12712371669241507->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['VASP'], ['ACTB', 'ACTG1']) --- 0.21926977687626775 vs. 0.004390321645899758±0.17570269545611006->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['CAV1', 'CAV2', 'CAV3'], ['FYN', 'SLK']) --- 0.0746450304259635 vs. -0.0016988312566406092±0.06611486013657976->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['BCAR1'], ['CRK', 'CRKL']) --- -0.19842893363646308 vs. 0.003988416500664342±0.1253851406988484->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['CRK', 'CRKL'], ['PXN']) --- 0.20588240172417813 vs. 0.009277748326994916±0.16081967272695105->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['RAP1A', 'RAP1B'], ['MAPK10', 'MAPK8', 'MAPK9']) --- 0.12553734128833818 vs. 0.005289163033608212±0.08413486642816607->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['RAP1A', 'RAP1B'], ['APBB1IP']) --- 0.12123317405561125 vs. 0.0001828292385463598±0.11907518893185506->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['SOS1', 'SOS2'], ['HRAS']) --- 0.17371776296725588 vs. -0.003986670530281049±0.12264377616077778->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['SOS1', 'SOS2'], ['HRAS', 'KRAS', 'NRAS']) --- 0.0852796290930165 vs. 0.008369361537718539±0.07282091404008748->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['CHAD', 'COL1A1', 'COL1A2', 'COL4A1', 'COL4A2', 'COL4A3', 'COL4A4', 'COL4A5', 'COL4A6', 'COL6A1', 'COL6A3', 'COL6A6', 'COL9A1', 'COL9A2', 'COMP', 'FN1', 'IBSP', 'LAMA1', 'LAMA2', 'LAMA3', 'LAMA4', 'LAMA5', 'LAMB1', 'LAMB2', 'LAMB3', 'LAMC1', 'LAMC2', 'LAMC3', 'RELN', 'SPP1', 'THBS1', 'THBS2', 'THBS3', 'THBS4', 'TNC', 'TNN', 'TNXB', 'VWF'], ['ITGA1', 'ITGA10', 'ITGA11', 'ITGA2', 'ITGA2B', 'ITGA3', 'ITGA4', 'ITGA5', 'ITGA7', 'ITGA8', 'ITGA9', 'ITGAV']) --- 0.016699747988213273 vs. -0.0014088087191029396±0.01734431922062581->Healthy are significantly more correlated\n",
      "[[('inhibition', '--|')]]: (['THEM4'], ['AKT1', 'AKT2', 'AKT3']) --- -0.13232879358640004 vs. 0.009172027431662318±0.11483860422222654->Diseased are significantly more correlated\n",
      "[[('expression', '-->')]]: (['NFKB1', 'RELA'], ['MYB']) --- 0.18284130431350276 vs. -0.004559646186629117±0.11626082969414483->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['RHEB'], ['MLST8']) --- 0.15520139090118806 vs. -0.010259634888438118±0.1595046625371213->Healthy are significantly more correlated\n",
      "[[('activation', '-->'), ('phosphorylation', '+p')]]: (['SYK'], ['CD19']) --- 0.36765584214504116 vs. 0.001275295641847623±0.1614938133286478->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['SYK'], ['PLCG2']) --- -0.16256167392068305 vs. 0.005268538142395257±0.16169242934341327->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['BAP1', 'MAGI1', 'MAGI2'], ['PTEN', 'TEP1']) --- 0.060272384815995346 vs. -0.003491298073097974±0.04327854454806435->Healthy are significantly more correlated\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[('inhibition', '--|')]]: (['CDKN1A'], ['CCND1', 'CCND2', 'CCND3', 'CCNE1', 'CCNE2']) --- -0.1583535277073205 vs. -0.0041794800302074045±0.08353359261529537->Diseased are significantly more correlated\n",
      "[[('inhibition', '--|')]]: (['CDKN1A'], ['CDK2', 'CDK4', 'CDK6']) --- 0.162264893950764 vs. 0.0036717273108048177±0.14510733633386375->Healthy are significantly more correlated\n",
      "[[('expression', '-->')]]: (['FOXO3'], ['BCL2L11']) --- 0.20347725297015354 vs. 0.001018255578093309±0.18630911363253655->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['MTOR'], ['RPS6KB1', 'RPS6KB2']) --- 0.15725876557519558 vs. -0.011964572142013168±0.11009777767693044->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['STK11'], ['PRKAA1', 'PRKAA2']) --- 0.35986670530281073 vs. 0.009175682081002542±0.13623688726622707->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['PRKG1', 'PRKG2'], ['BST1', 'CD38']) --- 0.1454071283685888 vs. 0.009686177919443645±0.1065689248847202->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['ADRB1', 'ADRB2', 'ADRB3'], ['GNAS']) --- -0.19248526997005694 vs. -0.008025906584802355±0.10709728325520289->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['GNAS'], ['ADCY1', 'ADCY2', 'ADCY3', 'ADCY4', 'ADCY5', 'ADCY6', 'ADCY7', 'ADCY8', 'ADCY9']) --- -0.08821657621521248 vs. 0.004510515651546733±0.0641876795237314->Diseased are significantly more correlated\n",
      "[[('indirect effect', '..>')]]: (['PRKACA', 'PRKACB', 'PRKACG'], ['VAMP2']) --- 0.16294246983176974 vs. 9.365343766675961e-05±0.1231141472692276->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['ADRA1A', 'ADRA1B'], ['CMC1', 'GNAQ']) --- 0.08516372066067808 vs. 0.0018341911798647919±0.0794335481196599->Healthy are significantly more correlated\n",
      "[[('activation', '-->'), ('indirect effect', '..>')]]: (['IL6R'], ['STAT3']) --- -0.25911330049261083 vs. 0.00046942915097071563±0.15308133558183473->Diseased are significantly more correlated\n",
      "[[('activation', '-->'), ('indirect effect', '..>')]]: (['EGFR', 'ERBB2', 'IGF1R', 'INSR'], ['PIK3CA', 'PIK3CB', 'PIK3CD', 'PIK3R1', 'PIK3R2', 'PIK3R3']) --- 0.047346355598986076 vs. -0.0044899683197403215±0.029892077867177316->Healthy are significantly more correlated\n",
      "[[('activation', '-->'), ('indirect effect', '..>')]]: (['IFNGR1', 'IFNGR2'], ['NFKB1', 'RELA']) --- -0.12436594013936275 vs. -0.0009407328983441305±0.08775225524360314->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['EGF', 'IGF1', 'INS'], ['EGFR', 'ERBB2', 'IGF1R', 'INSR']) --- 0.0639165578785695 vs. -0.0007254336619201082±0.037552348959978714->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['MKNK1', 'MKNK2'], ['EIF4E', 'EIF4E1B', 'EIF4E2']) --- 0.1013329469718922 vs. -0.00234635904185367±0.07502969253194694->Healthy are significantly more correlated\n",
      "[[('inhibition', '--|'), ('phosphorylation', '+p')]]: (['PDK1'], ['PDHA1', 'PDHA2', 'PDHB']) --- 0.10617709650124961 vs. -0.004994825578251202±0.10213392006560601->Healthy are significantly more correlated\n",
      "[[('inhibition', '--|'), ('indirect effect', '..>')]]: (['HSPB1'], ['NFKB1', 'RELA']) --- 0.12738436770184175 vs. -0.00742011562245041±0.10445745986405693->Healthy are significantly more correlated\n",
      "[[('activation', '-->'), ('indirect effect', '..>')]]: (['IL1R1', 'IL1R2'], ['NFKB1', 'RELA']) --- -0.11885777465716525 vs. 0.004266118462878571±0.09742220023730669->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['GNA13'], ['SRC']) --- 0.2860620110113011 vs. -0.029956534337873052±0.21347770473269098->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['F2R', 'F2RL3'], ['GNA13']) --- 0.17971602434077075 vs. -0.0019432048681541669±0.12122576206786613->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['APBB1IP'], ['TLN1', 'TLN2']) --- 0.21706751666183713 vs. -0.0002983125569399287±0.11145397157985586->Healthy are significantly more correlated\n",
      "[[('activation', '-->'), ('indirect effect', '..>')]]: (['RASGRP1', 'RASGRP2'], ['PLA2G4A', 'PLA2G4C', 'PLA2G4D', 'PLA2G4E', 'PLA2G4F']) --- -0.12171292785593295 vs. -0.001093954631649998±0.07785402892699118->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['ITGA2B', 'ITGB3'], ['GNA13']) --- -0.19681251811069256 vs. -0.0016403462478421793±0.13471777536645804->Diseased are significantly more correlated\n",
      "[[('binding/association', '---')]]: (['ITGA2B', 'ITGB3'], ['FGA', 'FGB']) --- 0.10782381918284553 vs. 0.005382842597150474±0.08446236224462521->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['FERMT3'], ['ITGA2B', 'ITGB3']) --- -0.1813387423935091 vs. 0.014257922487637376±0.1362571530676633->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['GNAI1', 'GNAI2', 'GNAI3'], ['PIK3CG', 'PIK3R5', 'PIK3R6']) --- -0.07366666784930428 vs. -0.003536357391589804±0.06203035378533995->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['MAPK11', 'MAPK12', 'MAPK13', 'MAPK14'], ['MAPK1', 'MAPK3']) --- 0.07619530570849029 vs. -0.006815415821501014±0.06836802090578957->Healthy are significantly more correlated\n",
      "[[('activation', '-->')]]: (['VWF'], ['GP9']) --- -0.4622268537416605 vs. 0.0002971922203887342±0.18398954808071089->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['VWF'], ['GP5']) --- -0.2514994493423496 vs. 0.03385711634545199±0.18880019917620766->Diseased are significantly more correlated\n",
      "[[('activation', '-->')]]: (['VWF'], ['GP1BA']) --- -0.18984279040123606 vs. 0.02709883234413724±0.16506626099572808->Diseased are significantly more correlated\n"
     ]
    }
   ],
   "source": [
    "outd = {}\n",
    "for s,t,d in G.edges(data=True):\n",
    "    subtypeName = [k for k,v in subtype2color.items() if d[\"color\"]==v][0]\n",
    "    sg = node2columns[s]\n",
    "    tg = node2columns[t]\n",
    "    if len(sg) == 0 or len(tg) == 0:\n",
    "        continue\n",
    "    taudiff = tauForPairDiff(genesHealthy, genesDiseased, sg, tg)\n",
    "    tt = sampleTauForPairDiffsDist(genes, sg, tg, numHealthy)\n",
    "    tauSampleDiff = pd.DataFrame(tt).mean().values[0]\n",
    "    tauSampleDiffStd = pd.DataFrame(tt).std().values[0]\n",
    "    if taudiff >= tauSampleDiff - tauSampleDiffStd and taudiff <= tauSampleDiff + tauSampleDiffStd:\n",
    "        pass#print(sg, tg, taudiff, tauSampleDiff, tauSampleDiffStd,\"->NO\")\n",
    "    else:\n",
    "        outd[f\"({s},{t})\"] = taudiff\n",
    "        if taudiff < tauSampleDiff:\n",
    "            print(f\"[{subtypeName}]: ({sg}, {tg}) --- {taudiff} vs. {tauSampleDiff}±{tauSampleDiffStd}->Diseased are significantly more correlated\")\n",
    "        else:\n",
    "            print(f\"[{subtypeName}]: ({sg}, {tg}) --- {taudiff} vs. {tauSampleDiff}±{tauSampleDiffStd}->Healthy are significantly more correlated\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "32b16bfb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from reportlab.lib.units import inch\n",
    "import PIL.Image\n",
    "from io import BytesIO\n",
    "import reportlab.pdfgen.canvas\n",
    "import reportlab.lib.colors\n",
    "\n",
    "def significantTau(tau):\n",
    "    return abs(tau) > 0.2\n",
    "\n",
    "def convertTauValue(tau):\n",
    "    return 3*math.exp(abs(tau)*3)\n",
    "\n",
    "def convertTauColor(tau):\n",
    "    if tau < 0:\n",
    "        return reportlab.lib.colors.red\n",
    "    else:\n",
    "        return reportlab.lib.colors.green\n",
    "\n",
    "#from __future__ import nested_scopes\n",
    "\n",
    "def get_temp_imagefilename(url):\n",
    "    \"\"\"Return filename of temporary file containing downloaded image.\n",
    "\n",
    "    Create a new temporary file to hold the image file at the passed URL\n",
    "    and return the filename.\n",
    "    \"\"\"\n",
    "    img = urlopen(url).read()\n",
    "    im = PIL.Image.open(BytesIO(img))\n",
    "    # im.transpose(Image.FLIP_TOP_BOTTOM)\n",
    "    f = tempfile.NamedTemporaryFile(delete=False, suffix=\".png\")\n",
    "    fname = f.name\n",
    "    f.close()\n",
    "    im.save(fname, \"PNG\")\n",
    "    return fname\n",
    "\n",
    "\n",
    "#def enhance_method(klass, method_name, replacement):\n",
    "#    'replace a method with an enhanced version'\n",
    "#    method = getattr(klass, method_name)\n",
    "#    def enhanced(*args, **kwds): return replacement(*args, **kwds)\n",
    "#    setattr(klass, method_name, enhanced)\n",
    "\n",
    "def new_draw(self, filename, colrel):\n",
    "        \"\"\"Add the map elements to the drawing.\"\"\"\n",
    "        # Instantiate the drawing, first\n",
    "        # size x_max, y_max for now - we can add margins, later\n",
    "        if self.import_imagemap:\n",
    "            # We're drawing directly on the image, so we set the canvas to the\n",
    "            # same size as the image\n",
    "            if os.path.isfile(self.pathway.image):\n",
    "                imfilename = self.pathway.image\n",
    "            else:\n",
    "                imfilename = get_temp_imagefilename(self.pathway.image)\n",
    "            im = PIL.Image.open(imfilename)\n",
    "            cwidth, cheight = im.size\n",
    "        else:\n",
    "            # No image, so we set the canvas size to accommodate visible\n",
    "            # elements\n",
    "            cwidth, cheight = (self.pathway.bounds[1][0], self.pathway.bounds[1][1])\n",
    "        # Instantiate canvas\n",
    "        self.drawing = reportlab.pdfgen.canvas.Canvas(\n",
    "            filename,\n",
    "            bottomup=0,\n",
    "            pagesize=(\n",
    "                cwidth * (1 + 2 * self.margins[0]),\n",
    "                cheight * (1 + 2 * self.margins[1]),\n",
    "            ),\n",
    "        )\n",
    "        self.drawing.setFont(self.fontname, self.fontsize)\n",
    "        # Transform the canvas to add the margins\n",
    "        self.drawing.translate(\n",
    "            self.margins[0] * self.pathway.bounds[1][0],\n",
    "            self.margins[1] * self.pathway.bounds[1][1],\n",
    "        )\n",
    "        # Add the map image, if required\n",
    "        if self.import_imagemap:\n",
    "            self.drawing.saveState()\n",
    "            self.drawing.scale(1, -1)\n",
    "            self.drawing.translate(0, -cheight)\n",
    "            self.drawing.drawImage(imfilename, 0, 0)\n",
    "            self.drawing.restoreState()\n",
    "        # Add the reactions, compounds and maps\n",
    "        # Maps go on first, to be overlaid by more information.\n",
    "        # By default, they're slightly transparent.\n",
    "        if self.show_maps:\n",
    "            self.__add_maps()\n",
    "        if self.show_reaction_entries:\n",
    "            KGMLCanvas._KGMLCanvas__add_reaction_entries(self)\n",
    "        if self.show_orthologs:\n",
    "            KGMLCanvas._KGMLCanvas__add_orthologs(self)\n",
    "        if self.show_compounds:\n",
    "            KGMLCanvas._KGMLCanvas__add_compounds(self)\n",
    "        if self.show_genes:\n",
    "            KGMLCanvas._KGMLCanvas__add_genes(self)\n",
    "        # TODO: complete draw_relations code\n",
    "        # if self.draw_relations:\n",
    "        #    self.__add_relations()\n",
    "        # Write the pathway map to PDF\n",
    "        for sg,tg,width,color in colrel:\n",
    "            self.drawing.setLineWidth(width)\n",
    "            self.drawing.setStrokeColor(color)\n",
    "            self.drawing.line(sg.x,sg.y,tg.x,tg.y)\n",
    "        self.drawing.save()\n",
    "        \n",
    "#enhance_method(KGMLCanvas, 'draw',  new_draw)\n",
    "setattr(KGMLCanvas, 'new_draw', new_draw)\n",
    "\n",
    "s = \"\"\n",
    "for pair in superpathwayDict[superpathway]:\n",
    "    pathwayID = pair[0]\n",
    "    pathway = pathwayFromFile(dataDirectory, organism, pathwayID)\n",
    "    \n",
    "    colrel = []\n",
    "    for relation in rel[pathwayID]:\n",
    "        source = element2id[set2can(relation.source)]\n",
    "        for targ in relation.target:\n",
    "            target = element2id[set2can(targ)]    \n",
    "            key = \"({},{})\".format(source,target)\n",
    "            if key in outd:\n",
    "                if significantTau(outd[key]):\n",
    "                    colrel += [(relation.sourceg, relation.targetg, convertTauValue(outd[key]), convertTauColor(outd[key]))]\n",
    "\n",
    "    canvas = KGMLCanvas(pathway)\n",
    "\n",
    "    for k in pathway.entries:\n",
    "        name = pathway.entries[k].name\n",
    "        canonicalName = set2can(words2set(name))\n",
    "        if canonicalName in element2id.keys():\n",
    "            node = element2id[canonicalName]\n",
    "            klass = cc[node]\n",
    "            if klass in nontrivialclass2col:\n",
    "                pathway.entries[k].graphics[0].bgcolor = nontrivialclass2col[klass]\n",
    "                pathway.entries[k].graphics[0].fgcolor = nontrivialclass2col[klass]\n",
    "            else:\n",
    "                pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "                pathway.enries[k].graphics[0].fgcolor = \"#FFFFFF\"\n",
    "            s += pathwayID + \"\\t\" + str(node) + \"\\t\" + str(klass) +\"\\t\"+pathway.entries[k].graphics[0].name+\"\\t\"+canonicalName + \"\\n\"\n",
    "    \n",
    "            \n",
    "    canvas.import_imagemap = True\n",
    "    pdfName = organism + pathwayID + \".pdf\"\n",
    "    pathlib.Path(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway + \"-col\")).mkdir(parents=True, exist_ok=True)\n",
    "    canvas.new_draw(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway+ \"-col\", pdfName), colrel)\n",
    "#with open(os.path.join(\"/Users/boldi/Desktop/\", organism, safeSuperpathway + \"-col\", \"all.tsv\"), \"wt\") as file:\n",
    "#    file.write(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "b1bb9776",
   "metadata": {},
   "outputs": [],
   "source": [
    "t=sampleTauForPairDiffsDist(genes, ['RAF1'], ['MAP2K1'], numHealthy, numberOfSamples = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "1789490d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<seaborn.axisgrid.FacetGrid at 0x12e1f71c0>"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeoAAAHpCAYAAABN+X+UAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAsZElEQVR4nO3de3QUZZ7/8U9CbnLphIBJwAACKhAFUdDQDI6KkYAZVw45By+IiAw6TGBGsqvIisIPRkDWFUaNsLoIepRF2AVXkVEhiKMSUOOwQsBIkDkBQyeLQAKM5Pr8/nDTY5NuJE1fHpL365w6h36quuv7NcQP9VR1VYQxxggAAFgpMtwFAAAA3whqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6glGWNUVVUlvlIOALANQS3pxIkTio+P14kTJ8JdCgAAHghqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYLCrcBQAInbH3TlTZkWNNxrt27qg1r68IQ0UAfg5BDbQiZUeOKXXMjCbjh9Y9HYZqAJwLpr4BALAYQQ0AgMXCGtRz5sxRRESEx9K3b1/3+tOnTysnJ0edOnVS+/btlZ2drfLyco/PKC0tVVZWltq2baukpCQ98sgjqqurC3UrAAAERdjPUV955ZXavHmz+3VU1N9Lmj59ut59912tXbtW8fHxmjp1qsaMGaNPP/1UklRfX6+srCylpKRo27ZtOnz4sO677z5FR0dr/vz5Ie8FAIBAC3tQR0VFKSUlpcl4ZWWlli9frlWrVmn48OGSpBUrVqhfv37avn27hgwZog8++EB79uzR5s2blZycrIEDB2revHmaMWOG5syZo5iYmFC3AwBAQIX9HPW+ffvUtWtX9erVS+PGjVNpaakkqbCwULW1tcrIyHBv27dvX3Xv3l0FBQWSpIKCAvXv31/JycnubTIzM1VVVaWioiKf+6yurlZVVZXHAgCAjcIa1Onp6Vq5cqXee+89LV26VAcOHNANN9ygEydOyOVyKSYmRgkJCR7vSU5OlsvlkiS5XC6PkG5c37jOlwULFig+Pt69dOvWLbCNAQAQIGGd+h41apT7zwMGDFB6erp69OihNWvW6KKLLgrafmfOnKnc3Fz366qqKsIaAGClsJ+j/qmEhARdccUVKikp0a233qqamhodP37c46i6vLzcfU47JSVFn332mcdnNF4V7u28d6PY2FjFxsYGvgEAPnFXNMA/YT9H/VMnT57U/v371aVLFw0aNEjR0dHKz893ry8uLlZpaamcTqckyel0ateuXaqoqHBvs2nTJjkcDqWlpYW8fgC+Nd4V7czFW3gD+LuwHlH/0z/9k26//Xb16NFDZWVlmj17ttq0aaO7775b8fHxmjRpknJzc5WYmCiHw6Fp06bJ6XRqyJAhkqQRI0YoLS1N48eP16JFi+RyuTRr1izl5ORwxAwAaBHCGtSHDh3S3Xffre+//14XX3yxhg0bpu3bt+viiy+WJC1evFiRkZHKzs5WdXW1MjMz9eKLL7rf36ZNG23YsEFTpkyR0+lUu3btNGHCBM2dOzdcLQEAEFBhDerVq1efdX1cXJzy8vKUl5fnc5sePXpo48aNgS4NAAArWHWOGgAAeCKoAQCwGEENAIDFCGoAACxm1Q1PAFw4uIEJEBoENQC/NN7A5EyH1j0dhmqAloupbwAALEZQAwBgMYIaAACLcY4agL4p3qthI0d7XcfFYUB4EdRAC+Triux9JfuV6mX7WhPp9cIwiYvDgHAjqIEWyNcV2UXzHwhDNQDOB+eoAQCwGEENAIDFmPoGEFZcyAacHUENIKy4kA04O6a+AQCwGEENAIDFmPoGcFa+ziH7+k42gMAiqAGcla9zyHwnGwgNpr4BALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFuIUogIAK5L3BfX0Wz6lGa0JQAwioQN4b3Ndn8ZxqtCZMfQMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxaLCXQAA/429d6LKjhxrMr6vZL9Sw1APgMAjqIELWNmRY0odM6PJeNH8B8JQDYBgYOobAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDFrgnrhwoWKiIjQww8/7B47ffq0cnJy1KlTJ7Vv317Z2dkqLy/3eF9paamysrLUtm1bJSUl6ZFHHlFdXV2IqwcAIDisCOrPP/9c//Zv/6YBAwZ4jE+fPl3vvPOO1q5dq48++khlZWUaM2aMe319fb2ysrJUU1Ojbdu26dVXX9XKlSv15JNPhroFAACCIuxBffLkSY0bN04vv/yyOnbs6B6vrKzU8uXL9eyzz2r48OEaNGiQVqxYoW3btmn79u2SpA8++EB79uzR66+/roEDB2rUqFGaN2+e8vLyVFNT43Of1dXVqqqq8lgAALBR2IM6JydHWVlZysjI8BgvLCxUbW2tx3jfvn3VvXt3FRQUSJIKCgrUv39/JScnu7fJzMxUVVWVioqKfO5zwYIFio+Pdy/dunULcFcAAARGWIN69erV+vLLL7VgwYIm61wul2JiYpSQkOAxnpycLJfL5d7mpyHduL5xnS8zZ85UZWWlezl48OB5dgIAQHCE7elZBw8e1O9//3tt2rRJcXFxId13bGysYmNjQ7pPAAD8EbYj6sLCQlVUVOjaa69VVFSUoqKi9NFHH+m5555TVFSUkpOTVVNTo+PHj3u8r7y8XCkpKZKklJSUJleBN75u3AYAgAtZ2IL6lltu0a5du7Rz5073MnjwYI0bN8795+joaOXn57vfU1xcrNLSUjmdTkmS0+nUrl27VFFR4d5m06ZNcjgcSktLC3lPAAAEWtimvjt06KCrrrrKY6xdu3bq1KmTe3zSpEnKzc1VYmKiHA6Hpk2bJqfTqSFDhkiSRowYobS0NI0fP16LFi2Sy+XSrFmzlJOTw9Q2AKBFCFtQn4vFixcrMjJS2dnZqq6uVmZmpl588UX3+jZt2mjDhg2aMmWKnE6n2rVrpwkTJmju3LlhrBoAgMCxKqi3bt3q8TouLk55eXnKy8vz+Z4ePXpo48aNQa4MAIDwCPv3qAEAgG8ENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAi1n1PWoAOBffFO/VsJGjm4x37dxRa15fEfqCgCAiqAFccGpNpFLHzGgyfmjd02GoBggupr4BALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWCwq3AUA+NHYeyeq7MixJuNdO3fUmtdXhKEiADYgqAFLlB05ptQxM5qMH1r3dBiqAWALghpAi/FN8V4NGzm6yTizEriQEdSA5XyFjyTtK9mv1NCWY7VaE8msBFocghqwnK/wkaSi+Q+EuBoAocZV3wAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFuNe30CI+XruNA/YAOANQQ2EmK/nTvOADQDeMPUNAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGJR4S4AaKnG3jtRZUeONRnfV7JfqWGoB8CFKaxH1EuXLtWAAQPkcDjkcDjkdDr1pz/9yb3+9OnTysnJUadOndS+fXtlZ2ervLzc4zNKS0uVlZWltm3bKikpSY888ojq6upC3QrQRNmRY0odM6PJUsPfTwDNENagTk1N1cKFC1VYWKgvvvhCw4cP1x133KGioiJJ0vTp0/XOO+9o7dq1+uijj1RWVqYxY8a4319fX6+srCzV1NRo27ZtevXVV7Vy5Uo9+eST4WoJAICACuvU9+233+7x+qmnntLSpUu1fft2paamavny5Vq1apWGDx8uSVqxYoX69eun7du3a8iQIfrggw+0Z88ebd68WcnJyRo4cKDmzZunGTNmaM6cOYqJiQlHWwAs803xXg0bObrJeNfOHbXm9RWhLwhoBmvOUdfX12vt2rU6deqUnE6nCgsLVVtbq4yMDPc2ffv2Vffu3VVQUKAhQ4aooKBA/fv3V3JysnubzMxMTZkyRUVFRbrmmmu87qu6ulrV1dXu11VVVcFrDEDY1ZpIpY6Z0WT80Lqnw1AN0Dxhv+p7165dat++vWJjY/Wb3/xG69evV1pamlwul2JiYpSQkOCxfXJyslwulyTJ5XJ5hHTj+sZ1vixYsEDx8fHupVu3boFtCgCAAAl7UPfp00c7d+7Ujh07NGXKFE2YMEF79uwJ6j5nzpypyspK93Lw4MGg7g8AAH+Ffeo7JiZGl112mSRp0KBB+vzzz/XHP/5Rd955p2pqanT8+HGPo+ry8nKlpKRIklJSUvTZZ595fF7jVeGN23gTGxur2NjYAHcCAEDghf2I+kwNDQ2qrq7WoEGDFB0drfz8fPe64uJilZaWyul0SpKcTqd27dqliooK9zabNm2Sw+FQWlpayGsHACDQwnpEPXPmTI0aNUrdu3fXiRMntGrVKm3dulXvv/++4uPjNWnSJOXm5ioxMVEOh0PTpk2T0+nUkCFDJEkjRoxQWlqaxo8fr0WLFsnlcmnWrFnKycnhiBkA0CKENagrKip033336fDhw4qPj9eAAQP0/vvv69Zbb5UkLV68WJGRkcrOzlZ1dbUyMzP14osvut/fpk0bbdiwQVOmTJHT6VS7du00YcIEzZ07N1wtAQAQUGEN6uXLl591fVxcnPLy8pSXl+dzmx49emjjxo2BLg0AACtYd44aAAD8HUENAIDFCGoAACzmV1D36tVL33//fZPx48ePq1evXuddFAAA+JFfQf3Xv/5V9fX1Tcarq6v13XffnXdRAADgR8266vvtt992/7nxu86N6uvrlZ+fr0svvTRgxQEA0No1K6hHjx4tSYqIiNCECRM81kVHR+vSSy/Vv/7rvwasOAAAWrtmBXVDQ4MkqWfPnvr888/VuXPnoBQFAAB+5NcNTw4cOBDoOgAAgBd+35ksPz9f+fn5qqiocB9pN3rllVfOuzAAAOBnUP+///f/NHfuXA0ePFhdunRRREREoOsCAADyM6iXLVumlStXavz48YGuBwBC5pvivRo2crTXdV07d9Sa11eEtiDAC7+CuqamRkOHDg10LQAQUrUmUqljZnhdd2jd0yGuBvDOrxue/PrXv9aqVasCXQsAADiDX0fUp0+f1ksvvaTNmzdrwIABio6O9lj/7LPPBqQ4AABaO7+C+quvvtLAgQMlSbt37/ZYx4VlAAAEjl9B/eGHHwa6DgAA4AWPuQQAwGJ+HVHffPPNZ53i3rJli98FAQCAv/MrqBvPTzeqra3Vzp07tXv37iYP6wAAAP7zK6gXL17sdXzOnDk6efLkeRUEAAD+LqDnqO+9917u8w0AQAAFNKgLCgoUFxcXyI8EAKBV82vqe8yYMR6vjTE6fPiwvvjiCz3xxBMBKQwAAPgZ1PHx8R6vIyMj1adPH82dO1cjRowISGEAAMDPoF6xgifKAAAQCn4FdaPCwkLt3btXknTllVfqmmuuCUhRAADgR34FdUVFhe666y5t3bpVCQkJkqTjx4/r5ptv1urVq3XxxRcHskYAAFotv676njZtmk6cOKGioiIdPXpUR48e1e7du1VVVaXf/e53ga4RAIBWy68j6vfee0+bN29Wv3793GNpaWnKy8vjYjIAAALIryPqhoaGJs+glqTo6Gg1NDScd1EAAOBHfgX18OHD9fvf/15lZWXuse+++07Tp0/XLbfcErDiAABo7fwK6hdeeEFVVVW69NJL1bt3b/Xu3Vs9e/ZUVVWVnn/++UDXCABAq+XXOepu3brpyy+/1ObNm/X1119Lkvr166eMjIyAFgcAQGvXrCPqLVu2KC0tTVVVVYqIiNCtt96qadOmadq0abruuut05ZVX6uOPPw5WrQAAtDrNCuolS5Zo8uTJcjgcTdbFx8froYce0rPPPhuw4gAAaO2aFdT/8z//o5EjR/pcP2LECBUWFp53UQAA4EfNOkddXl7u9WtZ7g+LitL//u//nndRgG3G3jtRZUeOeV3XtXNHrXmd+98DCI5mBfUll1yi3bt367LLLvO6/quvvlKXLl0CUhhgk7Ijx5Q6ZobXdYfWPR3iagC0Js2a+r7tttv0xBNP6PTp003W/fDDD5o9e7Z+9atfBaw4AABau2YdUc+aNUvr1q3TFVdcoalTp6pPnz6SpK+//lp5eXmqr6/X448/HpRCAQBojZoV1MnJydq2bZumTJmimTNnyhgjSYqIiFBmZqby8vKUnJwclEIBAGiNmn3Dkx49emjjxo06duyYSkpKZIzR5Zdfro4dOwajPgAAWjW/7kwmSR07dtR1110XyFoAAMAZ/LrXNwAACA2/j6gB/Oib4r0aNnJ0k/F9JfuVGvpyALQwBDVwnmpNpNfvWBfNfyAM1QBoaZj6BgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMb6eBQBe+Pp+PM8fR6gR1ADgha/vx/P8cYQaU98AAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYmEN6gULFui6665Thw4dlJSUpNGjR6u4uNhjm9OnTysnJ0edOnVS+/btlZ2drfLyco9tSktLlZWVpbZt2yopKUmPPPKI6urqQtkKAABBEdag/uijj5STk6Pt27dr06ZNqq2t1YgRI3Tq1Cn3NtOnT9c777yjtWvX6qOPPlJZWZnGjBnjXl9fX6+srCzV1NRo27ZtevXVV7Vy5Uo9+eST4WgJAICAigrnzt977z2P1ytXrlRSUpIKCwv1y1/+UpWVlVq+fLlWrVql4cOHS5JWrFihfv36afv27RoyZIg++OAD7dmzR5s3b1ZycrIGDhyoefPmacaMGZozZ45iYmKa7Le6ulrV1dXu11VVVcFtFAAAP1l1jrqyslKSlJiYKEkqLCxUbW2tMjIy3Nv07dtX3bt3V0FBgSSpoKBA/fv3V3JysnubzMxMVVVVqaioyOt+FixYoPj4ePfSrVu3YLUEAMB5sSaoGxoa9PDDD+sXv/iFrrrqKkmSy+VSTEyMEhISPLZNTk6Wy+Vyb/PTkG5c37jOm5kzZ6qystK9HDx4MMDdAAAQGGGd+v6pnJwc7d69W5988knQ9xUbG6vY2Nig7wcAgPNlxRH11KlTtWHDBn344YdKTU11j6ekpKimpkbHjx/32L68vFwpKSnubc68CrzxdeM2AABcqMJ6RG2M0bRp07R+/Xpt3bpVPXv29Fg/aNAgRUdHKz8/X9nZ2ZKk4uJilZaWyul0SpKcTqeeeuopVVRUKCkpSZK0adMmORwOpaWlhbYhAC3eN8V7NWzk6CbjXTt31JrXV4S+ILR4YQ3qnJwcrVq1Sv/93/+tDh06uM8px8fH66KLLlJ8fLwmTZqk3NxcJSYmyuFwaNq0aXI6nRoyZIgkacSIEUpLS9P48eO1aNEiuVwuzZo1Szk5OUxvAwi4WhOp1DEzmowfWvd0GKpBaxDWoF66dKkk6aabbvIYX7Fihe6//35J0uLFixUZGans7GxVV1crMzNTL774onvbNm3aaMOGDZoyZYqcTqfatWunCRMmaO7cuaFqAwCAoAn71PfPiYuLU15envLy8nxu06NHD23cuDGQpQEAYAUrLiYDAADeEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYzJrnUQM2GHvvRJUdOdZkfF/JfqV62R4Ago2gBn6i7Mgxr09GKpr/QBiqAQCmvgEAsBpBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALMa9vgEgAL4p3qthI0c3Ge/auaPWvL4i9AWhxSCoASAAak2k1we6HFr3dBiqQUvC1DcAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIvx9Sy0SmPvnaiyI8eajO8r2a/UMNQDAL4Q1GiVyo4c8/qd16L5D4ShGgDwjalvAAAsRlADAGAxghoAAIsR1AAAWIyLyQAgiHw9VUviyVo4NwQ1AASRr6dqSTxZC+eGqW8AACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwWFS4CwCCaey9E1V25FiT8X0l+5UahnoAoLkIarQIZwvkmx99qcl40fwHQlEWAJw3ghotQtmRY0odM6PJOIEM4ELHOWoAACxGUAMAYDGCGgAAixHUAABYjIvJACBMvineq2EjRzcZ79q5o9a8viL0BcFKBDUAhEmtifT6bYVD654OQzWwFVPfAABYLKxB/ec//1m33367unbtqoiICL311lse640xevLJJ9WlSxdddNFFysjI0L59+zy2OXr0qMaNGyeHw6GEhARNmjRJJ0+eDGEXAAAET1iD+tSpU7r66quVl5fndf2iRYv03HPPadmyZdqxY4fatWunzMxMnT592r3NuHHjVFRUpE2bNmnDhg3685//rAcffDBULQAAEFRhPUc9atQojRo1yus6Y4yWLFmiWbNm6Y477pAkvfbaa0pOTtZbb72lu+66S3v37tV7772nzz//XIMHD5YkPf/887rtttv0zDPPqGvXriHrBQCAYLD2HPWBAwfkcrmUkZHhHouPj1d6eroKCgokSQUFBUpISHCHtCRlZGQoMjJSO3bs8PnZ1dXVqqqq8lgAALCRtUHtcrkkScnJyR7jycnJ7nUul0tJSUke66OiopSYmOjexpsFCxYoPj7evXTr1i3A1QMAEBjWBnUwzZw5U5WVle7l4MGD4S4JAACvrA3qlJQUSVJ5ebnHeHl5uXtdSkqKKioqPNbX1dXp6NGj7m28iY2NlcPh8FgAALCRtUHds2dPpaSkKD8/3z1WVVWlHTt2yOl0SpKcTqeOHz+uwsJC9zZbtmxRQ0OD0tPTQ14zAACBFtarvk+ePKmSkhL36wMHDmjnzp1KTExU9+7d9fDDD+sPf/iDLr/8cvXs2VNPPPGEunbtqtGjR0uS+vXrp5EjR2ry5MlatmyZamtrNXXqVN11111c8Q3ggsWtRfFTYQ3qL774QjfffLP7dW5uriRpwoQJWrlypR599FGdOnVKDz74oI4fP65hw4bpvffeU1xcnPs9b7zxhqZOnapbbrlFkZGRys7O1nPPPRfyXgAgULi1KH4qrEF90003yRjjc31ERITmzp2ruXPn+twmMTFRq1atCkZ5AACEHQ/lAIALBFPirRNBDQAXCKbEWydrr/oGAAAENQAAViOoAQCwGEENAIDFCGoAACxGUAMAYDG+noULxth7J6rsyDGv6/aV7FdqiOsBgFAgqHHBKDtyzOt3SCWpaP4DIa4GAEKDqW8AACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAItxr29Yx9fDN3jwBoDWiKCGdXw9fIMHbwBojZj6BgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGLcmQwALnDfFO/VsJGjm4x37dxRa15fEfqCEFAENQBc4GpNpNfb7h5a93QYqkGgMfUNAIDFOKJG2PCULAD4eQQ1woanZAHAzyOoAaCF8nWRmcSFZhcSghoAWihfF5lJXGh2IeFiMgAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACzGnckAAG6+HpbDLUfDh6AGgFbI133A95Xs182PvtRknFuOhg9BDQCtkK/7gPP0OvtwjhoAAItxRI2g83XOa1/JfqWGoR4AuJAQ1Ai6siPHmGIDAD8x9Q0AgMUIagAALMbUNwLC13loiXPRAHA+CGoEhK/z0BLnogHgfBDUaBau4AaA0CKo0SxcwQ0AoUVQwyuOnAHADgQ1vOLIGQDsQFADAH6Wr4d48FSt4COoAQA/y9dDPHiqVvBxwxMAACxGUAMAYDGCGgAAixHUAABYjIvJWjm+Lw3gfHA1ePAR1K0c35cGcD64Gjz4COoW5GxPsCo9sF/de/ZuMs6RM4Bg8HWkLfk+2vb1/7DWfnROUF+AzjZdffOjL3l9T9H8BzhyBhAyvo60Jd9H275m+Fr70TlBfQFiuhoAWo8WE9R5eXn6l3/5F7lcLl199dV6/vnndf3114e7LADAGXxNi3MqzrsWEdRvvvmmcnNztWzZMqWnp2vJkiXKzMxUcXGxkpKSwl0eAOAnfE2LMyvoXYsI6meffVaTJ0/WxIkTJUnLli3Tu+++q1deeUWPPfZYk+2rq6tVXV3tfl1ZWSlJqqqqOu9aJkyeosPfH/e6rkunBL368tJzfo+v7evqalX7w6km46ah3uv42dYFe7yl7KO17jsU+2it+w7FPlrKvr/eu1tDMrKajB/66wGlXtrT6z58rQvU/4f91aFDB0VERDTvTeYCV11dbdq0aWPWr1/vMX7fffeZf/iHf/D6ntmzZxtJLCwsLCwsIV0qKyubnXMX/BH1kSNHVF9fr+TkZI/x5ORkff31117fM3PmTOXm5rpfNzQ06OjRo+rUqZP7XzpVVVXq1q2bDh48KIfDEbwGgqwl9NESepDowzb0YY+W0IN0bn106NCh2Z97wQe1P2JjYxUbG+sxlpCQ4HVbh8NxQf/FadQS+mgJPUj0YRv6sEdL6EEKfB8X/L2+O3furDZt2qi8vNxjvLy8XCkpKWGqCgCAwLjggzomJkaDBg1Sfn6+e6yhoUH5+flyOp1hrAwAgPPXIqa+c3NzNWHCBA0ePFjXX3+9lixZolOnTrmvAvdHbGysZs+e3WSK/ELTEvpoCT1I9GEb+rBHS+hBCl4fEcYYE9BPDJMXXnjBfcOTgQMH6rnnnlN6enq4ywIA4Ly0mKAGAKAluuDPUQMA0JIR1AAAWIygBgDAYgQ1AAAWI6j/z9GjRzVu3Dg5HA4lJCRo0qRJOnny5Fnf43K5NH78eKWkpKhdu3a69tpr9V//9V8hqtg7f/qQpIKCAg0fPlzt2rWTw+HQL3/5S/3www8hqNg7f/uQJGOMRo0apYiICL311lvBLfRnNLePo0ePatq0aerTp48uuugide/eXb/73e/cD44Jlby8PF166aWKi4tTenq6Pvvss7Nuv3btWvXt21dxcXHq37+/Nm7cGKJKz645fbz88su64YYb1LFjR3Xs2FEZGRk/23eoNPfn0Wj16tWKiIjQ6NGjg1vgOWhuD8ePH1dOTo66dOmi2NhYXXHFFVb8vWpuH0uWLHH/Pnfr1k3Tp0/X6dOnm7fTZt8dvIUaOXKkufrqq8327dvNxx9/bC677DJz9913n/U9t956q7nuuuvMjh07zP79+828efNMZGSk+fLLL0NUdVP+9LFt2zbjcDjMggULzO7du83XX39t3nzzTXP69OkQVd2UP300evbZZ82oUaOMpCYPawm15vaxa9cuM2bMGPP222+bkpISk5+fby6//HKTnZ0dsppXr15tYmJizCuvvGKKiorM5MmTTUJCgikvL/e6/aeffmratGljFi1aZPbs2WNmzZploqOjza5du0JWszfN7eOee+4xeXl55i9/+YvZu3evuf/++018fLw5dOhQiCv31Nw+Gh04cMBccskl5oYbbjB33HFHaIr1obk9VFdXm8GDB5vbbrvNfPLJJ+bAgQNm69atZufOnSGu3FNz+3jjjTdMbGyseeONN8yBAwfM+++/b7p06WKmT5/erP0S1MaYPXv2GEnm888/d4/96U9/MhEREea7777z+b527dqZ1157zWMsMTHRvPzyy0Gr9Wz87SM9Pd3MmjUrFCWeE3/7MMaYv/zlL+aSSy4xhw8fDntQn08fP7VmzRoTExNjamtrg1FmE9dff73Jyclxv66vrzddu3Y1CxYs8Lr92LFjTVZWlsdYenq6eeihh4Ja589pbh9nqqurMx06dDCvvvpqsEo8J/70UVdXZ4YOHWr+/d//3UyYMCHsQd3cHpYuXWp69eplampqQlXiOWluHzk5OWb48OEeY7m5ueYXv/hFs/bL1Ld+nPZNSEjQ4MGD3WMZGRmKjIzUjh07fL5v6NChevPNN3X06FE1NDRo9erVOn36tG666aYQVN2UP31UVFRox44dSkpK0tChQ5WcnKwbb7xRn3zySajKbsLfn8ff/vY33XPPPcrLy7PiPu/+9nGmyspKORwORUUF/0aCNTU1KiwsVEZGhnssMjJSGRkZKigo8PqegoICj+0lKTMz0+f2oeBPH2f629/+ptraWiUmJgarzJ/lbx9z585VUlKSJk2aFIoyz8qfHt5++205nU7l5OQoOTlZV111lebPn6/6+vpQld2EP30MHTpUhYWF7unxb7/9Vhs3btRtt93WrH23iFuIni+Xy6WkpCSPsaioKCUmJsrlcvl835o1a3TnnXeqU6dOioqKUtu2bbV+/XpddtllwS7ZK3/6+PbbbyVJc+bM0TPPPKOBAwfqtdde0y233KLdu3fr8ssvD3rdZ/L35zF9+nQNHTpUd9xxR7BLPCf+9vFTR44c0bx58/Tggw8Go0Sv+2vuY2NdLpfX7c+1x2Dwp48zzZgxQ127dm3yj5BQ8qePTz75RMuXL9fOnTtDUOHP86eHb7/9Vlu2bNG4ceO0ceNGlZSU6Le//a1qa2s1e/bsUJTdhD993HPPPTpy5IiGDRsmY4zq6ur0m9/8Rv/8z//crH236CPqxx57TBEREWddzvWX1psnnnhCx48f1+bNm/XFF18oNzdXY8eO1a5duwLYRXD7aGhokCQ99NBDmjhxoq655hotXrxYffr00SuvvBLINoLax9tvv60tW7ZoyZIlAa3Zm2D/vWpUVVWlrKwspaWlac6cOedfOM7ZwoULtXr1aq1fv15xcXHhLuecnThxQuPHj9fLL7+szp07h7scvzU0NCgpKUkvvfSSBg0apDvvvFOPP/64li1bFu7SmmXr1q2aP3++XnzxRX355Zdat26d3n33Xc2bN69Zn9Oij6j/8R//Uffff/9Zt+nVq5dSUlJUUVHhMV5XV6ejR4/6nELdv3+/XnjhBe3evVtXXnmlJOnqq6/Wxx9/rLy8vID+hQpmH126dJEkpaWleYz369dPpaWl/hftRTD72LJli/bv39/kueLZ2dm64YYbtHXr1vOo3FMw+2h04sQJjRw5Uh06dND69esVHR19vmWfE38eG5uSkmLdY2bP5/G3zzzzjBYuXKjNmzdrwIABwSzzZzW3j/379+uvf/2rbr/9dvdY4z/Go6KiVFxcrN69ewe36DP487Po0qWLoqOj1aZNG/dYv3795HK5VFNTo5iYmKDW7I0/fTzxxBMaP368fv3rX0uS+vfvr1OnTunBBx/U448/rsjIczxWbvbZ9Bao8aKfL774wj32/vvvn/Win6+++spIMnv27PEYHzFihJk8eXJQ6/XFnz4aGhpM165dm1xMNnDgQDNz5syg1uuLP30cPnzY7Nq1y2ORZP74xz+ab7/9NlSle/CnD2OMqaysNEOGDDE33nijOXXqVChK9XD99debqVOnul/X19ebSy655KwXk/3qV7/yGHM6nVZcTNacPowx5umnnzYOh8MUFBSEosRz0pw+fvjhhya/B3fccYcZPny42bVrl6murg5l6W7N/VnMnDnT9OjRw9TX17vHlixZYrp06RL0Ws+muX1ce+215tFHH/UYW7VqlbnoootMXV3dOe+XoP4/I0eONNdcc43ZsWOH+eSTT8zll1/u8TWaQ4cOmT59+pgdO3YYY4ypqakxl112mbnhhhvMjh07TElJiXnmmWdMRESEeffdd8PVRrP7MMaYxYsXG4fDYdauXWv27dtnZs2aZeLi4kxJSUk4WjDG+NfHmWTJ17Oa00dlZaVJT083/fv3NyUlJebw4cPupTm/2Odj9erVJjY21qxcudLs2bPHPPjggyYhIcG4XC5jjDHjx483jz32mHv7Tz/91ERFRZlnnnnG7N2718yePduar2c1p4+FCxeamJgY85//+Z8e/91PnDgRrhaMMc3v40w2XPXd3B5KS0tNhw4dzNSpU01xcbHZsGGDSUpKMn/4wx/C1YIxpvl9zJ4923To0MH8x3/8h/n222/NBx98YHr37m3Gjh3brP0S1P/n+++/N3fffbdp3769cTgcZuLEiR6/oAcOHDCSzIcffuge++abb8yYMWNMUlKSadu2rRkwYECTr2uFmj99GGPMggULTGpqqmnbtq1xOp3m448/DnHlnvzt46dsCOrm9vHhhx8aSV6XAwcOhKzu559/3nTv3t3ExMSY66+/3mzfvt297sYbbzQTJkzw2H7NmjXmiiuuMDExMebKK68M6z9Wf6o5ffTo0cPrf/fZs2eHvvAzNPfn8VM2BLUxze9h27ZtJj093cTGxppevXqZp556KmT/WD2b5vRRW1tr5syZY3r37m3i4uJMt27dzG9/+1tz7NixZu2Tx1wCAGCxFn3VNwAAFzqCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWOz/Ax2PWos1yUXjAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 500x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "sns.displot(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "f9ff8777",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "res = stats.normaltest(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "f440c07d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9031307461093279"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.statistic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1dc0352f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.6366308065065713"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "db51339e",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = stats.normaltest(np.random.normal(pd.DataFrame(t).mean().values[0], pd.DataFrame(t).std().values[0], 10000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "24ed7331",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7444207889823913"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.statistic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "3b6273dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.6892092214454659"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4571a000",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.444566792234135"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tauForPairDiff(genesHealthy, genesDiseased, ['RAF1'], ['MAP2K1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e6552d7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "758b5926",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes1 = random.sample(range(len(genes)), 3)\n",
    "genes2 = list(set(range(len(genes)))-set(genes1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "39cbd1a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>ADA</th>\n",
       "      <th>CDH2</th>\n",
       "      <th>AKT3</th>\n",
       "      <th>100008588</th>\n",
       "      <th>100008589</th>\n",
       "      <th>100009676</th>\n",
       "      <th>MED6</th>\n",
       "      <th>...</th>\n",
       "      <th>THOC1</th>\n",
       "      <th>REC8</th>\n",
       "      <th>RCE1</th>\n",
       "      <th>HNRNPDL</th>\n",
       "      <th>DMTF1</th>\n",
       "      <th>PPP4R1</th>\n",
       "      <th>SLC12A6</th>\n",
       "      <th>PTBP3</th>\n",
       "      <th>DGCR2</th>\n",
       "      <th>SCO2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>51</td>\n",
       "      <td>GSM1574474</td>\n",
       "      <td>5.736730</td>\n",
       "      <td>7.085667</td>\n",
       "      <td>5.395656</td>\n",
       "      <td>5.692972</td>\n",
       "      <td>12.011536</td>\n",
       "      <td>10.206174</td>\n",
       "      <td>6.420234</td>\n",
       "      <td>8.090269</td>\n",
       "      <td>...</td>\n",
       "      <td>6.614601</td>\n",
       "      <td>6.412066</td>\n",
       "      <td>6.002645</td>\n",
       "      <td>6.680613</td>\n",
       "      <td>7.256032</td>\n",
       "      <td>7.579093</td>\n",
       "      <td>5.924431</td>\n",
       "      <td>8.318820</td>\n",
       "      <td>7.352752</td>\n",
       "      <td>8.427214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>GSM1574440</td>\n",
       "      <td>5.906989</td>\n",
       "      <td>7.725843</td>\n",
       "      <td>5.255621</td>\n",
       "      <td>5.431474</td>\n",
       "      <td>11.494311</td>\n",
       "      <td>10.867244</td>\n",
       "      <td>6.153189</td>\n",
       "      <td>8.439548</td>\n",
       "      <td>...</td>\n",
       "      <td>7.079432</td>\n",
       "      <td>6.300575</td>\n",
       "      <td>5.953793</td>\n",
       "      <td>6.929206</td>\n",
       "      <td>7.306660</td>\n",
       "      <td>7.490425</td>\n",
       "      <td>5.373872</td>\n",
       "      <td>8.643313</td>\n",
       "      <td>7.323018</td>\n",
       "      <td>7.888204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>61</td>\n",
       "      <td>GSM1574484</td>\n",
       "      <td>5.606271</td>\n",
       "      <td>7.757515</td>\n",
       "      <td>5.618650</td>\n",
       "      <td>5.494501</td>\n",
       "      <td>11.678308</td>\n",
       "      <td>11.502444</td>\n",
       "      <td>5.985153</td>\n",
       "      <td>8.202394</td>\n",
       "      <td>...</td>\n",
       "      <td>7.030866</td>\n",
       "      <td>6.064598</td>\n",
       "      <td>5.797958</td>\n",
       "      <td>7.296972</td>\n",
       "      <td>7.784769</td>\n",
       "      <td>7.516425</td>\n",
       "      <td>5.852640</td>\n",
       "      <td>8.801789</td>\n",
       "      <td>6.655766</td>\n",
       "      <td>7.288676</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 21170 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0        name      A1BG       ADA      CDH2      AKT3  100008588   \n",
       "51          51  GSM1574474  5.736730  7.085667  5.395656  5.692972  12.011536  \\\n",
       "17          17  GSM1574440  5.906989  7.725843  5.255621  5.431474  11.494311   \n",
       "61          61  GSM1574484  5.606271  7.757515  5.618650  5.494501  11.678308   \n",
       "\n",
       "    100008589  100009676      MED6  ...     THOC1      REC8      RCE1   \n",
       "51  10.206174   6.420234  8.090269  ...  6.614601  6.412066  6.002645  \\\n",
       "17  10.867244   6.153189  8.439548  ...  7.079432  6.300575  5.953793   \n",
       "61  11.502444   5.985153  8.202394  ...  7.030866  6.064598  5.797958   \n",
       "\n",
       "     HNRNPDL     DMTF1    PPP4R1   SLC12A6     PTBP3     DGCR2      SCO2  \n",
       "51  6.680613  7.256032  7.579093  5.924431  8.318820  7.352752  8.427214  \n",
       "17  6.929206  7.306660  7.490425  5.373872  8.643313  7.323018  7.888204  \n",
       "61  7.296972  7.784769  7.516425  5.852640  8.801789  6.655766  7.288676  \n",
       "\n",
       "[3 rows x 21170 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.iloc[genes1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ef9adc72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>ADA</th>\n",
       "      <th>CDH2</th>\n",
       "      <th>AKT3</th>\n",
       "      <th>100008588</th>\n",
       "      <th>100008589</th>\n",
       "      <th>100009676</th>\n",
       "      <th>MED6</th>\n",
       "      <th>...</th>\n",
       "      <th>THOC1</th>\n",
       "      <th>REC8</th>\n",
       "      <th>RCE1</th>\n",
       "      <th>HNRNPDL</th>\n",
       "      <th>DMTF1</th>\n",
       "      <th>PPP4R1</th>\n",
       "      <th>SLC12A6</th>\n",
       "      <th>PTBP3</th>\n",
       "      <th>DGCR2</th>\n",
       "      <th>SCO2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>GSM1574423</td>\n",
       "      <td>5.513040</td>\n",
       "      <td>8.105445</td>\n",
       "      <td>5.538153</td>\n",
       "      <td>5.367775</td>\n",
       "      <td>11.146230</td>\n",
       "      <td>11.589834</td>\n",
       "      <td>5.901013</td>\n",
       "      <td>8.079373</td>\n",
       "      <td>...</td>\n",
       "      <td>7.198096</td>\n",
       "      <td>6.152714</td>\n",
       "      <td>6.291999</td>\n",
       "      <td>7.194089</td>\n",
       "      <td>7.440856</td>\n",
       "      <td>7.791873</td>\n",
       "      <td>5.683778</td>\n",
       "      <td>8.471226</td>\n",
       "      <td>7.106159</td>\n",
       "      <td>7.440375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>GSM1574424</td>\n",
       "      <td>5.238013</td>\n",
       "      <td>7.019328</td>\n",
       "      <td>5.188991</td>\n",
       "      <td>5.401298</td>\n",
       "      <td>11.728307</td>\n",
       "      <td>10.464450</td>\n",
       "      <td>6.265463</td>\n",
       "      <td>8.179442</td>\n",
       "      <td>...</td>\n",
       "      <td>7.045319</td>\n",
       "      <td>6.022443</td>\n",
       "      <td>6.045065</td>\n",
       "      <td>7.297708</td>\n",
       "      <td>7.821551</td>\n",
       "      <td>7.663454</td>\n",
       "      <td>5.675778</td>\n",
       "      <td>8.700451</td>\n",
       "      <td>6.962812</td>\n",
       "      <td>7.005557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>GSM1574425</td>\n",
       "      <td>5.708308</td>\n",
       "      <td>7.230270</td>\n",
       "      <td>5.507846</td>\n",
       "      <td>5.397361</td>\n",
       "      <td>12.047896</td>\n",
       "      <td>11.318551</td>\n",
       "      <td>6.224461</td>\n",
       "      <td>8.074019</td>\n",
       "      <td>...</td>\n",
       "      <td>7.213119</td>\n",
       "      <td>6.178724</td>\n",
       "      <td>5.782921</td>\n",
       "      <td>6.998591</td>\n",
       "      <td>7.510197</td>\n",
       "      <td>7.663987</td>\n",
       "      <td>5.664503</td>\n",
       "      <td>8.643570</td>\n",
       "      <td>7.205886</td>\n",
       "      <td>7.787051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>GSM1574426</td>\n",
       "      <td>5.740369</td>\n",
       "      <td>7.396263</td>\n",
       "      <td>5.442095</td>\n",
       "      <td>5.296547</td>\n",
       "      <td>11.650462</td>\n",
       "      <td>11.303002</td>\n",
       "      <td>6.022263</td>\n",
       "      <td>8.177189</td>\n",
       "      <td>...</td>\n",
       "      <td>7.098039</td>\n",
       "      <td>6.058239</td>\n",
       "      <td>5.673125</td>\n",
       "      <td>7.204400</td>\n",
       "      <td>7.893798</td>\n",
       "      <td>7.376775</td>\n",
       "      <td>5.750436</td>\n",
       "      <td>8.928652</td>\n",
       "      <td>6.890701</td>\n",
       "      <td>7.143114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>GSM1574427</td>\n",
       "      <td>5.932976</td>\n",
       "      <td>7.238479</td>\n",
       "      <td>6.143618</td>\n",
       "      <td>5.671180</td>\n",
       "      <td>11.718869</td>\n",
       "      <td>10.859258</td>\n",
       "      <td>6.023132</td>\n",
       "      <td>8.228727</td>\n",
       "      <td>...</td>\n",
       "      <td>6.648952</td>\n",
       "      <td>6.114541</td>\n",
       "      <td>5.814955</td>\n",
       "      <td>7.399123</td>\n",
       "      <td>7.605499</td>\n",
       "      <td>7.356980</td>\n",
       "      <td>5.543969</td>\n",
       "      <td>8.511676</td>\n",
       "      <td>6.663802</td>\n",
       "      <td>7.748388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>58</td>\n",
       "      <td>GSM1574481</td>\n",
       "      <td>5.827496</td>\n",
       "      <td>7.584883</td>\n",
       "      <td>5.710704</td>\n",
       "      <td>5.559310</td>\n",
       "      <td>11.847496</td>\n",
       "      <td>10.287512</td>\n",
       "      <td>6.426236</td>\n",
       "      <td>8.609280</td>\n",
       "      <td>...</td>\n",
       "      <td>7.129741</td>\n",
       "      <td>6.093136</td>\n",
       "      <td>6.077830</td>\n",
       "      <td>7.047893</td>\n",
       "      <td>7.181003</td>\n",
       "      <td>7.615644</td>\n",
       "      <td>5.663526</td>\n",
       "      <td>8.558329</td>\n",
       "      <td>7.015060</td>\n",
       "      <td>8.094692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>59</td>\n",
       "      <td>GSM1574482</td>\n",
       "      <td>5.884727</td>\n",
       "      <td>7.484607</td>\n",
       "      <td>5.790493</td>\n",
       "      <td>5.582019</td>\n",
       "      <td>11.553017</td>\n",
       "      <td>10.963190</td>\n",
       "      <td>5.950991</td>\n",
       "      <td>8.646998</td>\n",
       "      <td>...</td>\n",
       "      <td>7.091413</td>\n",
       "      <td>6.117294</td>\n",
       "      <td>5.996777</td>\n",
       "      <td>7.323042</td>\n",
       "      <td>7.529939</td>\n",
       "      <td>7.391352</td>\n",
       "      <td>5.715396</td>\n",
       "      <td>8.527401</td>\n",
       "      <td>7.160782</td>\n",
       "      <td>7.783217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>60</td>\n",
       "      <td>GSM1574483</td>\n",
       "      <td>5.571254</td>\n",
       "      <td>7.993836</td>\n",
       "      <td>5.530246</td>\n",
       "      <td>5.470003</td>\n",
       "      <td>10.881541</td>\n",
       "      <td>10.127892</td>\n",
       "      <td>6.354627</td>\n",
       "      <td>8.418685</td>\n",
       "      <td>...</td>\n",
       "      <td>6.993226</td>\n",
       "      <td>6.276569</td>\n",
       "      <td>6.158704</td>\n",
       "      <td>7.178020</td>\n",
       "      <td>7.428225</td>\n",
       "      <td>7.736230</td>\n",
       "      <td>5.452297</td>\n",
       "      <td>8.670611</td>\n",
       "      <td>7.251531</td>\n",
       "      <td>8.328839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>62</td>\n",
       "      <td>GSM1574485</td>\n",
       "      <td>5.808886</td>\n",
       "      <td>7.268714</td>\n",
       "      <td>5.222689</td>\n",
       "      <td>5.519814</td>\n",
       "      <td>12.081060</td>\n",
       "      <td>10.591896</td>\n",
       "      <td>6.233185</td>\n",
       "      <td>8.026523</td>\n",
       "      <td>...</td>\n",
       "      <td>7.125517</td>\n",
       "      <td>6.292087</td>\n",
       "      <td>5.787027</td>\n",
       "      <td>6.969160</td>\n",
       "      <td>7.399395</td>\n",
       "      <td>7.505281</td>\n",
       "      <td>5.655602</td>\n",
       "      <td>9.048492</td>\n",
       "      <td>7.017329</td>\n",
       "      <td>8.029034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>63</td>\n",
       "      <td>GSM1574486</td>\n",
       "      <td>5.676531</td>\n",
       "      <td>8.053514</td>\n",
       "      <td>4.959765</td>\n",
       "      <td>5.417831</td>\n",
       "      <td>11.353458</td>\n",
       "      <td>10.976332</td>\n",
       "      <td>6.008427</td>\n",
       "      <td>8.258550</td>\n",
       "      <td>...</td>\n",
       "      <td>7.006998</td>\n",
       "      <td>6.038474</td>\n",
       "      <td>6.342585</td>\n",
       "      <td>7.790379</td>\n",
       "      <td>7.493930</td>\n",
       "      <td>7.573584</td>\n",
       "      <td>5.577641</td>\n",
       "      <td>8.561699</td>\n",
       "      <td>7.219044</td>\n",
       "      <td>7.738660</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 21170 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0        name      A1BG       ADA      CDH2      AKT3  100008588   \n",
       "0            0  GSM1574423  5.513040  8.105445  5.538153  5.367775  11.146230  \\\n",
       "1            1  GSM1574424  5.238013  7.019328  5.188991  5.401298  11.728307   \n",
       "2            2  GSM1574425  5.708308  7.230270  5.507846  5.397361  12.047896   \n",
       "3            3  GSM1574426  5.740369  7.396263  5.442095  5.296547  11.650462   \n",
       "4            4  GSM1574427  5.932976  7.238479  6.143618  5.671180  11.718869   \n",
       "..         ...         ...       ...       ...       ...       ...        ...   \n",
       "58          58  GSM1574481  5.827496  7.584883  5.710704  5.559310  11.847496   \n",
       "59          59  GSM1574482  5.884727  7.484607  5.790493  5.582019  11.553017   \n",
       "60          60  GSM1574483  5.571254  7.993836  5.530246  5.470003  10.881541   \n",
       "62          62  GSM1574485  5.808886  7.268714  5.222689  5.519814  12.081060   \n",
       "63          63  GSM1574486  5.676531  8.053514  4.959765  5.417831  11.353458   \n",
       "\n",
       "    100008589  100009676      MED6  ...     THOC1      REC8      RCE1   \n",
       "0   11.589834   5.901013  8.079373  ...  7.198096  6.152714  6.291999  \\\n",
       "1   10.464450   6.265463  8.179442  ...  7.045319  6.022443  6.045065   \n",
       "2   11.318551   6.224461  8.074019  ...  7.213119  6.178724  5.782921   \n",
       "3   11.303002   6.022263  8.177189  ...  7.098039  6.058239  5.673125   \n",
       "4   10.859258   6.023132  8.228727  ...  6.648952  6.114541  5.814955   \n",
       "..        ...        ...       ...  ...       ...       ...       ...   \n",
       "58  10.287512   6.426236  8.609280  ...  7.129741  6.093136  6.077830   \n",
       "59  10.963190   5.950991  8.646998  ...  7.091413  6.117294  5.996777   \n",
       "60  10.127892   6.354627  8.418685  ...  6.993226  6.276569  6.158704   \n",
       "62  10.591896   6.233185  8.026523  ...  7.125517  6.292087  5.787027   \n",
       "63  10.976332   6.008427  8.258550  ...  7.006998  6.038474  6.342585   \n",
       "\n",
       "     HNRNPDL     DMTF1    PPP4R1   SLC12A6     PTBP3     DGCR2      SCO2  \n",
       "0   7.194089  7.440856  7.791873  5.683778  8.471226  7.106159  7.440375  \n",
       "1   7.297708  7.821551  7.663454  5.675778  8.700451  6.962812  7.005557  \n",
       "2   6.998591  7.510197  7.663987  5.664503  8.643570  7.205886  7.787051  \n",
       "3   7.204400  7.893798  7.376775  5.750436  8.928652  6.890701  7.143114  \n",
       "4   7.399123  7.605499  7.356980  5.543969  8.511676  6.663802  7.748388  \n",
       "..       ...       ...       ...       ...       ...       ...       ...  \n",
       "58  7.047893  7.181003  7.615644  5.663526  8.558329  7.015060  8.094692  \n",
       "59  7.323042  7.529939  7.391352  5.715396  8.527401  7.160782  7.783217  \n",
       "60  7.178020  7.428225  7.736230  5.452297  8.670611  7.251531  8.328839  \n",
       "62  6.969160  7.399395  7.505281  5.655602  9.048492  7.017329  8.029034  \n",
       "63  7.790379  7.493930  7.573584  5.577641  8.561699  7.219044  7.738660  \n",
       "\n",
       "[61 rows x 21170 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.iloc[genes2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "ba1d0515",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'(0,3)': -0.1533323674297305}"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "22f0966e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[-0.444566792234135,\n",
       " -0.27238481599536357,\n",
       " 0.2824108953926398,\n",
       " -0.28861199652274705,\n",
       " -0.23309185743262822,\n",
       " 0.20483917705013044,\n",
       " 0.221172606973824,\n",
       " -0.2592871631411185,\n",
       " -0.2326137351492321,\n",
       " 0.21926977687626775,\n",
       " 0.20588240172417813,\n",
       " 0.36765584214504116,\n",
       " 0.20347725297015354,\n",
       " 0.35986670530281073,\n",
       " -0.25911330049261083,\n",
       " 0.2860620110113011,\n",
       " 0.21706751666183713,\n",
       " -0.4622268537416605,\n",
       " -0.2514994493423496]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x for x in list(outd.values()) if abs(x)>0.2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f47ca313",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qf",
   "language": "python",
   "name": "qf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
