{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d4364c55",
   "metadata": {},
   "outputs": [],
   "source": [
    "from hypergraphs import *\n",
    "from kegg import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8fcb987c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from urllib.request import urlopen\n",
    "from urllib.parse import quote_plus\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "from Levenshtein import ratio\n",
    "from chemspipy import ChemSpider\n",
    "import logging\n",
    "import pathlib\n",
    "from Bio.KEGG.KGML.KGML_parser import read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fa19c64c",
   "metadata": {},
   "outputs": [],
   "source": [
    "superpathwayDict = {\n",
    "    \"Carbohydrate metabolism\":\n",
    "[\n",
    "\t(\"00010\", \"Glycolysis / Gluconeogenesis\"),\n",
    "\t(\"00020\", \"Citrate cycle (TCA cycle)\"),\n",
    "\t(\"00030\", \"Pentose phosphate pathway\"),\n",
    "\t(\"00040\", \"Pentose and glucuronate interconversions\"),\n",
    "\t(\"00051\", \"Fructose and mannose metabolism\"),\n",
    "\t(\"00052\", \"Galactose metabolism\"),\n",
    "\t(\"00053\", \"Ascorbate and aldarate metabolism\"),\n",
    "\t(\"00500\", \"Starch and sucrose metabolism\"),\n",
    "\t(\"00520\", \"Amino sugar and nucleotide sugar metabolism\"),\n",
    "\t(\"00620\", \"Pyruvate metabolism\"),\n",
    "\t(\"00630\", \"Glyoxylate and dicarboxylate metaboliscm\"),\n",
    "\t(\"00640\", \"Propanoate metabolism\"),\n",
    "\t(\"00650\", \"Butanoate metabolism\"),\n",
    "\t(\"00562\", \"Inositol phosphate metabolism\")\n",
    "],\n",
    "\"Energy metabolism\":\n",
    "[\n",
    "\t(\"00190\", \"Oxidative phosphorylation\"),\n",
    "\t(\"00910\", \"Nitrogen metabolism\"),\n",
    "\t(\"00920\", \"Sulfur metabolism\")\n",
    "],\n",
    "\"Lipid metabolism\":\n",
    "[\t(\"00061\", \"Fatty acid biosynthesis\"),\n",
    "\t(\"00062\", \"Fatty acid elongation\"),\t\n",
    "\t(\"00071\", \"Fatty acid degradation\"),\n",
    "\t(\"00100\", \"Steroid biosynthesis\"),\t\n",
    "\t(\"00120\", \"Primary bile acid biosynthesis\"),\n",
    "\t(\"00140\", \"Steroid hormone biosynthesis\"),\n",
    "\t(\"00561\", \"Glycerolipid metabolism\"),\n",
    "\t(\"00564\", \"Glycerophospholipid metabolism\"),\n",
    "\t(\"00565\", \"Ether lipid metabolism\"),\n",
    "\t(\"00600\", \"Sphingolipid metabolism\"),\n",
    "\t(\"00590\", \"Arachidonic acid metabolism\"),\n",
    "\t(\"00591\", \"Linoleic acid metabolism\"),\n",
    "\t(\"00592\", \"alpha-Linolenic acid metabolism\"),\n",
    "\t(\"01040\", \"Biosynthesis of unsaturated fatty acids\")\n",
    "],\n",
    "\"Nucleotide metabolism\":\n",
    "[\n",
    "\t(\"00230\", \"Purine metabolism\"),\n",
    "\t(\"00240\", \"Pyrimidine metabolism\")\n",
    "],\n",
    "\"Amino acid metabolism\":\n",
    "[\n",
    "\t(\"00250\", \"Alanine, aspartate and glutamate metabolism\"),\n",
    "\t(\"00260\", \"Glycine, serine and threonine metabolism\"),\n",
    "\t(\"00270\", \"Cysteine and methionine metabolism\"),\n",
    "\t(\"00280\", \"Valine, leucine and isoleucine degradation\"),\n",
    "\t(\"00290\", \"Valine, leucine and isoleucine biosynthesis\"),\n",
    "\t(\"00310\", \"Lysine degradation\"),\n",
    "\t(\"00220\", \"Arginine biosynthesis\"),\n",
    "\t(\"00330\", \"Arginine and proline metabolism\"),\n",
    "\t(\"00340\", \"Histidine metabolism\"),\n",
    "\t(\"00350\", \"Tyrosine metabolism\"),\n",
    "\t(\"00360\", \"Phenylalanine metabolism\"),\n",
    "\t(\"00380\", \"Tryptophan metabolism\"),\n",
    "\t(\"00400\", \"Phenylalanine, tyrosine and tryptophan biosynthesis\")\n",
    "],\n",
    "\"Metabolism of other amino acids\":\n",
    "[\n",
    "\t(\"00410\", \"beta-Alanine metabolism\"),\n",
    "\t(\"00430\", \"Taurine and hypotaurine metabolism\"),\n",
    "\t(\"00440\", \"Phosphonate and phosphinate metabolism\"),\n",
    "\t(\"00450\", \"Selenocompound metabolism\"),\n",
    "\t(\"00470\", \"D-Amino acid metabolism\"),\n",
    "\t(\"00480\", \"Glutathione metabolism\")\n",
    "],\n",
    "\"Glycan biosynthesis and metabolism\":\n",
    "[\n",
    "\t(\"00510\", \"N-Glycan biosynthesis\"),\n",
    "\t(\"00513\", \"Various types of N-glycan biosynthesis\"),\n",
    "\t(\"00512\", \"Mucin type O-glycan biosynthesis\"),\n",
    "\t(\"00515\", \"Mannose type O-glycan biosynthesis\"),\n",
    "\t(\"00514\", \"Other types of O-glycan biosynthesis\"),\n",
    "\t(\"00532\", \"Glycosaminoglycan biosynthesis - chondroitin sulfate / dermatan sulfate\"),\n",
    "\t(\"00534\", \"Glycosaminoglycan biosynthesis - heparan sulfate / heparin\"),\n",
    "\t(\"00533\", \"Glycosaminoglycan biosynthesis - keratan sulfate\"),\n",
    "\t(\"00531\", \"Glycosaminoglycan degradation\"),\n",
    "\t(\"00563\", \"Glycosylphosphatidylinositol (GPI)-anchor biosynthesis\"),\n",
    "\t(\"00601\", \"Glycosphingolipid biosynthesis - lacto and neolacto series\"),\n",
    "\t(\"00603\", \"Glycosphingolipid biosynthesis - globo and isoglobo series\"),\n",
    "\t(\"00604\", \"Glycosphingolipid biosynthesis - ganglio series\"),\n",
    "\t(\"00511\", \"Other glycan degradation\")\n",
    "],\n",
    "\"Metabolism of cofactors and vitamins\":\n",
    "[\n",
    "\t(\"00730\", \"Thiamine metabolism\"),\n",
    "\t(\"00740\", \"Riboflavin metabolism\"),\n",
    "\t(\"00750\", \"Vitamin B6 metabolism\"),\n",
    "\t(\"00760\", \"Nicotinate and nicotinamide metabolism\"),\n",
    "\t(\"00770\", \"Pantothenate and CoA biosynthesis\"),\n",
    "\t(\"00780\", \"Biotin metabolism\"),\n",
    "\t(\"00785\", \"Lipoic acid metabolism\"),\n",
    "\t(\"00790\", \"Folate biosynthesis\"),\n",
    "\t(\"00670\", \"One carbon pool by folate\"),\n",
    "\t(\"00830\", \"Retinol metabolism\"),\n",
    "\t(\"00860\", \"Porphyrin metabolism\"),\n",
    "\t(\"00130\", \"Ubiquinone and other terpenoid-quinone biosynthesis\")\n",
    "],\n",
    "\"Metabolism of terpenoids and polyketides\":\n",
    "[\n",
    "\t(\"00900\", \"Terpenoid backbone biosynthesis\")\n",
    "],\n",
    "\"Biosynthesis of other secondary metabolites\":\n",
    "[\n",
    "\t(\"00232\", \"Caffeine metabolism\"),\n",
    "\t(\"00524\", \"Neomycin, kanamycin and gentamicin biosynthesis\")\n",
    "],\n",
    "\"Xenobiotics biodegradation and metabolism\":\n",
    "[\n",
    "\t(\"00980\", \"Metabolism of xenobiotics by cytochrome P450\"),\n",
    "\t(\"00982\", \"Drug metabolism - cytochrome P450\"),\n",
    "\t(\"00983\", \"Drug metabolism - other enzymes\")\n",
    "]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "35f00122",
   "metadata": {},
   "outputs": [],
   "source": [
    "organismDict = {\"hsa\": \"Homo sapiens (human)\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1b4206df",
   "metadata": {},
   "outputs": [],
   "source": [
    "outputDirectory = \"/Users/boldi/Desktop/pw/\"\n",
    "dataDirectory = \"../../LaTeX/Data/KEGG-Pathways/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1fa56a6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def lratio(needle, haystack):\n",
    "    \"\"\"\n",
    "        Given a string needle and a list of strings haystack, returns the maximum \n",
    "        Levenshtein ratio between needle and the elements of haystack (case insensitive match). This value ranges\n",
    "        from 1 (the needle is present in the haystack) to 0.\n",
    "    \"\"\"\n",
    "    return max([ratio(needle.lower(), hay.lower()) for hay in haystack])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "597655a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_compound_KEGG(compound_name, compound_formula=None):\n",
    "    \"\"\"\n",
    "        Search for compound on KEGG. If compound_formula is provided, the formula is used instead.\n",
    "        \n",
    "        Returns ID, list of names as from KEGG, best Levenshtein ratio (w.r.t. the names).\n",
    "        If nothing is found, IDs and list are both None, and ratio is 0.0.\n",
    "    \"\"\"\n",
    "    if compound_formula is None:\n",
    "        url1 = \"https://www.kegg.jp/kegg-bin/search_ligand?query=\" + quote_plus(compound_name) + \"&column=entry%2Bname%2Bformula&DATABASE=compound&STEP=1000\"\n",
    "    else:\n",
    "        url1 = \"https://www.kegg.jp/kegg-bin/search_ligand?query=\" + quote_plus(compound_formula) + \"&column=entry%2Bname%2Bformula&DATABASE=compound&STEP=1000\"\n",
    "    response1 = requests.get(url1, allow_redirects=False)    \n",
    "    url2 = \"https://www.kegg.jp/kegg-bin/\" + response1.headers[\"Location\"]\n",
    "    response2 = requests.get(url2)\n",
    "    soup = BeautifulSoup(response2.text, \"html.parser\")\n",
    "    tds = [tds for tds in soup.find_all(\"td\", {\"class\": \"data1\"})]\n",
    "    cd = []\n",
    "    d = {}\n",
    "    for i in range(len(tds)//5):\n",
    "        compound_id = tds[5*i+1].getText()\n",
    "        names = tds[5*i+3].getText(strip=True, separator=\"|\").split(\"|\")\n",
    "        cd += [compound_id]\n",
    "        d[compound_id]=names\n",
    "    if cd:\n",
    "        i = max(range(len(cd)), key = lambda i: lratio(compound_name, d[cd[i]])) # Best match\n",
    "        return cd[i], d[cd[i]], lratio(compound_name, d[cd[i]])\n",
    "    else:\n",
    "        return None, None, 0.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "84c62e0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_compound_chemspider(compound_name, chemspider_api_key=\"ZOoKYwWPa9AIpeSu3f90LWZ9NHn5TY2A\"):\n",
    "    \"\"\"\n",
    "        Search for compound on ChemSpider, using a valid API key. \n",
    "        \n",
    "        Returns common name, molecular formula, best Levenshtein ratio (w.r.t. the names).\n",
    "        If nothing is found, the first two entries are both None, and ratio is 0.0.\n",
    "    \"\"\"\n",
    "\n",
    "    if chemspider_api_key is None:\n",
    "        return None, None, 0.0\n",
    "    cs = ChemSpider(chemspider_api_key)\n",
    "    results = cs.search(compound_name)\n",
    "    results.wait()\n",
    "    if results.status != \"Complete\":\n",
    "        logging.info(f\"Failed search from ChemSpider, status: {results.status}\")\n",
    "        logging.info(f\"Message: {results.message}\")\n",
    "        return None, None, 0.0\n",
    "    if len(results) == 0:\n",
    "        return None, None, 0.0\n",
    "    i = max(range(len(results)), key = lambda i: lratio(compound_name, results[i].common_name))\n",
    "    return results[i].common_name, results[i].molecular_formula, lratio(compound_name, results[i].common_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2ed76756",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_compound(compound_name):\n",
    "    \"\"\"\n",
    "        Search for compound on KEGG and (as a last resort, if no or insufficient match is found) on ChemSpider and\n",
    "        then again on KEGG through molecular formula. \n",
    "        \n",
    "        Returns ID, list of names as from KEGG, best Levenshtein ratio (w.r.t. the names).\n",
    "        If nothing is found, IDs and list are both None, and ratio is 0.0.\n",
    "    \"\"\"    \n",
    "    logging.info(f\"Looking for {compound_name} on KEGG\")\n",
    "    cid, cnames, keggratio = search_compound_KEGG(compound_name)\n",
    "    logging.info(f\"Best match {cnames} with ratio {keggratio}\")\n",
    "    if keggratio < 0.5: #Bad or no match\n",
    "        logging.info(f\"Insufficient! Trying ChemSpider\")\n",
    "        spname, spformula, spratio = search_compound_chemspider(compound_name)\n",
    "        logging.info(f\"ChemSpider best match {spname} with ratio {spratio} and formula {spformula}\")\n",
    "        if spformula is None:\n",
    "            return cid, cnames, keggratio\n",
    "        compound_formula = spformula.translate(str.maketrans(\"\", \"\", \"_{}\"))   # Remove special characters\n",
    "        logging.info(f\"Looking for {compound_formula} on KEGG\")\n",
    "        sid, snames, skeggratio = search_compound_KEGG(compound_name, compound_formula)\n",
    "        logging.info(f\"Best match {snames} with ratio {skeggratio}\")\n",
    "        if skeggratio > keggratio:\n",
    "            return sid, snames, skeggratio\n",
    "    return cid, cnames, keggratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8bfade83",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_list_to_csv(input_filename, output_filename):\n",
    "    basename = input_filename\n",
    "    with open(input_filename) as f:\n",
    "        compounds = f.read().splitlines()\n",
    "    with open(output_filename, \"w\") as f:\n",
    "        for compound in compounds:\n",
    "            if len(compound.strip()) == 0:\n",
    "                continue\n",
    "            cid, cnames, lr = search_compound(compound)\n",
    "            if cid is not None:\n",
    "                jnames = \" | \".join(cnames)\n",
    "                f.write(f\"{cid},\\\"{jnames}\\\",\\\"{compound}\\\",{lr}\\n\")\n",
    "            else:\n",
    "                f.write(f\"?,?,\\\"{compound}\\\",{lr}\\n\")\n",
    "            f.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "448a033d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_compound_csv(filename, min_levenshtein=0.9):\n",
    "    res = []\n",
    "    with open(filename, \"r\") as f:\n",
    "        csvreader = csv.reader(f)\n",
    "        for row in csvreader:\n",
    "            if float(row[3]) < min_levenshtein:\n",
    "                continue\n",
    "            res += [(row[0], row[2])]\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "88d1a057",
   "metadata": {},
   "outputs": [],
   "source": [
    "valA = \"#FF0000\"\n",
    "valB = \"#0000FF\"\n",
    "\n",
    "def writeKEGGpdfcolors(outputDirectory, organism, superpathway, superpathwayDict, compound2color):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    index = []\n",
    "    safeSuperpathway = makesafe(superpathway)\n",
    "    for pw in superpathwayDict[superpathway]:\n",
    "        countColor = {}\n",
    "        for v in set(compound2color.values()):\n",
    "            countColor[v] = set([])\n",
    "        try:\n",
    "            pathway = KGML_parser.read(kegg_get(organism + pw[0], \"kgml\"))\n",
    "        except:\n",
    "            print(\"Pathway\", pw, \"could not be downloaded: ignoring\")\n",
    "            continue\n",
    "        canvas = KGMLCanvas(pathway)\n",
    "        for k in pathway.entries:\n",
    "            t = pathway.entries[k].type\n",
    "            pathway.entries[k].graphics[0].bgcolor = \"#FFFFFF\"\n",
    "            if t == \"compound\":\n",
    "                compound = pathway.entries[k].name[4:]\n",
    "                if compound in compound2color:\n",
    "                    pathway.entries[k].graphics[0].bgcolor = compound2color[compound]\n",
    "                    pathway.entries[k].graphics[0].fgcolor = compound2color[compound]\n",
    "                    countColor[compound2color[compound]] |= set([compound])\n",
    "        canvas.import_imagemap = True\n",
    "        pdfName = organism + pw[0] + \".pdf\"\n",
    "        pathlib.Path(os.path.join(outputDirectory, organism, safeSuperpathway)).mkdir(parents=True, exist_ok=True)\n",
    "        canvas.draw(os.path.join(outputDirectory, organism, safeSuperpathway, pdfName))\n",
    "        index += [(pdfName, pw[1], len(countColor[valA]), len(countColor[valB]))]\n",
    "    return index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ae27a306",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeIndex(directory, index, title):\n",
    "    with open(os.path.join(directory, \"index.html\"), \"w\") as file:\n",
    "        file.write(\"<!DOCTYPE html>\\n<html>\\n\")\n",
    "        file.write(f\"\\t<h1>{title}</h1>\\n\")\n",
    "        file.write(\"\\t<table border=1>\\n\")\n",
    "        file.write(\"\\t\\t<thead><tr><th>Pathway<th>ListA<th>ListB</thead>\\n\")\n",
    "        file.write(\"\\t\\t<tbody>\\n\")\n",
    "        for link, anchor, cA, cB in index:\n",
    "            file.write(\"\\t\\t<tr><td><a href=\\\"{}\\\">{}</a><td>{}<td>{}\\n\".format(link, anchor, cA, cB))\n",
    "        file.write(\"\\t</tbody>\\n\")\n",
    "        file.write(\"\\t</table>\\n\")\n",
    "        file.write(\"</html>\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7fd75bda",
   "metadata": {},
   "outputs": [],
   "source": [
    "tA = read_compound_csv(\"/Users/boldi/Desktop/listA-complete.csv\")\n",
    "tB = read_compound_csv(\"/Users/boldi/Desktop/listB-complete.csv\")\n",
    "\n",
    "compound2color = {}\n",
    "for x in tA:\n",
    "    compound2color[x[0]] = valA\n",
    "for x in tB:\n",
    "    compound2color[x[0]] = valB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b9885cb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### PRODUCE all KEGG pdfs and index.html\n",
    "\n",
    "for organism in []: #organismDict.keys():\n",
    "    superIndex = []\n",
    "    for superpathway in superpathwayDict.keys():\n",
    "        safeSuperpathway = makesafe(superpathway)\n",
    "        index = writeKEGGpdfcolors(outputDirectory, organism, superpathway, superpathwayDict, compound2color)\n",
    "        sumA = sum([x[2] for x in index])\n",
    "        sumB = sum([x[3] for x in index])\n",
    "        writeIndex(os.path.join(outputDirectory, organism, safeSuperpathway), index, \n",
    "               f\"{organismDict[organism]}, {superpathway} (A: {sumA}, B: {sumB})\")\n",
    "        superIndex += [(safeSuperpathway, superpathway, sumA, sumB)]\n",
    "    sumA = sum([x[2] for x in superIndex])\n",
    "    sumB = sum([x[3] for x in superIndex])\n",
    "    writeIndex(os.path.join(outputDirectory, organism, \"index.html\"), superIndex, \n",
    "               f\"{organismDict[organism]} (A: {sumA}, B: {sumB})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "add37631",
   "metadata": {},
   "outputs": [],
   "source": [
    "### PRODUCE a dictionary compounds with the set of compounds appearing in the whole organism, in each\n",
    "### superpathway and in each pathway\n",
    "compounds = {}\n",
    "for organism in organismDict.keys():\n",
    "    compounds[organism] = set([])\n",
    "    for superpathway in superpathwayDict.keys():\n",
    "        compounds[superpathway] = set([])\n",
    "        for pw in superpathwayDict[superpathway]:\n",
    "            pathway = read(open(os.path.join(dataDirectory, organism, f\"{organism}{pw[0]}.xml\"), \"r\"))\n",
    "            compounds[pw[0]] = set([x.name[4:] for x in pathway.compounds])\n",
    "            compounds[superpathway] |= compounds[pw[0]]\n",
    "            compounds[organism] |= compounds[pw[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7483c7c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hsa,3023,58,15\n",
      "Carbohydrate metabolism,503,21,6\n",
      "00010,31,3,0\n",
      "00020,20,0,0\n",
      "00030,36,5,0\n",
      "00040,58,4,2\n",
      "00051,54,2,0\n",
      "00052,46,1,0\n",
      "00053,57,1,1\n",
      "00500,37,0,2\n",
      "00520,118,4,1\n",
      "00620,32,1,0\n",
      "00630,64,3,1\n",
      "00640,41,3,1\n",
      "00650,47,4,1\n",
      "00562,47,2,0\n",
      "Energy metabolism,67,5,3\n",
      "00190,16,1,3\n",
      "00910,19,2,0\n",
      "00920,33,2,0\n",
      "Lipid metabolism,659,8,2\n",
      "00061,58,1,0\n",
      "00062,40,0,0\n",
      "00071,50,1,0\n",
      "00100,57,1,0\n",
      "00120,48,1,0\n",
      "00140,99,0,0\n",
      "00561,38,2,2\n",
      "00564,56,3,2\n",
      "00565,25,0,0\n",
      "00600,27,0,0\n",
      "00590,79,0,0\n",
      "00591,28,0,0\n",
      "00592,44,0,0\n",
      "01040,74,0,0\n",
      "Nucleotide metabolism,159,9,2\n",
      "00230,101,5,1\n",
      "00240,64,4,1\n",
      "Amino acid metabolism,550,17,3\n",
      "00250,28,0,1\n",
      "00260,48,3,0\n",
      "00270,66,4,0\n",
      "00280,41,4,1\n",
      "00290,23,2,0\n",
      "00310,51,2,0\n",
      "00220,23,1,0\n",
      "00330,69,3,2\n",
      "00340,47,1,0\n",
      "00350,78,1,0\n",
      "00360,49,1,0\n",
      "00380,83,0,0\n",
      "00400,34,1,0\n",
      "Metabolism of other amino acids,220,8,2\n",
      "00410,32,1,2\n",
      "00430,24,0,0\n",
      "00440,53,1,0\n",
      "00450,27,2,0\n",
      "00470,69,3,0\n",
      "00480,38,4,1\n",
      "Glycan biosynthesis and metabolism,270,2,0\n",
      "00510,38,0,0\n",
      "00513,31,0,0\n",
      "00512,17,0,0\n",
      "00515,27,0,0\n",
      "00514,0,0,0\n",
      "00532,12,1,0\n",
      "00534,11,1,0\n",
      "00533,0,0,0\n",
      "00531,27,0,0\n",
      "00563,19,1,0\n",
      "00601,54,0,0\n",
      "00603,23,0,0\n",
      "00604,31,0,0\n",
      "00511,0,0,0\n",
      "Metabolism of cofactors and vitamins,497,12,4\n",
      "00730,31,0,1\n",
      "00740,24,2,1\n",
      "00750,29,1,0\n",
      "00760,55,2,1\n",
      "00770,30,1,0\n",
      "00780,29,1,0\n",
      "00785,13,0,0\n",
      "00790,58,1,0\n",
      "00670,9,0,0\n",
      "00830,24,0,0\n",
      "00860,144,3,0\n",
      "00130,71,1,1\n",
      "Metabolism of terpenoids and polyketides,46,2,0\n",
      "00900,46,2,0\n",
      "Biosynthesis of other secondary metabolites,103,1,0\n",
      "00232,22,0,0\n",
      "00524,81,1,0\n",
      "Xenobiotics biodegradation and metabolism,260,0,0\n",
      "00980,121,0,0\n",
      "00982,87,0,0\n",
      "00983,52,0,0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def printLine(key, compoundDict, listsOfInterest):\n",
    "    \"\"\"\n",
    "        Interrogate compoundDict with key, and obtain a set.\n",
    "        Print the key, the size of the set, and for each of the (key,value) pairs in the\n",
    "        dictionary, the size of the intersection between the set and the value\n",
    "    \"\"\"\n",
    "    whole = compoundDict[key]\n",
    "    nwhole = len(whole)\n",
    "    intersections = \",\".join([str(len(whole & value)) for key, value in listsOfInterest.items()])\n",
    "    print(f\"{key},{nwhole},{intersections}\")\n",
    "            \n",
    "tA = read_compound_csv(\"/Users/boldi/Desktop/listA-complete.csv\")\n",
    "tB = read_compound_csv(\"/Users/boldi/Desktop/listB-complete.csv\")\n",
    "setA = set([x[0] for x in tA])\n",
    "setB = set([x[0] for x in tB])\n",
    "\n",
    "\n",
    "        \n",
    "for organism in organismDict.keys():\n",
    "    printLine(organism, compounds, {'A': setA, 'B': setB})\n",
    "    for superpathway in superpathwayDict.keys():\n",
    "        printLine(superpathway, compounds, {'A': setA, 'B': setB})\n",
    "        for pw in superpathwayDict[superpathway]:\n",
    "            printLine(pw[0], compounds, {'A': setA, 'B': setB})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "856f7e0f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9d47f255",
   "metadata": {},
   "outputs": [],
   "source": [
    "H, sr2r, r2ss, rid2rname, mid2mname = read_hg_from_SBML(\"../../LaTeX/Data/KEGG-Pathways/hsa/hsa00040-sbml.xml\")\n",
    "G, dd = hg2g(H)\n",
    "cc = nx.closeness_centrality(G)\n",
    "\n",
    "ccA = []\n",
    "for x in setA:\n",
    "    for node in H.nodes():\n",
    "        if str(node)==x:\n",
    "            ccA += [cc[x]]\n",
    "\n",
    "ccB = []\n",
    "for x in setB:\n",
    "    for node in H.nodes():\n",
    "        if str(node)==x:\n",
    "            ccB += [cc[x]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "26a53e63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([0.0878048780487805], [0.0821566110397946])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccA, ccB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e8494df7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C00022',\n",
       " 'C00026',\n",
       " 'C00029',\n",
       " 'C00103',\n",
       " 'C00111',\n",
       " 'C00116',\n",
       " 'C00167',\n",
       " 'C00181',\n",
       " 'C00191',\n",
       " 'C00199',\n",
       " 'C00204',\n",
       " 'C00216',\n",
       " 'C00231',\n",
       " 'C00259',\n",
       " 'C00266',\n",
       " 'C00309',\n",
       " 'C00310',\n",
       " 'C00312',\n",
       " 'C00333',\n",
       " 'C00379',\n",
       " 'C00433',\n",
       " 'C00470',\n",
       " 'C00474',\n",
       " 'C00476',\n",
       " 'C00502',\n",
       " 'C00508',\n",
       " 'C00514',\n",
       " 'C00532',\n",
       " 'C00558',\n",
       " 'C00618',\n",
       " 'C00714',\n",
       " 'C00789',\n",
       " 'C00800',\n",
       " 'C00817',\n",
       " 'C00905',\n",
       " 'C01068',\n",
       " 'C01101',\n",
       " 'C01508',\n",
       " 'C01904',\n",
       " 'C02266',\n",
       " 'C02273',\n",
       " 'C02426',\n",
       " 'C02753',\n",
       " 'C03033',\n",
       " 'C03291',\n",
       " 'C03826',\n",
       " 'C04053',\n",
       " 'C04349',\n",
       " 'C04575',\n",
       " 'C05385',\n",
       " 'C05411',\n",
       " 'C05412',\n",
       " 'C06118',\n",
       " 'C06441',\n",
       " 'C14899',\n",
       " 'C15930',\n",
       " 'C20680',\n",
       " 'C22337'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compounds['00040']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f5b71b84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C03033 True\n",
      "C00191 True\n",
      "C00309 True\n",
      "C00199 True\n",
      "C00231 True\n",
      "C01068 True\n",
      "C00789 True\n",
      "C00310 True\n",
      "C00379 True\n",
      "C00312 True\n",
      "C00181 True\n",
      "C00800 True\n",
      "C00618 True\n",
      "C00103 True\n",
      "C00029 True\n",
      "C00167 True\n",
      "C02266 True\n"
     ]
    }
   ],
   "source": [
    "for x in H.nodes():\n",
    "    print(x, str(x) in compounds['00040'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "034f62c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "document = libsbml.readSBMLFromFile(\"../../LaTeX/Data/KEGG-Pathways/hsa/hsa00040-sbml.xml\")\n",
    "model = document.getModel()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7767b6c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C00022 <annotation>\n",
      "  <rdf:RDF xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\" xmlns:bqbiol=\"http://biomodels.net/biology-qualifiers/\">\n",
      "    <rdf:Description rdf:about=\"#meta_C00022\">\n",
      "      <bqbiol:is>\n",
      "        <rdf:Bag>\n",
      "          <rdf:li rdf:resource=\"http://identifiers.org/kegg.compound/C00022\"/>\n",
      "        </rdf:Bag>\n",
      "      </bqbiol:is>\n",
      "      <bqbiol:isDescribedBy>\n",
      "        <rdf:Bag>\n",
      "          <rdf:li rdf:resource=\"http://identifiers.org/obo.eco/ECO:0000313\"/>\n",
      "        </rdf:Bag>\n",
      "      </bqbiol:isDescribedBy>\n",
      "    </rdf:Description>\n",
      "  </rdf:RDF>\n",
      "</annotation>\n"
     ]
    }
   ],
   "source": [
    "species = model.getListOfSpecies()\n",
    "for specie in species:\n",
    "    if specie.getId() == \"C00022\":\n",
    "        print(specie.getId(), specie.getAnnotationString())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3c28d7c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Species K01184 \"K01184\">"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "effb8e57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "66 ko:K01184 ortholog\n",
      "67 cpd:C14899 compound\n",
      "68 ko:K03078 ortholog\n",
      "69 ko:K00880 ortholog\n",
      "70 cpd:C03033 compound\n",
      "71 hsa:2990 hsa:9365 gene\n",
      "72 ko:K01804 ortholog\n",
      "74 ko:K00853 ko:K24707 ortholog\n",
      "75 ko:K00039 ortholog\n",
      "76 hsa:55277 gene\n",
      "77 ko:K00853 ko:K24707 ortholog\n",
      "78 hsa:6120 hsa:729020 gene\n",
      "79 ko:K03077 ortholog\n",
      "80 ko:K21680 ortholog\n",
      "81 hsa:729920 gene\n",
      "82 ko:K03079 ortholog\n",
      "83 ko:K00007 ortholog\n",
      "84 hsa:9942 gene\n",
      "85 ko:K09988 ortholog\n",
      "86 ko:K22397 ortholog\n",
      "87 ko:K14275 ko:K22186 ko:K22396 ortholog\n",
      "88 ko:K00880 ortholog\n",
      "89 ko:K01820 ortholog\n",
      "90 ko:K05351 ortholog\n",
      "91 ko:K01805 ortholog\n",
      "92 ko:K14274 ortholog\n",
      "93 hsa:231 hsa:57016 gene\n",
      "94 ko:K25880 ortholog\n",
      "95 hsa:51181 gene\n",
      "96 hsa:231 hsa:57016 gene\n",
      "97 ko:K22185 ortholog\n",
      "99 hsa:51084 gene\n",
      "100 hsa:7360 gene\n",
      "101 hsa:10720 hsa:10941 hsa:54490 hsa:54575 hsa:54576 hsa:54577 hsa:54578 hsa:54579 hsa:54600 hsa:54657 hsa:54658 hsa:54659 hsa:574537 hsa:7363 hsa:7364 hsa:7365 hsa:7366 hsa:7367 hsa:79799 gene\n",
      "102 hsa:7358 gene\n",
      "104 ko:K16190 ortholog\n",
      "105 ko:K12447 ortholog\n",
      "106 ko:K01812 ortholog\n",
      "107 ko:K00040 ortholog\n",
      "108 ko:K01686 ko:K08323 ortholog\n",
      "109 ko:K01685 ko:K16849 ko:K16850 ortholog\n",
      "110 ko:K01812 ortholog\n",
      "111 ko:K00041 ortholog\n",
      "112 path:hsa00010 map\n",
      "113 path:hsa00562 map\n",
      "114 path:hsa00053 map\n",
      "115 path:hsa00040 map\n",
      "116 path:hsa00030 map\n",
      "117 path:hsa00052 map\n",
      "118 path:hsa00520 map\n",
      "119 path:hsa00500 map\n",
      "120 path:hsa00740 map\n",
      "123 ko:K08092 ortholog\n",
      "124 ko:K01213 ko:K22933 ortholog\n",
      "125 ko:K01184 ortholog\n",
      "126 ko:K01051 ortholog\n",
      "127 ko:K22539 ko:K01728 ko:K19551 ortholog\n",
      "128 ko:K01730 ortholog\n",
      "129 ko:K01815 ortholog\n",
      "130 ko:K00065 ortholog\n",
      "131 ko:K00848 ortholog\n",
      "132 ko:K01629 ortholog\n",
      "133 cpd:C00817 compound\n",
      "134 cpd:C00103 compound\n",
      "135 cpd:C00905 compound\n",
      "136 cpd:C00514 compound\n",
      "137 cpd:C00558 compound\n",
      "138 cpd:C00800 compound\n",
      "139 cpd:C05385 compound\n",
      "140 cpd:C00167 compound\n",
      "141 cpd:C00029 compound\n",
      "142 cpd:C00312 compound\n",
      "143 cpd:C02266 compound\n",
      "144 cpd:C00379 compound\n",
      "145 cpd:C00181 compound\n",
      "146 cpd:C01068 compound\n",
      "147 cpd:C00789 compound\n",
      "148 cpd:C00022 compound\n",
      "149 cpd:C03826 compound\n",
      "150 cpd:C00502 compound\n",
      "151 cpd:C00309 compound\n",
      "152 cpd:C00474 compound\n",
      "153 cpd:C00231 compound\n",
      "154 cpd:C00310 compound\n",
      "155 cpd:C00199 compound\n",
      "156 cpd:C00476 compound\n",
      "157 cpd:C00508 compound\n",
      "158 cpd:C00532 compound\n",
      "159 cpd:C03291 compound\n",
      "160 cpd:C01508 compound\n",
      "161 cpd:C01101 compound\n",
      "162 cpd:C00259 compound\n",
      "163 cpd:C01904 compound\n",
      "164 cpd:C00618 compound\n",
      "165 cpd:C00714 compound\n",
      "166 cpd:C00470 compound\n",
      "167 cpd:C02273 compound\n",
      "168 cpd:C06118 compound\n",
      "169 cpd:C04053 compound\n",
      "170 cpd:C04349 compound\n",
      "171 cpd:C05411 compound\n",
      "172 cpd:C04575 compound\n",
      "173 cpd:C05412 compound\n",
      "174 cpd:C06441 compound\n",
      "175 cpd:C00266 compound\n",
      "176 cpd:C00111 compound\n",
      "191 ko:K01731 ko:K22994 ortholog\n",
      "193 hsa:27294 gene\n",
      "194 cpd:C00204 compound\n",
      "196 path:hsa00030 map\n",
      "198 ko:K17738 ortholog\n",
      "199 ko:K17818 ortholog\n",
      "200 ko:K17818 ortholog\n",
      "201 ko:K18107 ortholog\n",
      "202 ko:K18102 ortholog\n",
      "203 ko:K18103 ortholog\n",
      "204 path:hsa00561 map\n",
      "207 cpd:C00116 compound\n",
      "208 ko:K18106 ortholog\n",
      "210 cpd:C00333 compound\n",
      "211 cpd:C15930 compound\n",
      "212 cpd:C20680 compound\n",
      "213 cpd:C02426 compound\n",
      "214 ko:K17743 ko:K26066 ortholog\n",
      "215 hsa:6652 gene\n",
      "216 hsa:10327 gene\n",
      "217 cpd:C00191 compound\n",
      "219 ko:K08322 ortholog\n",
      "221 cpd:C00433 compound\n",
      "222 ko:K14259 ortholog\n",
      "224 ko:K13877 ko:K19588 ortholog\n",
      "225 cpd:C00026 compound\n",
      "226 path:hsa00020 map\n",
      "228 ko:K19634 ortholog\n",
      "230 ko:K00594 ortholog\n",
      "232 ko:K21619 ortholog\n",
      "233 ko:K05352 ko:K21681 ortholog\n",
      "234 ko:K22187 ortholog\n",
      "237 cpd:C02753 compound\n",
      "238 ko:K24442 ortholog\n",
      "239 ko:K24443 ortholog\n",
      "241 cpd:C00216 compound\n",
      "246 ko:K01818 ortholog\n",
      "249 cpd:C22337 compound\n",
      "250 ko:K00879 ortholog\n",
      "253 cpd:C00266 compound\n",
      "254 ko:K01628 ortholog\n",
      "255 cpd:C00111 compound\n",
      "256 ko:K25881 ko:K26066 ortholog\n",
      "66 ko:K01184 ortholog\n",
      "67 cpd:C14899 compound\n",
      "68 ko:K03078 ortholog\n",
      "69 ko:K00880 ortholog\n",
      "70 cpd:C03033 compound\n",
      "71 hsa:2990 hsa:9365 gene\n",
      "72 ko:K01804 ortholog\n",
      "74 ko:K00853 ko:K24707 ortholog\n",
      "75 ko:K00039 ortholog\n",
      "76 hsa:55277 gene\n",
      "77 ko:K00853 ko:K24707 ortholog\n",
      "78 hsa:6120 hsa:729020 gene\n",
      "79 ko:K03077 ortholog\n",
      "80 ko:K21680 ortholog\n",
      "81 hsa:729920 gene\n",
      "82 ko:K03079 ortholog\n",
      "83 ko:K00007 ortholog\n",
      "84 hsa:9942 gene\n",
      "85 ko:K09988 ortholog\n",
      "86 ko:K22397 ortholog\n",
      "87 ko:K14275 ko:K22186 ko:K22396 ortholog\n",
      "88 ko:K00880 ortholog\n",
      "89 ko:K01820 ortholog\n",
      "90 ko:K05351 ortholog\n",
      "91 ko:K01805 ortholog\n",
      "92 ko:K14274 ortholog\n",
      "93 hsa:231 hsa:57016 gene\n",
      "94 ko:K25880 ortholog\n",
      "95 hsa:51181 gene\n",
      "96 hsa:231 hsa:57016 gene\n",
      "97 ko:K22185 ortholog\n",
      "99 hsa:51084 gene\n",
      "100 hsa:7360 gene\n",
      "101 hsa:10720 hsa:10941 hsa:54490 hsa:54575 hsa:54576 hsa:54577 hsa:54578 hsa:54579 hsa:54600 hsa:54657 hsa:54658 hsa:54659 hsa:574537 hsa:7363 hsa:7364 hsa:7365 hsa:7366 hsa:7367 hsa:79799 gene\n",
      "102 hsa:7358 gene\n",
      "104 ko:K16190 ortholog\n",
      "105 ko:K12447 ortholog\n",
      "106 ko:K01812 ortholog\n",
      "107 ko:K00040 ortholog\n",
      "108 ko:K01686 ko:K08323 ortholog\n",
      "109 ko:K01685 ko:K16849 ko:K16850 ortholog\n",
      "110 ko:K01812 ortholog\n",
      "111 ko:K00041 ortholog\n",
      "112 path:hsa00010 map\n",
      "113 path:hsa00562 map\n",
      "114 path:hsa00053 map\n",
      "115 path:hsa00040 map\n",
      "116 path:hsa00030 map\n",
      "117 path:hsa00052 map\n",
      "118 path:hsa00520 map\n",
      "119 path:hsa00500 map\n",
      "120 path:hsa00740 map\n",
      "123 ko:K08092 ortholog\n",
      "124 ko:K01213 ko:K22933 ortholog\n",
      "125 ko:K01184 ortholog\n",
      "126 ko:K01051 ortholog\n",
      "127 ko:K22539 ko:K01728 ko:K19551 ortholog\n",
      "128 ko:K01730 ortholog\n",
      "129 ko:K01815 ortholog\n",
      "130 ko:K00065 ortholog\n",
      "131 ko:K00848 ortholog\n",
      "132 ko:K01629 ortholog\n",
      "133 cpd:C00817 compound\n",
      "134 cpd:C00103 compound\n",
      "135 cpd:C00905 compound\n",
      "136 cpd:C00514 compound\n",
      "137 cpd:C00558 compound\n",
      "138 cpd:C00800 compound\n",
      "139 cpd:C05385 compound\n",
      "140 cpd:C00167 compound\n",
      "141 cpd:C00029 compound\n",
      "142 cpd:C00312 compound\n",
      "143 cpd:C02266 compound\n",
      "144 cpd:C00379 compound\n",
      "145 cpd:C00181 compound\n",
      "146 cpd:C01068 compound\n",
      "147 cpd:C00789 compound\n",
      "148 cpd:C00022 compound\n",
      "149 cpd:C03826 compound\n",
      "150 cpd:C00502 compound\n",
      "151 cpd:C00309 compound\n",
      "152 cpd:C00474 compound\n",
      "153 cpd:C00231 compound\n",
      "154 cpd:C00310 compound\n",
      "155 cpd:C00199 compound\n",
      "156 cpd:C00476 compound\n",
      "157 cpd:C00508 compound\n",
      "158 cpd:C00532 compound\n",
      "159 cpd:C03291 compound\n",
      "160 cpd:C01508 compound\n",
      "161 cpd:C01101 compound\n",
      "162 cpd:C00259 compound\n",
      "163 cpd:C01904 compound\n",
      "164 cpd:C00618 compound\n",
      "165 cpd:C00714 compound\n",
      "166 cpd:C00470 compound\n",
      "167 cpd:C02273 compound\n",
      "168 cpd:C06118 compound\n",
      "169 cpd:C04053 compound\n",
      "170 cpd:C04349 compound\n",
      "171 cpd:C05411 compound\n",
      "172 cpd:C04575 compound\n",
      "173 cpd:C05412 compound\n",
      "174 cpd:C06441 compound\n",
      "175 cpd:C00266 compound\n",
      "176 cpd:C00111 compound\n",
      "191 ko:K01731 ko:K22994 ortholog\n",
      "193 hsa:27294 gene\n",
      "194 cpd:C00204 compound\n",
      "196 path:hsa00030 map\n",
      "198 ko:K17738 ortholog\n",
      "199 ko:K17818 ortholog\n",
      "200 ko:K17818 ortholog\n",
      "201 ko:K18107 ortholog\n",
      "202 ko:K18102 ortholog\n",
      "203 ko:K18103 ortholog\n",
      "204 path:hsa00561 map\n",
      "207 cpd:C00116 compound\n",
      "208 ko:K18106 ortholog\n",
      "210 cpd:C00333 compound\n",
      "211 cpd:C15930 compound\n",
      "212 cpd:C20680 compound\n",
      "213 cpd:C02426 compound\n",
      "214 ko:K17743 ko:K26066 ortholog\n",
      "215 hsa:6652 gene\n",
      "216 hsa:10327 gene\n",
      "217 cpd:C00191 compound\n",
      "219 ko:K08322 ortholog\n",
      "221 cpd:C00433 compound\n",
      "222 ko:K14259 ortholog\n",
      "224 ko:K13877 ko:K19588 ortholog\n",
      "225 cpd:C00026 compound\n",
      "226 path:hsa00020 map\n",
      "228 ko:K19634 ortholog\n",
      "230 ko:K00594 ortholog\n",
      "232 ko:K21619 ortholog\n",
      "233 ko:K05352 ko:K21681 ortholog\n",
      "234 ko:K22187 ortholog\n",
      "237 cpd:C02753 compound\n",
      "238 ko:K24442 ortholog\n",
      "239 ko:K24443 ortholog\n",
      "241 cpd:C00216 compound\n",
      "246 ko:K01818 ortholog\n",
      "249 cpd:C22337 compound\n",
      "250 ko:K00879 ortholog\n",
      "253 cpd:C00266 compound\n",
      "254 ko:K01628 ortholog\n",
      "255 cpd:C00111 compound\n",
      "256 ko:K25881 ko:K26066 ortholog\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import matplotlib\n",
    "def gnagna(outputDirectory, organism, superpathway, superpathwayDict, compound2color):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    index = []\n",
    "    safeSuperpathway = makesafe(superpathway)\n",
    "    for pw in superpathwayDict[superpathway]:\n",
    "        countColor = {}\n",
    "        for v in set(compound2color.values()):\n",
    "            countColor[v] = set([])\n",
    "        try:\n",
    "            pathway = KGML_parser.read(kegg_get(\"hsa00040\", \"kgml\"))\n",
    "        except:\n",
    "            print(\"Pathway\", pw, \"could not be downloaded: ignoring\")\n",
    "            continue\n",
    "        canvas = KGMLCanvas(pathway)\n",
    "        for k in pathway.entries:\n",
    "            t = pathway.entries[k].type\n",
    "            print(k, pathway.entries[k].name, t)\n",
    "            pathway.entries[k].graphics[0].bgcolor = compound2color[t]\n",
    "            pathway.entries[k].graphics[0].fgcolor = compound2color[t]\n",
    "        for k in range(len(pathway.reaction_entries)):\n",
    "            pathway.reaction_entries[k].graphics[0].bgcolor = compound2color[\"reaction\"]\n",
    "            pathway.reaction_entries[k].graphics[0].fgcolor = compound2color[\"reaction\"]\n",
    "        canvas.import_imagemap = True\n",
    "        pdfName = organism + pw[0] + \".pdf\"\n",
    "        pathlib.Path(os.path.join(outputDirectory, organism, safeSuperpathway)).mkdir(parents=True, exist_ok=True)\n",
    "        canvas.draw(os.path.join(outputDirectory, organism, safeSuperpathway, pdfName))\n",
    "    return index\n",
    "\n",
    "\n",
    "gnagna(\"/tmp\", \"hsa\", \"Carbohydrate metabolism\", {\"Carbohydrate metabolism\": (\"00040\", \"Pentose and glucuronate interconversions\")},\n",
    "{\n",
    "    \"ortholog\": matplotlib.colors.cnames[\"blue\"], \n",
    "    \"compound\": matplotlib.colors.cnames[\"red\"],\n",
    "    \"map\": matplotlib.colors.cnames[\"yellow\"],\n",
    "    \"gene\": matplotlib.colors.cnames[\"brown\"],\n",
    "    \"reaction\": matplotlib.colors.cnames[\"green\"]\n",
    "}\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "a2fe3970",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entries:  149\n",
      "Orthologs:  62 Compounds:  60 Maps:  12 Genes:  15\n",
      "\n",
      "Entry node ID: 86\n",
      "Names: ko:K22397\n",
      "Type: ortholog\n",
      "Components: set()\n",
      "Reactions: rn:R01782\n",
      "Graphics elements: 1 [<Bio.KEGG.KGML.KGML_pathway.Graphics object at 0x13b9ebaf0>]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pathway = KGML_parser.read(kegg_get(\"hsa00040\", \"kgml\"))\n",
    "print(\"Entries: \",len(pathway.entries))\n",
    "print(\"Orthologs: \", len(pathway.orthologs), \"Compounds: \", len(pathway.compounds), \n",
    "      \"Maps: \", len(pathway.maps), \"Genes: \", len(pathway.genes))\n",
    "print()\n",
    "print(pathway.entries[86])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a85ed81c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entry node ID: 86\n",
      "Names: ko:K22397\n",
      "Type: ortholog\n",
      "Components: set()\n",
      "Reactions: rn:R01782\n",
      "Graphics elements: 1 [<Bio.KEGG.KGML.KGML_pathway.Graphics object at 0x13b31b280>]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(pathway.orthologs[12])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "9374ba6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 rn:R01478\n",
      "1 rn:R01526\n",
      "2 rn:R01529\n",
      "3 rn:R02921\n",
      "4 rn:R01639\n",
      "5 rn:R01758 rn:R01759\n",
      "6 rn:R01904\n",
      "7 rn:R01431\n",
      "8 rn:R02640\n",
      "9 rn:R00289\n",
      "10 rn:R01383\n",
      "11 rn:R00286\n",
      "12 rn:R01430\n",
      "13 rn:R01896\n",
      "14 rn:R01481\n",
      "\n",
      "0 rn:R01478\n",
      "1 rn:R01526\n",
      "2 rn:R01529\n",
      "3 rn:R02921\n",
      "4 rn:R01639\n",
      "5 rn:R01758 rn:R01759\n",
      "6 rn:R01904\n",
      "7 rn:R01431\n",
      "8 rn:R02640\n",
      "9 rn:R00289\n",
      "10 rn:R01383\n",
      "11 rn:R00286\n",
      "12 rn:R01430\n",
      "13 rn:R01896\n",
      "14 rn:R01481\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(pathway.reaction_entries)):\n",
    "    print(i, pathway.reaction_entries[i].reaction)\n",
    "    \n",
    "print()\n",
    "lpr=list(pathway.reactions)\n",
    "for i in range(len(lpr)):\n",
    "    print(i, lpr[i].name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "4e43d24e",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[63], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mpathway\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mentries\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mname)\n",
      "\u001b[0;31mKeyError\u001b[0m: 0"
     ]
    }
   ],
   "source": [
    "print(pathway.entries[0].name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "8a30895a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "66 ko:K01184\n",
      "67 cpd:C14899\n",
      "68 ko:K03078\n",
      "69 ko:K00880\n",
      "70 cpd:C03033\n",
      "71 hsa:2990 hsa:9365\n",
      "72 ko:K01804\n",
      "74 ko:K00853 ko:K24707\n",
      "75 ko:K00039\n",
      "76 hsa:55277\n",
      "77 ko:K00853 ko:K24707\n",
      "78 hsa:6120 hsa:729020\n",
      "79 ko:K03077\n",
      "80 ko:K21680\n",
      "81 hsa:729920\n",
      "82 ko:K03079\n",
      "83 ko:K00007\n",
      "84 hsa:9942\n",
      "85 ko:K09988\n",
      "86 ko:K22397\n",
      "87 ko:K14275 ko:K22186 ko:K22396\n",
      "88 ko:K00880\n",
      "89 ko:K01820\n",
      "90 ko:K05351\n",
      "91 ko:K01805\n",
      "92 ko:K14274\n",
      "93 hsa:231 hsa:57016\n",
      "94 ko:K25880\n",
      "95 hsa:51181\n",
      "96 hsa:231 hsa:57016\n",
      "97 ko:K22185\n",
      "99 hsa:51084\n",
      "100 hsa:7360\n",
      "101 hsa:10720 hsa:10941 hsa:54490 hsa:54575 hsa:54576 hsa:54577 hsa:54578 hsa:54579 hsa:54600 hsa:54657 hsa:54658 hsa:54659 hsa:574537 hsa:7363 hsa:7364 hsa:7365 hsa:7366 hsa:7367 hsa:79799\n",
      "102 hsa:7358\n",
      "104 ko:K16190\n",
      "105 ko:K12447\n",
      "106 ko:K01812\n",
      "107 ko:K00040\n",
      "108 ko:K01686 ko:K08323\n",
      "109 ko:K01685 ko:K16849 ko:K16850\n",
      "110 ko:K01812\n",
      "111 ko:K00041\n",
      "112 path:hsa00010\n",
      "113 path:hsa00562\n",
      "114 path:hsa00053\n",
      "115 path:hsa00040\n",
      "116 path:hsa00030\n",
      "117 path:hsa00052\n",
      "118 path:hsa00520\n",
      "119 path:hsa00500\n",
      "120 path:hsa00740\n",
      "123 ko:K08092\n",
      "124 ko:K01213 ko:K22933\n",
      "125 ko:K01184\n",
      "126 ko:K01051\n",
      "127 ko:K22539 ko:K01728 ko:K19551\n",
      "128 ko:K01730\n",
      "129 ko:K01815\n",
      "130 ko:K00065\n",
      "131 ko:K00848\n",
      "132 ko:K01629\n",
      "133 cpd:C00817\n",
      "134 cpd:C00103\n",
      "135 cpd:C00905\n",
      "136 cpd:C00514\n",
      "137 cpd:C00558\n",
      "138 cpd:C00800\n",
      "139 cpd:C05385\n",
      "140 cpd:C00167\n",
      "141 cpd:C00029\n",
      "142 cpd:C00312\n",
      "143 cpd:C02266\n",
      "144 cpd:C00379\n",
      "145 cpd:C00181\n",
      "146 cpd:C01068\n",
      "147 cpd:C00789\n",
      "148 cpd:C00022\n",
      "149 cpd:C03826\n",
      "150 cpd:C00502\n",
      "151 cpd:C00309\n",
      "152 cpd:C00474\n",
      "153 cpd:C00231\n",
      "154 cpd:C00310\n",
      "155 cpd:C00199\n",
      "156 cpd:C00476\n",
      "157 cpd:C00508\n",
      "158 cpd:C00532\n",
      "159 cpd:C03291\n",
      "160 cpd:C01508\n",
      "161 cpd:C01101\n",
      "162 cpd:C00259\n",
      "163 cpd:C01904\n",
      "164 cpd:C00618\n",
      "165 cpd:C00714\n",
      "166 cpd:C00470\n",
      "167 cpd:C02273\n",
      "168 cpd:C06118\n",
      "169 cpd:C04053\n",
      "170 cpd:C04349\n",
      "171 cpd:C05411\n",
      "172 cpd:C04575\n",
      "173 cpd:C05412\n",
      "174 cpd:C06441\n",
      "175 cpd:C00266\n",
      "176 cpd:C00111\n",
      "191 ko:K01731 ko:K22994\n",
      "193 hsa:27294\n",
      "194 cpd:C00204\n",
      "196 path:hsa00030\n",
      "198 ko:K17738\n",
      "199 ko:K17818\n",
      "200 ko:K17818\n",
      "201 ko:K18107\n",
      "202 ko:K18102\n",
      "203 ko:K18103\n",
      "204 path:hsa00561\n",
      "207 cpd:C00116\n",
      "208 ko:K18106\n",
      "210 cpd:C00333\n",
      "211 cpd:C15930\n",
      "212 cpd:C20680\n",
      "213 cpd:C02426\n",
      "214 ko:K17743 ko:K26066\n",
      "215 hsa:6652\n",
      "216 hsa:10327\n",
      "217 cpd:C00191\n",
      "219 ko:K08322\n",
      "221 cpd:C00433\n",
      "222 ko:K14259\n",
      "224 ko:K13877 ko:K19588\n",
      "225 cpd:C00026\n",
      "226 path:hsa00020\n",
      "228 ko:K19634\n",
      "230 ko:K00594\n",
      "232 ko:K21619\n",
      "233 ko:K05352 ko:K21681\n",
      "234 ko:K22187\n",
      "237 cpd:C02753\n",
      "238 ko:K24442\n",
      "239 ko:K24443\n",
      "241 cpd:C00216\n",
      "246 ko:K01818\n",
      "249 cpd:C22337\n",
      "250 ko:K00879\n",
      "253 cpd:C00266\n",
      "254 ko:K01628\n",
      "255 cpd:C00111\n",
      "256 ko:K25881 ko:K26066\n"
     ]
    }
   ],
   "source": [
    "for k,v in pathway.entries.items():\n",
    "    print(k,v.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "f025ba3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Bio.KEGG.KGML.KGML_pathway.Entry at 0x1395f05e0>"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway.reaction_entries[0].graphics[]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4c4b7da",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qf",
   "language": "python",
   "name": "qf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
